{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/pytorch_lightning/utilities/warnings.py:53: LightningDeprecationWarning: pytorch_lightning.utilities.warnings.rank_zero_deprecation has been deprecated in v1.6 and will be removed in v1.8. Use the equivalent function from the pytorch_lightning.utilities.rank_zero module instead.\n",
      "  new_rank_zero_deprecation(\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/pytorch_lightning/utilities/warnings.py:58: LightningDeprecationWarning: The `pytorch_lightning.loggers.base.rank_zero_experiment` is deprecated in v1.7 and will be removed in v1.9. Please use `pytorch_lightning.loggers.logger.rank_zero_experiment` instead.\n",
      "  return new_rank_zero_deprecation(*args, **kwargs)\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/phenograph/cluster.py:13: DeprecationWarning: Please use `spmatrix` from the `scipy.sparse` namespace, the `scipy.sparse.base` namespace is deprecated.\n",
      "  from scipy.sparse.base import spmatrix\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import scvi\n",
    "import scgen\n",
    "import rpy2\n",
    "import scib\n",
    "import json\n",
    "import torch\n",
    "import anndata\n",
    "import logging\n",
    "import warnings\n",
    "import scanorama\n",
    "import anndata2ri\n",
    "import matplotlib\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import scrublet as scr\n",
    "import doubletdetection\n",
    "import decoupler as dc\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from os import PathLike, fspath\n",
    "import rpy2.robjects as robjects\n",
    "from scipy.sparse import csr_matrix\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from rpy2.robjects.packages import importr\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:356: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:245: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.1\n",
      "-----\n",
      "OpenSSL                     22.0.0\n",
      "PIL                         9.2.0\n",
      "absl                        NA\n",
      "adjustText                  NA\n",
      "anndata2ri                  1.1\n",
      "annoy                       NA\n",
      "appnope                     0.1.2\n",
      "asttokens                   NA\n",
      "astunparse                  1.6.3\n",
      "attr                        21.4.0\n",
      "backcall                    0.2.0\n",
      "beta_ufunc                  NA\n",
      "binom_ufunc                 NA\n",
      "boto3                       1.26.32\n",
      "botocore                    1.29.32\n",
      "bottleneck                  1.3.5\n",
      "brotli                      NA\n",
      "certifi                     2022.09.24\n",
      "cffi                        1.15.1\n",
      "chex                        0.1.5\n",
      "cloudpickle                 2.2.0\n",
      "colorama                    0.4.4\n",
      "contextlib2                 NA\n",
      "cryptography                38.0.1\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dask                        2022.11.0\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.5.1\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "defusedxml                  0.7.1\n",
      "deprecate                   0.3.2\n",
      "deprecated                  1.2.13\n",
      "docrep                      0.3.2\n",
      "doubletdetection            4.2\n",
      "entrypoints                 0.4\n",
      "executing                   0.8.3\n",
      "fbpca                       NA\n",
      "flax                        0.6.1\n",
      "fsspec                      2022.10.0\n",
      "google                      NA\n",
      "h5py                        3.7.0\n",
      "hypergeom_ufunc             NA\n",
      "igraph                      0.10.2\n",
      "intervaltree                NA\n",
      "ipykernel                   6.15.2\n",
      "ipython_genutils            0.2.0\n",
      "ipywidgets                  7.6.5\n",
      "jax                         0.3.24\n",
      "jaxlib                      0.3.24\n",
      "jedi                        0.18.1\n",
      "jinja2                      3.1.2\n",
      "jmespath                    1.0.1\n",
      "joblib                      1.1.1\n",
      "kiwisolver                  1.4.2\n",
      "leidenalg                   0.9.0\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.1\n",
      "matplotlib                  3.5.3\n",
      "ml_collections              NA\n",
      "mpl_toolkits                NA\n",
      "msgpack                     1.0.3\n",
      "mudata                      0.1.2\n",
      "multipledispatch            0.6.0\n",
      "natsort                     7.1.1\n",
      "nbinom_ufunc                NA\n",
      "ncf_ufunc                   NA\n",
      "numba                       0.56.3\n",
      "numexpr                     2.8.3\n",
      "numpy                       1.23.3\n",
      "numpyro                     0.10.1\n",
      "opt_einsum                  v3.3.0\n",
      "optax                       0.1.3\n",
      "packaging                   21.3\n",
      "pandas                      1.4.4\n",
      "parso                       0.8.3\n",
      "pexpect                     4.8.0\n",
      "phenograph                  1.5.7\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "prompt_toolkit              3.0.20\n",
      "psutil                      5.9.0\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     11.0.0\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.6.0\n",
      "pydevd_concurrency_analyser NA\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.13.0\n",
      "pyparsing                   3.0.9\n",
      "pyro                        1.8.2+2e3bd02\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2022.1\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.10\n",
      "scanorama                   1.7.2\n",
      "scgen                       2.1.0\n",
      "scib                        1.0.4\n",
      "scipy                       1.9.3\n",
      "scrublet                    NA\n",
      "scvi                        0.19.0\n",
      "seaborn                     0.11.2\n",
      "session_info                1.0.0\n",
      "setuptools                  65.6.3\n",
      "six                         1.16.0\n",
      "sklearn                     1.1.3\n",
      "sortedcontainers            2.4.0\n",
      "sparse                      0.13.0\n",
      "stack_data                  0.2.0\n",
      "statsmodels                 0.13.5\n",
      "tensorboard                 2.10.1\n",
      "texttable                   1.6.4\n",
      "threadpoolctl               2.2.0\n",
      "tlz                         0.12.0\n",
      "toolz                       0.12.0\n",
      "torch                       1.12.1\n",
      "torchmetrics                0.10.2\n",
      "tornado                     6.2\n",
      "tqdm                        4.64.1\n",
      "traitlets                   5.1.1\n",
      "tree                        0.1.7\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "urllib3                     1.26.12\n",
      "wcwidth                     0.2.5\n",
      "wrapt                       1.14.1\n",
      "yaml                        5.4.1\n",
      "zmq                         23.2.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.6.0\n",
      "jupyter_client      7.3.5\n",
      "jupyter_core        4.11.2\n",
      "notebook            6.4.12\n",
      "-----\n",
      "Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:41:22) [Clang 13.0.1 ]\n",
      "macOS-13.0.1-arm64-arm-64bit\n",
      "-----\n",
      "Session information updated at 2023-05-05 11:26\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30cd9808",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(reticulate)\n",
    "    library(ggplot2)\n",
    "    library(tidyr)\n",
    "    library(dplyr)\n",
    "    library(purrr)\n",
    "    library(Seurat)\n",
    "    library(tibble)\n",
    "    library(magrittr) \n",
    "    library(forcats)\n",
    "    library(Matrix)\n",
    "    library(stats)\n",
    "    library(tester)\n",
    "    library(Seurat)\n",
    "    library(methods)\n",
    "    library(matrixStats)\n",
    "    library(edgeR)\n",
    "    library(DESeq2)\n",
    "    library(limma)\n",
    "    library(pbmcapply)\n",
    "    library(parallel)\n",
    "    library(lmerTest)\n",
    "    library(lme4)\n",
    "    library(glmmTMB)\n",
    "    library(blme)\n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1a6a687-9397-4cce-809f-0bca4c0072f6",
   "metadata": {},
   "source": [
    "## Table of contents:\n",
    "\n",
    "  * <a href=#Reading>1. Reading in the data</a>\n",
    "  * <a href=#Preprocessing>2. Systematic differential analysis of gene expression</a>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# **1. Reading in the data**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d40204d0-56d5-4984-a2af-1075b64098f1",
   "metadata": {},
   "source": [
    "## **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e763f1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'leng_etc'\n",
    "\n",
    "adata_annot = sc.read_h5ad(f'../data/processed/{save_prefix}/{save_prefix}_mapped_anndata.h5ad')\n",
    "adata_annot.obs_names_make_unique()\n",
    "adata_annot.var_names_make_unique()\n",
    "adata_annot.X = adata_annot.layers['counts'].copy()\n",
    "del adata_annot.obsm, adata_annot.layers, adata_annot.varm, adata_annot.uns, adata_annot.obsp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fb234cad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 41432 × 16585\n",
       "    obs: 'SampleID', 'PatientID', 'BrainRegion', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.clusters', 'clusterAssignment', 'clusterCellType', 'cell_type', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'predictions'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "53847a6f",
   "metadata": {},
   "source": [
    "## **Prepare metadata**\n",
    "\n",
    "Now we specify other related information\n",
    "\n",
    "Specify the following:\n",
    "\n",
    "- `metadata`: Path to metadata. Metadata must contain a column called `pathology.group` with the only unique groups being `no`, `early`, and `late`.\n",
    "\n",
    "- `map_meta`: whether to map metadata to obtain `pathology.group`. If False, it will be assumed that `pathology.group` exist in `adata.obs`\n",
    "\n",
    "- `test_names`: List of the different test names of interest.\n",
    "\n",
    "- `save_prefix`: Prefix for saving critical files. preferably chosen to be in the format `{source name}_{brain region}`. e.g `mathys_pfc`\n",
    "\n",
    "- `subject_id`: Name of the column containing Subject/Patient ID in metadata and `.obs`\n",
    "\n",
    "- `covariates`: This should be a list of additional confounding covariates (`not including pathology.group`) e.g `Sex`, `Sample Batch`, `Age` and other factors that might not be of interest but might have an effect on the pathological status. If you would not like to include any cofounders, please set `covaraites = ['None']`.\n",
    "\n",
    "- `filter_genes`: Specifies whether to filter genes using `gene_celltype_threshold` before before performing differential expression tests`\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a1ba83ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_meta = True\n",
    "filter_genes = \"TRUE\"\n",
    "subject_id = 'PatientID'                # for leng this is `PatientID` for mathys is 'Subject', and 'individualID' for allen\n",
    "gene_celltype_threshold = 0.05          # determines number of cells the gene must be expressed in \n",
    "covariates = ['None']                   # list of covariates to be accounted for in regression.\n",
    "test_names = ['late_vs_early', 'early_vs_no', 'late_vs_no', 'ad_vs_no']\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' \n",
    "\n",
    "\n",
    "deg_methods_to_run = robjects.ListVector({\n",
    "                                        'single_cell_methods': \"FALSE\",\n",
    "                                        'pseudo_bulk_methods': \"TRUE\",\n",
    "                                        'mixed_model_methods': \"FALSE\",\n",
    "                                        })\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a5d8a01b",
   "metadata": {},
   "source": [
    "## **Systematic differential analysis of gene expression**\n",
    "\n",
    "[**Hansruedi Mathys et. al.**](https://doi.org/10.1038/s41586-019-1195-2) compared gene expression levels between `AD-pathology and no-pathology individuals in a cell type manner. The differential expression analysis was assessed using two tests. \n",
    "\n",
    "- **First**, a cell-level analysis was performed using the Wilcoxon rank-sum test and FDR multiple-testing correction (`FDR-adjusted p-values`). \n",
    "\n",
    "- **Second**, a Poisson mixed model accounting for the individual of origin for nuclei and for unwanted sources of variability was performed using the R packages `lme4` and `RUV-seq`, respectively.\n",
    "\n",
    "\n",
    "However, extensive evidence suggests that approaches that take a cell-level view where cells are modeled/considered individually are often susceptible to pseudoreplication where inferential statistics is applied to biological replicates which are not statistically independent. As a result, failing to account for the inherent correlation of replicates (cells from the same individual) inflates the false discovery rate (FDR). [**Junttila et al. 2022**](https://doi.org/10.1093/bib/bbac286), [**Squair et al. 2021**](https://www.nature.com/articles/s41467-021-25960-2), [**Zimmerman et. al 2021**](https://www.nature.com/articles/s41467-021-21038-1)\n",
    "\n",
    "\n",
    "Indeed, these studies suggest that, batch effect correction or the aggregation of cell-type-specific expression values within an individual through either a sum, mean or random effect per individual, that is `pseudobulk generation`, should be applied prior to DGE analysis to account for within-sample correlations [**Zimmerman et. al 2021**](https://www.nature.com/articles/s41467-021-21038-1). Generally, both, pseudobulk methods with sum aggregation such as edgeR, DESeq2, or Limma [**Ritchie et al., 2015**](https://academic.oup.com/nar/article/43/7/e47/2414268) and mixed models such as MAST with random effect setting were found to be superior compared to naive methods, such as the popular Wilcoxon rank-sum test or Seurat’s latent models, which do not account for them [**Junttila et al. 2022**](https://doi.org/10.1093/bib/bbac286).\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e772597d",
   "metadata": {},
   "source": [
    "## **Methods**\n",
    "\n",
    "Here we implement a modified version of the R package, `Libra`, provided by [**Squair et al. 2021**](https://www.nature.com/articles/s41467-021-25960-2), which implementing all methods for DE analysis discussed in the study within a consistent interface. \n",
    "The source package is available from [GitHub](https://github.com/neurorestore/Libra) and as Supplementary Software. \n",
    "\n",
    "The source package implements a total of `22 unique differential expression methods` that can all be accessed from a single function. These methods encompass traditional single-cell methods as well as methods accounting for biological replicate including pseudobulk and mixed model methods. The code for this package has been largely inspired by the Seurat and Muscat packages. Please see the documentation of these packages for further information.\n",
    "\n",
    "### **Usage**\n",
    "\n",
    "The main function of `run_de` takes as input a preprocessed features-by-cells (e.g., genes-by-cells for scRNA-seq) matrix, and a data frame containing metadata associated with each cell, minimally including the cell type annotations, replicates, and sample labels to be predicted.\n",
    "\n",
    "The methods impolemented in this package are summarized as follows:\n",
    "\n",
    "__Single cell methods__\n",
    "\n",
    "- Wilcoxon Rank-Sum test\n",
    "- Likelihood ratio test\n",
    "- Student's t-test\n",
    "- Negative binomial linear model\n",
    "- Logistic regression\n",
    "- MAST\n",
    "\n",
    "__Pseudobulk methods__\n",
    "- edgeR-LRT\n",
    "- edgeR-QLF\n",
    "- DESeq2-LRT\n",
    "- DESeq2-Wald\n",
    "- limma-trend\n",
    "- limma-voom\n",
    "\n",
    "__Mixed model methods__\n",
    "- Linear mixed model\n",
    "- Linear mixed model-LRT\n",
    "- Negative binomial generalized linear mixed model\n",
    "- Negative binomial generalized linear mixed model-LRT\n",
    "- Negative binomial generalized linear mixed model with offset\n",
    "- Negative binomial generalized linear mixed model with offset-LRT\n",
    "- Poisson generalized linear mixed model\n",
    "- Poisson generalized linear mixed model-LRT\n",
    "- Poisson generalized linear mixed model with offset\n",
    "- Poisson generalized linear mixed model with offset-LRT\n",
    "\n",
    "By default the pipeline will use a pseudobulk approach, implementing the `edgeR` package with a likelihood ratio test (LRT) null hypothesis testing framework. Each of the 22 tests can be accessed through three key variables of the `run_de` function: `de_family`, `de_method`, and `de_type`. Their precise access arguments are summarized in the below table.\n",
    "\n",
    "| Method | de_family | de_method | de_type |\n",
    "|--------|-----------|-----------|---------|\n",
    "Wilcoxon Rank-Sum test | singlecell | wilcox | |\n",
    "Likelihood ratio test | singecell | bimod | |\n",
    "Student's t-test | singlecell | t | |\n",
    "Negative binomial linear model | singlecell | negbinom | |\n",
    "Logistic regression | singlecell | LR | |\n",
    "MAST | singlecell | MAST | |\n",
    "edgeR-LRT | pseudobulk | edgeR | LRT\n",
    "edgeR-QLF | pseudobulk | edgeR | QLF\n",
    "DESeq2-LRT | pseudobulk | DESeq2 | LRT\n",
    "DESeq2-Wald | pseudobulk | DESeq2 | Wald\n",
    "limma-trend | pseudobulk | limma | trend\n",
    "limma-voom | pseudobulk | limma | voom\n",
    "Linear mixed model | mixedmodel | linear | Wald\n",
    "Linear mixed model-LRT | mixedmodel | linear | LRT\n",
    "Negative binomial generalized linear mixed model | mixedmodel | negbinom | Wald\n",
    "Negative binomial generalized linear mixed model-LRT | mixedmodel | negbinom | LRT\n",
    "Negative binomial generalized linear mixed model with offset | mixedmodel | negbinom_offset | Wald\n",
    "Negative binomial generalized linear mixed model with offset-LRT | mixedmodel | negbinom_offset | LRT\n",
    "Poisson generalized linear mixed model | mixedmodel | poisson | Wald\n",
    "Poisson generalized linear mixed model-LRT | mixedmodel | poisson | LRT\n",
    "Poisson generalized linear mixed model with offset | mixedmodel | poisson_offset | Wald\n",
    "Poisson generalized linear mixed model with offset-LRT | mixedmodel | poisson_offset | LRT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2de08892",
   "metadata": {},
   "outputs": [],
   "source": [
    "single_cell_methods = ['wilcox', 'MAST']\n",
    "pseudo_bulk_methods = ['edgeR-QLF', 'DESeq2-Wald', 'limma-voom']\n",
    "mixed_model_methods = ['poisson-LRT', 'negbinom-LRT']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Consensus of Differential Expressed Transcripts**\n",
    "\n",
    "Considering these models have different specificity, we apply of the methods and genes that re,\n",
    "\n",
    "As in [**Hansruedi Mathys et. al.**](https://doi.org/10.1038/s41586-019-1195-2), the consistency of DEGs detected using the  `single-cell`, `mixed-model`, and the `pseudobulk` approaches will be assessed by comparing the directionality and rank of DEGs across all the methods. Considering these models have different specificity, we apply all of the methods. In [**Hansruedi Mathys et. al.**](https://doi.org/10.1038/s41586-019-1195-2) consistency in directionality was measured by counting the fraction of the top 1,000 DEGs (ranked by FDR scores) detected in cell-level analysis that showed consistent direction in the deep generative/mixed model.\n",
    "\n",
    "For analyses involving DEG counts, we used only genes that were significantly supported by all methods using the criteria\n",
    "\n",
    "- `FDR-corrected P < 0.05 in a pseudobulk test`,\n",
    "- `FDR-corrected P < 0.05 in a single-cell test`,\n",
    "- `FDR-corrected P < 0.05 in the Poisson mixed model` \n",
    "- `absolute log2(mean gene expression in AD category x/mean gene expression in AD category y) > 0.25`,\n",
    "\n",
    "\n",
    "when comparing  `AD-pathology` group to `no-pathology` such that the log foldchange is ;\n",
    "\n",
    "$$ Log_{2} ({Mean\\ Gene\\ Expression\\ in\\ AD\\ category\\ of\\ Cell\\ Type\\ x \\over Mean\\ Gene\\ Expression\\ in\\ Normal\\ category\\ of\\ Cell\\ Type\\ x})$$"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "edb67647",
   "metadata": {},
   "source": [
    "## **Map Metadata**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d3301215",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = pd.read_csv(metadata)\n",
    "meta = meta.astype(str)\n",
    "mapping = dict(zip(meta[subject_id], meta['pathology.group']))\n",
    "adata_annot.obs['pathology.group'] = adata_annot.obs[subject_id].map(mapping)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "866e1208",
   "metadata": {},
   "source": [
    "## **Loading data into memory**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c09d6bfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtering genes in Excitatory...\n",
      "filtering genes in Inhibitory...\n",
      "filtering genes in Astrocyte...\n",
      "filtering genes in Microglia...\n",
      "filtering genes in Oligodendrocyte...\n",
      "filtering genes in OPC...\n"
     ]
    }
   ],
   "source": [
    "pdata = dc.get_pseudobulk(adata_annot,\n",
    "                          sample_col=subject_id,\n",
    "                          groups_col='cell_type',\n",
    "                          layer=None,\n",
    "                          mode='sum',\n",
    "                          min_cells=0,\n",
    "                          min_counts=0\n",
    "                         )\n",
    "pdata\n",
    "\n",
    "genes_to_keep = {}\n",
    "\n",
    "celltypes = ['Excitatory', 'Inhibitory', 'Astrocyte', 'Microglia', 'Oligodendrocyte', 'OPC']\n",
    "\n",
    "for cell_type in celltypes:\n",
    "    print(f'filtering genes in {cell_type}...')\n",
    "    genes_to_keep[cell_type] = dc.filter_by_prop(pdata[pdata.obs['cell_type'] == cell_type].copy(), min_prop=gene_celltype_threshold)\n",
    "\n",
    "genes_to_keep_list = robjects.ListVector(genes_to_keep)\n",
    "\n",
    "del pdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3a679ec8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:356: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:245: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/scRNA_seq_meta_analysis/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 16585 41432 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(16585): SAMD11 NOC2L ... S100B PRMT2\n",
      "rowData names(12): mt n_cells_by_counts ... mean std\n",
      "colnames(41432): EC2_AAACCTGAGGATGCGT EC2_AAACCTGAGTCAATAG ...\n",
      "  EC10_TTTGTCATCTATCGCC EC10_TTTGTCATCTCTGCTG\n",
      "colData names(24): SampleID PatientID ... predictions pathology.group\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "[1] \"loaded data into memory for recursive use\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_annot -i genes_to_keep_list -i deg_methods_to_run -i celltypes\n",
    "\n",
    "print(adata_annot)\n",
    "\n",
    "print('loaded data into memory for recursive use')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c10e5b17",
   "metadata": {},
   "source": [
    "#### **Pseduo-bulk Differential Expression Analysis with custom scripts adapted from [**Squair et al. 2021**](https://www.nature.com/articles/s41467-021-25960-2)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "66f7e120",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Excitatory...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Excitatory\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Inhibitory...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Inhibitory\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Astrocyte...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Astrocyte\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Microglia...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Microglia\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in Oligodendrocyte...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"Oligodendrocyte\"\n",
      "[1] \"........................\"\n",
      "[1] \"--------------------------------\"\n",
      "[1] \"Estimating DEGs in OPC...\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_EARLY with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for EARLY_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for LATE_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method EDGER-QLF\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method DESEQ2-WALD\"\n",
      "[1] \"OPC\"\n",
      "[1] \"----------------------------------\"\n",
      "[1] \"Obtaining statistics for AD_VS_NO with pseudo-bulk method LIMMA-VOOM\"\n",
      "[1] \"OPC\"\n",
      "[1] \"........................\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i covariates -i test_names -i subject_id -i pseudo_bulk_methods -i gene_celltype_threshold -o pseudobulk_degs\n",
    "\n",
    "library(scuttle)\n",
    "\n",
    "source('../scripts/functions/deg_functions/run_de.R')\n",
    "source('../scripts/functions/deg_functions/pseudobulk_de.R')\n",
    "source('../scripts/functions/deg_functions/check_inputs.R')\n",
    "source('../scripts/functions/deg_functions/to_pseudobulk.R')\n",
    "source('../scripts/functions/deg_functions/mixedmodel_de.R')\n",
    "source('../scripts/functions/deg_functions/singlecell_de.R')\n",
    "\n",
    "# celltypes <- c('Excitatory', 'Inhibitory', 'Astrocyte', 'Microglia', 'Oligodendrocyte', 'OPC')\n",
    "\n",
    "if ('None' %in% covariates){\n",
    "    latent_vars = NULL\n",
    "} else {\n",
    "    latent_vars = covariates\n",
    "}\n",
    "\n",
    "\n",
    "if (eval(deg_methods_to_run[['pseudo_bulk_methods']])){ \n",
    "    pseudobulk_degs <- list()\n",
    "    for (celltype in celltypes){\n",
    "\n",
    "        pseudobulk_degs[[celltype]] <- list()\n",
    "\n",
    "        # filter out celltypes\n",
    "        print('--------------------------------')\n",
    "        print(paste0('Estimating DEGs in ', celltype, '...'))\n",
    "        sce_cell <- adata_annot[, adata_annot$cell_type == celltype]\n",
    "        # print('----------------------------------')\n",
    "        # print(paste0('Dimensions before filtering genes: ', 'n_genes: ',\n",
    "        #      dim(assay(sce_cell))[1], ', n_cells: ', dim(assay(sce_cell))[2], '...'))\n",
    "\n",
    "        # assay(sce_cell, 'counts') <- assay(sce_cell)\n",
    "        \n",
    "        # Aggregate across cluster-sample groups\n",
    "        # aggr_counts <- aggregateAcrossCells(sce_cell, ids=colData(sce_cell)[, c(subject_id)],)\n",
    "\n",
    "        # create an edgeR object with counts and grouping factor\n",
    "        # y <- DGEList(assay(aggr_counts, \"counts\"), group = colnames(aggr_counts))\n",
    "        # filter out genes with low counts\n",
    "        # keep <- filterByExpr(y, min.prop = gene_celltype_threshold)\n",
    "        # y <- y[keep, , keep.lib.sizes=FALSE]\n",
    "        # Subset the `SingleCellExperiment` object to include only genes with > 0.1 detection rate/frequency\n",
    "        sce_cell <- sce_cell[genes_to_keep_list[[celltype]], ]\n",
    "                            \n",
    "        # print('----------------------------------')\n",
    "        # print(paste0('Dimensions after filtering genes: ', 'n_genes: ',\n",
    "        #      dim(assay(sce_cell))[1], ', n_cells: ', dim(assay(sce_cell))[2], '...'))\n",
    "\n",
    "        # assay(sce_cell) <- assay(sce_cell, \"counts\")\n",
    "        meta = colData(sce_cell)\n",
    "\n",
    "        for (test_name in test_names){\n",
    "            # print('---------------------------------')\n",
    "            # print(paste0('Running differential expression test for: ', test_name))\n",
    "            sce <- sce_cell\n",
    "\n",
    "            pseudobulk_degs[[celltype]][[test_name]] <- list()\n",
    "\n",
    "            # if (test_name == 'ad_vs_no'){\n",
    "            #         #sce$pathology.group = ifelse(sce$pathology.group == \"no\", \"no\", \"ad\")\n",
    "            #         ref_level <- tail(strsplit('ad_vs_no', \"_vs_\")[[1]], n = 1)\n",
    "            #     } else if (test_name %in% c('late_vs_no', 'late_vs_early', 'early_vs_no')){\n",
    "            #         #sce <- sce[, colData(sce)$pathology.group %in% c(strsplit(test_name, \"_vs_\")[[1]])]\n",
    "            #         ref_level <- tail(strsplit(test_name, \"_vs_\")[[1]], n = 1)\n",
    "            #     }\n",
    "            \n",
    "            ref_level = 'no'\n",
    "            meta$pathology.group <- factor(meta$pathology.group)\n",
    "            meta$pathology.group <- relevel(meta$pathology.group, ref = ref_level)\n",
    "\n",
    "            for(de_method in pseudo_bulk_methods){\n",
    "                print('----------------------------------')\n",
    "                print(paste0('Obtaining statistics for ', toupper(test_name), ' with pseudo-bulk method ', toupper(de_method)))\n",
    "                pseudobulk_degs[[celltype]][[test_name]][[de_method]] = run_de(sce,\n",
    "                                                                                meta = meta,\n",
    "                                                                                replicate_col = subject_id,\n",
    "                                                                                cell_type_col = 'cell_type',\n",
    "                                                                                label_col = 'pathology.group',\n",
    "                                                                                latent_vars = latent_vars,\n",
    "                                                                                test_name = test_name,\n",
    "                                                                                ref_level = ref_level,\n",
    "                                                                                min_cells = 3,\n",
    "                                                                                min_reps = 2,\n",
    "                                                                                min_features = 0,\n",
    "                                                                                de_family = 'pseudobulk',\n",
    "                                                                                de_method = strsplit(de_method, \"-\")[[1]][1],\n",
    "                                                                                de_type = strsplit(de_method, \"-\")[[1]][2],\n",
    "                                                                                n_threads = 2)\n",
    "            }\n",
    "        }\n",
    "\n",
    "        print('........................')\n",
    "            \n",
    "    }\n",
    "}    \n",
    "else{\n",
    "\n",
    "    print(\"Not evaluating DEGs with Pseudobulk Methods.\")\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6bf927c1",
   "metadata": {},
   "source": [
    "##### **Save Results**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "00b0b482",
   "metadata": {},
   "outputs": [],
   "source": [
    "if eval(dict(zip(deg_methods_to_run.names, list(deg_methods_to_run)))['pseudo_bulk_methods'][0].capitalize()):\n",
    "    for test_name in test_names:\n",
    "        for deg_method in pseudo_bulk_methods:\n",
    "            with pd.ExcelWriter(f\"../results/{test_name}/{save_prefix}_{deg_method}_degs.xlsx\") as writer:\n",
    "                for cell_type in celltypes:\n",
    "                    \n",
    "                    df = pseudobulk_degs[cell_type][test_name][deg_method].sort_values(by='p_val_adj')\n",
    "                    df['abs_logFC'] = abs(df['avg_logFC'])\n",
    "                    df['direction'] = df['avg_logFC'].apply(lambda x: \"up\" if x>0 else \"down\")\n",
    "                    df.to_excel(writer, sheet_name=cell_type, na_rep='NA')\n",
    "else:\n",
    "    print('No result saved for Pseudobulk Methods')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ae412e1f",
   "metadata": {},
   "source": [
    "### **Mixed Models**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5a7d8857",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Not evaluating DEGs with Mixed-model Methods.\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i covariates -i test_names -i subject_id -i gene_celltype_threshold -i mixed_model_methods -o mixedmodel_degs\n",
    "\n",
    "source('../scripts/functions/deg_functions/run_de.R')\n",
    "source('../scripts/functions/deg_functions/pseudobulk_de.R')\n",
    "source('../scripts/functions/deg_functions/check_inputs.R')\n",
    "source('../scripts/functions/deg_functions/to_pseudobulk.R')\n",
    "source('../scripts/functions/deg_functions/mixedmodel_de.R')\n",
    "source('../scripts/functions/deg_functions/singlecell_de.R')\n",
    "\n",
    "# celltypes <- c('Excitatory', 'Inhibitory', 'Astrocyte', 'Microglia', 'Oligodendrocyte, OPC')\n",
    "\n",
    "\n",
    "if ('None' %in% covariates){\n",
    "    latent_vars = NULL\n",
    "} else {\n",
    "    latent_vars = covariates\n",
    "}\n",
    "\n",
    "if (eval(deg_methods_to_run[['mixed_model_methods']])){ \n",
    "    mixedmodel_degs <- list()\n",
    "    for (celltype in celltypes){\n",
    "\n",
    "        mixedmodel_degs[[celltype]] <- list()\n",
    "\n",
    "        # filter out celltypes\n",
    "        print('--------------------------------')\n",
    "        print(paste0('Estimating DEGs in ', celltype, '...'))\n",
    "        sce_cell <- adata_annot[, adata_annot$cell_type == celltype]\n",
    "        # print('----------------------------------')\n",
    "        # print(paste0('Dimensions before filtering genes: ', 'n_genes: ',\n",
    "        #      dim(assay(sce_cell))[1], ', n_cells: ', dim(assay(sce_cell))[2], '...'))\n",
    "\n",
    "        assay(sce_cell, 'counts') <- assay(sce_cell)\n",
    "\n",
    "        # Calculate the detection rate/frequency for each gene\n",
    "        detection_rate <- rowMeans(counts(sce_cell) > 0)\n",
    "        # Subset the `SingleCellExperiment` object to include only genes with >0.1 detection rate/frequency\n",
    "        sce_cell <- sce_cell[detection_rate > gene_celltype_threshold, ]\n",
    "        # print('----------------------------------')\n",
    "        # print(paste0('Dimensions after filtering genes: ', 'n_genes: ',\n",
    "        #      dim(assay(sce_cell))[1], ', n_cells: ', dim(assay(sce_cell))[2], '...'))\n",
    "\n",
    "        # assay(sce_cell) <- assay(sce_cell, \"counts\")\n",
    "        meta = colData(sce_cell)\n",
    "\n",
    "        for (test_name in test_names){\n",
    "            # print('---------------------------------')\n",
    "            # print(paste0('Running differential expression test for: ', test_name))\n",
    "            sce <- sce_cell\n",
    "\n",
    "            mixedmodel_degs[[celltype]][[test_name]] <- list()\n",
    "\n",
    "            if (test_name == 'ad_vs_no'){\n",
    "                    sce$pathology.group = ifelse(sce$pathology.group == \"no\", \"no\", \"ad\")\n",
    "                    ref_level <- tail(strsplit('ad_vs_no', \"_vs_\")[[1]], n = 1)\n",
    "                } else if (test_name %in% c('late_vs_no', 'late_vs_early', 'early_vs_no')){\n",
    "                    sce <- sce[, colData(sce)$pathology.group %in% c(strsplit(test_name, \"_vs_\")[[1]])]\n",
    "                    ref_level <- tail(strsplit(test_name, \"_vs_\")[[1]], n = 1)\n",
    "                }\n",
    "            \n",
    "            meta$pathology.group <- factor(meta$pathology.group)\n",
    "            meta$pathology.group <- relevel(meta$pathology.group, ref = ref_level)\n",
    "\n",
    "            for(de_method in mixed_model_methods){\n",
    "                print('----------------------------------')\n",
    "                print(paste0('Obtaining statistics for ', toupper(test_name), ' with single-cell method ', toupper(de_method)))\n",
    "                mixedmodel_degs[[celltype]][[test_name]][[de_method]] = run_de(sce,\n",
    "                                                                                meta = meta,\n",
    "                                                                                replicate_col = subject_id,\n",
    "                                                                                cell_type_col = 'cell_type',\n",
    "                                                                                label_col = 'pathology.group',\n",
    "                                                                                latent_vars = latent_vars,\n",
    "                                                                                min_cells = 3,\n",
    "                                                                                min_reps = 2,\n",
    "                                                                                min_features = 0,\n",
    "                                                                                de_family = 'mixedmodel',\n",
    "                                                                                de_method = strsplit(de_method, \"-\")[[1]][1],\n",
    "                                                                                de_type = strsplit(de_method, \"-\")[[1]][2],\n",
    "                                                                                n_threads = 2)\n",
    "            }\n",
    "        }\n",
    "        print('........................')\n",
    "    }\n",
    "}    \n",
    "else{\n",
    "\n",
    "    print(\"Not evaluating DEGs with Mixed-model Methods.\")\n",
    "    mixedmodel_degs = NULL\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa3f7229",
   "metadata": {},
   "source": [
    "##### **Save Results**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2106faf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No result saved for Mixed-model Methods\n"
     ]
    }
   ],
   "source": [
    "if eval(dict(zip(deg_methods_to_run.names, list(deg_methods_to_run)))['mixed_model_methods'][0].capitalize()):\n",
    "    for test_name in test_names:\n",
    "        for deg_method in mixed_model_methods:\n",
    "            with pd.ExcelWriter(f\"../results/{test_name}/{save_prefix}_{deg_method}_degs.xlsx\") as writer:\n",
    "                for cell_type in celltypes:\n",
    "                    \n",
    "                    df = mixedmodel_degs[cell_type][test_name][deg_method].sort_values(by='p_val_adj')\n",
    "                    df['abs_logFC'] = abs(df['avg_logFC'])\n",
    "                    df['direction'] = df['avg_logFC'].apply(lambda x: \"up\" if x>0 else \"down\")\n",
    "                    df.to_excel(writer, sheet_name=cell_type, na_rep='NA')\n",
    "else:\n",
    "    print('No result saved for Mixed-model Methods')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4974ce28",
   "metadata": {},
   "source": [
    "### **Single-cell Models**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "36127617",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Not evaluating DEGs with Single-cell Methods.\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i covariates -i test_names -i subject_id -i gene_celltype_threshold -i single_cell_methods -o singlecell_degs\n",
    "\n",
    "source('../scripts/functions/deg_functions/run_de.R')\n",
    "source('../scripts/functions/deg_functions/pseudobulk_de.R')\n",
    "source('../scripts/functions/deg_functions/check_inputs.R')\n",
    "source('../scripts/functions/deg_functions/to_pseudobulk.R')\n",
    "source('../scripts/functions/deg_functions/mixedmodel_de.R')\n",
    "source('../scripts/functions/deg_functions/singlecell_de.R')\n",
    "\n",
    "#celltypes <- c('Excitatory', 'Inhibitory', 'Astrocyte', 'Microglia', 'Oligodendrocyte', 'OPC')\n",
    "#de_methods <- c('wilcox', 'bimod', 't', 'negbinom', 'LR', 'MAST')\n",
    "\n",
    "\n",
    "if ('None' %in% covariates){\n",
    "    latent_vars = NULL\n",
    "} else {\n",
    "    latent_vars = covariates\n",
    "}\n",
    "\n",
    "if (eval(deg_methods_to_run[['single_cell_methods']])){ \n",
    "    singlecell_degs <- list()\n",
    "    for (celltype in celltypes){\n",
    "\n",
    "        singlecell_degs[[celltype]] <- list()\n",
    "\n",
    "        # filter out celltypes\n",
    "        print('--------------------------------')\n",
    "        print(paste0('Estimating DEGs in ', celltype, '...'))\n",
    "        sce_cell <- adata_annot[, adata_annot$cell_type == celltype]\n",
    "        # print('----------------------------------')\n",
    "        # print(paste0('Dimensions before filtering genes: ', 'n_genes: ',\n",
    "        #      dim(assay(sce_cell))[1], ', n_cells: ', dim(assay(sce_cell))[2], '...'))\n",
    "\n",
    "        # Calculate the detection rate/frequency for each gene\n",
    "        detection_rate <- rowMeans(counts(sce_cell) > 0)\n",
    "        # Subset the `SingleCellExperiment` object to include only genes with >0.1 detection rate/frequency\n",
    "        sce_cell <- sce_cell[detection_rate > gene_celltype_threshold, ]\n",
    "        # print('----------------------------------')\n",
    "        # print(paste0('Dimensions after filtering genes: ', 'n_genes: ',\n",
    "        #      dim(assay(sce_cell))[1], ', n_cells: ', dim(assay(sce_cell))[2], '...'))\n",
    "\n",
    "        # assay(sce_cell) <- assay(sce_cell, \"counts\")\n",
    "        meta = colData(sce_cell)\n",
    "\n",
    "        for (test_name in test_names){\n",
    "            # print('---------------------------------')\n",
    "            # print(paste0('Running differential expression test for: ', test_name))\n",
    "            sce <- sce_cell\n",
    "\n",
    "            singlecell_degs[[celltype]][[test_name]] <- list()\n",
    "\n",
    "            if (test_name == 'ad_vs_no'){\n",
    "                    sce$pathology.group = ifelse(sce$pathology.group == \"no\", \"no\", \"ad\")\n",
    "                    ref_level <- tail(strsplit('ad_vs_no', \"_vs_\")[[1]], n = 1)\n",
    "                } else if (test_name %in% c('late_vs_no', 'late_vs_early', 'early_vs_no')){\n",
    "                    sce <- sce[, colData(sce)$pathology.group %in% c(strsplit(test_name, \"_vs_\")[[1]])]\n",
    "                    ref_level <- tail(strsplit(test_name, \"_vs_\")[[1]], n = 1)\n",
    "                }\n",
    "            \n",
    "            meta$pathology.group <- factor(meta$pathology.group)\n",
    "            meta$pathology.group <- relevel(meta$pathology.group, ref = ref_level)\n",
    "\n",
    "            for(de_method in single_cell_methods){\n",
    "                print('----------------------------------')\n",
    "                print(paste0('Obtaining statistics for ', toupper(test_name), ' with single-cell method ', toupper(de_method)))\n",
    "                singlecell_degs[[celltype]][[test_name]][[de_method]] = run_de(sce,\n",
    "                                                                                meta = meta,\n",
    "                                                                                replicate_col = subject_id,\n",
    "                                                                                cell_type_col = 'cell_type',\n",
    "                                                                                label_col = 'pathology.group',\n",
    "                                                                                latent_vars = latent_vars,\n",
    "                                                                                min_cells = 3,\n",
    "                                                                                min_reps = 2,\n",
    "                                                                                min_features = 0,\n",
    "                                                                                de_family = 'singlecell',\n",
    "                                                                                de_method = de_method,\n",
    "                                                                                de_type = NULL,\n",
    "                                                                                n_threads = 2)\n",
    "            }\n",
    "        }\n",
    "\n",
    "        print('........................')\n",
    "    }\n",
    "\n",
    "}    \n",
    "else{\n",
    "\n",
    "    print(\"Not evaluating DEGs with Single-cell Methods.\")\n",
    "    singlecell_degs = NULL\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b18d571a",
   "metadata": {},
   "source": [
    "##### **Save Results**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "02c5cbb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No result saved for Mixed-model Methods\n"
     ]
    }
   ],
   "source": [
    "if eval(dict(zip(deg_methods_to_run.names, list(deg_methods_to_run)))['single_cell_methods'][0].capitalize()):\n",
    "    for test_name in test_names:\n",
    "        for deg_method in single_cell_methods:\n",
    "            with pd.ExcelWriter(f\"../results/{test_name}/{save_prefix}_{deg_method}_degs.xlsx\") as writer:\n",
    "                for cell_type in celltypes:\n",
    "                    \n",
    "                    df = singlecell_degs[cell_type][test_name][deg_method].sort_values(by='p_val_adj')\n",
    "                    df['abs_logFC'] = abs(df['avg_logFC'])\n",
    "                    df['direction'] = df['avg_logFC'].apply(lambda x: \"up\" if x>0 else \"down\")\n",
    "                    df.to_excel(writer, sheet_name=cell_type, na_rep='NA')\n",
    "else:\n",
    "    print('No result saved for Mixed-model Methods')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "90c66c6f",
   "metadata": {},
   "source": [
    "## **Consensus of Differential Expressed Transcripts**\n",
    "\n",
    "The consistency of DEGs detected using the cell-level `Wilcoxon rank-sum` analysis model with those obtained with the `MAST` generalized mixed model and the `scANVI` deep generative model was assessed by comparing the directionality and rank of DEGs in the two models. Consistency in directionality was measured by counting the fraction of the top 1,000 DEGs (ranked by FDR scores) detected in cell-level analysis that showed consistent direction in the deep generative/mixed model.\n",
    "\n",
    "For analyses involving DEG counts, [**Hansuredi et. al. 2019**](https://www.nature.com/articles/s41586-019-1195-2) used only genes that were significantly supported by both models using the criteria\n",
    "\n",
    "- `FDR-corrected P < 0.01 in a two sided Wilcoxon-rank sum test`,\n",
    "- `absolute log2(mean gene expression in AD category x/mean gene expression in AD category y) > 0.25`,\n",
    "- `FDR-corrected P < 0.05 in the Poisson mixed model` \n",
    "\n",
    "Motivated by [**Zhou. et. al. 2020**](https://doi.org/10.1038/s41591-019-0695-9) who used estimated DEGs between conditions using the `MAST algorithm of the Seurat package in R`, we include the following filtering criteria for the \n",
    "\n",
    "- `DEGs in the generalized mixed effects model are obatined by filtering genes for log2(fold change) > 0.1, P < 0.05`. \n",
    "- `DEGs in the deep generative model are obatined by filtering genes for`  $$\\ln(Bayes\\ Factor) > 2.3 $$"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "21d83925",
   "metadata": {},
   "source": [
    "### **Save Data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa6a283b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
