{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/pytorch_lightning/utilities/warnings.py:53: LightningDeprecationWarning: pytorch_lightning.utilities.warnings.rank_zero_deprecation has been deprecated in v1.6 and will be removed in v1.8. Use the equivalent function from the pytorch_lightning.utilities.rank_zero module instead.\n",
      "  new_rank_zero_deprecation(\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/pytorch_lightning/utilities/warnings.py:58: LightningDeprecationWarning: The `pytorch_lightning.loggers.base.rank_zero_experiment` is deprecated in v1.7 and will be removed in v1.9. Please use `pytorch_lightning.loggers.logger.rank_zero_experiment` instead.\n",
      "  return new_rank_zero_deprecation(*args, **kwargs)\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/phenograph/cluster.py:13: DeprecationWarning: Please use `spmatrix` from the `scipy.sparse` namespace, the `scipy.sparse.base` namespace is deprecated.\n",
      "  from scipy.sparse.base import spmatrix\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import scvi\n",
    "import scgen\n",
    "import rpy2\n",
    "import scib\n",
    "import json\n",
    "import torch\n",
    "import anndata\n",
    "import logging\n",
    "import warnings\n",
    "import scanorama\n",
    "import anndata2ri\n",
    "import matplotlib\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import scrublet as scr\n",
    "import doubletdetection\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from os import PathLike, fspath\n",
    "import rpy2.robjects as robjects\n",
    "from scipy.sparse import csr_matrix\n",
    "from rpy2.robjects import pandas2ri\n",
    "from matplotlib.pyplot import rcParams\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:351: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/numpy2ri.py:226: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.\n",
      "  warnings.warn('The global conversion available with activate() '\n",
      "/Users/temitopeleke/miniforge3/envs/mathys_reproduce_arm64/lib/python3.10/site-packages/rpy2/robjects/conversion.py:28: DeprecationWarning: The use of converter in module rpy2.robjects.conversion is deprecated. Use rpy2.robjects.conversion.get_conversion() instead of rpy2.robjects.conversion.converter.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.1\n",
      "-----\n",
      "OpenSSL                     22.0.0\n",
      "PIL                         9.2.0\n",
      "absl                        NA\n",
      "adjustText                  NA\n",
      "anndata2ri                  1.1\n",
      "annoy                       NA\n",
      "appnope                     0.1.2\n",
      "asttokens                   NA\n",
      "astunparse                  1.6.3\n",
      "attr                        21.4.0\n",
      "backcall                    0.2.0\n",
      "beta_ufunc                  NA\n",
      "binom_ufunc                 NA\n",
      "boto3                       1.26.32\n",
      "botocore                    1.29.32\n",
      "bottleneck                  1.3.5\n",
      "brotli                      NA\n",
      "certifi                     2022.09.24\n",
      "cffi                        1.15.1\n",
      "chex                        0.1.5\n",
      "cloudpickle                 2.2.0\n",
      "colorama                    0.4.4\n",
      "contextlib2                 NA\n",
      "cryptography                38.0.1\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dask                        2022.11.0\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.5.1\n",
      "decorator                   5.1.1\n",
      "defusedxml                  0.7.1\n",
      "deprecate                   0.3.2\n",
      "deprecated                  1.2.13\n",
      "docrep                      0.3.2\n",
      "doubletdetection            4.2\n",
      "entrypoints                 0.4\n",
      "executing                   0.8.3\n",
      "fbpca                       NA\n",
      "flax                        0.6.1\n",
      "fsspec                      2022.10.0\n",
      "google                      NA\n",
      "h5py                        3.7.0\n",
      "hypergeom_ufunc             NA\n",
      "igraph                      0.10.2\n",
      "intervaltree                NA\n",
      "ipykernel                   6.15.2\n",
      "ipython_genutils            0.2.0\n",
      "ipywidgets                  7.6.5\n",
      "jax                         0.3.24\n",
      "jaxlib                      0.3.24\n",
      "jedi                        0.18.1\n",
      "jinja2                      3.1.2\n",
      "jmespath                    1.0.1\n",
      "joblib                      1.1.1\n",
      "kiwisolver                  1.4.2\n",
      "leidenalg                   0.9.0\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.1\n",
      "matplotlib                  3.5.3\n",
      "ml_collections              NA\n",
      "mpl_toolkits                NA\n",
      "msgpack                     1.0.3\n",
      "mudata                      0.2.0\n",
      "multipledispatch            0.6.0\n",
      "natsort                     7.1.1\n",
      "nbinom_ufunc                NA\n",
      "ncf_ufunc                   NA\n",
      "numba                       0.56.3\n",
      "numexpr                     2.8.3\n",
      "numpy                       1.23.3\n",
      "numpyro                     0.10.1\n",
      "opt_einsum                  v3.3.0\n",
      "optax                       0.1.3\n",
      "packaging                   21.3\n",
      "pandas                      1.4.4\n",
      "parso                       0.8.3\n",
      "pexpect                     4.8.0\n",
      "phenograph                  1.5.7\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "prompt_toolkit              3.0.20\n",
      "psutil                      5.9.0\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.6.0\n",
      "pydevd_concurrency_analyser NA\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.13.0\n",
      "pyparsing                   3.0.9\n",
      "pyro                        1.8.2+2e3bd02\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2022.1\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.5\n",
      "scanorama                   1.7.2\n",
      "scgen                       2.1.0\n",
      "scib                        1.0.4\n",
      "scipy                       1.9.3\n",
      "scrublet                    NA\n",
      "scvi                        0.19.0\n",
      "seaborn                     0.11.2\n",
      "session_info                1.0.0\n",
      "setuptools                  65.6.3\n",
      "six                         1.16.0\n",
      "sklearn                     1.1.3\n",
      "sortedcontainers            2.4.0\n",
      "sparse                      0.13.0\n",
      "stack_data                  0.2.0\n",
      "statsmodels                 0.13.5\n",
      "tensorboard                 2.10.1\n",
      "texttable                   1.6.4\n",
      "threadpoolctl               2.2.0\n",
      "tlz                         0.12.0\n",
      "toolz                       0.12.0\n",
      "torch                       1.12.1\n",
      "torchmetrics                0.10.2\n",
      "tornado                     6.2\n",
      "tqdm                        4.64.1\n",
      "traitlets                   5.1.1\n",
      "tree                        0.1.7\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "urllib3                     1.26.12\n",
      "wcwidth                     0.2.5\n",
      "wrapt                       1.14.1\n",
      "yaml                        5.4.1\n",
      "zmq                         23.2.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.6.0\n",
      "jupyter_client      7.3.5\n",
      "jupyter_core        4.11.2\n",
      "notebook            6.4.12\n",
      "-----\n",
      "Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:41:22) [Clang 13.0.1 ]\n",
      "macOS-13.0.1-arm64-arm-64bit\n",
      "-----\n",
      "Session information updated at 2023-03-15 14:04\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30cd9808",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(reticulate)\n",
    "    library(ggplot2)\n",
    "    library(tidyr)\n",
    "    library(dplyr)\n",
    "    library(purrr)\n",
    "    library(Seurat)\n",
    "    library(tibble)\n",
    "    library(magrittr) # needs to be run every time you start R and want to use %>%\n",
    "})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1a6a687-9397-4cce-809f-0bca4c0072f6",
   "metadata": {},
   "source": [
    "## Table of contents:\n",
    "\n",
    "  * <a href=#Reading>1. Reading in the data</a>\n",
    "  * <a href=#Preprocessing>2. Systematic differential analysis of gene expression</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# 1. Reading in the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d40204d0-56d5-4984-a2af-1075b64098f1",
   "metadata": {},
   "source": [
    "### [Mathys et. al. 2019](https://doi.org/10.1038/s41586-019-1195-2) (Prefrontal Cortex)\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e763f1ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_annot = sc.read_h5ad('../data/processed/adata_annotated.h5ad')\n",
    "adata_annot.X = csr_matrix(adata_annot.layers['counts'].toarray())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6263d762",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 65619 × 16590\n",
       "    obs: 'projid', 'fastq', 'Subject', 'sample', 'libraryid', 'study', 'age_death', 'educ', 'msex_x', 'gpath_x', 'amyloid_x', 'plaq_n_x', 'cogdx_x', 'pathologic diagnosis of AD', 'amyloid_y', 'plaq_n_y', 'nft', 'tangles', 'cogn_global_lv', 'gpath_y', 'gpath_3neocort', 'amyloid.group', 'caa_4gp', 'ceradsc', 'braaksc', 'niareagansc', 'cogdx_y', 'msex_y', 'pathology.group', 'sampleid', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'cell_type'\n",
       "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'Subject_colors', 'amyloid.group_colors', 'braaksc_colors', 'cell_type_colors', 'dendrogram_louvain_0.5', 'hvg', 'log1p', 'louvain', 'louvain_0.5_colors', 'louvain_1.0_colors', 'msex_x_colors', 'neighbors', 'ov_t_test_rank_genes_0.5', 'pathology.group_colors', 'pca', 'predicted_doublet_colors', 'scrublet', 't_test_rank_genes_0.5', 'tsne', 'umap'\n",
       "    obsm: 'X_pca', 'X_tsne', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'log'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a5d8a01b",
   "metadata": {},
   "source": [
    "## 2.4 Systematic differential analysis of gene expression\n",
    "\n",
    "[**Hansruedi Mathys et. al.**](https://doi.org/10.1038/s41586-019-1195-2) compared gene expression levels between `AD-pathology and no-pathology individuals in a cell type manner. The differential expression analysis was assessed using two tests. \n",
    "\n",
    "- **First**, a cell-level analysis was performed using the Wilcoxon rank-sum test and FDR multiple-testing correction (`FDR-adjusted p-values`). \n",
    "\n",
    "- **Second**, a Poisson mixed model accounting for the individual of origin for nuclei and for unwanted sources of variability was performed using the R packages `lme4` and `RUV-seq`, respectively.\n",
    "\n",
    "\n",
    "Next, we use the ` Wilcoxon rank-sum test` in `scapany.tl.rank_genes_group` comparing `AD-pathology` group to `no-pathology` such that the log foldchange is ;\n",
    "\n",
    "$$ Log_{2} ({Mean\\ Gene\\ Expression\\ in\\ AD\\ category\\ of\\ Cell\\ Type\\ x \\over Mean\\ Gene\\ Expression\\ in\\ Normal\\ category\\ of\\ Cell\\ Type\\ x})$$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c568575c",
   "metadata": {},
   "source": [
    "##### Group Cells (Pathology / No-Pathology)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "12fa36cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_annot.obs['disease_group'] = None\n",
    "adata_annot.obs.loc[adata_annot.obs['pathology.group'] == 'no-pathology', 'disease_group'] = 'no-pathology'\n",
    "adata_annot.obs.loc[adata_annot.obs['pathology.group'].isin(['early-pathology', 'late-pathology']), 'disease_group'] = 'AD-pathology'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3830d142",
   "metadata": {},
   "source": [
    "##### Load DEGs from Cell-type Differential Expression with Wilcoxon Rank-Sum Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "64fddc81",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_annot.uns['log1p'] = {'base': None}\n",
    "\n",
    "adata_sub = {}\n",
    "degs_t_test = {}\n",
    "for cell_type in adata_annot.obs.cell_type.unique():\n",
    "\n",
    "    # adata_sub[cell_type] = adata_annot[adata_annot.obs.cell_type==cell_type]\n",
    "    # adata_sub[cell_type].X = adata_sub[cell_type].layers['counts'].toarray()\n",
    "    degs_t_test[cell_type] = pd.read_excel('../results/ad_vs_no/t_test_degs.xlsx', \n",
    "                                   sheet_name=cell_type)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0f84051c",
   "metadata": {},
   "source": [
    "Reported DEGs\n",
    "\n",
    "Next, we load the reported DEGs from [**Hansruedi Mathys et. al.**](https://doi.org/10.1038/s41586-019-1195-2) for comparison in a later step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f184cab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mathys_degs = {}\n",
    "for key in adata_annot.obs.cell_type.unique():\n",
    "    try:\n",
    "        mathys_degs[key] = pd.read_excel(f'../data/raw/mathys_pfc_from_paper/degs/ad_vs_no/{key.lower()}_degs.xlsx')\n",
    "        mathys_degs[key].rename({'Unnamed: 0': 'names'}, inplace=True)\n",
    "        mathys_degs[key]['IndModel.absFC'] = np.abs(mathys_degs[key]['IndModel.FC'])\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf3e9017",
   "metadata": {},
   "source": [
    "### Deep Generative Models for Detecting Differential Expression\n",
    "\n",
    "Although we provide `MAST` implementation for all cell clusters using randomly sampled cells with less than 25000 cells in the AD-pathology and no-pathology groups, this parametric based approach scales very poorly to larger clusters (e.g `excitatory` with >20k cells). Thus, we instead employ a deep learning based approaches described by [**Pierre Boyeau et. al. 2022**](https://www.biorxiv.org/content/10.1101/794289v1) & [**Lopez et. al. 2018**](https://www.nature.com/articles/s41592-018-0229-2) and implemented in [**scvi-tools**](https://scvi-tools.org/). \n",
    "\n",
    "[**scvi-tools**](https://scvi-tools.org/) provides a suite of probabilistic single-cell variational inference tools for single-cell omics data.A semi-supervised model called `scANVI` (single-cell ANnotation using Variational Inference) uses cell type knowledge for a portion of the cells contained in the data sets to infer the states of the other cells. As a result, in addition to Differential Gene Expression analysis, scANVI can assist in annotating a data set of unlabelled cells from manually annotated atlases.\n",
    "\n",
    "In the particular case of single-cell RNA-seq data, existing differential expression models often model that the mean expression level \n",
    ". as a linear function of the cell-state and batch assignments. These models face two notable limitations to detect differences in expression between cell-states in large-scale scRNA-seq datasets. First, such linear assumptions may not capture complex batch effects existing in such datasets accurately. When comparing two given states  and  in a large dataset, these models may also struggle to leverage data of a related state present in the data.\n",
    "\n",
    "Existing differential expression models frequently represent the mean expression level in the specific situation of single-cell RNA-seq data as a linear function of the cell-state and batch assignments. To identify changes in expression between cell states in large-scale scRNA-seq datasets, these models have two significant drawbacks. First off, such linear hypotheses might not accurately represent complex batch effects present in such datasets. These models could have trouble utilizing data from a related state that is included in the data when comparing two given states and in a big dataset.\n",
    "\n",
    "These issues might not apply to deep generative models. To capture batch effects on expression, the majority of scvi-tools models employ intricate nonlinear mappings. They can use a lot of data and amortization to better identify shared relationships between features. Deep generative models therefore have desirable characteristics for differential expression in big data.\n",
    "\n",
    "In addition, these models can\n",
    "\n",
    "- approximate population-specific normalized expression levels\n",
    "\n",
    "- detect biologically relevant features\n",
    "\n",
    "- provide easy-to-interpret predictions\n",
    "\n",
    "\n",
    "Please see [**this page**](https://docs.scvi-tools.org/en/stable/user_guide/background/differential_expression.html) for further information \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0613a885",
   "metadata": {},
   "source": [
    "### Explanation of `change` mode and Bayes Factors\n",
    "\n",
    "Explanation adapted from the scVi [`differential_expression_score` docstring](https://github.com/YosefLab/scVI/blob/05920d1f85daa362d4fb694e588ab090bc84e207/scvi/inference/posterior.py#L640).\n",
    "\n",
    "`scANVI` provides several ways to formulate the competing hypotheses from the effect-sizes equivalent to `log2 Fold-Change` to detect Differentially Expressed features. To avoid detecting features of little practical interest, e.g., when expression differences between conditions are significant but very subtle, we set the `mode` hyperparameter to `change`. In this formulation, we consider null hypotheses instead, such that\n",
    "\n",
    "$$|\\beta|\\ \\leq\\ \\delta\\$$\n",
    "\n",
    "Where, delta ($\\delta$) is an hyperparameter that is equivalent to the `log2 Fold-Change` threshold, but can also be estimated by the model in a data-driven fashion. The **`change`** mode follows the protocol described in [Boyeau et al, bioRxiv 2019. doi: 10.1101/794289](https://doi.org/10.1101/794289)\n",
    "\n",
    "Performing differential expression using the `change` mode consists in estimating an effect size random variable (e.g., log fold-change) and \n",
    "performing Bayesian hypothesis testing on this variable. \n",
    "\n",
    "The `change function` computes the effect size variable `r` based two inputs corresponding to the normalized means in both populations \n",
    "\n",
    "#### Hypotheses:\n",
    "\n",
    "$M_1: r \\in R_0$ (effect size r in region inducing differential expression)\n",
    "\n",
    "\n",
    "$M_2: r \\notin R_0$  (no differential expression)\n",
    "\n",
    "To characterize the region $R_0$, the user has two choices. \n",
    "\n",
    "##### Option 1) Specify an interval\n",
    "    \n",
    "A common case is when the region $[-\\delta, \\delta]$ does not induce differential expression. If the user specifies a threshold delta, we suppose that $R_0 = \\mathbb{R} \\backslash [-\\delta, \\delta]$\n",
    "    \n",
    "##### Option 2) Specify an indicator function\n",
    "\n",
    "Specify an specific indicator function $ \\mathbb{1} : \\mathbb{R} \\mapsto \\{0, 1\\}  \\text{  s.t.  }  r \\in R_0 \\iff \\mathbb{1}(r) = 1$. Decision-making can then be based on the estimates of $p(M_1 | x_1, x_2)$\n",
    "\n",
    "Differential Expression between conditions 1 and 2 for each gene can then be based on the Bayes factors:\n",
    "\n",
    "$$\n",
    "\\text{Natural Log Bayes Factor for gene g in conditions 1 and 2} = \\ln ( {BF^g_{12}) = \\ln(\\frac{ p(M^g_1 | x_1, x_2)}{p(M^g_2 | x_1, x_2)}})\n",
    "$$\n",
    "\n",
    "**The scvi `differential_expression_score` returns the _natural logarithm_ of the Bayes Factor. This is $\\ln(BF_{10})$ in the table below.**\n",
    "\n",
    "Bayes factors vs. p-values has been expertly reviewed by Leonhard Held and Manuela Ott in [On p-Values and Bayes Factors](https://doi.org/10.1146/annurev-statistics-031017-100307), as well as a shorter overview in [this blog post](https://www.nicebread.de/what-does-a-bayes-factor-feel-like/) by Felix Schönbrodt. \n",
    "\n",
    "A common interpretation table is copied below.  \n",
    "In our notation, $BF_{10}$ is $BF^g_{12}$, $H_0$ is $M^g_1$ and $H_1$ is $M^g_2$\n",
    "    \n",
    "To compute the gene specific Bayes factors using masks idx1 and idx2 we sample the Posterior in the following way:\n",
    "\n",
    "1. The posterior ($q(z_A | x_A)$ and $q(z_B | x_B)$) is sampled `n_samples` times for each subpopulation\n",
    "    \n",
    "2. For computation efficiency (posterior sampling is quite expensive), instead of\n",
    "    comparing element-wise the obtained samples, we can permute posterior samples.\n",
    "    \n",
    "\n",
    "### **Interpreting Bayes factors**\n",
    "\n",
    "A common interpretation table is copied below.  \n",
    "\n",
    "In our notation, $BF_{10}$ is $BF^g_{12}$, $H_0$ is $M^g_1$ and $H_1$ is $M^g_2$\n",
    "    \n",
    "    \n",
    "| Bayes factor $BF_{10}$ |  $\\ln(BF_{10})$        | Interpretation              |\n",
    "|------------------------|------------------------|-----------------------------|  \n",
    "| > 100                  | > 4.60                 | Extreme evidence for H1     | \n",
    "| 30 – 100               | (3.4, 4.6)             | Very strong evidence for H1 | \n",
    "| 10 – 30                | (2.3, 3.4)             | Strong evidence for H1      | \n",
    "| 3 – 10                 | (1.1, 2.3)             | Moderate evidence for H1    | \n",
    "| 1 – 3                  | (0 , 1.1)              | Anecdotal evidence for H1   | \n",
    "| 1                      | 0                      | No evidence                 | \n",
    "| 1/3 – 1                | (-1.1, 0)              | Anecdotal evidence for H0   | \n",
    "| 1/3 – 1/10             | (-2.30, -1.1)          | Moderate evidence for H0    | \n",
    "| 1/10 – 1/30            | (-3.4, -2.30)          | Strong evidence for H0      | \n",
    "| 1/30 – 1/100           | (-4.6, -3.4)           | Very strong evidence for H0 | \n",
    "| < 1/100                | < -4.6                 | Extreme evidence for H0     | \n",
    "\n",
    "Please note that the `bayes_factor` value returned by scVI is $\\ln(BF_{10})$. The difference in interpreting $\\ln(BF_{10})$, $\\ln(BF_{01})$, $BF_{10}$ and $BF_{01})$ is nicely explained in the blog post [A short taxonomy of Bayes factors](https://www.nicebread.de/a-short-taxonomy-of-bayes-factors/) by Felix Schönbrodt.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb2fa71e",
   "metadata": {},
   "source": [
    "#### Interpretation of differential expression results\n",
    "\n",
    "- `bayes_factor`: The bayes factor for condition 1 having a higher expression than condition 2\n",
    "\n",
    "- `proba_de`: p-value for condition **i** having higher expression than the other.\n",
    "\n",
    "- `raw_mean`**`i`**: the average number of UMI counts sampled condition **i** during the DE. Because scvi is a generative model, it is generating new cells and counting how many times each gene is sampled to compute the DE.\n",
    "\n",
    "- `raw_normalized_mean`**`i`**: the average library size normalized expression for that group (divide by library size, multiply by 10,000).\n",
    "\n",
    "- `non_zeros_proportion`**`i`**: proportion of non-zero expression in the group/condition **i**\n",
    "\n",
    "- `scale_`**`i`**: average scVI inferred gene expression scale in group/condition **i**. This is the same value that can be obtained using \n",
    "`model.get_normalized_expression()` using `library_size=1`, equivalent to the \"expression frequency\" of the gene: what fraction of the UMIs sampled from that cell would belong to that gene. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b13b0aa5",
   "metadata": {},
   "source": [
    "##### Set-up model and train for All Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7889680e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_scanvi(adata: dict,\n",
    "                layer: str='counts',\n",
    "                labels_key: str='cell_type',\n",
    "                unlabeled_category: str='unknown',\n",
    "                batch_key: str='Subject',\n",
    "                categorical_covariate_keys: list=['Subject'],\n",
    "                train: bool=True,\n",
    "                train_kwargs: dict =  dict(use_gpu=False, batch_size=64, early_stopping=True, \n",
    "                                            max_epochs=100, early_stopping_min_delta=100,\n",
    "                                            early_stopping_patience=15,\n",
    "                                            early_stopping_mode='min')                 \n",
    "                ):\n",
    "\n",
    "    degs_scanvi = {}\n",
    "\n",
    "    for key in adata_sub.keys():\n",
    "\n",
    "        adata_temp = adata[key].copy()\n",
    "\n",
    "        scvi.model.SCANVI.setup_anndata(adata_temp, layer=layer, labels_key=labels_key,\n",
    "                                        unlabeled_category=unlabeled_category, batch_key=batch_key,\n",
    "                                        categorical_covariate_keys=categorical_covariate_keys,\n",
    "                                        )\n",
    "\n",
    "        logger = TensorBoardLogger(save_dir=f'../models/scanvi_{key}_degs/')\n",
    "\n",
    "        train_kwargs['logger'] = logger\n",
    "\n",
    "        if train:\n",
    "            model = scvi.model.SCANVI(adata_temp, gene_likelihood='zinb')\n",
    "\n",
    "            print(f'Training model for {key}...')\n",
    "\n",
    "            model.train(**train_kwargs)\n",
    "            model.save('../models/', prefix=f'scanvi_{key.lower()}_', overwrite=True)\n",
    "        else:\n",
    "            model = scvi.model.SCANVI.load('../models/', adata=adata_temp, prefix=f'scanvi_{key.lower()}_',)\n",
    "        \n",
    "        adata[key].obsm[\"X_scANVI\"] = model.get_latent_representation()\n",
    "\n",
    "        print(f'Estimating DEGs for {key} cluster...')\n",
    "\n",
    "        degs_scanvi[key] = model.differential_expression(adata=adata_temp, groupby='disease_group',\n",
    "                                                         group1='AD-pathology', group2='no-pathology', mode='change',\n",
    "                                                         delta=None, batch_size=None, all_stats=True, batch_correction=True,\n",
    "                                                         fdr_target=0.05, silent=False, **dict(n_samples=5000))\n",
    "                                        \n",
    "        degs_scanvi[key].reset_index(inplace=True)\n",
    "        degs_scanvi[key].rename(columns={'index': 'names'}, inplace=True)\n",
    "        degs_scanvi[key]['odds_ratio'] = np.exp(degs_scanvi[key]['bayes_factor'])\n",
    "        \n",
    "        print(\"\\n\")\n",
    "\n",
    "    return degs_scanvi\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "789c700e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# degs_scanvi = train_scanvi(adata=adata_sub, layer='counts', labels_key='disease_group', batch_key='Subject',\n",
    "#                             unlabeled_category='unknown', categorical_covariate_keys=['Subject',], train=True,\n",
    "#                             train_kwargs=dict(use_gpu=False, batch_size=64, early_stopping=True, max_epochs=100, \n",
    "#                                             early_stopping_min_delta=100, early_stopping_patience=15, early_stopping_mode='min') \n",
    "#                             )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90c66c6f",
   "metadata": {},
   "source": [
    "#### Consensus of Differential Expressed Transcripts\n",
    "\n",
    "The consistency of DEGs detected using the cell-level `Wilcoxon rank-sum` analysis model with those obtained with the `MAST` generalized mixed model and the `scANVI` deep generative model was assessed by comparing the directionality and rank of DEGs in the two models. Consistency in directionality was measured by counting the fraction of the top 1,000 DEGs (ranked by FDR scores) detected in cell-level analysis that showed consistent direction in the deep generative/mixed model.\n",
    "\n",
    "For analyses involving DEG counts, [**Hansuredi et. al. 2019**](https://www.nature.com/articles/s41586-019-1195-2) used only genes that were significantly supported by both models using the criteria\n",
    "\n",
    "- `FDR-corrected P < 0.01 in a two sided Wilcoxon-rank sum test`,\n",
    "- `absolute log2(mean gene expression in AD category x/mean gene expression in AD category y) > 0.25`,\n",
    "- `FDR-corrected P < 0.05 in the Poisson mixed model` \n",
    "\n",
    "Motivated by [**Zhou. et. al. 2020**](https://doi.org/10.1038/s41591-019-0695-9) who used estimated DEGs between conditions using the `MAST algorithm of the Seurat package in R`, we include the following filtering criteria for the \n",
    "\n",
    "- `DEGs in the generalized mixed effects model are obatined by filtering genes for log2(fold change) > 0.1, P < 0.05`. \n",
    "- `DEGs in the deep generative model are obatined by filtering genes for`  $$\\ln(Bayes\\ Factor) > 2.3 $$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "58d73849",
   "metadata": {},
   "outputs": [],
   "source": [
    "mast_overlap = {}\n",
    "mast_consistency  = {}\n",
    "\n",
    "for key in adata_annot.obs.cell_type.unique():\n",
    "    mast_overlap[key] = pd.merge(degs_t_test[key], degs_mast[key], how='outer', on=['names'])\n",
    "\n",
    "    mast_consistency[key] = len(list(set(list(degs_t_test[key]['names'][:1000])) &\n",
    "                            set(mast_overlap[key][((mast_overlap[key]['avg_log2FC']>0) & (mast_overlap[key]['logfoldchanges']>0))|\n",
    "                            ((mast_overlap[key]['avg_log2FC']<0) & (mast_overlap[key]['logfoldchanges']<0))].names)))/1000\n",
    "\n",
    "    mast_overlap[key]['bulk_log2FC'] = np.log2((mast_overlap[key]['pct.1']+0.005)/\n",
    "                                            (mast_overlap[key]['pct.2']+0.005)) + mast_overlap[key]['avg_log2FC']\n",
    "\n",
    "    mast_overlap[key]['p_val_adj_DEGs.Ind.Model'] = np.NaN\n",
    "\n",
    "    mast_overlap[key]['DEGs.Ind.Model'] = False\n",
    "    mast_overlap[key].loc[(mast_overlap[key]['pvals_adj'] < 0.01) & \n",
    "                        (mast_overlap[key]['abs_logfoldchanges'] > 0.25), 'DEGs.Ind.Model'] = True\n",
    "\n",
    "    pvals = mast_overlap[key].loc[mast_overlap[key]['DEGs.Ind.Model'], 'p_val_adj'].tolist()\n",
    "                        \n",
    "    try:\n",
    "        mast_overlap[key].loc[mast_overlap[key]['DEGs.Ind.Model'],\n",
    "                        'p_val_adj_DEGs.Ind.Model'] = multipletests(pvals, alpha=0.05, method='fdr_bh')[1]\n",
    "        \n",
    "        mast_overlap[key]['DEGs.Ind.Mixed.Model'] = False\n",
    "        mast_overlap[key].loc[(mast_overlap[key]['p_val_adj_DEGs.Ind.Model'] < 0.05), 'DEGs.Ind.Mixed.Model'] = True\n",
    "        \n",
    "    except ZeroDivisionError:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "bc2f07f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>abs_logfoldchanges</th>\n",
       "      <th>Direction_t_test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>RASGEF1B</td>\n",
       "      <td>30.155712</td>\n",
       "      <td>9.027398e-200</td>\n",
       "      <td>1.428766e-195</td>\n",
       "      <td>1.758242</td>\n",
       "      <td>1.758242</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>LINGO1</td>\n",
       "      <td>29.114767</td>\n",
       "      <td>2.334517e-186</td>\n",
       "      <td>1.847420e-182</td>\n",
       "      <td>1.433463</td>\n",
       "      <td>1.433463</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>SLC26A3</td>\n",
       "      <td>24.047064</td>\n",
       "      <td>8.959428e-128</td>\n",
       "      <td>4.726696e-124</td>\n",
       "      <td>1.057443</td>\n",
       "      <td>1.057443</td>\n",
       "      <td>up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>BEX1</td>\n",
       "      <td>-23.102089</td>\n",
       "      <td>4.411363e-118</td>\n",
       "      <td>1.396373e-114</td>\n",
       "      <td>-0.872032</td>\n",
       "      <td>0.872032</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>SPARCL1</td>\n",
       "      <td>-23.105984</td>\n",
       "      <td>4.031128e-118</td>\n",
       "      <td>1.396373e-114</td>\n",
       "      <td>-1.169944</td>\n",
       "      <td>1.169944</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15822</th>\n",
       "      <td>15822</td>\n",
       "      <td>ZNF580</td>\n",
       "      <td>-0.100866</td>\n",
       "      <td>9.196568e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-1.700520</td>\n",
       "      <td>1.700520</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15823</th>\n",
       "      <td>15823</td>\n",
       "      <td>SSX2</td>\n",
       "      <td>-0.100854</td>\n",
       "      <td>9.196663e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-1.212336</td>\n",
       "      <td>1.212336</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15824</th>\n",
       "      <td>15824</td>\n",
       "      <td>RPEL1</td>\n",
       "      <td>-0.100728</td>\n",
       "      <td>9.197667e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-0.302013</td>\n",
       "      <td>0.302013</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15825</th>\n",
       "      <td>15825</td>\n",
       "      <td>DMRTC2</td>\n",
       "      <td>-0.109230</td>\n",
       "      <td>9.130199e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-0.307460</td>\n",
       "      <td>0.307460</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15826</th>\n",
       "      <td>15826</td>\n",
       "      <td>TMPRSS11D</td>\n",
       "      <td>-0.262470</td>\n",
       "      <td>7.929593e-01</td>\n",
       "      <td>9.998724e-01</td>\n",
       "      <td>-0.443457</td>\n",
       "      <td>0.443457</td>\n",
       "      <td>down</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15827 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0      names     scores          pvals      pvals_adj  \\\n",
       "0               0   RASGEF1B  30.155712  9.027398e-200  1.428766e-195   \n",
       "1               1     LINGO1  29.114767  2.334517e-186  1.847420e-182   \n",
       "2               2    SLC26A3  24.047064  8.959428e-128  4.726696e-124   \n",
       "3               3       BEX1 -23.102089  4.411363e-118  1.396373e-114   \n",
       "4               4    SPARCL1 -23.105984  4.031128e-118  1.396373e-114   \n",
       "...           ...        ...        ...            ...            ...   \n",
       "15822       15822     ZNF580  -0.100866   9.196568e-01   9.998724e-01   \n",
       "15823       15823       SSX2  -0.100854   9.196663e-01   9.998724e-01   \n",
       "15824       15824      RPEL1  -0.100728   9.197667e-01   9.998724e-01   \n",
       "15825       15825     DMRTC2  -0.109230   9.130199e-01   9.998724e-01   \n",
       "15826       15826  TMPRSS11D  -0.262470   7.929593e-01   9.998724e-01   \n",
       "\n",
       "       logfoldchanges  abs_logfoldchanges Direction_t_test  \n",
       "0            1.758242            1.758242               up  \n",
       "1            1.433463            1.433463               up  \n",
       "2            1.057443            1.057443               up  \n",
       "3           -0.872032            0.872032             down  \n",
       "4           -1.169944            1.169944             down  \n",
       "...               ...                 ...              ...  \n",
       "15822       -1.700520            1.700520             down  \n",
       "15823       -1.212336            1.212336             down  \n",
       "15824       -0.302013            0.302013             down  \n",
       "15825       -0.307460            0.307460             down  \n",
       "15826       -0.443457            0.443457             down  \n",
       "\n",
       "[15827 rows x 8 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "degs_t_test['Excitatory']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "067fb3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize empty dictionaries to store the results\n",
    "mast_overlap = {}\n",
    "mast_consistency = {}\n",
    "\n",
    "# Loop through each cell type\n",
    "for key in adata_annot.obs.cell_type.unique():\n",
    "    # Merge DEGs detected by t-test and MAST algorithms based on their gene names\n",
    "    mast_overlap[key] = pd.merge(degs_t_test[key], degs_mast[key], how='outer', on=['names'])\n",
    "\n",
    "    # Calculate the consistency in directionality of the top 1000 DEGs between t-test and MAST algorithms\n",
    "    mast_consistency[key] = len(list(set(list(degs_t_test[key]['names'][:1000])) &\n",
    "                            set(mast_overlap[key][((mast_overlap[key]['avg_log2FC']>0) & (mast_overlap[key]['logfoldchanges']>0))|\n",
    "                            ((mast_overlap[key]['avg_log2FC']<0) & (mast_overlap[key]['logfoldchanges']<0))].names)))/1000\n",
    "\n",
    "    # Calculate the bulk log2FC of each gene by taking the log2 ratio of the percentage of cells expressing the gene in two conditions\n",
    "    # and adding the average log2FC calculated from MAST and t-test algorithms\n",
    "    mast_overlap[key]['bulk_log2FC'] = np.log2((mast_overlap[key]['pct.1']+0.005)/\n",
    "                                            (mast_overlap[key]['pct.2']+0.005)) + mast_overlap[key]['avg_log2FC']\n",
    "\n",
    "    # Initialize columns for adjusted p-values and DEG identification in the merged table\n",
    "    mast_overlap[key]['p_val_adj_DEGs.Ind.Model'] = np.NaN\n",
    "    mast_overlap[key]['DEGs.Ind.Model'] = False\n",
    "    \n",
    "    # Mark genes as DEGs if they meet certain criteria based on FDR-corrected p-values and absolute log2 fold change\n",
    "    mast_overlap[key].loc[(mast_overlap[key]['pvals_adj'] < 0.01) & \n",
    "                        (mast_overlap[key]['abs_logfoldchanges'] > 0.25), 'DEGs.Ind.Model'] = True\n",
    "\n",
    "    # Get the p-values of identified DEGs\n",
    "    pvals = mast_overlap[key].loc[mast_overlap[key]['DEGs.Ind.Model'], 'p_val_adj'].tolist()\n",
    "                        \n",
    "    try:\n",
    "        # Apply FDR correction to the p-values of identified DEGs\n",
    "        mast_overlap[key].loc[mast_overlap[key]['DEGs.Ind.Model'],\n",
    "                        'p_val_adj_DEGs.Ind.Model'] = multipletests(pvals, alpha=0.05, method='fdr_bh')[1]\n",
    "        \n",
    "        # Identify DEGs in the mixed model based on the FDR-corrected p-values\n",
    "        mast_overlap[key]['DEGs.Ind.Mixed.Model'] = False\n",
    "        mast_overlap[key].loc[(mast_overlap[key]['p_val_adj_DEGs.Ind.Model'] < 0.05), 'DEGs.Ind.Mixed.Model'] = True\n",
    "        \n",
    "    except ZeroDivisionError:\n",
    "        # Skip if no DEGs are found\n",
    "        continue\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c63fd2dd",
   "metadata": {},
   "source": [
    "View some DEGs from mathys data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d0bdf0ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>IndModel.adj.pvals</th>\n",
       "      <th>no.pathology.mean</th>\n",
       "      <th>pathology.mean</th>\n",
       "      <th>IndModel.FC</th>\n",
       "      <th>MixedModel.z</th>\n",
       "      <th>MixedModel.p</th>\n",
       "      <th>DEGs.Ind.Model</th>\n",
       "      <th>DEGs.Ind.Mix.models</th>\n",
       "      <th>IndModel.absFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RASGEF1B</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.195138</td>\n",
       "      <td>5.950515</td>\n",
       "      <td>0.897136</td>\n",
       "      <td>6.127535</td>\n",
       "      <td>8.925109e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.897136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NGFRAP1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.837262</td>\n",
       "      <td>1.160450</td>\n",
       "      <td>-0.662872</td>\n",
       "      <td>-11.221532</td>\n",
       "      <td>3.196845e-29</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.662872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LINGO1</td>\n",
       "      <td>6.259506e-298</td>\n",
       "      <td>2.884947</td>\n",
       "      <td>4.381583</td>\n",
       "      <td>0.602907</td>\n",
       "      <td>5.689687</td>\n",
       "      <td>1.272728e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.602907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BEX1</td>\n",
       "      <td>8.203347e-283</td>\n",
       "      <td>1.582531</td>\n",
       "      <td>1.039826</td>\n",
       "      <td>-0.605891</td>\n",
       "      <td>-9.953163</td>\n",
       "      <td>2.442933e-23</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.605891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SLC26A3</td>\n",
       "      <td>2.121809e-276</td>\n",
       "      <td>1.750697</td>\n",
       "      <td>3.262202</td>\n",
       "      <td>0.897916</td>\n",
       "      <td>6.540229</td>\n",
       "      <td>6.142477e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.897916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13004</th>\n",
       "      <td>TNFRSF25</td>\n",
       "      <td>7.630976e-03</td>\n",
       "      <td>0.238967</td>\n",
       "      <td>0.312709</td>\n",
       "      <td>0.388009</td>\n",
       "      <td>4.637332</td>\n",
       "      <td>3.529353e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.388009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13025</th>\n",
       "      <td>RP11-618P17.4</td>\n",
       "      <td>8.239413e-03</td>\n",
       "      <td>0.310741</td>\n",
       "      <td>0.410157</td>\n",
       "      <td>0.400463</td>\n",
       "      <td>4.307534</td>\n",
       "      <td>1.650848e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.400463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13068</th>\n",
       "      <td>FZD1</td>\n",
       "      <td>8.869463e-03</td>\n",
       "      <td>0.011419</td>\n",
       "      <td>0.020506</td>\n",
       "      <td>0.844692</td>\n",
       "      <td>4.687224</td>\n",
       "      <td>2.769360e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.844692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13105</th>\n",
       "      <td>PDZD4</td>\n",
       "      <td>9.460903e-03</td>\n",
       "      <td>0.254717</td>\n",
       "      <td>0.398181</td>\n",
       "      <td>0.644528</td>\n",
       "      <td>8.728511</td>\n",
       "      <td>2.580469e-18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.644528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13120</th>\n",
       "      <td>BCAR1</td>\n",
       "      <td>9.861804e-03</td>\n",
       "      <td>0.190384</td>\n",
       "      <td>0.256937</td>\n",
       "      <td>0.432499</td>\n",
       "      <td>8.244239</td>\n",
       "      <td>1.662150e-16</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.432499</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>756 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Unnamed: 0  IndModel.adj.pvals  no.pathology.mean  pathology.mean  \\\n",
       "0           RASGEF1B        0.000000e+00           3.195138        5.950515   \n",
       "1            NGFRAP1        0.000000e+00           1.837262        1.160450   \n",
       "2             LINGO1       6.259506e-298           2.884947        4.381583   \n",
       "3               BEX1       8.203347e-283           1.582531        1.039826   \n",
       "4            SLC26A3       2.121809e-276           1.750697        3.262202   \n",
       "...              ...                 ...                ...             ...   \n",
       "13004       TNFRSF25        7.630976e-03           0.238967        0.312709   \n",
       "13025  RP11-618P17.4        8.239413e-03           0.310741        0.410157   \n",
       "13068           FZD1        8.869463e-03           0.011419        0.020506   \n",
       "13105          PDZD4        9.460903e-03           0.254717        0.398181   \n",
       "13120          BCAR1        9.861804e-03           0.190384        0.256937   \n",
       "\n",
       "       IndModel.FC  MixedModel.z  MixedModel.p  DEGs.Ind.Model  \\\n",
       "0         0.897136      6.127535  8.925109e-10            True   \n",
       "1        -0.662872    -11.221532  3.196845e-29            True   \n",
       "2         0.602907      5.689687  1.272728e-08            True   \n",
       "3        -0.605891     -9.953163  2.442933e-23            True   \n",
       "4         0.897916      6.540229  6.142477e-11            True   \n",
       "...            ...           ...           ...             ...   \n",
       "13004     0.388009      4.637332  3.529353e-06            True   \n",
       "13025     0.400463      4.307534  1.650848e-05            True   \n",
       "13068     0.844692      4.687224  2.769360e-06            True   \n",
       "13105     0.644528      8.728511  2.580469e-18            True   \n",
       "13120     0.432499      8.244239  1.662150e-16            True   \n",
       "\n",
       "       DEGs.Ind.Mix.models  IndModel.absFC  \n",
       "0                     True        0.897136  \n",
       "1                     True        0.662872  \n",
       "2                     True        0.602907  \n",
       "3                     True        0.605891  \n",
       "4                     True        0.897916  \n",
       "...                    ...             ...  \n",
       "13004                 True        0.388009  \n",
       "13025                 True        0.400463  \n",
       "13068                 True        0.844692  \n",
       "13105                 True        0.644528  \n",
       "13120                 True        0.432499  \n",
       "\n",
       "[756 rows x 10 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mathys_degs['Excitatory'][(mathys_degs['Excitatory']['DEGs.Ind.Model'])&\n",
    "                           (mathys_degs['Excitatory']['DEGs.Ind.Mix.models'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "166d56dd",
   "metadata": {},
   "source": [
    "View some DEGs from re-analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "48671b12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>abs_logfoldchanges</th>\n",
       "      <th>Direction_t_test</th>\n",
       "      <th>p_val</th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>pct.1</th>\n",
       "      <th>pct.2</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>diff</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bulk_log2FC</th>\n",
       "      <th>p_val_adj_DEGs.Ind.Model</th>\n",
       "      <th>DEGs.Ind.Model</th>\n",
       "      <th>DEGs.Ind.Mixed.Model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GFAP</td>\n",
       "      <td>9.283654</td>\n",
       "      <td>1.637632e-20</td>\n",
       "      <td>2.129904e-16</td>\n",
       "      <td>1.128408</td>\n",
       "      <td>1.128408</td>\n",
       "      <td>up</td>\n",
       "      <td>7.630713e-23</td>\n",
       "      <td>0.316605</td>\n",
       "      <td>0.559</td>\n",
       "      <td>0.411</td>\n",
       "      <td>1.265935e-18</td>\n",
       "      <td>0.148</td>\n",
       "      <td>1.265935e-18</td>\n",
       "      <td>0.755717</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>-8.830657</td>\n",
       "      <td>1.040630e-18</td>\n",
       "      <td>6.767220e-15</td>\n",
       "      <td>-1.276099</td>\n",
       "      <td>1.276099</td>\n",
       "      <td>down</td>\n",
       "      <td>5.718907e-13</td>\n",
       "      <td>-0.142494</td>\n",
       "      <td>0.898</td>\n",
       "      <td>0.942</td>\n",
       "      <td>9.487666e-09</td>\n",
       "      <td>-0.044</td>\n",
       "      <td>9.487666e-09</td>\n",
       "      <td>-0.211132</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>PRKG1</td>\n",
       "      <td>8.440312</td>\n",
       "      <td>3.164897e-17</td>\n",
       "      <td>1.029066e-13</td>\n",
       "      <td>0.988118</td>\n",
       "      <td>0.988118</td>\n",
       "      <td>up</td>\n",
       "      <td>3.192618e-19</td>\n",
       "      <td>0.213602</td>\n",
       "      <td>0.761</td>\n",
       "      <td>0.685</td>\n",
       "      <td>5.296554e-15</td>\n",
       "      <td>0.076</td>\n",
       "      <td>5.296554e-15</td>\n",
       "      <td>0.364350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.0</td>\n",
       "      <td>TENM2</td>\n",
       "      <td>-8.466842</td>\n",
       "      <td>2.521362e-17</td>\n",
       "      <td>1.029066e-13</td>\n",
       "      <td>-0.734385</td>\n",
       "      <td>0.734385</td>\n",
       "      <td>down</td>\n",
       "      <td>1.081682e-16</td>\n",
       "      <td>-0.458044</td>\n",
       "      <td>0.498</td>\n",
       "      <td>0.639</td>\n",
       "      <td>1.794511e-12</td>\n",
       "      <td>-0.141</td>\n",
       "      <td>1.794511e-12</td>\n",
       "      <td>-0.814546</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.0</td>\n",
       "      <td>RASGEF1B</td>\n",
       "      <td>7.965983</td>\n",
       "      <td>1.639156e-15</td>\n",
       "      <td>4.263772e-12</td>\n",
       "      <td>0.476628</td>\n",
       "      <td>0.476628</td>\n",
       "      <td>up</td>\n",
       "      <td>3.981474e-14</td>\n",
       "      <td>0.509415</td>\n",
       "      <td>0.629</td>\n",
       "      <td>0.512</td>\n",
       "      <td>6.605266e-10</td>\n",
       "      <td>0.117</td>\n",
       "      <td>6.605266e-10</td>\n",
       "      <td>0.803734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6394</th>\n",
       "      <td>6394.0</td>\n",
       "      <td>MMAB</td>\n",
       "      <td>-0.689976</td>\n",
       "      <td>4.902093e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>-0.416674</td>\n",
       "      <td>0.416674</td>\n",
       "      <td>down</td>\n",
       "      <td>2.127330e-01</td>\n",
       "      <td>-0.039396</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.064</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.014</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.366560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6416</th>\n",
       "      <td>6416.0</td>\n",
       "      <td>UBE2J1</td>\n",
       "      <td>-0.713720</td>\n",
       "      <td>4.754006e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>-0.203220</td>\n",
       "      <td>0.203220</td>\n",
       "      <td>down</td>\n",
       "      <td>8.506776e-02</td>\n",
       "      <td>-0.027932</td>\n",
       "      <td>0.068</td>\n",
       "      <td>0.083</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.015</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.297539</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6430</th>\n",
       "      <td>6430.0</td>\n",
       "      <td>JUP</td>\n",
       "      <td>-0.699657</td>\n",
       "      <td>4.841416e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>-0.592525</td>\n",
       "      <td>0.592525</td>\n",
       "      <td>down</td>\n",
       "      <td>7.678922e-02</td>\n",
       "      <td>-0.038278</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.040</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.014</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.575935</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6474</th>\n",
       "      <td>6474.0</td>\n",
       "      <td>TMTC2</td>\n",
       "      <td>-0.661018</td>\n",
       "      <td>5.086006e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>0.053248</td>\n",
       "      <td>0.053248</td>\n",
       "      <td>up</td>\n",
       "      <td>1.571967e-02</td>\n",
       "      <td>-0.079056</td>\n",
       "      <td>0.498</td>\n",
       "      <td>0.528</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.030</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.162633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6501</th>\n",
       "      <td>6501.0</td>\n",
       "      <td>DYNC1I1</td>\n",
       "      <td>-0.677187</td>\n",
       "      <td>4.982873e-01</td>\n",
       "      <td>9.998465e-01</td>\n",
       "      <td>-0.427949</td>\n",
       "      <td>0.427949</td>\n",
       "      <td>down</td>\n",
       "      <td>8.080137e-02</td>\n",
       "      <td>-0.051140</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.051</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.015</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>-0.500943</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>797 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0     names    scores         pvals     pvals_adj  \\\n",
       "0            0.0      GFAP  9.283654  1.637632e-20  2.129904e-16   \n",
       "1            1.0     NRXN1 -8.830657  1.040630e-18  6.767220e-15   \n",
       "2            2.0     PRKG1  8.440312  3.164897e-17  1.029066e-13   \n",
       "3            3.0     TENM2 -8.466842  2.521362e-17  1.029066e-13   \n",
       "4            4.0  RASGEF1B  7.965983  1.639156e-15  4.263772e-12   \n",
       "...          ...       ...       ...           ...           ...   \n",
       "6394      6394.0      MMAB -0.689976  4.902093e-01  9.998465e-01   \n",
       "6416      6416.0    UBE2J1 -0.713720  4.754006e-01  9.998465e-01   \n",
       "6430      6430.0       JUP -0.699657  4.841416e-01  9.998465e-01   \n",
       "6474      6474.0     TMTC2 -0.661018  5.086006e-01  9.998465e-01   \n",
       "6501      6501.0   DYNC1I1 -0.677187  4.982873e-01  9.998465e-01   \n",
       "\n",
       "      logfoldchanges  abs_logfoldchanges Direction_t_test         p_val  \\\n",
       "0           1.128408            1.128408               up  7.630713e-23   \n",
       "1          -1.276099            1.276099             down  5.718907e-13   \n",
       "2           0.988118            0.988118               up  3.192618e-19   \n",
       "3          -0.734385            0.734385             down  1.081682e-16   \n",
       "4           0.476628            0.476628               up  3.981474e-14   \n",
       "...              ...                 ...              ...           ...   \n",
       "6394       -0.416674            0.416674             down  2.127330e-01   \n",
       "6416       -0.203220            0.203220             down  8.506776e-02   \n",
       "6430       -0.592525            0.592525             down  7.678922e-02   \n",
       "6474        0.053248            0.053248               up  1.571967e-02   \n",
       "6501       -0.427949            0.427949             down  8.080137e-02   \n",
       "\n",
       "      avg_log2FC  pct.1  pct.2     p_val_adj   diff           FDR  \\\n",
       "0       0.316605  0.559  0.411  1.265935e-18  0.148  1.265935e-18   \n",
       "1      -0.142494  0.898  0.942  9.487666e-09 -0.044  9.487666e-09   \n",
       "2       0.213602  0.761  0.685  5.296554e-15  0.076  5.296554e-15   \n",
       "3      -0.458044  0.498  0.639  1.794511e-12 -0.141  1.794511e-12   \n",
       "4       0.509415  0.629  0.512  6.605266e-10  0.117  6.605266e-10   \n",
       "...          ...    ...    ...           ...    ...           ...   \n",
       "6394   -0.039396  0.050  0.064  1.000000e+00 -0.014  1.000000e+00   \n",
       "6416   -0.027932  0.068  0.083  1.000000e+00 -0.015  1.000000e+00   \n",
       "6430   -0.038278  0.026  0.040  1.000000e+00 -0.014  1.000000e+00   \n",
       "6474   -0.079056  0.498  0.528  1.000000e+00 -0.030  1.000000e+00   \n",
       "6501   -0.051140  0.036  0.051  1.000000e+00 -0.015  1.000000e+00   \n",
       "\n",
       "      bulk_log2FC  p_val_adj_DEGs.Ind.Model  DEGs.Ind.Model  \\\n",
       "0        0.755717                       NaN            True   \n",
       "1       -0.211132                       NaN            True   \n",
       "2        0.364350                       NaN            True   \n",
       "3       -0.814546                       NaN            True   \n",
       "4        0.803734                       NaN            True   \n",
       "...           ...                       ...             ...   \n",
       "6394    -0.366560                       NaN           False   \n",
       "6416    -0.297539                       NaN           False   \n",
       "6430    -0.575935                       NaN           False   \n",
       "6474    -0.162633                       NaN           False   \n",
       "6501    -0.500943                       NaN           False   \n",
       "\n",
       "      DEGs.Ind.Mixed.Model  \n",
       "0                    False  \n",
       "1                    False  \n",
       "2                    False  \n",
       "3                    False  \n",
       "4                    False  \n",
       "...                    ...  \n",
       "6394                 False  \n",
       "6416                 False  \n",
       "6430                 False  \n",
       "6474                 False  \n",
       "6501                 False  \n",
       "\n",
       "[797 rows x 19 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mast_overlap['Astrocyte'][(mast_overlap['Excitatory']['DEGs.Ind.Model'])&\n",
    "                           (mast_overlap['Excitatory']['DEGs.Ind.Mixed.Model'])]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "21d83925",
   "metadata": {},
   "source": [
    "### **Save Data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "11cfe608",
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.ExcelWriter(\"../results/ad_vs_no/mixed_model_degs.xlsx\") as writer:\n",
    "    for cell_type in adata_annot.obs.cell_type.unique():\n",
    "        mast_overlap[cell_type].to_excel(writer, sheet_name=cell_type, na_rep='NA')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6 ('mathys_reproduce_arm64')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "cab05a1ffa065c3a8c064e722daf73ead9f84ae54f7be3a13f1e06eccdd3df7e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
