{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import scipy\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import decoupler as dc\n",
    "import scrublet as scr\n",
    "import decoupler as dc\n",
    "from scipy import sparse\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from typing import Optional, Union\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.1\n",
      "-----\n",
      "OpenSSL                     22.0.0\n",
      "PIL                         9.2.0\n",
      "absl                        NA\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.2\n",
      "asttokens                   NA\n",
      "astunparse                  1.6.3\n",
      "attr                        21.4.0\n",
      "backcall                    0.2.0\n",
      "beta_ufunc                  NA\n",
      "binom_ufunc                 NA\n",
      "boto3                       1.26.32\n",
      "botocore                    1.29.32\n",
      "bottleneck                  1.3.5\n",
      "brotli                      NA\n",
      "certifi                     2022.09.24\n",
      "cffi                        1.15.1\n",
      "cloudpickle                 2.2.0\n",
      "colorama                    0.4.4\n",
      "cryptography                38.0.1\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dask                        2022.11.0\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.5.1\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "defusedxml                  0.7.1\n",
      "deprecate                   0.3.2\n",
      "entrypoints                 0.4\n",
      "executing                   0.8.3\n",
      "fsspec                      2022.10.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.7.0\n",
      "hypergeom_ufunc             NA\n",
      "igraph                      0.10.2\n",
      "ipykernel                   6.15.2\n",
      "ipython_genutils            0.2.0\n",
      "ipywidgets                  7.6.5\n",
      "jedi                        0.18.1\n",
      "jinja2                      3.1.2\n",
      "jmespath                    1.0.1\n",
      "joblib                      1.1.1\n",
      "kiwisolver                  1.4.2\n",
      "leidenalg                   0.9.0\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.1\n",
      "matplotlib                  3.5.3\n",
      "mpl_toolkits                NA\n",
      "natsort                     7.1.1\n",
      "nbinom_ufunc                NA\n",
      "ncf_ufunc                   NA\n",
      "numba                       0.56.3\n",
      "numexpr                     2.8.3\n",
      "numpy                       1.23.3\n",
      "packaging                   21.3\n",
      "pandas                      1.4.4\n",
      "parso                       0.8.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "prompt_toolkit              3.0.20\n",
      "psutil                      5.9.0\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     11.0.0\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.6.0\n",
      "pydevd_concurrency_analyser NA\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.13.0\n",
      "pyparsing                   3.0.9\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2022.1\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.10\n",
      "scipy                       1.9.3\n",
      "scrublet                    NA\n",
      "seaborn                     0.11.2\n",
      "session_info                1.0.0\n",
      "setuptools                  65.6.3\n",
      "six                         1.16.0\n",
      "sklearn                     1.1.3\n",
      "sparse                      0.13.0\n",
      "stack_data                  0.2.0\n",
      "statsmodels                 0.13.5\n",
      "tensorboard                 2.10.1\n",
      "texttable                   1.6.4\n",
      "threadpoolctl               2.2.0\n",
      "tlz                         0.12.0\n",
      "toolz                       0.12.0\n",
      "torch                       1.12.1\n",
      "torchmetrics                0.10.2\n",
      "tornado                     6.2\n",
      "tqdm                        4.64.1\n",
      "traitlets                   5.1.1\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "urllib3                     1.26.12\n",
      "wcwidth                     0.2.5\n",
      "yaml                        5.4.1\n",
      "zmq                         23.2.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.6.0\n",
      "jupyter_client      7.3.5\n",
      "jupyter_core        4.11.2\n",
      "notebook            6.4.12\n",
      "-----\n",
      "Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:41:22) [Clang 13.0.1 ]\n",
      "macOS-13.0.1-arm64-arm-64bit\n",
      "-----\n",
      "Session information updated at 2023-06-16 20:06\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# **hdWGCNA in single-cell data**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0c5d2f8d",
   "metadata": {},
   "source": [
    "Here we use `hdWGCNA` to perform co-expression network analysis on single-cell data. \n",
    "\n",
    "`hdWGCNA` is an R package for performing `weighted gene co-expression network analysis (WGCNA)` in high dimensional transcriptomics data such as single-cell RNA-seq or spatial transcriptomics. hdWGCNA is highly modular and can construct co-expression networks across multi-scale cellular and spatial hierarchies. hdWGNCA identifies robust modules of inerconnected genes, and provides context for these modules through various biological knowledge sources. hdWGCNA requires data formatted as Seurat objects, one of the most ubiquitous formats for single-cell data. \n",
    "\n",
    "`Note: hdWGCNA is under active development, so you may run into errors and small typos. We welcome users to write GitHub issues to report bugs, ask for help, and to request potential enhancements.`\n",
    "\n",
    "Explore the full capabilites of hWGNA in the following publications:\n",
    "\n",
    "- [**Morabito et al. bioRxiv 2022**](https://www.biorxiv.org/content/10.1101/2022.09.22.509094v1)\n",
    "- [**Morabito & Miyoshi et al. Nature Genetics 2021**](https://www.nature.com/articles/s41588-021-00894-z)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1b4733e8",
   "metadata": {},
   "source": [
    "Here, apply hdWGCNA to the processed single-nucleus RNA-seq (snRNA-seq) datasets. These dataset have already been fully processed using a standard single-cell transcritpomics analysis pipelines (please see the following notebooks, `preprocessing_data_integration`, `preprocessing_quality_control`, `preprocessing_cluster_annotation`). \n",
    "\n",
    "\n",
    "If you would like to apply this notebook to your own dataset, you first need to satisfy the following prerequisites:\n",
    "\n",
    "A single-cell or single-nucleus transcriptomics dataset in Seurat format.\n",
    "Normalize the gene expression matrix NormalizeData.\n",
    "Identify highly variable genes VariableFeatures.\n",
    "Scale the normalized expression data ScaleData\n",
    "Perform dimensionality reduction RunPCA and batch correction if needed RunHarmony.\n",
    "Non-linear dimensionality reduction RunUMAP for visualizations.\n",
    "Group cells into clusters (FindNeighbors and FindClusters).\n",
    "An example of running the prerequisite data processing steps can be found in the Seurat Guided Clustering Tutorial.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8b87d243",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(WGCNA)\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "\n",
    "\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "\n",
    "\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "af425574",
   "metadata": {},
   "source": [
    "## **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9697f83a",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'leng_etc'\n",
    "\n",
    "adata_annot = sc.read_h5ad(f'../data/processed/{save_prefix}/{save_prefix}_mapped_anndata.h5ad')\n",
    "adata_annot.obs_names_make_unique()\n",
    "adata_annot.var_names_make_unique()\n",
    "adata_annot.X = adata_annot.layers['counts'].copy()\n",
    "del adata_annot.obsm, adata_annot.layers, adata_annot.varm, adata_annot.uns, adata_annot.obsp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d8504101",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_meta = True\n",
    "deg_method =  'DESeq2-Wald'\n",
    "test_names = ['late_vs_early', 'early_vs_no', 'late_vs_no', 'ad_vs_no']\n",
    "filter_genes = \"TRUE\"\n",
    "subject_id = 'PatientID'                # for leng this is `PatientID` for mathys is 'Subject', and 'individualID' for allen\n",
    "gene_celltype_threshold = 0.05          # determines number of cells the gene must be expressed in \n",
    "covariates = ['None']                   # list of covariates to be accounted for in regression.\n",
    "gene_selection = 'custom'               # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "celltypes = [\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\"] #[\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\"]\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' \n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "\n",
    "gene_selection = 'custom'   # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "                            # Posible values are \"custom\", \"fraction\", \"variable\"\n",
    "                            # If custom, a list of genes must be passed.\n",
    "\n",
    "gene_set_select = 'diff_exp'      # If gene_selection = 'custom'. This specifies how to obtain the list of\n",
    "                                  # genes to pass into `SetupForWGCNA`. # The posible values are \"diff_exp\", \"overlap\", \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3888c7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# map the pathology group to teh subject id in .obs\n",
    "\n",
    "adata_annot.obs['pathology.group'] = adata_annot.obs[subject_id].map(dict(zip(meta[subject_id].astype(str), meta['pathology.group'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "839f795a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 16585 41432 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(16585): SAMD11 NOC2L ... S100B PRMT2\n",
      "rowData names(12): mt n_cells_by_counts ... mean std\n",
      "colnames(41432): EC2_AAACCTGAGGATGCGT EC2_AAACCTGAGTCAATAG ...\n",
      "  EC10_TTTGTCATCTATCGCC EC10_TTTGTCATCTCTGCTG\n",
      "colData names(24): SampleID PatientID ... predictions pathology.group\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "[1] \"loaded data into memory for recursive use\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_annot -i subject_id -i gene_celltype_threshold -i celltypes -i test_names -i save_prefix -i gene_selection\n",
    "\n",
    "print(adata_annot)\n",
    "print('loaded data into memory for recursive use')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "74c5d20a",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8a3955a2",
   "metadata": {},
   "source": [
    "## **Set up Seurat object for WGCNA**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2d154d2e",
   "metadata": {},
   "source": [
    "Before running hdWGCNA, we first have to set up the Seurat object. Most of the information computed by hdWGCNA is stored in the Seurat object’s `@misc slot`. \n",
    "\n",
    "A single Seurat object can hold multiple hdWGCNA experiments, for example representing different cell types in the same single-cell dataset. Notably, since we consider hdWGCNA to be a downstream data analysis step, we do not need to subset the Seurat object after `SetupForWGCNA` has been run.\n",
    "\n",
    "Here we will set up the Seurat object using the `SetupForWGCNA` function, specifying the name of the hdWGNCA experiment. This function also selects the genes that will be used for WGCNA. We can aslo select genes using three different approaches using the `gene_select` parameter:\n",
    "\n",
    "\n",
    "- `variable`: use the genes stored in the Seurat object’s `VariableFeatures`.\n",
    "\n",
    "- `fraction`: use genes that are expressed in a certain fraction of cells for in the whole dataset or in each group of cells, specified by group.by.\n",
    "\n",
    "- `custom`: use genes that are specified in a custom list."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3e7607cd",
   "metadata": {},
   "source": [
    "For our specific case, we will either select genes that are expressed in at least 5% of cells within each cell type or specify a list of custom genes. Additionally, we will rename our hdWGCNA experiment to enhance clarity.\n",
    "\n",
    "Here, the `gene_selection` parameter is set to `custom`, we employ the genes present in all the overlapping pathways in a cell type-specific and study-specific manner.\n",
    "\n",
    "To accomplish this, we load the list of overlapping pathways that was previously saved in `../results/{test_name}/{test_name}_pathway_overlap.csv` from the `pathway_meta_analysis.ipynb` notebook. Additionally, we load the genes from the `.gmt` file saved in `../results/gsva/{study}/{cur_celltype}/gs.gmt` from the `pathway_analysis.ipynb` notebook."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c4ddf8bf",
   "metadata": {},
   "source": [
    "##### **Select Gene Set**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7234ec55",
   "metadata": {},
   "source": [
    "To obtain the gene sets required for the `SetupForWGCNA` function, we consider the `gene_set_select` variable when `gene_select` is set to `custom`. The `gene_set_select` variable determines how the `gene_list` passed into the `SetupForWGCNA` function will be obtained. \n",
    "\n",
    "There are three options available:\n",
    "\n",
    "- `diff_exp`: This option utilizes the genes present in the differentially expressed pathways, previously computed and saved to the `../results/{test_name}/{save_prefix}_differentially_expressed_pathways.csv` file.\n",
    "\n",
    "- `overlap`: Here, the genes used are those expressed in pathways that overlap across different studies. The pathway overlap information is retrieved from the `../results/{test_name}/{test_name}_pathway_overlap.csv` file.\n",
    "\n",
    "- `all`: With this option, all genes in the pathways stored in the `../results/gsva/{save_prefix}/{celltype}/gs.gmt` file are included.\n",
    "\n",
    "By specifying the appropriate gene_select and gene_set_select values, you can customize the gene selection process for your hdWGCNA experiment.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "703041ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 916\n",
      "[1] 668\n",
      "[1] 646\n",
      "[1] 425\n",
      "[1] 329\n",
      "[1] 694\n",
      "[1] 2416\n",
      "[1] 1943\n",
      "[1] 1234\n",
      "[1] 634\n",
      "[1] 1004\n",
      "[1] 1263\n",
      "[1] 3149\n",
      "[1] 2547\n",
      "[1] 1415\n",
      "[1] 706\n",
      "[1] 1091\n",
      "[1] 1446\n",
      "[1] 3139\n",
      "[1] 2526\n",
      "[1] 1448\n",
      "[1] 770\n",
      "[1] 1222\n",
      "[1] 1615\n"
     ]
    }
   ],
   "source": [
    "%%R -o geneSets -i gene_set_select\n",
    "\n",
    "# iterate over celltypes and obtain \n",
    "geneSets <- list()\n",
    "for(test_name in test_names){\n",
    "\n",
    "  geneSets[[test_name]] <- list()\n",
    "  for (cur_celltype in celltypes){\n",
    "    # get gene sets\n",
    "    geneSets[[test_name]][[cur_celltype]] <- getGmt(file.path('../results/gsva/', save_prefix, '/', cur_celltype, 'gs.gmt'))\n",
    "    geneSets[[test_name]][[cur_celltype]] <- geneIds(geneSets[[test_name]][[cur_celltype]])\n",
    "\n",
    "    if (gene_set_select=='overlap'){\n",
    "      pathways <- read.csv(paste0('../results/',test_name, '/', test_name, '_pathway_overlap.csv'))\n",
    "      pathways <- pathways$pathway\n",
    "    }else if (gene_set_select=='diff_exp'){\n",
    "      pathways <- read.csv(paste0('../results/', test_name, '/', save_prefix, '_differentially_expressed_pathways.csv'))\n",
    "      pathways <- subset(pathways, P.Value < 0.05 & celltype == cur_celltype)$pathway\n",
    "    }else if (gene_set_select=='all'){\n",
    "      pathways <- names(geneSets[[test_name]][[cur_celltype]])\n",
    "    }\n",
    "\n",
    "    geneSets[[test_name]][[cur_celltype]] <-  geneSets[[test_name]][[cur_celltype]][names(geneSets[[test_name]][[cur_celltype]]) %in% pathways]\n",
    "    geneSets[[test_name]][[cur_celltype]] <- unique(unlist(geneSets[[test_name]][[cur_celltype]], recursive = FALSE))\n",
    "\n",
    "    print(length(geneSets[[test_name]][[cur_celltype]]))\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f2d4d1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of geneSets object into Rpy2 object \n",
    "\n",
    "geneSets =  robjects.ListVector(\n",
    "                        {\n",
    "                            test_name: robjects.ListVector(\n",
    "                                    {cell_type: geneSets[test_name][cell_type]\n",
    "                                                 \n",
    "                                    for cell_type in celltypes\n",
    "                                    }\n",
    "                                )\n",
    "                            for test_name in test_names\n",
    "                        }\n",
    "\n",
    "                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "845fa1c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An object of class Seurat \n",
      "16585 features across 41432 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 0 variable features)\n",
      "[1] \"Creating hdWGNA Experiment for EXCITATORY\"\n",
      "An object of class Seurat \n",
      "16585 features across 10449 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"3139 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for INHIBITORY\"\n",
      "An object of class Seurat \n",
      "16585 features across 5207 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"2526 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for ASTROCYTE\"\n",
      "An object of class Seurat \n",
      "16585 features across 5851 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"1448 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for MICROGLIA\"\n",
      "An object of class Seurat \n",
      "16585 features across 5892 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"770 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for OLIGODENDROCYTE\"\n",
      "An object of class Seurat \n",
      "16585 features across 9606 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"1222 WGCNA Genes\"\n",
      "[1] \"Creating hdWGNA Experiment for OPC\"\n",
      "An object of class Seurat \n",
      "16585 features across 3801 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"1615 WGCNA Genes\"\n"
     ]
    }
   ],
   "source": [
    "%%R -o seurat_obj -i geneSets\n",
    "\n",
    "seurat_obj <- as.Seurat(adata_annot, counts = \"X\", data = \"X\")\n",
    "\n",
    "print(seurat_obj)\n",
    "\n",
    "# Perform dimensionality reduction and plot\n",
    "\n",
    "seurat_obj <- FindVariableFeatures(seurat_obj)\n",
    "seurat_obj <- ScaleData(seurat_obj)\n",
    "seurat_obj <- RunPCA(seurat_obj)\n",
    "seurat_obj <- RunHarmony(seurat_obj, group.by.vars = subject_id)\n",
    "seurat_obj <- RunUMAP(seurat_obj, reduction='harmony', n.neighbors=15, dims=1:30, min.dist=0.1)\n",
    "\n",
    "p <- DimPlot(seurat_obj, group.by = \"cell_type\", label = TRUE) +\n",
    "     umap_theme() + ggtitle(\"leng\") + NoLegend()\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/UMAP/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "# Save plot to PDF\n",
    "pdf(file = paste0(fig_dir, save_prefix, \"_integrated_umap.pdf\"), width = 4, height = 4, useDingbats = FALSE)\n",
    "print(p)\n",
    "dev.off()\n",
    "\n",
    "# create a hdWGCNA experiment for each celltype\n",
    "seurat_dat <- seurat_obj\n",
    "seurat_obj <- list()\n",
    "\n",
    "\n",
    "for (cur_cell_type in celltypes){\n",
    "\n",
    "  seurat_obj[[cur_cell_type]] <- subset(seurat_dat, cell_type == cur_cell_type)\n",
    "\n",
    "  print(paste0('Creating hdWGNA Experiment for ', toupper(cur_cell_type)))\n",
    "\n",
    "  if (gene_selection == 'custom') {\n",
    "    seurat_obj[[cur_cell_type]] <- SetupForWGCNA(\n",
    "      seurat_obj[[cur_cell_type]],\n",
    "      gene_select = \"custom\",                                            # the gene selection approach\n",
    "      gene_list = as.vector(geneSets[['ad_vs_no']][[cur_cell_type]]),    # list of genes to be included\n",
    "      group.by = 'cell_type',                                            # grouping parameter\n",
    "      wgcna_name = toupper(substr(cur_cell_type, 1, 3))                  # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  } else {\n",
    "    seurat_obj[[cur_cell_type]] <- SetupForWGCNA(\n",
    "      seurat_obj[[cur_cell_type]],\n",
    "      gene_select = \"fraction\",                               # the gene selection approach\n",
    "      fraction = gene_celltype_threshold,                     # fraction of cells for gene inclusion\n",
    "      group.by = 'cell_type',                                 # grouping parameter\n",
    "      wgcna_name = toupper(substr(cur_cell_type, 1, 3))       # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  }\n",
    "\n",
    "  print(seurat_obj[[cur_cell_type]])\n",
    "  print(paste0(length(GetWGCNAGenes(seurat_obj[[cur_cell_type]])), \" WGCNA Genes\"))\n",
    "\n",
    "}\n",
    "\n",
    "rm(seurat_dat)\n",
    "rm(adata_annot)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9e89e217",
   "metadata": {},
   "source": [
    "Send list to R interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ff771eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "\n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ccd0e590",
   "metadata": {},
   "source": [
    "## **Construct metacells**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "07813598",
   "metadata": {},
   "source": [
    "After setting up our Seurat object, the first step in running the hdWGCNA pipeine in hdWGCNA is to construct metacells from the single-cell dataset. \n",
    "\n",
    "Briefly, `metacells are aggregates of small groups of similar cells originating from the same biological sample of origin.` The k-Nearest Neighbors (KNN) algorithm is used to identify groups of similar cells to aggregate, and then the average or summed expression of these cells is computed, thus yielding a metacell gene expression matrix. `The sparsity of the metacell expression matrix is considerably reduced` when compared to the original expression matrix, and therefore it is preferable to use. We were originally motivated to use metacells in place of the original single cells because correlation network approaches such as WGCNA are sensitive to data sparsity. Furthermore, single-cell epigenomic approaches, such as Cicero, employ a similar metacell aggregation approach prior to constructing co-accessibility networks.\n",
    "\n",
    "hdWGCNA includes a function MetacellsByGroups to construct metacell expression matrices given a single-cell dataset. This function constructs a new Seurat object for the metacell dataset which is stored internally in the hdWGCNA experiment. \n",
    "\n",
    "The `group.by` parameter determines which groups metacells will be constructed in. \n",
    "\n",
    "`We only want to construct metacells from cells that came from the same biological sample of origin, so it is critical to pass that information to hdWGCNA via the group.by parameter.` \n",
    "\n",
    "Additionally, we usually construct metacells for each cell type separately. `Thus, in this example, we are grouping by Sample and cell_type to achieve the desired result.`\n",
    "\n",
    "The number of cells to be aggregated k should be tuned based on the size of the input dataset, in general a lower number for k can be used for small datasets. `Generally use k values between 20 and 75 for 40,039 cells`, ranging from 890 to 8,188 in each biological sample, and here we used k=25. The amount of allowable overlap between metacells can be tuned using the max_shared argument.\n",
    "\n",
    "Note: The metacell aggregation approach does not yield good results for extremely underrepresented cell types. For example, the brain vascular cells (pericytes and endothelial cells), thus we have excluded them from this analysis. \n",
    "\n",
    "`MetacellsByGroups `has a parameter min_cells to exclude groups that are smaller than a specified number of cells. Errors are likely to arise if the selected value for min_cells is too low.\n",
    "\n",
    "`Here we construct metacells and normalize in each hdWGCNA experiment`\n",
    "\n",
    "The resulting expression matrix is obtained using the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e92860fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# construct metacells in each group for each wgnca experiment\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "  print(paste0('Constructing MetaCells in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  seurat_obj[[cell_type]] <- MetacellsByGroups(\n",
    "      \n",
    "    seurat_obj = seurat_obj[[cell_type]],\n",
    "    group.by = c(\"cell_type\", subject_id), # specify the columns in seurat_obj@meta.data to group by\n",
    "    reduction = 'harmony',  # select the dimensionality reduction to perform KNN on\n",
    "    k = 25, # nearest-neighbors parameter\n",
    "    max_shared = 10, # maximum number of shared cells between two metacells\n",
    "    ident.group = 'cell_type', # set the Idents of the metacell seurat object\n",
    "    wgcna_name = toupper(substr(cell_type, 1, 3)),        # the name of the hdWGCNA experiment\n",
    "    )\n",
    "\n",
    "    # normalize metacell expression matrix:\n",
    "  seurat_obj[[cell_type]] <- NormalizeMetacells(seurat_obj[[cell_type]],  wgcna_name = toupper(substr(cell_type, 1, 3)))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "db92cb94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b70e5121",
   "metadata": {},
   "source": [
    "## **Process the Metacell Seurat Object**\n",
    "\n",
    "Since we store the Metacell expression information as its own Seurat object, we can run Seurat functions on the metacell data. We can get the metacell object from the hdWGCNA experiment using `GetMetacellObject.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "46d9dccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o metacell_obj\n",
    "\n",
    "# metacell_obj <- list()\n",
    "# for (cell_type in celltypes){\n",
    "#     print(paste0('Obtaining MetaCells in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "#     metacell_obj[[cell_type]] <- GetMetacellObject(seurat_obj[[cell_type]] , \n",
    "#                                         wgcna_name = toupper(substr(cell_type, 1, 3))       # the name of the hdWGCNA experiment\n",
    "#                                      )\n",
    "#     print(metacell_obj[[cell_type]])\n",
    "# }"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a6c0058a",
   "metadata": {},
   "source": [
    "**`OPTIONAL`**: Here we use wrapper functions to process the metacell object and visualize one of the aggregated expression profiles in two dimensions with UMAP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "db7b36f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "# for (cell_type in celltypes){\n",
    "\n",
    "#     print(paste0('Processing MetaCells in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "#     seurat_obj[[cell_type]]  <- NormalizeMetacells(seurat_obj[[cell_type]])\n",
    "#     seurat_obj[[cell_type]]  <- ScaleMetacells(seurat_obj[[cell_type]] , features=VariableFeatures(seurat_obj[[cell_type]]))\n",
    "#     seurat_obj[[cell_type]]  <- RunPCAMetacells(seurat_obj[[cell_type]] , features=VariableFeatures(seurat_obj[[cell_type]]))\n",
    "#     seurat_obj[[cell_type]]  <- RunHarmonyMetacells(seurat_obj[[cell_type]] , group.by.vars=subject_id)\n",
    "#     seurat_obj[[cell_type]]  <- RunUMAPMetacells(seurat_obj[[cell_type]] , reduction='harmony', dims=1:15)\n",
    "\n",
    "\n",
    "#     p1 <- DimPlotMetacells(seurat_obj[[cell_type]] , group.by='cell_type') + umap_theme() + ggtitle(\"Cell Type\")\n",
    "#     p2 <- DimPlotMetacells(seurat_obj[[cell_type]] , group.by=subject_id) + umap_theme() + ggtitle(subject_id)\n",
    "\n",
    "#     p1 | p2\n",
    "\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6085ce6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # convert nested list of Seurat object into Rpy2 object \n",
    "# seurat_obj =  robjects.ListVector(\n",
    "#                         {\n",
    "#                             cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "#                             for cell_type in celltypes\n",
    "#                         }\n",
    "#                     )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fac55be8",
   "metadata": {},
   "source": [
    "## **Co-expression network analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f03072ae",
   "metadata": {},
   "source": [
    "### **Set up the expression matrix**\n",
    "\n",
    "Here we specify the expression matrix that we will use for network analysis. We only include a single cell-type of interest per run/simulation, so we have to subset our expression data prior to constructing the network. hdWGCNA includes the `SetDatExpr` function to store the transposed expression matrix for a given group of cells that will be used for downstream network analysis.\n",
    "\n",
    "The metacell expression matrix is used by default `(use_metacells=TRUE)`, but hdWGCNA does allow for the single-cell expression matrix to be used if desired.. This function allows the user to specify which slot to take the expression matrix from, for example if the user wanted to apply SCTransform normalization instead of NormalizeData."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7830296b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "    print(paste0('Setting up expression matrix in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    seurat_obj[[cell_type]] <- SetDatExpr(\n",
    "      seurat_obj[[cell_type]],\n",
    "      group_name = cell_type, # the name of the group of interest in the group.by column\n",
    "      group.by = 'cell_type', # the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups\n",
    "      assay = 'originalexp', # using RNA assay\n",
    "      slot = 'data', # using normalized data\n",
    "      wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3c40b7ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a70cc5c4",
   "metadata": {},
   "source": [
    "### **Select soft-power threshold**\n",
    "\n",
    "Next we will select the “soft power threshold”. This is an extremely important step in the hdWGNCA pipleine (and for vanilla WGCNA). hdWGCNA constructs a gene-gene correlation adjacency matrix to infer co-expression relationships between genes. The correlations are raised to a power to reduce the amount of noise present in the correlation matrix, thereby retaining the strong connections and removing the weak connections. Therefore, it is critical to determine a proper value for the soft power threshold.\n",
    "\n",
    "We use the function `TestSoftPowers` to perform a parameter sweep for different soft power thresholds. This function helps us to guide our choice in a soft power threshold for constructing the co-expression network by inspecting the resulting network topology for different power values. The co-expression network should have a scale-free topology, therefore the TestSoftPowers function models how closely the co-expression network resembles a scale-free graph at different soft power thresholds. Furthermore, we use the function `PlotSoftPowers` to visualize the results of the parameter sweep.\n",
    "\n",
    "The following code performs the parameter sweep and outputs a summary figure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e63133c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for EXCITATORY\"\n",
      "pickSoftThreshold: will use block size 3139.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 3139 of 3139\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.   max.k.\n",
      "1      1 0.097200  8.570          0.939 1.62e+03  1.61e+03 1760.000\n",
      "2      2 0.000242 -0.189          0.903 8.41e+02  8.35e+02 1010.000\n",
      "3      3 0.017100 -1.070          0.865 4.43e+02  4.36e+02  584.000\n",
      "4      4 0.071400 -1.640          0.831 2.35e+02  2.30e+02  345.000\n",
      "5      5 0.205000 -2.140          0.837 1.27e+02  1.22e+02  207.000\n",
      "6      6 0.392000 -2.580          0.869 6.88e+01  6.48e+01  129.000\n",
      "7      7 0.629000 -3.310          0.937 3.79e+01  3.49e+01   84.600\n",
      "8      8 0.763000 -3.660          0.973 2.11e+01  1.89e+01   56.800\n",
      "9      9 0.841000 -3.630          0.978 1.19e+01  1.03e+01   39.000\n",
      "10    10 0.896000 -3.500          0.988 6.81e+00  5.61e+00   28.000\n",
      "11    12 0.938000 -3.110          0.991 2.33e+00  1.71e+00   15.200\n",
      "12    14 0.939000 -2.760          0.960 8.48e-01  5.39e-01    8.780\n",
      "13    16 0.944000 -2.410          0.955 3.30e-01  1.73e-01    5.290\n",
      "14    18 0.420000 -3.150          0.356 1.38e-01  5.62e-02    3.300\n",
      "15    20 0.418000 -2.870          0.353 6.20e-02  1.87e-02    2.120\n",
      "16    22 0.413000 -2.610          0.338 2.97e-02  6.34e-03    1.400\n",
      "17    24 0.408000 -2.410          0.330 1.51e-02  2.18e-03    0.937\n",
      "18    26 0.406000 -2.270          0.330 8.12e-03  7.73e-04    0.640\n",
      "19    28 0.406000 -2.140          0.328 4.57e-03  2.73e-04    0.443\n",
      "20    30 0.406000 -2.060          0.325 2.68e-03  9.78e-05    0.311\n",
      "  Power    SFT.R.sq      slope truncated.R.sq    mean.k.  median.k.    max.k.\n",
      "1     1 0.097212732  8.5665292      0.9392948 1615.85173 1611.37525 1762.6752\n",
      "2     2 0.000242359 -0.1893038      0.9028249  841.10702  834.97333 1006.2185\n",
      "3     3 0.017068677 -1.0733057      0.8650761  442.53645  436.46276  583.9217\n",
      "4     4 0.071420555 -1.6392786      0.8309044  235.35216  229.65196  344.5984\n",
      "5     5 0.204792114 -2.1440086      0.8372495  126.54559  121.79149  206.8771\n",
      "6     6 0.391725190 -2.5750560      0.8685661   68.81462   64.82735  129.1036\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for INHIBITORY\"\n",
      "pickSoftThreshold: will use block size 2526.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 2526 of 2526\n",
      "   Power SFT.R.sq slope truncated.R.sq  mean.k. median.k.   max.k.\n",
      "1      1   0.2290 30.90          0.971 1.28e+03  1.28e+03 1320.000\n",
      "2      2   0.0250  4.62          0.995 6.55e+02  6.55e+02  718.000\n",
      "3      3   0.0697 -4.35          0.943 3.41e+02  3.40e+02  412.000\n",
      "4      4   0.2260 -4.85          0.853 1.80e+02  1.78e+02  246.000\n",
      "5      5   0.4500 -4.73          0.809 9.65e+01  9.44e+01  153.000\n",
      "6      6   0.7070 -4.74          0.861 5.25e+01  5.04e+01  103.000\n",
      "7      7   0.8780 -4.47          0.936 2.90e+01  2.72e+01   71.200\n",
      "8      8   0.9500 -4.07          0.975 1.63e+01  1.48e+01   50.800\n",
      "9      9   0.9800 -3.63          0.990 9.31e+00  8.11e+00   37.200\n",
      "10    10   0.9860 -3.27          0.991 5.42e+00  4.47e+00   27.800\n",
      "11    12   0.9810 -2.71          0.981 1.95e+00  1.40e+00   16.300\n",
      "12    14   0.9810 -2.31          0.977 7.61e-01  4.47e-01   10.000\n",
      "13    16   0.9360 -2.11          0.923 3.26e-01  1.47e-01    6.450\n",
      "14    18   0.9550 -1.89          0.950 1.52e-01  4.96e-02    4.270\n",
      "15    20   0.9730 -1.70          0.967 7.74e-02  1.72e-02    2.900\n",
      "16    22   0.9670 -1.60          0.958 4.22e-02  6.06e-03    2.020\n",
      "17    24   0.9550 -1.53          0.942 2.45e-02  2.20e-03    1.450\n",
      "18    26   0.9530 -1.48          0.939 1.49e-02  8.03e-04    1.050\n",
      "19    28   0.3410 -1.94          0.179 9.44e-03  2.99e-04    0.775\n",
      "20    30   0.3460 -1.90          0.186 6.18e-03  1.13e-04    0.577\n",
      "  Power   SFT.R.sq     slope truncated.R.sq    mean.k.  median.k.    max.k.\n",
      "1     1 0.22878670 30.942107      0.9706114 1276.41255 1276.53340 1323.2012\n",
      "2     2 0.02500548  4.617186      0.9951043  655.01042  655.05676  718.2887\n",
      "3     3 0.06973207 -4.349877      0.9425338  341.04724  340.01272  411.8652\n",
      "4     4 0.22550308 -4.846107      0.8531333  180.14687  178.21514  245.5160\n",
      "5     5 0.45023281 -4.731304      0.8085812   96.54854   94.37552  153.4788\n",
      "6     6 0.70692783 -4.738279      0.8614692   52.52167   50.42041  102.7052\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for ASTROCYTE\"\n",
      "pickSoftThreshold: will use block size 1448.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 1448 of 1448\n",
      "   Power SFT.R.sq slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1   0.1680 26.60          0.917 729.0000  7.29e+02 755.000\n",
      "2      2   0.0189  4.01          0.938 374.0000  3.74e+02 412.000\n",
      "3      3   0.0140 -1.69          0.865 196.0000  1.95e+02 236.000\n",
      "4      4   0.3820 -5.42          0.812 105.0000  1.03e+02 147.000\n",
      "5      5   0.7770 -5.64          0.915  56.8000  5.45e+01  95.200\n",
      "6      6   0.9320 -4.71          0.969  31.5000  2.93e+01  64.200\n",
      "7      7   0.9530 -3.83          0.970  17.8000  1.59e+01  44.700\n",
      "8      8   0.9740 -3.26          0.985  10.3000  8.68e+00  32.700\n",
      "9      9   0.9690 -2.87          0.975   6.1100  4.80e+00  24.500\n",
      "10    10   0.9730 -2.54          0.979   3.7100  2.67e+00  18.700\n",
      "11    12   0.9830 -2.13          0.992   1.4800  8.61e-01  11.600\n",
      "12    14   0.9720 -1.87          0.980   0.6540  2.84e-01   7.640\n",
      "13    16   0.9650 -1.71          0.971   0.3200  9.64e-02   5.350\n",
      "14    18   0.9170 -1.63          0.910   0.1710  3.34e-02   3.880\n",
      "15    20   0.9430 -1.54          0.945   0.0987  1.19e-02   2.900\n",
      "16    22   0.9510 -1.47          0.951   0.0603  4.43e-03   2.210\n",
      "17    24   0.3060 -2.03          0.152   0.0387  1.65e-03   1.720\n",
      "18    26   0.3110 -1.95          0.161   0.0258  6.32e-04   1.370\n",
      "19    28   0.3190 -1.92          0.172   0.0177  2.46e-04   1.100\n",
      "20    30   0.3040 -2.36          0.186   0.0125  9.82e-05   0.894\n",
      "  Power   SFT.R.sq     slope truncated.R.sq   mean.k. median.k.    max.k.\n",
      "1     1 0.16750414 26.571853      0.9173866 728.60674 729.46114 755.16117\n",
      "2     2 0.01887285  4.007011      0.9377111 374.32586 374.31808 412.09703\n",
      "3     3 0.01399579 -1.688283      0.8652777 195.98827 194.85179 236.31643\n",
      "4     4 0.38184101 -5.420995      0.8115396 104.54442 102.53990 146.65252\n",
      "5     5 0.77678941 -5.639128      0.9150038  56.83173  54.50147  95.20584\n",
      "6     6 0.93156296 -4.710366      0.9693086  31.50811  29.28296  64.16252\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for MICROGLIA\"\n",
      "pickSoftThreshold: will use block size 770.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 770 of 770\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1    0.243  22.90          0.919 389.0000  3.89e+02 405.000\n",
      "2      2    0.394 -11.80          0.726 200.0000  1.99e+02 231.000\n",
      "3      3    0.755  -9.65          0.795 105.0000  1.03e+02 141.000\n",
      "4      4    0.883  -7.04          0.872  56.4000  5.42e+01  90.600\n",
      "5      5    0.950  -5.21          0.941  30.8000  2.86e+01  60.500\n",
      "6      6    0.968  -3.98          0.960  17.2000  1.52e+01  41.900\n",
      "7      7    0.977  -3.17          0.971   9.8000  8.13e+00  29.900\n",
      "8      8    0.959  -2.64          0.950   5.7400  4.39e+00  22.000\n",
      "9      9    0.930  -2.30          0.916   3.4600  2.38e+00  16.500\n",
      "10    10    0.912  -2.04          0.899   2.1500  1.31e+00  12.700\n",
      "11    12    0.859  -1.78          0.821   0.9100  4.01e-01   8.440\n",
      "12    14    0.886  -1.58          0.853   0.4370  1.27e-01   5.950\n",
      "13    16    0.922  -1.40          0.900   0.2350  4.05e-02   4.330\n",
      "14    18    0.914  -1.32          0.891   0.1380  1.32e-02   3.230\n",
      "15    20    0.948  -1.22          0.933   0.0870  4.37e-03   2.460\n",
      "16    22    0.899  -1.23          0.870   0.0576  1.50e-03   1.890\n",
      "17    24    0.275  -1.82          0.166   0.0396  5.37e-04   1.480\n",
      "18    26    0.275  -1.73          0.166   0.0281  1.91e-04   1.170\n",
      "19    28    0.276  -1.64          0.172   0.0203  6.98e-05   0.925\n",
      "20    30    0.278  -1.60          0.172   0.0150  2.53e-05   0.740\n",
      "  Power  SFT.R.sq      slope truncated.R.sq   mean.k. median.k.    max.k.\n",
      "1     1 0.2427599  22.900364      0.9189858 388.64962 389.29902 405.01432\n",
      "2     2 0.3942961 -11.814406      0.7263499 200.39520 198.85059 231.01223\n",
      "3     3 0.7547536  -9.652827      0.7950857 105.28076 103.31325 141.14366\n",
      "4     4 0.8832620  -7.040327      0.8716344  56.36006  54.24201  90.55177\n",
      "5     5 0.9497925  -5.205451      0.9406263  30.77290  28.63355  60.51956\n",
      "6     6 0.9679015  -3.978966      0.9596361  17.16459  15.23221  41.90786\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "pickSoftThreshold: will use block size 1222.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 1222 of 1222\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1 0.162000  9.850         0.7900 6.22e+02  6.24e+02 666.000\n",
      "2      2 0.000415 -0.243         0.9640 3.20e+02  3.20e+02 376.000\n",
      "3      3 0.248000 -4.670         0.9770 1.67e+02  1.65e+02 223.000\n",
      "4      4 0.603000 -6.390         0.9630 8.76e+01  8.59e+01 137.000\n",
      "5      5 0.824000 -6.290         0.9760 4.65e+01  4.47e+01  86.900\n",
      "6      6 0.909000 -5.660         0.9730 2.50e+01  2.35e+01  57.000\n",
      "7      7 0.927000 -5.090         0.9590 1.36e+01  1.24e+01  38.500\n",
      "8      8 0.420000 -7.340         0.3220 7.49e+00  6.63e+00  26.700\n",
      "9      9 0.431000 -6.430         0.3390 4.18e+00  3.55e+00  18.900\n",
      "10    10 0.431000 -5.610         0.3410 2.37e+00  1.91e+00  13.700\n",
      "11    12 0.437000 -4.550         0.3600 8.02e-01  5.70e-01   7.580\n",
      "12    14 0.420000 -4.820         0.3520 2.93e-01  1.71e-01   4.430\n",
      "13    16 0.417000 -4.150         0.3600 1.16e-01  5.23e-02   2.700\n",
      "14    18 0.414000 -4.230         0.3430 5.02e-02  1.64e-02   1.700\n",
      "15    20 0.405000 -3.810         0.3470 2.36e-02  5.20e-03   1.100\n",
      "16    22 0.399000 -3.490         0.3460 1.19e-02  1.66e-03   0.721\n",
      "17    24 0.392000 -3.260         0.3370 6.44e-03  5.50e-04   0.482\n",
      "18    26 0.356000 -2.960         0.2350 3.68e-03  1.83e-04   0.326\n",
      "19    28 0.273000 -2.170         0.0676 2.22e-03  6.18e-05   0.224\n",
      "20    30 0.397000 -3.180         0.2900 1.40e-03  2.10e-05   0.157\n",
      "  Power     SFT.R.sq      slope truncated.R.sq   mean.k. median.k.    max.k.\n",
      "1     1 0.1621741209  9.8519684      0.7898342 622.04678 623.61210 666.23559\n",
      "2     2 0.0004146616 -0.2427322      0.9637355 320.25778 320.27596 376.21071\n",
      "3     3 0.2478216285 -4.6669855      0.9769484 166.59288 165.43575 222.66338\n",
      "4     4 0.6029631320 -6.3939872      0.9633159  87.56382  85.87396 136.78114\n",
      "5     5 0.8239142611 -6.2921684      0.9757621  46.52002  44.73118  86.94639\n",
      "6     6 0.9085379805 -5.6595562      0.9728051  24.99379  23.46289  57.03477\n",
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for OPC\"\n",
      "pickSoftThreshold: will use block size 1615.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 1615 of 1615\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1    0.105   7.60          0.615 8.18e+02  8.22e+02 873.000\n",
      "2      2    0.230  -6.92          0.966 4.21e+02  4.21e+02 499.000\n",
      "3      3    0.755 -10.50          0.970 2.19e+02  2.17e+02 297.000\n",
      "4      4    0.879  -8.71          0.968 1.16e+02  1.13e+02 182.000\n",
      "5      5    0.942  -7.36          0.976 6.19e+01  5.96e+01 115.000\n",
      "6      6    0.957  -6.07          0.977 3.36e+01  3.18e+01  74.000\n",
      "7      7    0.956  -5.17          0.970 1.85e+01  1.70e+01  48.800\n",
      "8      8    0.955  -4.45          0.967 1.03e+01  9.19e+00  33.100\n",
      "9      9    0.971  -3.81          0.990 5.86e+00  5.00e+00  22.900\n",
      "10    10    0.981  -3.27          0.998 3.38e+00  2.75e+00  16.100\n",
      "11    12    0.983  -2.56          0.987 1.20e+00  8.42e-01   8.860\n",
      "12    14    0.969  -2.22          0.978 4.63e-01  2.65e-01   5.800\n",
      "13    16    0.946  -1.95          0.948 1.98e-01  8.51e-02   3.930\n",
      "14    18    0.964  -1.69          0.965 9.37e-02  2.80e-02   2.720\n",
      "15    20    0.943  -1.54          0.929 4.88e-02  9.32e-03   1.920\n",
      "16    22    0.947  -1.46          0.942 2.76e-02  3.20e-03   1.510\n",
      "17    24    0.961  -1.41          0.968 1.68e-02  1.09e-03   1.260\n",
      "18    26    0.355  -1.81          0.214 1.07e-02  3.90e-04   1.090\n",
      "19    28    0.324  -2.13          0.158 7.22e-03  1.39e-04   0.954\n",
      "20    30    0.402  -2.85          0.330 5.04e-03  4.98e-05   0.854\n",
      "  Power  SFT.R.sq      slope truncated.R.sq   mean.k. median.k.    max.k.\n",
      "1     1 0.1049761   7.603194      0.6147123 817.90315 821.64664 872.79351\n",
      "2     2 0.2299508  -6.918313      0.9658169 420.50477 421.18094 498.92936\n",
      "3     3 0.7551021 -10.452395      0.9700059 219.14806 217.15849 296.71906\n",
      "4     4 0.8787178  -8.714714      0.9678526 115.72909 112.95008 182.07790\n",
      "5     5 0.9418695  -7.358418      0.9762142  61.92829  59.64547 114.74715\n",
      "6     6 0.9568173  -6.068219      0.9769726  33.59058  31.75942  74.04528\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/SoftPower/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "  \n",
    "  print(paste0('Estinmating Soft-Power Threshold in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # Test different soft powers:\n",
    "  seurat_obj[[cell_type]] <- TestSoftPowers(\n",
    "    seurat_obj[[cell_type]],\n",
    "    networkType = 'signed', # you can also use \"unsigned\" or \"signed hybrid\"\n",
    "    setDatExpr = TRUE,\n",
    "    group.by = 'cell_type',\n",
    "    group_name = cell_type,\n",
    "  )\n",
    "\n",
    "  # plot the results:\n",
    "  plot_list <- PlotSoftPowers(seurat_obj[[cell_type]], \n",
    "                              # wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment\n",
    "                              )\n",
    "\n",
    "  # assemble with patchwork\n",
    "  pdf(paste0(fig_dir, cell_type, '_SoftPower.pdf'), width=12, height=8)\n",
    "  print(wrap_plots(plot_list, ncol=2))\n",
    "  dev.off()\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f5b7f474",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2ad0b902",
   "metadata": {},
   "source": [
    "The general guidance for WGCNA and hdWGCNA is to `pick the lowest soft power threshold that has a Scale Free Topology Model Fit greater than or equal to 0.8`. Later on, the `ConstructNetwork `will automatically select the soft power threshold if we do not provide one.\n",
    "\n",
    "The output table from the parameter sweep is stored in the hdWGCNA experiment and can be accessed using the GetPowerTable function for further inspection:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "62c54e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# power_table <- GetPowerTable(seurat_obj)\n",
    "# head(power_table)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7c6251e4",
   "metadata": {},
   "source": [
    "### **Construct co-expression network**\n",
    "\n",
    "We now have everything that we need to construct our co-expression network. Here we use the hdWGCNA function ConstructNetwork, which calls the WGCNA function blockwiseConsensusModules under the hood. This function has quite a few parameters to play with if for advanced use cases, but we use default parameters that work well with many single-cell datasets. \n",
    "\n",
    "The parameters for blockwiseConsensusModules can be passed directly to ConstructNetwork with the same parameter names.\n",
    "\n",
    "The following code construtcts the co-expression network using the soft power threshold selected above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "14831950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for EXCITATORY\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 9\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for INHIBITORY\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 7\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for ASTROCYTE\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 6\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for MICROGLIA\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 4\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 5\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for OPC\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 4\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Excitatory\"      \"Inhibitory\"      \"Astrocyte\"       \"Microglia\"      \n",
      "[5] \"Oligodendrocyte\" \"OPC\"            \n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "  \n",
    "  print(paste0('Constructing co-expression network in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  tryCatch({\n",
    "    # construct co-expression network:\n",
    "    seurat_obj[[cell_type]] <- ConstructNetwork(\n",
    "      seurat_obj[[cell_type]], \n",
    "      soft_power = NULL, # Set to NULL so that the ConstructNetwork function obtains it automatically (Scale Free Topology Model Fit >= 0.8)\n",
    "      setDatExpr = FALSE,\n",
    "      overwrite_tom = TRUE,\n",
    "      tom_name = toupper(substr(cell_type, 1, 3)) # name of the topological overlap matrix written to disk set as the name of the hdWGCNA experiment \n",
    "    )\n",
    "  }, error = function(e) {\n",
    "    print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "    print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "    # Code to handle the error condition\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "  },  \n",
    "  message = function(m) {\n",
    "    print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "    print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "    # Code to handle the error condition\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "  }) \n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "print(names(seurat_obj))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b8e77379",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "49676183",
   "metadata": {},
   "source": [
    "hdWGCNA also includes a function PlotDendrogram to visualize the WGCNA dendrogram, a common visualization to show the different co-expression modules resulting from the network analysis. Each leaf on the dendrogram represents a single gene, and the color at the bottom indicates the co-expression module assignment.\n",
    "\n",
    "\n",
    "`Importantly, the “grey” module consists of genes that were not grouped into any co-expression module. The grey module should be ignored for all downstream analysis and interpretation.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8da7dcca",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/dendrogram/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "tryCatch({\n",
    "    for (cell_type in names(seurat_obj)){\n",
    "        # print(paste0('Constructing co-expression networkd in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "        pdf(paste0(fig_dir, cell_type, \"_dendro.pdf\"), height=2, width=4)\n",
    "        PlotDendrogram(seurat_obj[[cell_type]], main='hdWGCNA Dendrogram')\n",
    "        dev.off()\n",
    "    }\n",
    "}, error = function(e){\n",
    "    NULL\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ed9baba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a5dfb822",
   "metadata": {},
   "source": [
    "### **`OPTIONAl`: inspect the topoligcal overlap matrix (TOM)**\n",
    "\n",
    "hdWGCNA represents the co-expression network as a `topoligcal overlap matrix (TOM)`. This is a square matrix of genes by genes, where each value is the topoligcal overlap between the genes. The TOM is written to the disk when running ConstructNetwork, and we can load it into R using the GetTOM function. \n",
    "\n",
    "We may wish to inspect the TOM for custom downstream analyses.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fbea29f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj -o TOM\n",
    "\n",
    "# TOM <- GetTOM(seurat_obj)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5e1c23ce",
   "metadata": {},
   "source": [
    "### **Module Eigengenes and Connectivity**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "08779263",
   "metadata": {},
   "source": [
    "### Compute harmonized module eigengenes\n",
    "\n",
    "`Module Eigengenes (MEs)` are a commonly used metric to summarize the gene expression profile of an entire co-expression module. Briefly, module eigengenes are computed by performing principal component analysis (PCA) on the subset of the gene expression matrix comprising each module. `The first PC of each of these PCA matrices are the MEs.`\n",
    "\n",
    "Dimensionality reduction techniques are a very hot topic in single-cell genomics. It is well known that technical artifacts can muddy the analysis of single-cell datasets, and over the years there have been many methods that aim to reduce the effects of these artifacts. Therefore it stands to reason that `MEs would be subject to these technical artifacts as well`, and hdWGCNA seeks to alleviate these effects.\n",
    "\n",
    "`hdWGCNA includes a function ModuleEigengenes` to compute module eigengenes in single cells. Additionally, we can apply Harmony batch correction to the MEs, yielding harmonized module eigengenes (hMEs).\n",
    "\n",
    "The following code performs the module eigengene computation `harmonizing by the Sample of origin using the group.by.vars parameter.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "df854ccb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"blue\"\n",
      "[1] \"grey\"\n",
      "[1] \"brown\"\n",
      "[1] \"green\"\n",
      "[1] \"yellow\"\n",
      "[1] \"----------------\"\n",
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"brown\"\n",
      "[1] \"blue\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"grey\"\n",
      "[1] \"green\"\n",
      "[1] \"yellow\"\n",
      "[1] \"----------------\"\n",
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"brown\"\n",
      "[1] \"grey\"\n",
      "[1] \"yellow\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"blue\"\n",
      "[1] \"----------------\"\n",
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"grey\"\n",
      "[1] \"blue\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"----------------\"\n",
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"grey\"\n",
      "[1] \"blue\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"brown\"\n",
      "[1] \"----------------\"\n",
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for OPC\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"grey\"\n",
      "[1] \"blue\"\n",
      "[1] \"----------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Estimating module eigen-genes in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "    # need to run ScaleData first or else harmony throws an error:\n",
    "    seurat_obj[[cell_type]] <- ScaleData(seurat_obj[[cell_type]], features=VariableFeatures(seurat_obj[[cell_type]]))\n",
    "\n",
    "    # compute all MEs in the full single-cell dataset\n",
    "    seurat_obj[[cell_type]] <- ModuleEigengenes(\n",
    "    seurat_obj[[cell_type]],\n",
    "    group.by.vars=subject_id\n",
    "    )\n",
    "    print('----------------')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "05f2fc35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "05cd78cf",
   "metadata": {},
   "source": [
    "The ME matrices are stored as a matrix where each row is a cell and each column is a module. This matrix can be extracted from the Seurat object using the GetMEs function, which retrieves the hMEs by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "93840f67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -o hMEs -o MEs\n",
    "\n",
    "hMEs <- list()\n",
    "MEs <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    print(paste0('Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "    # harmonized module eigengenes:\n",
    "    hMEs[[cell_type]] <- GetMEs(seurat_obj[[cell_type]])\n",
    "\n",
    "    # module eigengenes:\n",
    "    MEs[[cell_type]] <- GetMEs(seurat_obj[[cell_type]], harmonized=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8075bc3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c011bbb2",
   "metadata": {},
   "source": [
    "### Compute module connectivity"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e9f2e4ad",
   "metadata": {},
   "source": [
    "In co-expression network analysis, we often want to focus on the `“hub genes”`,`those which are highly connected within each module.` Therefore we wish to determine the `eigengene-based connectivity, also known as kME`, of each gene. \n",
    "\n",
    "hdWGCNA includes the ModuleConnectivity to compute the kME values in the full single-cell dataset, rather than the metacell dataset. This function essentially computes pairwise correlations between genes and module eigengenes. kME can be computed for all cells in the dataset, but we recommend computing kME in the cell type or group that was previously used to run ConstructNetwork."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9ebaef5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Estimating module connectivity in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # compute eigengene-based connectivity (kME):\n",
    "  seurat_obj[[cell_type]] <- ModuleConnectivity(\n",
    "    seurat_obj[[cell_type]],\n",
    "    group.by = 'cell_type', \n",
    "    group_name = cell_type\n",
    "  )\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "72891253",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "95b46026",
   "metadata": {},
   "source": [
    "For convenience, we re-name the hdWGCNA modules to indicate that they are from the Excitatory neuron group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a52dff32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Renaming modules in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Renaming modules in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Renaming modules in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Renaming modules in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Renaming modules in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Renaming modules in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Renaming modules in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # rename the modules\n",
    "  seurat_obj[[cell_type]] <- ResetModuleNames(\n",
    "    seurat_obj[[cell_type]],\n",
    "    new_name = paste0(toupper(substr(cell_type, 1, 3)), \"-M\"),\n",
    "    wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment,\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0f353a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "88ad5a62",
   "metadata": {},
   "source": [
    "We also reset module colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5e7726f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Resetting module colors in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Resetting module colors in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Resetting module colors in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Resetting module colors in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Resetting module colors in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Resetting module colors in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj \n",
    "\n",
    "\n",
    "\n",
    "library(MetBrewer)\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Resetting module colors in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "  mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "    distinct %>% arrange(module) %>% .$color\n",
    "  n_colors <- length(mod_colors) -1\n",
    "\n",
    "  new_colors <- paste0(met.brewer(\"Signac\", n=n_colors,))\n",
    "  new_colors <- sample(new_colors)\n",
    "  seurat_obj[[cell_type]] <- ResetModuleColors(seurat_obj[[cell_type]], new_colors)\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6b194589",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ae334b97",
   "metadata": {},
   "source": [
    "We can visualize the genes in each module ranked by kME using the PlotKMEs function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "03423498",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer dplyr::select over any other package.\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"EXC-M1\"\n",
      "[1] \"EXC-M2\"\n",
      "[1] \"EXC-M3\"\n",
      "[1] \"EXC-M4\"\n",
      "[1] \"EXC-M5\"\n",
      "[1] \"EXC-M1\"\n",
      "[1] \"EXC-M2\"\n",
      "[1] \"EXC-M3\"\n",
      "[1] \"EXC-M4\"\n",
      "[1] \"EXC-M5\"\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n",
      "[1] \"INH-M5\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n",
      "[1] \"INH-M5\"\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"AST-M1\"\n",
      "[1] \"AST-M2\"\n",
      "[1] \"AST-M3\"\n",
      "[1] \"AST-M4\"\n",
      "[1] \"AST-M1\"\n",
      "[1] \"AST-M2\"\n",
      "[1] \"AST-M3\"\n",
      "[1] \"AST-M4\"\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"MIC-M1\"\n",
      "[1] \"MIC-M2\"\n",
      "[1] \"MIC-M1\"\n",
      "[1] \"MIC-M2\"\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for OPC\"\n",
      "[1] \"OPC-M1\"\n",
      "[1] \"OPC-M2\"\n",
      "[1] \"OPC-M1\"\n",
      "[1] \"OPC-M2\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "library(conflicted)\n",
    "conflicts_prefer(dplyr::select)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/ModuleConnectivity/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # plot genes ranked by kME for each module\n",
    "\n",
    "    p <- PlotKMEs(seurat_obj[[cell_type]], ncol=5)\n",
    "\n",
    "    pdf(paste0(fig_dir, cell_type, '_KMEs.pdf'), width=12, height=8)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c6b1f8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b8d0658b",
   "metadata": {},
   "source": [
    "### Getting the module assignment table\n",
    "\n",
    "hdWGCNA allows for easy access of the module assignment table using the GetModules function. \n",
    "\n",
    "This table consists of three columns: \n",
    "\n",
    "- gene_name stores the gene’s symbol or ID, \n",
    "- module stores the gene’s module assignment, and \n",
    "- color stores a color mapping for each module, \n",
    "\n",
    "which is used in many downstream plotting steps. If ModuleConnectivity has been called on this hdWGCNA experiment, this table will have additional columns for the kME of each module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "9121be04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for EXCITATORY\"\n",
      "       gene_name module   color kME_EXC-M1 kME_EXC-M2  kME_grey\n",
      "EXOSC6    EXOSC6 EXC-M1 #de597c  0.3794672  0.3060506 0.3469334\n",
      "MSH2        MSH2 EXC-M2 #d8443c  0.3304868  0.4888776 0.4649081\n",
      "MSH6        MSH6   grey    grey  0.1996748  0.2825429 0.2755604\n",
      "RNF168    RNF168 EXC-M2 #d8443c  0.2663690  0.3499940 0.3503186\n",
      "RNF8        RNF8   grey    grey  0.2570769  0.2802334 0.2949134\n",
      "ACVR1      ACVR1   grey    grey  0.2307577  0.2289589 0.2499007\n",
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for INHIBITORY\"\n",
      "        gene_name module   color kME_INH-M1 kME_INH-M2 kME_INH-M3\n",
      "PAPSS1     PAPSS1 INH-M1 #de597c  0.2860051  0.2879507  0.2595689\n",
      "PODXL2     PODXL2 INH-M2 #d8443c  0.2004125  0.4882851  0.1889793\n",
      "SULT1A1   SULT1A1 INH-M3 #f4c40f  0.1892712  0.1662369  0.2461429\n",
      "SULT4A1   SULT4A1 INH-M2 #d8443c  0.2903661  0.4260990  0.2396406\n",
      "TPST1       TPST1   grey    grey  0.4821423  0.3861112  0.5108684\n",
      "AK9           AK9   grey    grey  0.3156525  0.2333540  0.3101160\n",
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for ASTROCYTE\"\n",
      "       gene_name module   color kME_AST-M1  kME_grey kME_AST-M2\n",
      "MNAT1      MNAT1 AST-M1 #1f6e9c  0.2972768 0.2658918  0.1700399\n",
      "POLR2B    POLR2B   grey    grey  0.1884088 0.2018828  0.1930746\n",
      "POLR2J    POLR2J AST-M2 #fe9b00  0.1480585 0.1642837  0.2594579\n",
      "RNMT        RNMT AST-M3 #f4c40f  0.2133691 0.2228074  0.1411242\n",
      "SUPT5H    SUPT5H   grey    grey  0.1387223 0.1479779  0.1369685\n",
      "ALDOA      ALDOA AST-M4 #d8443c  0.1814327 0.2018110  0.2213227\n",
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for MICROGLIA\"\n",
      "        gene_name module   color   kME_grey kME_MIC-M1  kME_MIC-M2\n",
      "CPEB3       CPEB3   grey    grey 0.29906450 0.24718947  0.28002436\n",
      "RC3H1       RC3H1   grey    grey 0.23369725 0.19870243  0.22921364\n",
      "ZFP36       ZFP36 MIC-M1 #d8443c 0.05784061 0.12857447 -0.01558296\n",
      "ZFP36L1   ZFP36L1   grey    grey 0.17480063 0.15462661  0.18862674\n",
      "ZFP36L2   ZFP36L2 MIC-M2 #fe9b00 0.23111197 0.16420265  0.34102401\n",
      "ALDOA       ALDOA   grey    grey 0.13354265 0.07320078  0.10251597\n",
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "      gene_name module   color  kME_grey kME_OLI-M1 kME_OLI-M2\n",
      "AK9         AK9   grey    grey 0.2409851 0.14587166  0.2183170\n",
      "ALDOA     ALDOA OLI-M1 #fe9b00 0.2230344 0.23438719  0.2387688\n",
      "BPGM       BPGM OLI-M2 #f4c40f 0.1488220 0.08926321  0.1703043\n",
      "ENO2       ENO2 OLI-M1 #fe9b00 0.2126610 0.25790748  0.2012188\n",
      "PKM         PKM OLI-M1 #fe9b00 0.1969123 0.27911885  0.1801871\n",
      "ACTB       ACTB OLI-M1 #fe9b00 0.3259575 0.34831814  0.3559341\n",
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for OPC\"\n",
      "       gene_name module   color kME_OPC-M1  kME_grey kME_OPC-M2\n",
      "ADAM17    ADAM17 OPC-M1 #d8443c  0.2885748 0.2897438  0.2310888\n",
      "BCL2        BCL2   grey    grey  0.2040306 0.2012497  0.1541597\n",
      "HDAC4      HDAC4 OPC-M1 #d8443c  0.3107843 0.3163975  0.2573867\n",
      "HDAC9      HDAC9 OPC-M1 #d8443c  0.4147849 0.4095207  0.3241053\n",
      "MEF2C      MEF2C OPC-M1 #d8443c  0.3418663 0.3377940  0.2665605\n",
      "PLCG2      PLCG2 OPC-M1 #d8443c  0.2180198 0.2240565  0.1754748\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Obtain Module Assignment Table in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # get the module assignment table:\n",
    "    modules <- GetModules(seurat_obj[[cell_type]])\n",
    "\n",
    "    # show the first 6 columns:\n",
    "    print(head(modules[,1:6]))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "21bdbbf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "48d2e707",
   "metadata": {},
   "source": [
    "A table of the top N hub genes sorted by kME can be extracted using the GetHubGenes function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3a0b8898",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for EXCITATORY\"\n",
      "  gene_name module       kME\n",
      "1      DNM1 EXC-M1 0.7226235\n",
      "2      PRNP EXC-M1 0.7229455\n",
      "3     YWHAG EXC-M1 0.7238225\n",
      "4  HSP90AB1 EXC-M1 0.7250436\n",
      "5     GPM6A EXC-M1 0.7394659\n",
      "6      RTN3 EXC-M1 0.7448993\n",
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for INHIBITORY\"\n",
      "  gene_name module       kME\n",
      "1     SLIT2 INH-M1 0.6811985\n",
      "2    CACNB4 INH-M1 0.6839641\n",
      "3      SYT1 INH-M1 0.6872877\n",
      "4      GRM5 INH-M1 0.6916851\n",
      "5  IL1RAPL1 INH-M1 0.6983711\n",
      "6    LRRTM4 INH-M1 0.7120656\n",
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for ASTROCYTE\"\n",
      "  gene_name module       kME\n",
      "1    SLC1A2 AST-M1 0.5529484\n",
      "2     AUTS2 AST-M1 0.5545391\n",
      "3     PDE4D AST-M1 0.5567502\n",
      "4      GPC5 AST-M1 0.6146770\n",
      "5     ERBB4 AST-M1 0.6404520\n",
      "6    CTNNA2 AST-M1 0.6518460\n",
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for MICROGLIA\"\n",
      "  gene_name module       kME\n",
      "1    RAB11A MIC-M1 0.3340226\n",
      "2       B2M MIC-M1 0.3436526\n",
      "3     TAOK3 MIC-M1 0.3467247\n",
      "4     RUNX1 MIC-M1 0.3468044\n",
      "5      GAB2 MIC-M1 0.3529157\n",
      "6     PRKCA MIC-M1 0.3544359\n",
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "  gene_name module       kME\n",
      "1       MBP OLI-M1 0.4051883\n",
      "2      CHN1 OLI-M1 0.4114154\n",
      "3    SNAP25 OLI-M1 0.4236483\n",
      "4     GPM6A OLI-M1 0.4456058\n",
      "5     NEGR1 OLI-M1 0.4597219\n",
      "6   SLC26A3 OLI-M1 0.4656504\n",
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for OPC\"\n",
      "  gene_name module       kME\n",
      "1     ERBB4 OPC-M1 0.6514086\n",
      "2     NPAS3 OPC-M1 0.6659624\n",
      "3    ADGRL3 OPC-M1 0.6795928\n",
      "4    PTPRZ1 OPC-M1 0.6797307\n",
      "5    LRRTM4 OPC-M1 0.6882198\n",
      "6    PCDH15 OPC-M1 0.6918980\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Obtain Hub Genes in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # get hub genes\n",
    "    hub_df <- GetHubGenes(seurat_obj[[cell_type]], n_hubs = 10)\n",
    "\n",
    "    print(head(hub_df))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8f478886",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "27f2d80d",
   "metadata": {},
   "source": [
    "## **Compute hub gene signature scores**\n",
    "\n",
    "Gene scoring analysis is a popular method in single-cell transcriptomics for computing a score for the overall signature of a set of genes. Seurat implements their own gene scoring technique using the `AddModuleScore` function, but there are also alternative approaches such as UCell.\n",
    "\n",
    "hdWGCNA includes the function `ModuleExprScore` to compute gene scores for a give number of genes for each module, using either the Seurat or UCell algorithm. Gene scoring is an alternative way of summarizing the expression of a module from computing the module eigengene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e1f93a8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for EXCITATORY\"\n",
      "Selecting by kME_EXC-M1\n",
      "Selecting by kME_EXC-M2\n",
      "Selecting by kME_EXC-M3\n",
      "Selecting by kME_EXC-M4\n",
      "Selecting by kME_EXC-M5\n",
      "Selecting by kME_EXC-M1\n",
      "Selecting by kME_EXC-M2\n",
      "Selecting by kME_EXC-M3\n",
      "Selecting by kME_EXC-M4\n",
      "Selecting by kME_EXC-M5\n",
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for INHIBITORY\"\n",
      "Selecting by kME_INH-M1\n",
      "Selecting by kME_INH-M2\n",
      "Selecting by kME_INH-M3\n",
      "Selecting by kME_INH-M4\n",
      "Selecting by kME_INH-M5\n",
      "Selecting by kME_INH-M1\n",
      "Selecting by kME_INH-M2\n",
      "Selecting by kME_INH-M3\n",
      "Selecting by kME_INH-M4\n",
      "Selecting by kME_INH-M5\n",
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for ASTROCYTE\"\n",
      "Selecting by kME_AST-M1\n",
      "Selecting by kME_AST-M2\n",
      "Selecting by kME_AST-M3\n",
      "Selecting by kME_AST-M4\n",
      "Selecting by kME_AST-M1\n",
      "Selecting by kME_AST-M2\n",
      "Selecting by kME_AST-M3\n",
      "Selecting by kME_AST-M4\n",
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for MICROGLIA\"\n",
      "Selecting by kME_MIC-M1\n",
      "Selecting by kME_MIC-M2\n",
      "Selecting by kME_MIC-M1\n",
      "Selecting by kME_MIC-M2\n",
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n",
      "Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n",
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for OPC\"\n",
      "Selecting by kME_OPC-M1\n",
      "Selecting by kME_OPC-M2\n",
      "Selecting by kME_OPC-M1\n",
      "Selecting by kME_OPC-M2\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Compute hub gene scores in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "  # compute gene scoring for the top 25 hub genes by kME for each module\n",
    "  # with Seurat method\n",
    "  seurat_obj[[cell_type]] <- ModuleExprScore(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_genes = 25,\n",
    "    method='Seurat'\n",
    "  )\n",
    "\n",
    "  # compute gene scoring for the top 25 hub genes by kME for each module\n",
    "  # with UCell method\n",
    "  library(UCell)\n",
    "  seurat_obj[[cell_type]] <- ModuleExprScore(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_genes = 25,\n",
    "    method='UCell'\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "34de1c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "03dc3f32",
   "metadata": {},
   "source": [
    "## **Basic Visualization**\n",
    "\n",
    "Here we visualize some results using hdWGCNA, and we employ some of Seurat’s built-in plotting tools to visualize our hdWGCNA results.\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "368c8716",
   "metadata": {},
   "source": [
    "### **Module Feature Plots**\n",
    "\n",
    "FeaturePlot is a commonly used Seurat visualization to show a feature of interest directly on the dimensionality reduction. hdWGCNA includes the ModuleFeaturePlot function to consruct FeaturePlots for each co-expression module colored by each module’s uniquely assigned color."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c5865383",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module Feature Plots in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"EXC-M1\"\n",
      "[1] \"EXC-M2\"\n",
      "[1] \"EXC-M3\"\n",
      "[1] \"EXC-M4\"\n",
      "[1] \"EXC-M5\"\n",
      "[1] \"Module Feature Plots in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n",
      "[1] \"INH-M5\"\n",
      "[1] \"Module Feature Plots in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"AST-M1\"\n",
      "[1] \"AST-M2\"\n",
      "[1] \"AST-M3\"\n",
      "[1] \"AST-M4\"\n",
      "[1] \"Module Feature Plots in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"MIC-M1\"\n",
      "[1] \"MIC-M2\"\n",
      "[1] \"Module Feature Plots in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n",
      "[1] \"Module Feature Plots in hdWGCNA Experiment for OPC\"\n",
      "[1] \"OPC-M1\"\n",
      "[1] \"OPC-M2\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/FeaturePlots/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Module Feature Plots in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "     \n",
    "  # make a featureplot of hMEs for each module\n",
    "  plot_list <- ModuleFeaturePlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    features='hMEs', # plot the hMEs\n",
    "    order=TRUE, # order so the points with highest hMEs are on top\n",
    "    # raster=TRUE, \n",
    "    # raster_dpi=400, \n",
    "    # alpha=1,\n",
    "    # restrict_range=FALSE, \n",
    "    # raster_scale=0.25,\n",
    "  )\n",
    "\n",
    "  plot_list <- lapply(1:length(plot_list), function(x){\n",
    "    plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_featureplot_MEs.pdf\"), height=10, width=5)\n",
    "  print(wrap_plots(plot_list, ncol=3))\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7d1a4b43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2e09999a",
   "metadata": {},
   "source": [
    "We can also plot the hub gene signature score using the same function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "07825465",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module Feature Plots for HUB GENES in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"EXC-M1\"\n",
      "[1] \"EXC-M2\"\n",
      "[1] \"EXC-M3\"\n",
      "[1] \"EXC-M4\"\n",
      "[1] \"EXC-M5\"\n",
      "[1] \"Module Feature Plots for HUB GENES in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n",
      "[1] \"INH-M5\"\n",
      "[1] \"Module Feature Plots for HUB GENES in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"AST-M1\"\n",
      "[1] \"AST-M2\"\n",
      "[1] \"AST-M3\"\n",
      "[1] \"AST-M4\"\n",
      "[1] \"Module Feature Plots for HUB GENES in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"MIC-M1\"\n",
      "[1] \"MIC-M2\"\n",
      "[1] \"Module Feature Plots for HUB GENES in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n",
      "[1] \"Module Feature Plots for HUB GENES in hdWGCNA Experiment for OPC\"\n",
      "[1] \"OPC-M1\"\n",
      "[1] \"OPC-M2\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/FeaturePlots/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Module Feature Plots for HUB GENES in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "     \n",
    "  # make a featureplot of hMEs for each module\n",
    "  plot_list <- ModuleFeaturePlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    features='scores',  # plot the hub gene scores\n",
    "    order='shuffle', # order so cells are shuffled\n",
    "    ucell = TRUE, # depending on Seurat vs UCell for gene scoring\n",
    "    # raster=TRUE, \n",
    "    # raster_dpi=400, \n",
    "    # alpha=1,\n",
    "    # restrict_range=FALSE, \n",
    "    # raster_scale=0.25,\n",
    "  )\n",
    "\n",
    "  plot_list <- lapply(1:length(plot_list), function(x){\n",
    "    plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_featureplot_hubs.pdf\"), height=10, width=5)\n",
    "  print(wrap_plots(plot_list, ncol=3))\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "dbbce473",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "219d5785",
   "metadata": {},
   "source": [
    "### **Module Correlations**\n",
    "\n",
    "hdWGCNA includes the ModuleCorrelogram function to visualize the correlation between each module based on their hMEs, MEs, or hub gene scores using the R package corrplot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "9ac695b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module Correlation in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Module Correlation in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Module Correlation in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Module Correlation in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Module Correlation in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Module Correlation in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/ModuleCorrelations/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Module Correlation in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    # plot module correlagram\n",
    "    pdf(paste0(fig_dir, cell_type, \"_correlogram.pdf\"), height=10, width=5)\n",
    "    ModuleCorrelogram(seurat_obj[[cell_type]])\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "29692fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3cf396dd",
   "metadata": {},
   "source": [
    "### **Seurat plotting functions**\n",
    "\n",
    "\n",
    "The base Seurat plotting functions are also great for visualizing hdWGCNA outputs. Here we demonstrate plotting hMEs using DotPlot and VlnPlot. The key to using Seurat’s plotting functions to visualize the hdWGCNA data is to add it into the Seurat object’s @meta.data slot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4dd017bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -o mods\n",
    "\n",
    "mods <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    # get hMEs from seurat object\n",
    "    MEs <- GetMEs(seurat_obj[[cell_type]], harmonized=TRUE)\n",
    "    mods[[cell_type]] <- colnames(MEs); \n",
    "    mods[[cell_type]] <- mods[[cell_type]][mods[[cell_type]] != 'grey']\n",
    "    \n",
    "    \n",
    "    # add hMEs to Seurat meta-data:\n",
    "    seurat_obj[[cell_type]]@meta.data <- cbind(seurat_obj[[cell_type]]@meta.data, MEs)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "6c8c3009",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )\n",
    "\n",
    "\n",
    "# convert nested list of Seurat object into Rpy2 object \n",
    "mods =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: mods[cell_type]\n",
    "                     \n",
    "                            for cell_type in mods.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c1c37b40",
   "metadata": {},
   "source": [
    "Now we can easily use Seurat’s DotPlot function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4f45a3f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"DotPlot for average expression of MEs in hdWGCNA Experiment for EXCITATORY\"\n",
      "Scale for colour is already present.\n",
      "Adding another scale for colour, which will replace the existing scale.\n",
      "[1] \"DotPlot for average expression of MEs in hdWGCNA Experiment for INHIBITORY\"\n",
      "Scale for colour is already present.\n",
      "Adding another scale for colour, which will replace the existing scale.\n",
      "[1] \"DotPlot for average expression of MEs in hdWGCNA Experiment for ASTROCYTE\"\n",
      "Scale for colour is already present.\n",
      "Adding another scale for colour, which will replace the existing scale.\n",
      "[1] \"DotPlot for average expression of MEs in hdWGCNA Experiment for MICROGLIA\"\n",
      "Scale for colour is already present.\n",
      "Adding another scale for colour, which will replace the existing scale.\n",
      "[1] \"DotPlot for average expression of MEs in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Scale for colour is already present.\n",
      "Adding another scale for colour, which will replace the existing scale.\n",
      "[1] \"DotPlot for average expression of MEs in hdWGCNA Experiment for OPC\"\n",
      "Scale for colour is already present.\n",
      "Adding another scale for colour, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -i mods -o seurat_obj\n",
    "\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/DotPlot/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('DotPlot for average expression of MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # plot with Seurat's DotPlot function\n",
    "  p <- DotPlot(seurat_obj[[cell_type]], features=mods[[cell_type]], group.by = 'cell_type')\n",
    "\n",
    "  # flip the x/y axes, rotate the axis labels, and change color scheme:\n",
    "  p <- p +\n",
    "    coord_flip() +\n",
    "    RotatedAxis() +\n",
    "    scale_color_gradient2(high='red', mid='grey95', low='blue')\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_dotplot.pdf\"), height=10, width=5)\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "31592f99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b1b75ca0",
   "metadata": {},
   "source": [
    "Here is another example where we use Seurat’s VlnPlot function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "5c23ebda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "# for (cell_type in names(seurat_obj)){\n",
    "\n",
    "#   print(paste0('ViolinPlot for average expression of MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "#   # Plot INH-M4 hME using Seurat VlnPlot function\n",
    "#   p <- VlnPlot(\n",
    "#     seurat_obj,\n",
    "#     features = as.vector(mods[[cell_type]]),\n",
    "#     group.by = 'cell_type',\n",
    "#     pt.size = 0 # don't show actual data points\n",
    "#   )\n",
    "\n",
    "#   # add box-and-whisker plots on top:\n",
    "#   p <- p + geom_boxplot(width=.25, fill='white')\n",
    "\n",
    "#   # change axis labels and remove legend:\n",
    "#   p <- p + xlab('') + ylab('hME') + NoLegend()\n",
    "\n",
    "#   # plot output\n",
    "#   pdf(paste0(\"../results/hdWGCNA/ViolinPlot/\", save_prefix, '_', cell_type, \"_violinplot.pdf\"), height=10, width=5)\n",
    "#   print(p)\n",
    "#   dev.off()\n",
    "\n",
    "# }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "8246a79f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # convert nested list of Seurat object into Rpy2 object \n",
    "# seurat_obj =  robjects.ListVector(\n",
    "#                         {\n",
    "#                             cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "#                             for cell_type in seurat_obj.keys()\n",
    "#                         }\n",
    "#                     )\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b657e6a1",
   "metadata": {},
   "source": [
    "## **Overlap With AD Genes**\n",
    "\n",
    "Next on the co-expression UMAP, we show&quantify the overlap between hub genes that have a known association with AD using the AD databases obtained from the [**KEGG Alzheimer's pathway**](https://www.genome.jp/pathway/hsa05010) and [**Harmonizome (Mayaanlab)**](https://maayanlab.cloud/Harmonizome/gene_set/Alzheimer+Disease/dbGAP+Gene-Trait+Associations) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "964ea985",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEGG_paths = pathway_analyses.read_pathways('../data/pathway_databases/KEGG_2019_Human.txt')\n",
    "KEGG_genes = [gene for gene in list(KEGG_paths[(KEGG_paths[0].str.startswith('Alzheimer disease'))].iloc[:, 1:].values[0]) if str(gene)!=\"nan\"]\n",
    "mayaanlab_genes = pd.read_csv('../data/pathway_databases/AD_genes.csv').Symbol.to_list()\n",
    "KEGG_genes.extend(mayaanlab_genes)\n",
    "\n",
    "AD_genes = list(set(KEGG_genes))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "61d42932",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -i AD_genes\n",
    "\n",
    "# load modules\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/GeneOverlap/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Estimating AD gene overlap in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "  genome.size <- dim(seurat_obj[[cell_type]])[1]\n",
    "  # genome.size <- length(GetWGCNAGenes(seurat_obj[[cell_type]]))\n",
    "  \n",
    "  overlap_df <- do.call(rbind, lapply(mods, function(cur_mod){\n",
    "\n",
    "  cur_genes <- modules %>% subset(module == cur_mod) %>% .$gene_name\n",
    "\n",
    "  cur_overlap <- testGeneOverlap(newGeneOverlap(\n",
    "      cur_genes,\n",
    "      as.vector(AD_genes),\n",
    "      genome.size=genome.size\n",
    "  ))\n",
    "\n",
    "  cur_overlap <- data.frame(\n",
    "    'odds.ratio' = cur_overlap@odds.ratio,\n",
    "    'pval' = cur_overlap@pval,\n",
    "    'Jaccard' = cur_overlap@Jaccard,\n",
    "    'size_intersection' = length(cur_overlap@intersection),\n",
    "    'module' = cur_mod\n",
    "  )\n",
    "\n",
    "    cur_overlap\n",
    "\n",
    "  })) %>% as.data.frame()\n",
    "\n",
    "  overlap_df <- overlap_df %>% mutate(fdr=p.adjust(pval, method='fdr'))\n",
    "  overlap_df <- overlap_df %>% subset(module != 'grey')\n",
    "\n",
    "  ################################################################################\n",
    "  # Plot as a lollipop\n",
    "  ################################################################################\n",
    "\n",
    "  overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)\n",
    "  overlap_df <- overlap_df %>% arrange(odds.ratio, descending=TRUE)\n",
    "  overlap_df$module <- factor(as.character(overlap_df$module), levels=as.character(overlap_df$module))\n",
    "\n",
    "  mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "    distinct\n",
    "  cp <- mod_colors$color; names(cp) <- mod_colors$module\n",
    "\n",
    "  p <- overlap_df %>%\n",
    "    ggplot(aes(y=module, x=odds.ratio, size= size_intersection, color=module)) +\n",
    "    geom_segment(aes(y=module, yend=module, x=0, xend=odds.ratio), size=0.5, color='grey') +\n",
    "    geom_point() +\n",
    "    geom_point(shape=overlap_df$shape, color='black', fill=NA) +\n",
    "    scale_color_manual(values=cp, guide='none') +\n",
    "    ylab('') + xlab(\"Odds ratio\") +\n",
    "    scale_x_continuous(breaks = seq(0, 19, 2)) +\n",
    "    labs(size='Size\\nintersection') +\n",
    "    ggtitle('Overlap with AD genes') +\n",
    "\n",
    "    theme(\n",
    "      panel.border = element_rect(size=1, color='black', fill=NA),\n",
    "      axis.line.y = element_blank(),\n",
    "      axis.line.x = element_blank(),\n",
    "      plot.title = element_text(hjust=0.5, face='plain')\n",
    "    )\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_AD_overlap.pdf'), width=4, height=3.5)\n",
    "  print(p)\n",
    "  dev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "071e1163",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4cf51263",
   "metadata": {},
   "source": [
    "## **Overlap With DEGs**\n",
    "\n",
    "We quantify the overlap between hub genes and DEGs obtained from previous analysis and saved at `../results/{test_name}/{save_prefix}_{deg_method}_degs.xlsx`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "fd63a2f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer base::intersect over any other package.\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for EXCITATORY\"\n",
      "New names:\n",
      "• `` -> `...1`\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for INHIBITORY\"\n",
      "New names:\n",
      "• `` -> `...1`\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for ASTROCYTE\"\n",
      "New names:\n",
      "• `` -> `...1`\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for MICROGLIA\"\n",
      "New names:\n",
      "• `` -> `...1`\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "New names:\n",
      "• `` -> `...1`\n",
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for OPC\"\n",
      "New names:\n",
      "• `` -> `...1`\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -i deg_method\n",
    "\n",
    "# load modules\n",
    "conflicts_prefer(base::intersect)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/GeneOverlap/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (test in c('ad_vs_no')){\n",
    "  for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Estimating AD gene overlap in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    wgcna_genes <- GetWGCNAGenes(seurat_obj[[cell_type]])\n",
    "    hub_genes <- GetHubGenes(seurat_obj[[cell_type]], 25)\n",
    "    degs <- read_excel(paste0('../results/', test, '/', save_prefix, '_', deg_method, '_degs.xlsx'), sheet = cell_type)\n",
    "    degs <- intersect(hub_genes$gene_name, subset(degs, p_val_adj<0.01 & abs_logFC>0.25)$gene)\n",
    "\n",
    "    modules <- GetModules(seurat_obj[[cell_type]])\n",
    "    mods <- levels(modules$module)\n",
    "    genome.size <- dim(seurat_obj[[cell_type]])[1]\n",
    "    \n",
    "    # genome.size <- length(GetWGCNAGenes(seurat_obj[[cell_type]]))\n",
    "\n",
    "    overlap_df <- do.call(rbind, lapply(mods, function(cur_mod){\n",
    "\n",
    "    cur_genes <- modules %>% subset(module == cur_mod) %>% .$gene_name\n",
    "\n",
    "    cur_overlap <- testGeneOverlap(newGeneOverlap(\n",
    "        cur_genes,\n",
    "        as.vector(degs),\n",
    "        genome.size=genome.size\n",
    "    ))\n",
    "\n",
    "    cur_overlap <- data.frame(\n",
    "      'odds.ratio' = cur_overlap@odds.ratio,\n",
    "      'pval' = cur_overlap@pval,\n",
    "      'Jaccard' = cur_overlap@Jaccard,\n",
    "      'size_intersection' = length(cur_overlap@intersection),\n",
    "      'module' = cur_mod\n",
    "    )\n",
    "\n",
    "      cur_overlap\n",
    "\n",
    "    })) %>% as.data.frame()\n",
    "\n",
    "    overlap_df <- overlap_df %>% mutate(fdr=p.adjust(pval, method='fdr'))\n",
    "    overlap_df <- overlap_df %>% subset(module != 'grey')\n",
    "\n",
    "    ################################################################################\n",
    "    # Plot as a lollipop\n",
    "    ################################################################################\n",
    "\n",
    "    overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)\n",
    "    overlap_df <- overlap_df %>% arrange(odds.ratio, descending=TRUE)\n",
    "    overlap_df$module <- factor(as.character(overlap_df$module), levels=as.character(overlap_df$module))\n",
    "\n",
    "    mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "      distinct\n",
    "    cp <- mod_colors$color; names(cp) <- mod_colors$module\n",
    "\n",
    "    p <- overlap_df %>%\n",
    "      ggplot(aes(y=module, x=odds.ratio, size= size_intersection, color=module)) +\n",
    "      geom_segment(aes(y=module, yend=module, x=0, xend=odds.ratio), size=0.5, color='grey') +\n",
    "      geom_point() +\n",
    "      geom_point(shape=overlap_df$shape, color='black', fill=NA) +\n",
    "      scale_color_manual(values=cp, guide='none') +\n",
    "      ylab('') + xlab(\"Odds ratio\") +\n",
    "      scale_x_continuous(breaks = seq(0, 19, 2)) +\n",
    "      labs(size='Size\\nintersection') +\n",
    "      ggtitle('Overlap with DEGs') +\n",
    "\n",
    "      theme(\n",
    "        panel.border = element_rect(size=1, color='black', fill=NA),\n",
    "        axis.line.y = element_blank(),\n",
    "        axis.line.x = element_blank(),\n",
    "        plot.title = element_text(hjust=0.5, face='plain')\n",
    "      )\n",
    "\n",
    "    pdf(paste0(fig_dir, test, '_', cell_type, '_DEG_overlap.pdf'), width=4, height=3.5)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "1ddecbaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3bc33b36",
   "metadata": {},
   "source": [
    "# **Save Seurat Object**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c0739db5",
   "metadata": {},
   "source": [
    "This wraps up the critical analysis steps for hdWGCNA, let's save our output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "431d9ac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for EXCITATORY\"\n",
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for ASTROCYTE\"\n",
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for MICROGLIA\"\n",
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for OPC\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Saving hdWGCNA object in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    saveRDS(seurat_obj[[cell_type]], file=paste0(dat_dir, cell_type, '_hdWGCNA_object.rds'))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "082b2591",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
