{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import scipy\n",
    "import logging\n",
    "import warnings\n",
    "import anndata\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import decoupler as dc\n",
    "import scrublet as scr\n",
    "import decoupler as dc\n",
    "from scipy import sparse\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from typing import Optional, Union\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.1\n",
      "-----\n",
      "OpenSSL                     22.0.0\n",
      "PIL                         9.2.0\n",
      "absl                        NA\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.2\n",
      "asttokens                   NA\n",
      "astunparse                  1.6.3\n",
      "attr                        21.4.0\n",
      "backcall                    0.2.0\n",
      "beta_ufunc                  NA\n",
      "binom_ufunc                 NA\n",
      "boto3                       1.26.32\n",
      "botocore                    1.29.32\n",
      "bottleneck                  1.3.5\n",
      "brotli                      NA\n",
      "certifi                     2022.09.24\n",
      "cffi                        1.15.1\n",
      "cloudpickle                 2.2.0\n",
      "colorama                    0.4.4\n",
      "cryptography                38.0.1\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dask                        2022.11.0\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.5.1\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "defusedxml                  0.7.1\n",
      "deprecate                   0.3.2\n",
      "entrypoints                 0.4\n",
      "executing                   0.8.3\n",
      "fsspec                      2022.10.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.7.0\n",
      "hypergeom_ufunc             NA\n",
      "igraph                      0.10.2\n",
      "ipykernel                   6.15.2\n",
      "ipython_genutils            0.2.0\n",
      "ipywidgets                  7.6.5\n",
      "jedi                        0.18.1\n",
      "jinja2                      3.1.2\n",
      "jmespath                    1.0.1\n",
      "joblib                      1.1.1\n",
      "kiwisolver                  1.4.2\n",
      "leidenalg                   0.9.0\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.1\n",
      "matplotlib                  3.5.3\n",
      "mpl_toolkits                NA\n",
      "natsort                     7.1.1\n",
      "nbinom_ufunc                NA\n",
      "ncf_ufunc                   NA\n",
      "numba                       0.56.3\n",
      "numexpr                     2.8.3\n",
      "numpy                       1.23.3\n",
      "packaging                   21.3\n",
      "pandas                      1.4.4\n",
      "parso                       0.8.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "prompt_toolkit              3.0.20\n",
      "psutil                      5.9.0\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     11.0.0\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.6.0\n",
      "pydevd_concurrency_analyser NA\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.13.0\n",
      "pyparsing                   3.0.9\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2022.1\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.10\n",
      "scipy                       1.9.3\n",
      "scrublet                    NA\n",
      "seaborn                     0.11.2\n",
      "session_info                1.0.0\n",
      "setuptools                  65.6.3\n",
      "six                         1.16.0\n",
      "sklearn                     1.1.3\n",
      "sparse                      0.13.0\n",
      "stack_data                  0.2.0\n",
      "statsmodels                 0.13.5\n",
      "tensorboard                 2.10.1\n",
      "texttable                   1.6.4\n",
      "threadpoolctl               2.2.0\n",
      "tlz                         0.12.0\n",
      "toolz                       0.12.0\n",
      "torch                       1.12.1\n",
      "torchmetrics                0.10.2\n",
      "tornado                     6.2\n",
      "tqdm                        4.64.1\n",
      "traitlets                   5.1.1\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "urllib3                     1.26.12\n",
      "wcwidth                     0.2.5\n",
      "yaml                        5.4.1\n",
      "zmq                         23.2.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.6.0\n",
      "jupyter_client      7.3.5\n",
      "jupyter_core        4.11.2\n",
      "notebook            6.4.12\n",
      "-----\n",
      "Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:41:22) [Clang 13.0.1 ]\n",
      "macOS-13.0.1-arm64-arm-64bit\n",
      "-----\n",
      "Session information updated at 2023-06-16 14:49\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# **Consensus network analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2fb39a9a",
   "metadata": {},
   "source": [
    "Here, we perform consensus co-expression network analysis using `hdWGCNA`. \n",
    "\n",
    "Consensus co-expression network analysis differs from the standard co-expression network analysis workflow by `constructing individual networks across distinct datasets`, and `then computing an integrated co-expression network`. \n",
    "\n",
    "This framework can be used to `identify networks that are conserved across a variety of biological conditions`, and it can also be used as a way to construct a unified network from any number of different datasets. \n",
    "\n",
    "Here, use this framework to identify consensus coexpression network of genes across celltypes (from the three studies) in AD.\n",
    "\n",
    "- [**Morabito et al. bioRxiv 2022**](https://www.biorxiv.org/content/10.1101/2022.09.22.509094v1)\n",
    "- [**Morabito & Miyoshi et al. Nature Genetics 2021**](https://www.nature.com/articles/s41588-021-00894-z)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "8b87d243",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(WGCNA)\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "    library(readxl)\n",
    "\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "\n",
    "\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c00d65ea",
   "metadata": {},
   "source": [
    "## **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d8504101",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_meta = True\n",
    "deg_method =  'DESeq2-Wald'\n",
    "test_names = ['late_vs_early', 'early_vs_no', 'late_vs_no', 'ad_vs_no']\n",
    "filter_genes = \"TRUE\"\n",
    "studies = ['allen_mtg', 'leng_sfg', 'leng_etc']\n",
    "subject_id = {'allen_mtg': 'individualID',\n",
    "             'leng_sfg': 'PatientID',\n",
    "             'leng_etc': 'PatientID'}      # for leng this is `PatientID` for mathys is 'Subject', and 'individualID' for allen\n",
    "\n",
    "gene_celltype_threshold = 0.05      # determines number of cells the gene must be expressed in \n",
    "covariates = ['None']               # list of covariates to be accounted for in regression.\n",
    "gene_selection = 'custom'           # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "celltypes = [\"Inhibitory\"]          #[\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\"]\n",
    "gene_selection = 'custom'       # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "                                # Posible values are \"custom\", \"fraction\", \"variable\"\n",
    "                                # If custom, a list of genes must be passed.\n",
    "\n",
    "gene_set_select = 'overlap'      # If gene_selection = 'custom'. This specifies how to obtain the list of\n",
    "                                  # genes to pass into `SetupForWGCNA`. # The posible values are \"diff_exp\", \"overlap\", \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1e74ee31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------\n",
      "Fetching ALLEN_MTG data\n",
      "----------------------------\n",
      "AnnData object with n_obs × n_vars = 41366 × 17950\n",
      "    obs: 'ID', 'name', 'versionNumber', 'contentType', 'dataFileSizeBytes', 'createdBy', 'createdOn', 'modifiedBy', 'modifiedOn', 'parentId', 'synapseURL', 'dataFileMD5Hex', 'assay_x', 'consortium', 'dataSubtype', 'dataType', 'fileFormat', 'grant', 'individualID_x', 'isModelSystem', 'isMultiSpecimen', 'isMultiome', 'libraryPrep', 'nucleicAcidSource_x', 'organ_x', 'platform', 'readLength', 'resourceType', 'runType', 'sex_x', 'species_x', 'specimenID', 'study', 'tissue_x', 'path', 'error', 'individualID_y', 'specimenIdSource', 'organ_y', 'tissue_y', 'BrodmannArea', 'sampleStatus', 'tissueWeight', 'tissueVolume', 'nucleicAcidSource_y', 'cellType', 'fastingState', 'isPostMortem', 'samplingAge', 'samplingAgeUnits', 'visitNumber', 'assay_y', 'individualID', 'dataset', 'individualIdSource', 'species_y', 'sex_y', 'race', 'ethnicity', 'yearsEducation', 'ageDeath', 'causeDeath', 'mannerDeath', 'apoe4Status', 'apoeGenotype', 'pmi', 'pH', 'brainWeight', 'diagnosis', 'diagnosisCriteria', 'CERAD', 'Braak', 'Cognitive status', 'Age of dementia diagnosis', 'Consensus clinical diagnosis', 'CASI score', 'Interval from last CASI', 'MMSE score', 'Interval from last MMSE', 'MoCA score', 'Interval from last MoCA', 'ADNC', 'Thal phase', 'Overall CAA Score', 'Lewy body disease pathology', 'Microinfarcts in screening sections', 'Atherosclerosis', 'Arteriolosclerosis', 'LATE-NC stage', 'pathology.group', 'sample_id', 'Class', 'Subclass', 'Supertype', 'barcodes', 'cell_labels', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'predictions'\n",
      "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "----------------------------\n",
      "Fetching LENG_SFG data\n",
      "----------------------------\n",
      "AnnData object with n_obs × n_vars = 43607 × 16755\n",
      "    obs: 'SampleID', 'PatientID', 'BrainRegion', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.clusters', 'clusterAssignment', 'clusterCellType', 'cell_type', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'predictions', 'study', 'pathology.group'\n",
      "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "----------------------------\n",
      "Fetching LENG_ETC data\n",
      "----------------------------\n",
      "AnnData object with n_obs × n_vars = 29167 × 16585\n",
      "    obs: 'SampleID', 'PatientID', 'BrainRegion', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.clusters', 'clusterAssignment', 'clusterCellType', 'cell_type', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'doublet_score', 'predicted_doublet', 'louvain_0.5', 'louvain_1.0', 'predictions', 'study', 'pathology.group'\n",
      "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
      "----------------------------\n",
      "Concatenated data\n",
      "----------------------------\n",
      "AnnData object with n_obs × n_vars = 114140 × 18168\n",
      "    obs: 'ID', 'name', 'versionNumber', 'contentType', 'dataFileSizeBytes', 'createdBy', 'createdOn', 'modifiedBy', 'modifiedOn', 'parentId', 'synapseURL', 'dataFileMD5Hex', 'assay_x', 'consortium', 'dataSubtype', 'dataType', 'fileFormat', 'grant', 'individualID_x', 'isModelSystem', 'isMultiSpecimen', 'isMultiome', 'libraryPrep', 'nucleicAcidSource_x', 'organ_x', 'platform', 'readLength', 'resourceType', 'runType', 'sex_x', 'species_x', 'specimenID', 'study', 'tissue_x', 'path', 'error', 'individualID_y', 'specimenIdSource', 'organ_y', 'tissue_y', 'BrodmannArea', 'sampleStatus', 'tissueWeight', 'tissueVolume', 'nucleicAcidSource_y', 'cellType', 'fastingState', 'isPostMortem', 'samplingAge', 'samplingAgeUnits', 'visitNumber', 'assay_y', 'individualID', 'dataset', 'individualIdSource', 'species_y', 'sex_y', 'race', 'ethnicity', 'yearsEducation', 'ageDeath', 'causeDeath', 'mannerDeath', 'apoe4Status', 'apoeGenotype', 'pmi', 'pH', 'brainWeight', 'diagnosis', 'diagnosisCriteria', 'CERAD', 'Braak', 'Cognitive status', 'Age of dementia diagnosis', 'Consensus clinical diagnosis', 'CASI score', 'Interval from last CASI', 'MMSE score', 'Interval from last MMSE', 'MoCA score', 'Interval from last MoCA', 'ADNC', 'Thal phase', 'Overall CAA Score', 'Lewy body disease pathology', 'Microinfarcts in screening sections', 'Atherosclerosis', 'Arteriolosclerosis', 'LATE-NC stage', 'pathology.group', 'sample_id', 'Class', 'Subclass', 'Supertype', 'barcodes', 'cell_labels', 'n_genes', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'total_counts_mt', 'pct_counts_mt', 'louvain_0.5', 'louvain_1.0', 'cell_type', 'predictions', 'SampleID', 'PatientID', 'BrainRegion', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.clusters', 'clusterAssignment', 'clusterCellType', 'doublet_score', 'predicted_doublet'\n"
     ]
    }
   ],
   "source": [
    "adatas = []\n",
    "for save_prefix in studies:\n",
    "    \n",
    "    print('----------------------------')\n",
    "    print(f'Fetching {save_prefix.upper()} data')\n",
    "    print('----------------------------')\n",
    "\n",
    "    adata = sc.read_h5ad(f'../data/processed/{save_prefix}/{save_prefix}_mapped_anndata.h5ad')\n",
    "    adata.obs['study'] = save_prefix\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.X = adata.layers['counts'].copy()\n",
    "\n",
    "    metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' \n",
    "    meta = pd.read_csv(metadata, encoding_errors='ignore')  \n",
    "\n",
    "    # map the pathology group to the subject id in .obs\n",
    "\n",
    "    adata.obs['pathology.group'] = adata.obs[subject_id[save_prefix]].map(dict(zip(meta[subject_id[save_prefix]].astype(str),meta['pathology.group'])))\n",
    "    \n",
    "    adata = adata[adata.obs['pathology.group'].isin(['early', 'late'])]\n",
    "\n",
    "    del adata.obsm, adata.layers, \\\n",
    "        adata.varm, adata.uns, \\\n",
    "        adata.obsp\n",
    "    \n",
    "    adatas.append(adata)\n",
    "    \n",
    "    print(adata)\n",
    "\n",
    "print('----------------------------')\n",
    "print(f'Concatenated data')\n",
    "print('----------------------------')\n",
    "\n",
    "adata_annot = anndata.concat(adatas, join='outer')\n",
    "adata_annot.obs = adata_annot.obs.replace(np.nan, None)\n",
    "print(adata_annot)\n",
    "\n",
    "del adatas, adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86ef226f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of anndata object into Rpy2 object \n",
    "subject_id =  robjects.ListVector(\n",
    "                        {\n",
    "                            save_prefix: subject_id[save_prefix]\n",
    "                     \n",
    "                            for save_prefix in studies\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "839f795a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 18168 114140 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(18168): A1BG A1CF ... ZZEF1 ZZZ3\n",
      "rowData names(0):\n",
      "colnames(114140): AAAGGTAGTACGGCAA AACCAACGTGGATCGA ...\n",
      "  EC10_TTTGTCATCTATCGCC EC10_TTTGTCATCTCTGCTG\n",
      "colData names(119): ID name ... doublet_score predicted_doublet\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "[1] \"loaded data into memory for recursive use\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_annot -i subject_id -i gene_celltype_threshold -i celltypes -i test_names -i studies -i gene_selection\n",
    "\n",
    "print(adata_annot)\n",
    "print('loaded data into memory for recursive use')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "74c5d20a",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "646870b9",
   "metadata": {},
   "source": [
    "## **Consensus network analysis of celltypes in AD**\n",
    "\n",
    "Here, we perform consensus network analysis between all three datastes in AD. We will also follow the “standard” workflow (not consensus), and compare the resulting gene module assignments.\n",
    "\n",
    "First, we setup the seurat object for hdWGCNA and we construct metacells. This part of the workflow is the same as the standard hdWGCNA workflow, but we must include the `relevant metadata for consensus network analysis when running MetacellsByGroups`. \n",
    "\n",
    "The loaded Seurat object has a metadata column called `pathology.group`, which contains a `early` or `late` AD identifiers depending on the pathological stage of the donor where a given cell originated from, so we must include  in our group.by list within MetacellsByGroups in order to run consensus network analysis.\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8a3955a2",
   "metadata": {},
   "source": [
    "## **Set up Seurat object for WGCNA**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3e7607cd",
   "metadata": {},
   "source": [
    "For our specific case, we will either select genes that are expressed in at least 5% of cells within each cell type or specify a list of custom genes. Additionally, we will rename our hdWGCNA experiment to enhance clarity.\n",
    "\n",
    "Here, the `gene_selection` parameter is set to `custom`, we employ the genes present in all the overlapping pathways in a cell type-specific and study-specific manner.\n",
    "\n",
    "To accomplish this, we load the list of overlapping pathways that was previously saved in `../results/{test_name}/{test_name}_pathway_overlap.csv` from the `pathway_meta_analysis.ipynb` notebook. Additionally, we load the genes from the `.gmt` file saved in `../results/gsva/{study}/{cur_celltype}/gs.gmt` from the `pathway_analysis.ipynb` notebook."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c4ddf8bf",
   "metadata": {},
   "source": [
    "##### **Gene Set Selection**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "83560708",
   "metadata": {},
   "source": [
    "To obtain the gene sets required for the `SetupForWGCNA` function, we consider the `gene_set_select` variable when `gene_select` is set to `custom`. The `gene_set_select` variable determines how the `gene_list` passed into the `SetupForWGCNA` function will be obtained. \n",
    "\n",
    "There are three options available:\n",
    "\n",
    "- `diff_exp`: This option utilizes the genes present in the differentially expressed pathways, previously computed and saved to the `../results/{test_name}/{save_prefix}_differentially_expressed_pathways.csv` file.\n",
    "\n",
    "- `overlap`: Here, the genes used are those expressed in pathways that overlap across different studies. The pathway overlap information is retrieved from the `../results/{test_name}/{test_name}_pathway_overlap.csv` file.\n",
    "\n",
    "- `all`: With this option, all genes in the pathways stored in the `../results/gsva/{save_prefix}/{celltype}/gs.gmt` file are included.\n",
    "\n",
    "By specifying the appropriate `gene_select` and `gene_set_select` values, you can customize the gene selection process for your hdWGCNA experiment.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "703041ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o geneSets -i gene_set_select\n",
    "\n",
    "# iterate over celltypes and obtain \n",
    "geneSets <- list()\n",
    "for(test_name in test_names){\n",
    "\n",
    "  geneSets[[test_name]] <- list()\n",
    "  for (cur_celltype in celltypes){\n",
    "\n",
    "    geneSets[[test_name]][[cur_celltype]] <- list()\n",
    "    for (study in studies){\n",
    "        # get gene sets\n",
    "        geneSets[[test_name]][[cur_celltype]][[study]] <- getGmt(file.path('../results/gsva/', study, '/', cur_celltype, 'gs.gmt'))\n",
    "        geneSets[[test_name]][[cur_celltype]][[study]] <- geneIds(geneSets[[test_name]][[cur_celltype]][[study]])\n",
    "\n",
    "        if (gene_set_select=='overlap'){\n",
    "          pathways <- read.csv(paste0('../results/',test_name, '/', test_name, '_pathway_overlap.csv'))\n",
    "          pathways <- pathways$pathway\n",
    "        }else if (gene_set_select=='diff_exp'){\n",
    "          pathways <- read.csv(paste0('../results/', test_name, '/', study, '_differentially_expressed_pathways.csv'))\n",
    "          pathways <- subset(pathways, P.Value < 0.05 & celltype == cur_celltype)$pathway\n",
    "        }else if (gene_set_select=='all'){\n",
    "          pathways <- names(geneSets[[test_name]][[cur_celltype]][[study]])\n",
    "        }\n",
    "        geneSets[[test_name]][[cur_celltype]][[study]] <-  geneSets[[test_name]][[cur_celltype]][[study]][names(geneSets[[test_name]][[cur_celltype]][[study]]) %in% pathways]\n",
    "        geneSets[[test_name]][[cur_celltype]][[study]] <- unlist(geneSets[[test_name]][[cur_celltype]][[study]], recursive = FALSE)\n",
    "    }\n",
    "    geneSets[[test_name]][[cur_celltype]] <- unique(unlist(geneSets[[test_name]][[cur_celltype]], recursive = FALSE))\n",
    "  }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f2d4d1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of geneSets object into Rpy2 object \n",
    "\n",
    "geneSets =  robjects.ListVector(\n",
    "                        {\n",
    "                            test_name: robjects.ListVector(\n",
    "                                    {cell_type: geneSets[test_name][cell_type]\n",
    "                                                 \n",
    "                                    for cell_type in celltypes\n",
    "                                    }\n",
    "                                )\n",
    "                            for test_name in test_names\n",
    "                        }\n",
    "\n",
    "                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "845fa1c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An object of class Seurat \n",
      "18168 features across 114140 samples within 1 assay \n",
      "Active assay: originalexp (18168 features, 0 variable features)\n",
      "[1] \"Creating hdWGNA Experiment for INHIBITORY\"\n",
      "An object of class Seurat \n",
      "18168 features across 18139 samples within 1 assay \n",
      "Active assay: originalexp (18168 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"4547 WGCNA Genes\"\n"
     ]
    }
   ],
   "source": [
    "%%R -o seurat_obj -i geneSets\n",
    "\n",
    "seurat_obj <- as.Seurat(adata_annot, counts = \"X\", data = \"X\")\n",
    "\n",
    "print(seurat_obj)\n",
    "\n",
    "# Perform dimensionality reduction and plot\n",
    "\n",
    "seurat_obj <- FindVariableFeatures(seurat_obj)\n",
    "seurat_obj <- ScaleData(seurat_obj)\n",
    "seurat_obj <- RunPCA(seurat_obj)\n",
    "seurat_obj <- RunHarmony(seurat_obj, group.by.vars = \"study\")\n",
    "seurat_obj <- RunUMAP(seurat_obj, reduction='harmony', n.neighbors=15, dims=1:30, min.dist=0.1)\n",
    "\n",
    "p <- DimPlot(seurat_obj, group.by = \"cell_type\", label = TRUE) +\n",
    "     umap_theme() + ggtitle(\"all_studies\") + NoLegend()\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/UMAP/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "# Save plot to PDF\n",
    "pdf(file = paste0(fig_dir, \"Integrated_umap_all_studies.pdf\"), width = 4, height = 4, useDingbats = FALSE)\n",
    "print(p)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "# create a hdWGCNA experiment for each celltype\n",
    "seurat_dat <- seurat_obj\n",
    "seurat_obj <- list()\n",
    "\n",
    "for (cur_cell_type in celltypes){\n",
    "\n",
    "  seurat_obj[[cur_cell_type]] <- subset(seurat_dat, cell_type == cur_cell_type)\n",
    "\n",
    "  print(paste0('Creating hdWGNA Experiment for ', toupper(cur_cell_type)))\n",
    "\n",
    "  if (gene_selection == 'custom'){\n",
    "    seurat_obj[[cur_cell_type]] <- SetupForWGCNA(\n",
    "      seurat_obj[[cur_cell_type]],\n",
    "      gene_select = \"custom\",                                 # the gene selection approach\n",
    "      gene_list = as.vector(geneSets[['ad_vs_no']][[cur_cell_type]]),    # list of genes to be included\n",
    "      group.by = 'study',                                 # grouping parameter\n",
    "      wgcna_name = paste0(toupper(substr(cur_cell_type, 1, 3)), '_consensus')       # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  } else {\n",
    "    seurat_obj[[cur_cell_type]] <- SetupForWGCNA(\n",
    "      seurat_obj[[cur_cell_type]],\n",
    "      gene_select = \"fraction\",                                                 # the gene selection approach\n",
    "      fraction = 0.05,                                                          # fraction of cells for gene inclusion\n",
    "      group.by = 'study',                                                       # grouping parameter\n",
    "      wgcna_name = paste0(toupper(substr(cur_cell_type, 1, 3)), '_consensus')       # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  }\n",
    "  print(seurat_obj[[cur_cell_type]])\n",
    "  print(paste0(length(GetWGCNAGenes(seurat_obj[[cur_cell_type]])), \" WGCNA Genes\"))\n",
    "\n",
    "}\n",
    "\n",
    "rm(seurat_dat)\n",
    "rm(adata_annot)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9e89e217",
   "metadata": {},
   "source": [
    "Send list to R interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ff771eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "\n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ccd0e590",
   "metadata": {},
   "source": [
    "## **Construct metacells**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "07813598",
   "metadata": {},
   "source": [
    "After setting up our Seurat object, we construct metacells from the single-cell dataset.\n",
    "\n",
    "First, we setup the seurat object for hdWGCNA and we construct metacells. This part of the workflow is the same as the standard hdWGCNA workflow, but we must include the `relevant metadata for consensus network analysis when running MetacellsByGroups`. \n",
    "\n",
    "The loaded Seurat object has a metadata column called `pathology.group`, which contains a `early` or `late` AD identifiers depending on the pathological stage of the donor where a given cell originated from, so we must include  in our group.by list within MetacellsByGroups in order to run consensus network analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e92860fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for INHIBITORY\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# construct metacells in each group for each wgnca experiment\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "  print(paste0('Constructing MetaCells in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  seurat_obj[[cell_type]] <- MetacellsByGroups(\n",
    "      \n",
    "    seurat_obj = seurat_obj[[cell_type]],\n",
    "    group.by = c(\"pathology.group\", \"study\", \"cell_type\"), # specify the columns in seurat_obj@meta.data to group by\n",
    "    reduction = 'harmony',  # select the dimensionality reduction to perform KNN on\n",
    "    k = 25, # nearest-neighbors parameter\n",
    "    max_shared = 10, # maximum number of shared cells between two metacells\n",
    "    ident.group = 'cell_type', # set the Idents of the metacell seurat object\n",
    "    wgcna_name = paste0(toupper(substr(cell_type, 1, 3)), '_consensus')        # the name of the hdWGCNA experiment\n",
    "    )\n",
    "\n",
    "    # normalize metacell expression matrix:\n",
    "  seurat_obj[[cell_type]] <- NormalizeMetacells(seurat_obj[[cell_type]],  \n",
    "                            wgcna_name = paste0(toupper(substr(cell_type, 1, 3)), '_consensus'))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "db92cb94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fac55be8",
   "metadata": {},
   "source": [
    "## **Co-expression network analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "bb462c2f",
   "metadata": {},
   "source": [
    "Next we have to set up the expression dataset for consensus network analysis. In the standard workflow, we use the function `SetDatExpr` to set up the expression matrix that will be used for network analyis. Instead of `SetDatExpr`, here we use `SetMultiExpr` to set up a separate expression matrix for each `study`. This allows us to perform network analysis individually on these expression matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6b1c6066",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for INHIBITORY\"\n",
      "  ..Excluding 164 genes from the calculation due to too many missing samples or zero variance.\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# set up a gene expression matrix list for each study\n",
    "# construct metacells in each group for each wgnca experiment\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "  print(paste0('Setting up expression matrix in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  seurat_obj[[cell_type]] <- SetMultiExpr(\n",
    "      seurat_obj[[cell_type]],\n",
    "      group_name = cell_type, # the name of the group of interest in the group.by column\n",
    "      group.by = 'cell_type', # the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups  \n",
    "      assay = 'originalexp', # using RNA assay\n",
    "      slot = 'data', # using normalized data\n",
    "      wgcna_name = paste0(toupper(substr(cell_type, 1, 3)), '_consensus'), # the name of the hdWGCNA experiment\n",
    "      multi.group.by = \"study\",\n",
    "      multi_groups = NULL # this parameter can be used to select a subset of groups in the multi.group.by column\n",
    "    )\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3c40b7ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a70cc5c4",
   "metadata": {},
   "source": [
    "### **Select soft-power threshold**\n",
    "\n",
    "Now that we have expression matrices for all studies, we identify an appropriate soft power threshold for each of these matrices separately. Instead of the `TestSoftPowers` function which we use in the standard hdWGCNA workflow, `we use TestSoftPowersConsensus`, which will perform the test for each of the expression matrices. When we plot the results with  `PlotSoftPowers`, we get a nested list of plots for each dataset, and we can assemble the plots using patchwork.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e63133c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"allen_mtg\"\n",
      "pickSoftThreshold: will use block size 4383.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 4383 of 4383\n",
      "   Power SFT.R.sq slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1  0.17600 26.10          0.881 2.21e+03  2.21e+03 2300.00\n",
      "2      2  0.09080  9.87          0.952 1.13e+03  1.13e+03 1240.00\n",
      "3      3  0.00128 -0.71          0.860 5.82e+02  5.78e+02  697.00\n",
      "4      4  0.30300 -7.31          0.766 3.04e+02  2.98e+02  413.00\n",
      "5      5  0.78700 -9.32          0.926 1.60e+02  1.55e+02  256.00\n",
      "6      6  0.92200 -7.85          0.974 8.53e+01  8.10e+01  165.00\n",
      "7      7  0.93800 -6.22          0.972 4.62e+01  4.26e+01  110.00\n",
      "8      8  0.94600 -4.92          0.970 2.54e+01  2.25e+01   76.60\n",
      "9      9  0.96400 -3.97          0.979 1.42e+01  1.19e+01   55.10\n",
      "10    10  0.95200 -3.40          0.968 8.13e+00  6.38e+00   40.80\n",
      "11    12  0.97500 -2.56          0.987 2.87e+00  1.86e+00   24.50\n",
      "12    14  0.97300 -2.14          0.992 1.14e+00  5.51e-01   16.90\n",
      "13    16  0.93700 -1.95          0.981 5.22e-01  1.68e-01   12.40\n",
      "14    18  0.93500 -1.76          0.985 2.70e-01  5.20e-02    9.48\n",
      "15    20  0.93300 -1.65          0.981 1.56e-01  1.65e-02    7.47\n",
      "16    22  0.38100 -2.18          0.351 9.71e-02  5.29e-03    6.03\n",
      "17    24  0.38700 -2.08          0.362 6.43e-02  1.75e-03    4.94\n",
      "18    26  0.34400 -2.43          0.264 4.45e-02  5.89e-04    4.12\n",
      "19    28  0.40200 -1.99          0.368 3.19e-02  2.01e-04    3.47\n",
      "20    30  0.40500 -1.94          0.374 2.35e-02  7.00e-05    2.95\n",
      "[1] \"leng_etc\"\n",
      "pickSoftThreshold: will use block size 4383.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 4383 of 4383\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.   max.k.\n",
      "1      1   0.1570  38.20          0.941 2.20e+03  2.20e+03 2270.000\n",
      "2      2   0.0167   6.37          0.988 1.11e+03  1.11e+03 1190.000\n",
      "3      3   0.2360 -16.60          0.951 5.67e+02  5.66e+02  644.000\n",
      "4      4   0.5690 -16.40          0.884 2.91e+02  2.90e+02  356.000\n",
      "5      5   0.7290 -13.30          0.856 1.50e+02  1.49e+02  201.000\n",
      "6      6   0.8600 -11.40          0.895 7.83e+01  7.70e+01  118.000\n",
      "7      7   0.9340  -9.44          0.934 4.11e+01  4.01e+01   71.400\n",
      "8      8   0.9550  -7.81          0.948 2.18e+01  2.09e+01   44.700\n",
      "9      9   0.9870  -6.13          0.983 1.16e+01  1.10e+01   28.800\n",
      "10    10   0.9880  -5.00          0.986 6.28e+00  5.82e+00   19.200\n",
      "11    12   0.9860  -3.64          0.984 1.90e+00  1.65e+00   10.000\n",
      "12    14   0.9750  -2.93          0.971 6.07e-01  4.78e-01    6.090\n",
      "13    16   0.9800  -2.42          0.985 2.10e-01  1.41e-01    4.010\n",
      "14    18   0.9700  -2.07          0.984 8.09e-02  4.26e-02    2.780\n",
      "15    20   0.9680  -1.82          0.990 3.52e-02  1.31e-02    2.010\n",
      "16    22   0.4170  -2.24          0.381 1.73e-02  4.12e-03    1.490\n",
      "17    24   0.4170  -2.03          0.383 9.54e-03  1.31e-03    1.120\n",
      "18    26   0.4200  -1.90          0.392 5.72e-03  4.28e-04    0.863\n",
      "19    28   0.3460  -2.10          0.221 3.65e-03  1.42e-04    0.671\n",
      "20    30   0.3860  -2.12          0.374 2.43e-03  4.78e-05    0.528\n",
      "[1] \"leng_sfg\"\n",
      "pickSoftThreshold: will use block size 4383.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 4383 of 4383\n",
      "   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1    0.078  21.20          0.955 2.20e+03  2.20e+03 2280.00\n",
      "2      2    0.038  -8.10          0.994 1.12e+03  1.12e+03 1210.00\n",
      "3      3    0.113  -8.73          0.917 5.70e+02  5.68e+02  653.00\n",
      "4      4    0.564 -12.90          0.903 2.94e+02  2.91e+02  371.00\n",
      "5      5    0.847 -12.10          0.945 1.52e+02  1.50e+02  221.00\n",
      "6      6    0.922  -9.50          0.971 7.98e+01  7.76e+01  136.00\n",
      "7      7    0.966  -7.50          0.984 4.22e+01  4.04e+01   87.60\n",
      "8      8    0.983  -6.05          0.989 2.26e+01  2.11e+01   58.40\n",
      "9      9    0.978  -5.03          0.979 1.22e+01  1.11e+01   40.40\n",
      "10    10    0.983  -4.18          0.982 6.68e+00  5.87e+00   29.00\n",
      "11    12    0.940  -3.16          0.925 2.10e+00  1.66e+00   16.30\n",
      "12    14    0.959  -2.46          0.950 7.21e-01  4.82e-01   10.10\n",
      "13    16    0.966  -2.04          0.963 2.76e-01  1.42e-01    6.70\n",
      "14    18    0.941  -1.79          0.937 1.21e-01  4.29e-02    4.79\n",
      "15    20    0.964  -1.58          0.968 6.06e-02  1.31e-02    3.65\n",
      "16    22    0.965  -1.48          0.981 3.42e-02  4.09e-03    2.85\n",
      "17    24    0.951  -1.41          0.964 2.11e-02  1.30e-03    2.28\n",
      "18    26    0.959  -1.36          0.981 1.39e-02  4.18e-04    1.85\n",
      "19    28    0.957  -1.33          0.967 9.63e-03  1.37e-04    1.52\n",
      "20    30    0.338  -1.58          0.149 6.89e-03  4.54e-05    1.27\n",
      "[1] \"here\"\n",
      "[1] 1\n",
      "[1] 6\n",
      "# A tibble: 6 × 7\n",
      "  Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.\n",
      "  <dbl>    <dbl>  <dbl>          <dbl>   <dbl>     <dbl>  <dbl>\n",
      "1     1  0.176   26.1            0.881  2213.     2214.   2303.\n",
      "2     2  0.0908   9.87           0.952  1129.     1128.   1242.\n",
      "3     3  0.00128 -0.710          0.860   582.      578.    697.\n",
      "4     4  0.303   -7.31           0.766   304.      298.    413.\n",
      "5     5  0.787   -9.32           0.926   160.      155.    256.\n",
      "6     6  0.922   -7.85           0.974    85.3      81.0   165.\n",
      "[1] 2\n",
      "[1] 6\n",
      "# A tibble: 6 × 7\n",
      "  Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.\n",
      "  <dbl>    <dbl>  <dbl>          <dbl>   <dbl>     <dbl>  <dbl>\n",
      "1     1   0.157   38.2           0.941  2200.     2201.   2266.\n",
      "2     2   0.0167   6.37          0.988  1113.     1112.   1191.\n",
      "3     3   0.236  -16.6           0.951   567.      566.    644.\n",
      "4     4   0.569  -16.4           0.884   291.      290.    356.\n",
      "5     5   0.729  -13.3           0.856   150.      149.    201.\n",
      "6     6   0.860  -11.4           0.895    78.3      77.0   118.\n",
      "[1] 3\n",
      "[1] 5\n",
      "# A tibble: 6 × 7\n",
      "  Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.\n",
      "  <dbl>    <dbl>  <dbl>          <dbl>   <dbl>     <dbl>  <dbl>\n",
      "1     1   0.0780  21.2           0.955  2202.     2203.   2280.\n",
      "2     2   0.0380  -8.10          0.994  1116.     1115.   1209.\n",
      "3     3   0.113   -8.73          0.917   570.      568.    653.\n",
      "4     4   0.564  -12.9           0.903   294.      291.    371.\n",
      "5     5   0.847  -12.1           0.945   152.      150.    221.\n",
      "6     6   0.922   -9.50          0.971    79.8      77.6   136.\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/SoftPower/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "  \n",
    "  print(paste0('Estinmating Soft-Power Threshold in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # Test different soft powers:\n",
    "  seurat_obj[[cell_type]] <- TestSoftPowersConsensus(\n",
    "    seurat_obj[[cell_type]],\n",
    "    group.by = 'cell_type',\n",
    "    group_name = cell_type,\n",
    "  )\n",
    "\n",
    "  # plot the results:\n",
    "  plot_list <- PlotSoftPowers(seurat_obj[[cell_type]], \n",
    "                              # wgcna_name = paste0(toupper(substr(cell_type, 1, 3)), '_consensus') # the name of the hdWGCNA experiment\n",
    "                              )\n",
    "\n",
    "  # get just the scale-free topology fit plot for each group\n",
    "  consensus_groups <- unique(seurat_obj[[cell_type]]$study)\n",
    "\n",
    "  p_list <- lapply(1:length(consensus_groups), function(i){\n",
    "    cur_group <- consensus_groups[[i]]\n",
    "    plot_list[[i]][[1]] + ggtitle(paste0('Study: ', cur_group)) + theme(plot.title=element_text(hjust=0.5))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_consensus_SoftPower.pdf'), width=10, height=8)\n",
    "  print(wrap_plots(p_list, ncol=length(studies)))\n",
    "  dev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f5b7f474",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2ad0b902",
   "metadata": {},
   "source": [
    "The general guidance for WGCNA and hdWGCNA is to `pick the lowest soft power threshold that has a Scale Free Topology Model Fit greater than or equal to 0.8`. Later on, the `ConstructNetwork `will automatically select the soft power threshold if we do not provide one.\n",
    "\n",
    "The output table from the parameter sweep is stored in the hdWGCNA experiment and can be accessed using the GetPowerTable function for further inspection:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "62c54e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# power_table <- GetPowerTable(seurat_obj)\n",
    "# head(power_table)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7c6251e4",
   "metadata": {},
   "source": [
    "### **Construct co-expression network**\n",
    "\n",
    "Now we construct the co-expression network and identify gene modules using the `ConstructNetwork` function, making sure to specify `consensus=TRUE`. \n",
    "\n",
    "Indicating `consensus=TRUE tells hdWGCNA to construct a separate network for each expression matrix`, followed by integrating the networks and identifying gene modules. Depending on the results of `TestSoftPowersConsensus`, we can supply a different soft power threshold for each dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "494fe710",
   "metadata": {},
   "outputs": [],
   "source": [
    "soft_powers = input('Look at the file saved to \"../results/hdWGCNA/SoftPower/Consensus/Consensus_SoftPower\" \\\n",
    "                    and input the softpowers for each study starting from left to right, separated by comma with no space: ').strip(\" \").split(',')\n",
    "\n",
    "soft_powers = [int(i) for i in soft_powers]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "14831950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for INHIBITORY\"\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Inhibitory\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -i soft_powers\n",
    "\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "  \n",
    "  print(paste0('Constructing co-expression network in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  tryCatch({\n",
    "    # construct co-expression network:\n",
    "    seurat_obj[[cell_type]] <- ConstructNetwork(\n",
    "      seurat_obj[[cell_type]], \n",
    "      soft_power = as.vector(soft_powers), # Set to NULL so that the ConstructNetwork function obtains it automatically (Scale Free Topology Model Fit >= 0.8)\n",
    "      setDatExpr = FALSE,\n",
    "      overwrite_tom = TRUE,\n",
    "      consensus = TRUE,\n",
    "      tom_name = paste0(toupper(substr(cell_type, 1, 3)), '_consensus') # name of the topological overlap matrix written to disk set as the name of the hdWGCNA experiment \n",
    "    )\n",
    "  }, error = function(e) {\n",
    "    print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "    print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "    # Code to handle the error condition\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "  },  \n",
    "  message = function(m) {\n",
    "    print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "    print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "    # Code to handle the error condition\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "  }) \n",
    "  \n",
    "}\n",
    "\n",
    "print(names(seurat_obj))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b8e77379",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "49676183",
   "metadata": {},
   "source": [
    "hdWGCNA also includes a function PlotDendrogram to visualize the WGCNA dendrogram, a common visualization to show the different co-expression modules resulting from the network analysis. Each leaf on the dendrogram represents a single gene, and the color at the bottom indicates the co-expression module assignment.\n",
    "\n",
    "\n",
    "`Importantly, the “grey” module consists of genes that were not grouped into any co-expression module. The grey module should be ignored for all downstream analysis and interpretation.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8da7dcca",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/dendrogram/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "tryCatch({\n",
    "    for (cell_type in names(seurat_obj)){\n",
    "        # print(paste0('Constructing co-expression networkd in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "        pdf(paste0(fig_dir, cell_type, \"_consensus_dendrogram.pdf\"), height=2, width=4)\n",
    "        PlotDendrogram(seurat_obj[[cell_type]], main='hdWGCNA Dendrogram')\n",
    "        dev.off()\n",
    "    }\n",
    "}, error = function(e){\n",
    "    NULL\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ed9baba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a5dfb822",
   "metadata": {},
   "source": [
    "### **Optional: Compare Consensus network results to the standard hdWGCNA workflow**\n",
    "\n",
    "### **Standard hdWGCNA workflow**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fbea29f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Running standard hdWGCNA workflow in hdWGCNA Experiment for INHIBITORY\"\n",
      "pickSoftThreshold: will use block size 4547.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 4547 of 4547\n",
      "   Power SFT.R.sq slope truncated.R.sq  mean.k. median.k. max.k.\n",
      "1      1    0.308 26.20          0.963 2310.000  2.31e+03 2410.0\n",
      "2      2    0.121 -7.39          0.858 1190.000  1.19e+03 1380.0\n",
      "3      3    0.640 -9.25          0.775  626.000  6.17e+02  839.0\n",
      "4      4    0.872 -6.98          0.886  334.000  3.24e+02  543.0\n",
      "5      5    0.946 -5.37          0.936  182.000  1.71e+02  370.0\n",
      "6      6    0.961 -4.16          0.949  101.000  9.14e+01  265.0\n",
      "7      7    0.953 -3.36          0.941   57.400  4.92e+01  198.0\n",
      "8      8    0.949 -2.81          0.936   33.600  2.66e+01  153.0\n",
      "9      9    0.938 -2.42          0.923   20.300  1.46e+01  122.0\n",
      "10    10    0.930 -2.14          0.911   12.700  8.01e+00   99.9\n",
      "11    12    0.885 -1.81          0.855    5.620  2.47e+00   71.4\n",
      "12    14    0.868 -1.57          0.841    2.910  7.86e-01   54.3\n",
      "13    16    0.874 -1.42          0.850    1.740  2.57e-01   43.1\n",
      "14    18    0.851 -1.34          0.838    1.150  8.67e-02   35.2\n",
      "15    20    0.834 -1.29          0.836    0.811  2.99e-02   29.4\n",
      "16    22    0.845 -1.23          0.855    0.603  1.05e-02   24.9\n",
      "17    24    0.855 -1.19          0.874    0.463  3.74e-03   21.3\n",
      "18    26    0.856 -1.18          0.883    0.364  1.37e-03   18.5\n",
      "19    28    0.838 -1.18          0.863    0.291  5.05e-04   16.1\n",
      "20    30    0.857 -1.15          0.892    0.237  1.91e-04   14.1\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 4\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "$mar\n",
      "[1] 1 5 0 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj \n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/dendrogram/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "  \n",
    "  print(paste0('Running standard hdWGCNA workflow in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  # setup new hdWGCNA experiment\n",
    "  seurat_obj[[cell_type]] <- SetupForWGCNA(\n",
    "    seurat_obj[[cell_type]],\n",
    "    gene_select = \"custom\",\n",
    "    gene_list = as.vector(geneSets[['ad_vs_no']][[cell_type]]),\n",
    "    wgcna_name = paste0(toupper(substr(cell_type, 1, 3)), '_standard'),\n",
    "    metacell_location = paste0(toupper(substr(cell_type, 1, 3)), '_consensus') # use the same metacells\n",
    "  )\n",
    "  seurat_obj[[cell_type]] <- NormalizeMetacells(seurat_obj[[cell_type]])\n",
    "\n",
    "  # construct network\n",
    "  seurat_obj[[cell_type]] <- SetDatExpr(seurat_obj[[cell_type]], group_name = cell_type, group.by = \"cell_type\")\n",
    "  seurat_obj[[cell_type]] <- TestSoftPowers(seurat_obj[[cell_type]])\n",
    "  seurat_obj[[cell_type]] <- ConstructNetwork(seurat_obj[[cell_type]], soft_power=NULL, overwrite_tom = TRUE,\n",
    "                              tom_name = paste0(toupper(substr(cell_type, 1, 3)), '_standard'))\n",
    "\n",
    "  # plot the dendrogram\n",
    "  pdf(paste0(fig_dir, cell_type, \"_standard_dendrogram.pdf\"), height=2, width=4)\n",
    "  print(PlotDendrogram(seurat_obj[[cell_type]], main='ASC standard Dendrogram'))\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f51be4d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a1171d43",
   "metadata": {},
   "source": [
    "We can plot the gene module assignments from the standard workflow under those from the consensus analysis in the same dendrogram plot as a rough comparison.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "bd0a74f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Comparing module assignments in hdWGCNA Experiment for INHIBITORY\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj \n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/dendrogram/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "  \n",
    "  print(paste0('Comparing module assignments in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  # get both sets of modules\n",
    "  modules <- GetModules(seurat_obj[[cell_type]], wgcna_name=paste0(toupper(substr(cell_type, 1, 3)), '_standard'))\n",
    "  consensus_modules <- GetModules(seurat_obj[[cell_type]], wgcna_name=paste0(toupper(substr(cell_type, 1, 3)), '_consensus'))\n",
    "\n",
    "  # get consensus dendrogram\n",
    "  net <- GetNetworkData(seurat_obj[[cell_type]], wgcna_name=paste0(toupper(substr(cell_type, 1, 3)), '_consensus'))\n",
    "  dendro <- net$dendrograms[[1]]\n",
    "\n",
    "  # get the gene and module color for consensus\n",
    "  consensus_genes <- consensus_modules$gene_name\n",
    "  consensus_colors <- consensus_modules$color\n",
    "  names(consensus_colors) <- consensus_genes\n",
    "\n",
    "  # get the gene and module color for standard\n",
    "  genes <- modules$gene_name\n",
    "  colors <- modules$color\n",
    "  names(colors) <- genes\n",
    "\n",
    "  # re-order the genes to match the consensus genes\n",
    "  colors <- colors[consensus_genes]\n",
    "\n",
    "  # set up dataframe for plotting\n",
    "  color_df <- data.frame(\n",
    "    consensus = consensus_colors,\n",
    "    standard = colors\n",
    "  )\n",
    "\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_consensus_&_standard_dendrogram.pdf\"), height=2, width=4)\n",
    "  # plot dendrogram using WGCNA function\n",
    "  WGCNA::plotDendroAndColors(\n",
    "    net$dendrograms[[1]],\n",
    "    color_df,\n",
    "    groupLabels=colnames(color_df),\n",
    "    dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05,\n",
    "    main = \"AD consensus dendrogram\",\n",
    "  )\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f0831855",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0c1e4545",
   "metadata": {},
   "source": [
    "It is clear to see that much of the co-expression structure is missing when using the standard workflow when the dataset contains vast differences coming from the individual studies. We can take the consensus network and perform any of the downstream hdWGCNA analysis steps.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5e1c23ce",
   "metadata": {},
   "source": [
    "## **Module Eigengenes and Connectivity**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "08779263",
   "metadata": {},
   "source": [
    "### **Compute harmonized module eigengenes**\n",
    "\n",
    "`Module Eigengenes (MEs)` are a commonly used metric to summarize the gene expression profile of an entire co-expression module. Briefly, module eigengenes are computed by performing principal component analysis (PCA) on the subset of the gene expression matrix comprising each module. `The first PC of each of these PCA matrices are the MEs.`\n",
    "\n",
    "Dimensionality reduction techniques are a very hot topic in single-cell genomics. It is well known that technical artifacts can muddy the analysis of single-cell datasets, and over the years there have been many methods that aim to reduce the effects of these artifacts. Therefore it stands to reason that `MEs would be subject to these technical artifacts as well`, and hdWGCNA seeks to alleviate these effects.\n",
    "\n",
    "`hdWGCNA includes a function ModuleEigengenes` to compute module eigengenes in single cells. Additionally, we can apply Harmony batch correction to the MEs, yielding harmonized module eigengenes (hMEs).\n",
    "\n",
    "The following code performs the module eigengene computation `harmonizing by the Sample of origin using the group.by.vars parameter.`"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c011bbb2",
   "metadata": {},
   "source": [
    "### **Compute module connectivity**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e9f2e4ad",
   "metadata": {},
   "source": [
    "In co-expression network analysis, we often want to focus on the `“hub genes”`,`those which are highly connected within each module.` Therefore we wish to determine the `eigengene-based connectivity, also known as kME`, of each gene. \n",
    "\n",
    "hdWGCNA includes the ModuleConnectivity to compute the kME values in the full single-cell dataset, rather than the metacell dataset. This function essentially computes pairwise correlations between genes and module eigengenes. kME can be computed for all cells in the dataset, but we recommend computing kME in the cell type or group that was previously used to run ConstructNetwork."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "df854ccb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"yellow\"\n",
      "[1] \"grey\"\n",
      "[1] \"blue\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"brown\"\n",
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"----------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    # change active hdWGCNA experiment to consensus\n",
    "    \n",
    "    seurat_obj[[cell_type]] <- SetActiveWGCNA(seurat_obj[[cell_type]],\n",
    "                             wgcna_name=paste0(toupper(substr(cell_type, 1, 3)), '_consensus'))\n",
    "\n",
    "    print(paste0('Estimating module eigen-genes in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "    # need to run ScaleData first or else harmony throws an error:\n",
    "    # seurat_obj[[cell_type]] <- ScaleData(seurat_obj[[cell_type]], features=VariableFeatures(seurat_obj[[cell_type]]))\n",
    "\n",
    "    # compute all MEs in the full single-cell dataset\n",
    "    seurat_obj[[cell_type]] <- ModuleEigengenes(\n",
    "    seurat_obj[[cell_type]],\n",
    "    group.by.vars='study'\n",
    "    )\n",
    "\n",
    "    print(paste0('Estimating module connectivity in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # compute eigengene-based connectivity (kME):\n",
    "    seurat_obj[[cell_type]] <- ModuleConnectivity(\n",
    "        seurat_obj[[cell_type]],\n",
    "        group.by = 'cell_type', \n",
    "        group_name = cell_type\n",
    "    )\n",
    "    \n",
    "    print('----------------')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "05f2fc35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "95b46026",
   "metadata": {},
   "source": [
    "For convenience, we re-name the hdWGCNA modules to indicate that they are from the Excitatory neuron group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a52dff32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Renaming modules in hdWGCNA Experiment for INHIBITORY\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Renaming modules in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # rename the modules\n",
    "  seurat_obj[[cell_type]] <- ResetModuleNames(\n",
    "    seurat_obj[[cell_type]],\n",
    "    new_name = paste0(toupper(substr(cell_type, 1, 3)), \"-M\"),\n",
    "    wgcna_name = paste0(toupper(substr(cell_type, 1, 3)), '_consensus') # the name of the hdWGCNA experiment,\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0f353a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "88ad5a62",
   "metadata": {},
   "source": [
    "We also reset module colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5e7726f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Resetting module colors in hdWGCNA Experiment for INHIBITORY\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj \n",
    "\n",
    "\n",
    "\n",
    "library(MetBrewer)\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Resetting module colors in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "  mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "    distinct %>% arrange(module) %>% .$color\n",
    "  n_colors <- length(mod_colors) -1\n",
    "\n",
    "  new_colors <- paste0(met.brewer(\"Signac\", n=n_colors,))\n",
    "  new_colors <- sample(new_colors)\n",
    "  seurat_obj[[cell_type]] <- ResetModuleColors(seurat_obj[[cell_type]], new_colors)\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6b194589",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ae334b97",
   "metadata": {},
   "source": [
    "We can visualize the genes in each module ranked by kME using the PlotKMEs function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "03423498",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer dplyr::select over any other package.\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "library(conflicted)\n",
    "conflicts_prefer(dplyr::select)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/ModuleConnectivity/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # plot genes ranked by kME for each module\n",
    "\n",
    "    p <- PlotKMEs(seurat_obj[[cell_type]], ncol=5)\n",
    "\n",
    "    pdf(paste0(fig_dir, cell_type, 'consensus_KMEs.pdf'), width=12, height=8)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c6b1f8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "27f2d80d",
   "metadata": {},
   "source": [
    "## **Compute hub gene signature scores**\n",
    "\n",
    "Gene scoring analysis is a popular method in single-cell transcriptomics for computing a score for the overall signature of a set of genes. Seurat implements their own gene scoring technique using the `AddModuleScore` function, but there are also alternative approaches such as UCell.\n",
    "\n",
    "hdWGCNA includes the function `ModuleExprScore` to compute gene scores for a give number of genes for each module, using either the Seurat or UCell algorithm. Gene scoring is an alternative way of summarizing the expression of a module from computing the module eigengene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e1f93a8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for INHIBITORY\"\n",
      "Selecting by kME_INH-M1\n",
      "Selecting by kME_INH-M2\n",
      "Selecting by kME_INH-M3\n",
      "Selecting by kME_INH-M4\n",
      "Selecting by kME_INH-M1\n",
      "Selecting by kME_INH-M2\n",
      "Selecting by kME_INH-M3\n",
      "Selecting by kME_INH-M4\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Compute hub gene scores in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "  # compute gene scoring for the top 25 hub genes by kME for each module\n",
    "  # with Seurat method\n",
    "  seurat_obj[[cell_type]] <- ModuleExprScore(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_genes = 25,\n",
    "    method='Seurat'\n",
    "  )\n",
    "  \n",
    "  # compute gene scoring for the top 25 hub genes by kME for each module\n",
    "  # with UCell method\n",
    "  library(UCell)\n",
    "  seurat_obj[[cell_type]] <- ModuleExprScore(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_genes = 25,\n",
    "    method='UCell'\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "34de1c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "03dc3f32",
   "metadata": {},
   "source": [
    "## **Basic Visualization**\n",
    "\n",
    "Here we visualize some results using hdWGCNA, and we employ some of Seurat’s built-in plotting tools to visualize our hdWGCNA results.\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "368c8716",
   "metadata": {},
   "source": [
    "### **Module Feature Plots**\n",
    "\n",
    "FeaturePlot is a commonly used Seurat visualization to show a feature of interest directly on the dimensionality reduction. hdWGCNA includes the ModuleFeaturePlot function to consruct FeaturePlots for each co-expression module colored by each module’s uniquely assigned color."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c5865383",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module Feature Plots in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/FeaturePlots/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Module Feature Plots in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "     \n",
    "  # make a featureplot of hMEs for each module\n",
    "  plot_list <- ModuleFeaturePlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    features='hMEs', # plot the hMEs\n",
    "    order=TRUE, # order so the points with highest hMEs are on top\n",
    "    # raster=TRUE, \n",
    "    # raster_dpi=400, \n",
    "    # alpha=1,\n",
    "    # restrict_range=FALSE, \n",
    "    # raster_scale=0.25,\n",
    "  )\n",
    "\n",
    "  plot_list <- lapply(1:length(plot_list), function(x){\n",
    "    plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_consensus_featureplot_MEs.pdf\"), height=10, width=5)\n",
    "  print(wrap_plots(plot_list, ncol=3))\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7d1a4b43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2e09999a",
   "metadata": {},
   "source": [
    "We can also plot the hub gene signature score using the same function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "07825465",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module Feature Plots for HUB GENES in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] \"INH-M1\"\n",
      "[1] \"INH-M2\"\n",
      "[1] \"INH-M3\"\n",
      "[1] \"INH-M4\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/FeaturePlots/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Module Feature Plots for HUB GENES in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "     \n",
    "  # make a featureplot of hMEs for each module\n",
    "  plot_list <- ModuleFeaturePlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    features='scores',  # plot the hub gene scores\n",
    "    order='shuffle', # order so cells are shuffled\n",
    "    ucell = TRUE, # depending on Seurat vs UCell for gene scoring\n",
    "    # raster=TRUE, \n",
    "    # raster_dpi=400, \n",
    "    # alpha=1,\n",
    "    # restrict_range=FALSE, \n",
    "    # raster_scale=0.25,\n",
    "  )\n",
    "\n",
    "  plot_list <- lapply(1:length(plot_list), function(x){\n",
    "    plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_consensus_featureplot_hubs.pdf\"), height=10, width=5)\n",
    "  print(wrap_plots(plot_list, ncol=3))\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "dbbce473",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2a1cecc7",
   "metadata": {},
   "source": [
    "## **Overlap With AD Genes**\n",
    "\n",
    "Next on the co-expression UMAP, we show&quantify the overlap between hub genes that have a known association with AD using the AD databases obtained from the [**KEGG Alzheimer's pathway**](https://www.genome.jp/pathway/hsa05010) and [**Harmonizome (Mayaanlab)**](https://maayanlab.cloud/Harmonizome/gene_set/Alzheimer+Disease/dbGAP+Gene-Trait+Associations) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "0cd6ac66",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEGG_paths = pathway_analyses.read_pathways('../data/pathway_databases/KEGG_2019_Human.txt')\n",
    "KEGG_genes = [gene for gene in list(KEGG_paths[(KEGG_paths[0].str.startswith('Alzheimer disease'))].iloc[:, 1:].values[0]) if str(gene)!=\"nan\"]\n",
    "mayaanlab_genes = pd.read_csv('../data/pathway_databases/AD_genes.csv').Symbol.to_list()\n",
    "KEGG_genes.extend(mayaanlab_genes)\n",
    "\n",
    "AD_genes = list(set(KEGG_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "56d87a08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for INHIBITORY\"\n",
      "[1] 18168\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -i AD_genes\n",
    "\n",
    "# load modules\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/GeneOverlap/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Estimating AD gene overlap in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "  genome.size <- dim(seurat_obj[[cell_type]])[1]\n",
    "\n",
    "\n",
    "  overlap_df <- do.call(rbind, lapply(mods, function(cur_mod){\n",
    "\n",
    "  cur_genes <- modules %>% subset(module == cur_mod) %>% .$gene_name\n",
    "\n",
    "  cur_overlap <- testGeneOverlap(newGeneOverlap(\n",
    "      cur_genes,\n",
    "      as.vector(AD_genes),\n",
    "      genome.size=genome.size\n",
    "  ))\n",
    "\n",
    "  cur_overlap <- data.frame(\n",
    "    'odds.ratio' = cur_overlap@odds.ratio,\n",
    "    'pval' = cur_overlap@pval,\n",
    "    'Jaccard' = cur_overlap@Jaccard,\n",
    "    'size_intersection' = length(cur_overlap@intersection),\n",
    "    'module' = cur_mod\n",
    "  )\n",
    "\n",
    "    cur_overlap\n",
    "\n",
    "  })) %>% as.data.frame()\n",
    "\n",
    "  overlap_df <- overlap_df %>% mutate(fdr=p.adjust(pval, method='fdr'))\n",
    "  overlap_df <- overlap_df %>% subset(module != 'grey')\n",
    "\n",
    "  ################################################################################\n",
    "  # Plot as a lollipop\n",
    "  ################################################################################\n",
    "\n",
    "  overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)\n",
    "  overlap_df <- overlap_df %>% arrange(odds.ratio, descending=TRUE)\n",
    "  overlap_df$module <- factor(as.character(overlap_df$module), levels=as.character(overlap_df$module))\n",
    "\n",
    "  mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "    distinct\n",
    "  cp <- mod_colors$color; names(cp) <- mod_colors$module\n",
    "\n",
    "  p <- overlap_df %>%\n",
    "    ggplot(aes(y=module, x=odds.ratio, size= size_intersection, color=module)) +\n",
    "    geom_segment(aes(y=module, yend=module, x=0, xend=odds.ratio), size=0.5, color='grey') +\n",
    "    geom_point() +\n",
    "    geom_point(shape=overlap_df$shape, color='black', fill=NA) +\n",
    "    scale_color_manual(values=cp, guide='none') +\n",
    "    ylab('') + xlab(\"Odds ratio\") +\n",
    "    scale_x_continuous(breaks = seq(0, 19, 2)) +\n",
    "    labs(size='Size\\nintersection') +\n",
    "    ggtitle('Overlap with AD genes') +\n",
    "\n",
    "    theme(\n",
    "      panel.border = element_rect(size=1, color='black', fill=NA),\n",
    "      axis.line.y = element_blank(),\n",
    "      axis.line.x = element_blank(),\n",
    "      plot.title = element_text(hjust=0.5, face='plain')\n",
    "    )\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_consensus_AD_overlap.pdf'), width=4, height=3.5)\n",
    "  print(p)\n",
    "  dev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "644b2b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c94606d3",
   "metadata": {},
   "source": [
    "## **Overlap With DEGs**\n",
    "\n",
    "We quantify the overlap between hub genes and DEGs obtained from previous analysis and saved at `../results/{test_name}/{save_prefix}_{deg_method}_degs.xlsx`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "27e94e42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating AD gene overlap in hdWGCNA Experiment for INHIBITORY\"\n",
      "New names:\n",
      "• `` -> `...1`\n",
      "New names:\n",
      "• `` -> `...1`\n",
      "New names:\n",
      "• `` -> `...1`\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -i deg_method\n",
    "\n",
    "# load modules\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/GeneOverlap/Consensus/\")\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (test in c('ad_vs_no')){\n",
    "  \n",
    "  for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Estimating AD gene overlap in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    degs <- list()\n",
    "\n",
    "    for (study in studies){\n",
    "      wgcna_genes <- GetWGCNAGenes(seurat_obj[[cell_type]])\n",
    "      hub_genes <- GetHubGenes(seurat_obj[[cell_type]], 25)\n",
    "      degs[[study]] <- read_excel(paste0('../results/', test, '/', study, '_', deg_method, '_degs.xlsx'), sheet = cell_type)\n",
    "      degs[[study]] <- intersect(hub_genes$gene_name, subset(degs, p_val_adj<0.01 & abs_logFC>0.25)$gene)\n",
    "    }\n",
    "\n",
    "    degs <- unique(unlist(degs, recursive=FALSE))\n",
    "\n",
    "    modules <- GetModules(seurat_obj[[cell_type]])\n",
    "    mods <- levels(modules$module)\n",
    "    genome.size <- dim(seurat_obj[[cell_type]])[1]\n",
    "\n",
    "    overlap_df <- do.call(rbind, lapply(mods, function(cur_mod){\n",
    "\n",
    "    cur_genes <- modules %>% subset(module == cur_mod) %>% .$gene_name\n",
    "\n",
    "    cur_overlap <- testGeneOverlap(newGeneOverlap(\n",
    "        cur_genes,\n",
    "        as.vector(degs),\n",
    "        genome.size=genome.size\n",
    "    ))\n",
    "\n",
    "    cur_overlap <- data.frame(\n",
    "      'odds.ratio' = cur_overlap@odds.ratio,\n",
    "      'pval' = cur_overlap@pval,\n",
    "      'Jaccard' = cur_overlap@Jaccard,\n",
    "      'size_intersection' = length(cur_overlap@intersection),\n",
    "      'module' = cur_mod\n",
    "    )\n",
    "\n",
    "      cur_overlap\n",
    "\n",
    "    })) %>% as.data.frame()\n",
    "\n",
    "    overlap_df <- overlap_df %>% mutate(fdr=p.adjust(pval, method='fdr'))\n",
    "    overlap_df <- overlap_df %>% subset(module != 'grey')\n",
    "\n",
    "    ################################################################################\n",
    "    # Plot as a lollipop\n",
    "    ################################################################################\n",
    "\n",
    "    overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)\n",
    "    overlap_df <- overlap_df %>% arrange(odds.ratio, descending=TRUE)\n",
    "    overlap_df$module <- factor(as.character(overlap_df$module), levels=as.character(overlap_df$module))\n",
    "\n",
    "    mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "      distinct\n",
    "    cp <- mod_colors$color; names(cp) <- mod_colors$module\n",
    "\n",
    "    p <- overlap_df %>%\n",
    "      ggplot(aes(y=module, x=odds.ratio, size= size_intersection, color=module)) +\n",
    "      geom_segment(aes(y=module, yend=module, x=0, xend=odds.ratio), size=0.5, color='grey') +\n",
    "      geom_point() +\n",
    "      geom_point(shape=overlap_df$shape, color='black', fill=NA) +\n",
    "      scale_color_manual(values=cp, guide='none') +\n",
    "      ylab('') + xlab(\"Odds ratio\") +\n",
    "      scale_x_continuous(breaks = seq(0, 19, 2)) +\n",
    "      labs(size='Size\\nintersection') +\n",
    "      ggtitle('Overlap with DEGs') +\n",
    "\n",
    "      theme(\n",
    "        panel.border = element_rect(size=1, color='black', fill=NA),\n",
    "        axis.line.y = element_blank(),\n",
    "        axis.line.x = element_blank(),\n",
    "        plot.title = element_text(hjust=0.5, face='plain')\n",
    "      )\n",
    "\n",
    "    pdf(paste0(fig_dir, test, '_', cell_type, '_consensus_DEG_overlap.pdf'), width=4, height=3.5)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "  }\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0e283197",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e1975403",
   "metadata": {},
   "source": [
    "# **Save Seurat Object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "816dbc14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for ASTROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/SeuratObject/Consensus/\")\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Saving hdWGCNA object in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    saveRDS(seurat_obj[[cell_type]], file=paste0(dat_dir, cell_type, '_hdWGCNA_object.rds'))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23f14b4e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
