{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62042f97-a76e-4b34-805f-3f1ac81e7288",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import scipy\n",
    "import logging\n",
    "import warnings\n",
    "import anndata2ri\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import decoupler as dc\n",
    "import scrublet as scr\n",
    "import decoupler as dc\n",
    "from scipy import sparse\n",
    "from anndata import AnnData\n",
    "from tabnanny import verbose\n",
    "import matplotlib.pyplot as plt\n",
    "from gsva_prep import prep_gsva\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from typing import Optional, Union\n",
    "from matplotlib.pyplot import rcParams\n",
    "from functions import pathway_analyses\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.model_selection import train_test_split\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from rpy2.robjects.conversion import localconverter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c36a2bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sys_dpi(width, height, diag):\n",
    "    '''\n",
    "    obtain dpi of system\n",
    "    \n",
    "    w: width in pixels (if unsure, go vist `whatismyscreenresolution.net`)\n",
    "    h: height in pixels\n",
    "    d: diagonal in inches\n",
    "    '''\n",
    "    w_inches = (diag**2/ (1 + height**2/width**2))**0.5\n",
    "    return round(width/w_inches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "447b6d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.1\n",
      "-----\n",
      "OpenSSL                     22.0.0\n",
      "PIL                         9.2.0\n",
      "absl                        NA\n",
      "anndata2ri                  1.1\n",
      "appnope                     0.1.2\n",
      "asttokens                   NA\n",
      "astunparse                  1.6.3\n",
      "attr                        21.4.0\n",
      "backcall                    0.2.0\n",
      "beta_ufunc                  NA\n",
      "binom_ufunc                 NA\n",
      "boto3                       1.26.32\n",
      "botocore                    1.29.32\n",
      "bottleneck                  1.3.5\n",
      "brotli                      NA\n",
      "certifi                     2022.09.24\n",
      "cffi                        1.15.1\n",
      "cloudpickle                 2.2.0\n",
      "colorama                    0.4.4\n",
      "cryptography                38.0.1\n",
      "cycler                      0.10.0\n",
      "cython_runtime              NA\n",
      "dask                        2022.11.0\n",
      "dateutil                    2.8.2\n",
      "debugpy                     1.5.1\n",
      "decorator                   5.1.1\n",
      "decoupler                   1.4.0\n",
      "defusedxml                  0.7.1\n",
      "deprecate                   0.3.2\n",
      "entrypoints                 0.4\n",
      "executing                   0.8.3\n",
      "fsspec                      2022.10.0\n",
      "functions                   NA\n",
      "google                      NA\n",
      "gsva_prep                   NA\n",
      "h5py                        3.7.0\n",
      "hypergeom_ufunc             NA\n",
      "igraph                      0.10.2\n",
      "ipykernel                   6.15.2\n",
      "ipython_genutils            0.2.0\n",
      "ipywidgets                  7.6.5\n",
      "jedi                        0.18.1\n",
      "jinja2                      3.1.2\n",
      "jmespath                    1.0.1\n",
      "joblib                      1.1.1\n",
      "kiwisolver                  1.4.2\n",
      "leidenalg                   0.9.0\n",
      "llvmlite                    0.39.1\n",
      "louvain                     0.8.0\n",
      "markupsafe                  2.1.1\n",
      "matplotlib                  3.5.3\n",
      "mpl_toolkits                NA\n",
      "natsort                     7.1.1\n",
      "nbinom_ufunc                NA\n",
      "ncf_ufunc                   NA\n",
      "numba                       0.56.3\n",
      "numexpr                     2.8.3\n",
      "numpy                       1.23.3\n",
      "packaging                   21.3\n",
      "pandas                      1.4.4\n",
      "parso                       0.8.3\n",
      "pexpect                     4.8.0\n",
      "pickleshare                 0.7.5\n",
      "pkg_resources               NA\n",
      "prompt_toolkit              3.0.20\n",
      "psutil                      5.9.0\n",
      "ptyprocess                  0.7.0\n",
      "pure_eval                   0.2.2\n",
      "pyarrow                     11.0.0\n",
      "pycparser                   2.21\n",
      "pydev_ipython               NA\n",
      "pydevconsole                NA\n",
      "pydevd                      2.6.0\n",
      "pydevd_concurrency_analyser NA\n",
      "pydevd_file_utils           NA\n",
      "pydevd_plugins              NA\n",
      "pydevd_tracing              NA\n",
      "pygments                    2.13.0\n",
      "pyparsing                   3.0.9\n",
      "pytorch_lightning           1.7.7\n",
      "pytz                        2022.1\n",
      "pytz_deprecation_shim       NA\n",
      "rich                        NA\n",
      "rpy2                        3.5.10\n",
      "scipy                       1.9.3\n",
      "scrublet                    NA\n",
      "seaborn                     0.11.2\n",
      "session_info                1.0.0\n",
      "setuptools                  65.6.3\n",
      "six                         1.16.0\n",
      "sklearn                     1.1.3\n",
      "sparse                      0.13.0\n",
      "stack_data                  0.2.0\n",
      "statsmodels                 0.13.5\n",
      "tensorboard                 2.10.1\n",
      "texttable                   1.6.4\n",
      "threadpoolctl               2.2.0\n",
      "tlz                         0.12.0\n",
      "toolz                       0.12.0\n",
      "torch                       1.12.1\n",
      "torchmetrics                0.10.2\n",
      "tornado                     6.2\n",
      "tqdm                        4.64.1\n",
      "traitlets                   5.1.1\n",
      "typing_extensions           NA\n",
      "tzlocal                     NA\n",
      "urllib3                     1.26.12\n",
      "wcwidth                     0.2.5\n",
      "yaml                        5.4.1\n",
      "zmq                         23.2.0\n",
      "zoneinfo                    NA\n",
      "-----\n",
      "IPython             8.6.0\n",
      "jupyter_client      7.3.5\n",
      "jupyter_core        4.11.2\n",
      "notebook            6.4.12\n",
      "-----\n",
      "Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:41:22) [Clang 13.0.1 ]\n",
      "macOS-13.0.1-arm64-arm-64bit\n",
      "-----\n",
      "Session information updated at 2023-06-14 18:45\n"
     ]
    }
   ],
   "source": [
    "# # Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# # Automatically convert rpy2 outputs to pandas dataframes\n",
    "# pandas2ri.activate()\n",
    "# anndata2ri.activate()\n",
    "# %load_ext rpy2.ipython\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=PendingDeprecationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "rcParams['figure.dpi'] = get_sys_dpi(1512, 982, 14.125)\n",
    "#rcParams['figure.figsize']=(4,4) #rescale figures\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "10be21f0-000d-48d7-9e97-4051ec9da0bd",
   "metadata": {},
   "source": [
    "# **hdWGCNA in single-cell data**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0c5d2f8d",
   "metadata": {},
   "source": [
    "Here we use `hdWGCNA` to perform co-expression network analysis on single-cell data. \n",
    "\n",
    "`hdWGCNA` is an R package for performing `weighted gene co-expression network analysis (WGCNA)` in high dimensional transcriptomics data such as single-cell RNA-seq or spatial transcriptomics. hdWGCNA is highly modular and can construct co-expression networks across multi-scale cellular and spatial hierarchies. hdWGNCA identifies robust modules of inerconnected genes, and provides context for these modules through various biological knowledge sources. hdWGCNA requires data formatted as Seurat objects, one of the most ubiquitous formats for single-cell data. \n",
    "\n",
    "`Note: hdWGCNA is under active development, so you may run into errors and small typos. We welcome users to write GitHub issues to report bugs, ask for help, and to request potential enhancements.`\n",
    "\n",
    "Explore the full capabilites of hWGNA in the following publications:\n",
    "\n",
    "- [**Morabito et al. bioRxiv 2022**](https://www.biorxiv.org/content/10.1101/2022.09.22.509094v1)\n",
    "- [**Morabito & Miyoshi et al. Nature Genetics 2021**](https://www.nature.com/articles/s41588-021-00894-z)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1b4733e8",
   "metadata": {},
   "source": [
    "Here, apply hdWGCNA to the processed single-nucleus RNA-seq (snRNA-seq) datasets. These dataset have already been fully processed using a standard single-cell transcritpomics analysis pipelines (please see the following notebooks, `preprocessing_data_integration`, `preprocessing_quality_control`, `preprocessing_cluster_annotation`). \n",
    "\n",
    "\n",
    "If you would like to apply this notebook to your own dataset, you first need to satisfy the following prerequisites:\n",
    "\n",
    "A single-cell or single-nucleus transcriptomics dataset in Seurat format.\n",
    "Normalize the gene expression matrix NormalizeData.\n",
    "Identify highly variable genes VariableFeatures.\n",
    "Scale the normalized expression data ScaleData\n",
    "Perform dimensionality reduction RunPCA and batch correction if needed RunHarmony.\n",
    "Non-linear dimensionality reduction RunUMAP for visualizations.\n",
    "Group cells into clusters (FindNeighbors and FindClusters).\n",
    "An example of running the prerequisite data processing steps can be found in the Seurat Guided Clustering Tutorial.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8b87d243",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    }
   ],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "    library(WGCNA)\n",
    "    library(Matrix)\n",
    "    library(viridis)\n",
    "    library(harmony)\n",
    "    library(ggpubr)\n",
    "    library(tictoc)\n",
    "    library(RColorBrewer)\n",
    "    library(Hmisc)\n",
    "    library(corrplot)\n",
    "    library(grid)\n",
    "    library(gridExtra)\n",
    "    library(igraph)\n",
    "    library(ggrepel)\n",
    "\n",
    "\n",
    "    # single-cell analysis package\n",
    "    library(Seurat)\n",
    "\n",
    "    # plotting and data science packages\n",
    "    library(tidyverse)\n",
    "    library(cowplot)\n",
    "    library(patchwork)\n",
    "\n",
    "    # co-expression network analysis packages:\n",
    "    library(WGCNA)\n",
    "    library(hdWGCNA)\n",
    "\n",
    "    # gene enrichment packages\n",
    "    library(enrichR)\n",
    "    library(GeneOverlap)\n",
    "\n",
    "\n",
    "    library(GSEABase)\n",
    "    library(GSVA) \n",
    "# needs to be run every time you start R and want to use %>%\n",
    "})\n",
    "\n",
    "# using the cowplot theme for ggplot\n",
    "theme_set(theme_cowplot())\n",
    "\n",
    "# set random seed for reproducibility\n",
    "set.seed(12345)\n",
    "\n",
    "# optionally enable multithreading\n",
    "# enableWGCNAThreads(nThreads = 4)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "af425574",
   "metadata": {},
   "source": [
    "## **Prepare data**\n",
    "\n",
    "Now, we load the preprocessed and annotated data for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9697f83a",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_prefix = 'leng_etc'\n",
    "\n",
    "adata_annot = sc.read_h5ad(f'../data/processed/{save_prefix}/{save_prefix}_mapped_anndata.h5ad')\n",
    "adata_annot.obs_names_make_unique()\n",
    "adata_annot.var_names_make_unique()\n",
    "adata_annot.X = adata_annot.layers['counts'].copy()\n",
    "del adata_annot.obsm, adata_annot.layers, adata_annot.varm, adata_annot.uns, adata_annot.obsp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d8504101",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_meta = True\n",
    "deg_method =  'DESeq2-Wald'\n",
    "test_names = ['late_vs_early', 'early_vs_no', 'late_vs_no', 'ad_vs_no']\n",
    "filter_genes = \"TRUE\"\n",
    "subject_id = 'PatientID'             # for leng this is `PatientID` for mathys is 'Subject', and 'individualID' for allen\n",
    "gene_celltype_threshold = 0.05          # determines number of cells the gene must be expressed in \n",
    "covariates = ['None']                   # list of covariates to be accounted for in regression.\n",
    "gene_selection = 'custom'               # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "celltypes = [\"Oligodendrocyte\"] #[\"Excitatory\", \"Inhibitory\", \"Astrocyte\", \"Microglia\", \"Oligodendrocyte\", \"OPC\"]\n",
    "metadata = f'../data/raw/{save_prefix}/{save_prefix}_metadata.csv' \n",
    "meta = pd.read_csv(metadata, encoding_errors='ignore')\n",
    "\n",
    "gene_selection = 'custom'   # specifies the gene selection method when setting up seurat object for WGCNA. The th\n",
    "                            # Posible values are \"custom\", \"fraction\", \"variable\"\n",
    "                            # If custom, a list of genes must be passed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3888c7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# map the pathology group to teh subject id in .obs\n",
    "\n",
    "adata_annot.obs['pathology.group'] = adata_annot.obs[subject_id].map(dict(zip(meta[subject_id].astype(str), meta['pathology.group'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "839f795a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 16585 41432 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(16585): SAMD11 NOC2L ... S100B PRMT2\n",
      "rowData names(12): mt n_cells_by_counts ... mean std\n",
      "colnames(41432): EC2_AAACCTGAGGATGCGT EC2_AAACCTGAGTCAATAG ...\n",
      "  EC10_TTTGTCATCTATCGCC EC10_TTTGTCATCTCTGCTG\n",
      "colData names(24): SampleID PatientID ... predictions pathology.group\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "[1] \"loaded data into memory for recursive use\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i adata_annot -i subject_id -i gene_celltype_threshold -i celltypes -i test_names -i save_prefix -i gene_selection\n",
    "\n",
    "print(adata_annot)\n",
    "print('loaded data into memory for recursive use')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "74c5d20a",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_annot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8a3955a2",
   "metadata": {},
   "source": [
    "## **Set up Seurat object for WGCNA**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2d154d2e",
   "metadata": {},
   "source": [
    "Before running hdWGCNA, we first have to set up the Seurat object. Most of the information computed by hdWGCNA is stored in the Seurat object’s `@misc slot`. \n",
    "\n",
    "A single Seurat object can hold multiple hdWGCNA experiments, for example representing different cell types in the same single-cell dataset. Notably, since we consider hdWGCNA to be a downstream data analysis step, we do not need to subset the Seurat object after `SetupForWGCNA` has been run.\n",
    "\n",
    "Here we will set up the Seurat object using the `SetupForWGCNA` function, specifying the name of the hdWGNCA experiment. This function also selects the genes that will be used for WGCNA. We can aslo select genes using three different approaches using the gene_select parameter:\n",
    "\n",
    "\n",
    "- `variable`: use the genes stored in the Seurat object’s `VariableFeatures`.\n",
    "\n",
    "- `fraction`: use genes that are expressed in a certain fraction of cells for in the whole dataset or in each group of cells, specified by group.by.\n",
    "\n",
    "- `custom`: use genes that are specified in a custom list."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3e7607cd",
   "metadata": {},
   "source": [
    "For our specific case, we will either select genes that are expressed in at least 5% of cells within each cell type or specify a list of custom genes. Additionally, we will rename our hdWGCNA experiment to enhance clarity.\n",
    "\n",
    "Here, the `gene_selection` parameter is set to `custom`, we employ the genes present in all the overlapping pathways in a cell type-specific and study-specific manner.\n",
    "\n",
    "To accomplish this, we load the list of overlapping pathways that was previously saved in `../results/{test_name}/{test_name}_pathway_overlap.csv` from the `pathway_meta_analysis.ipynb` notebook. Additionally, we load the genes from the `.gmt` file saved in `../results/gsva/{study}/{cur_celltype}/gs.gmt` from the `pathway_analysis.ipynb` notebook."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c4ddf8bf",
   "metadata": {},
   "source": [
    "##### **Overlapping Pathways**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9e481a4c",
   "metadata": {},
   "source": [
    "Obtain list of overlapping pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0f695cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get list of overlapping pathways and save to R                        \n",
    "overlap_paths = {}\n",
    "\n",
    "for test_name in test_names:\n",
    "    overlap_paths[test_name] = pd.read_csv(f'../results/{test_name}/{test_name}_pathway_overlap.csv')\n",
    "\n",
    "try:\n",
    "    list_of_overlapping_paths = robjects.ListVector({test_name: robjects.vectors.StrVector(overlap_paths[test_name].pathway.to_list()) for test_name in test_names})\n",
    "except AttributeError:\n",
    "    temp_tests = []\n",
    "    for test_name in test_names:\n",
    "        if len(overlap_paths.pathway)>0:\n",
    "            temp_tests.append(test_name)\n",
    "        else:\n",
    "            print(f\"AttributeError occurred for test_name: {test_name}\")\n",
    "    list_of_overlapping_paths = robjects.ListVector({test_name: robjects.vectors.StrVector(overlap_paths[test_name].pathway.to_list()) for test_name in temp_tests})\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7234ec55",
   "metadata": {},
   "source": [
    "Obtain the genesets in the overlapping pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "703041ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 206\n",
      "[1] 827\n",
      "[1] 1141\n",
      "[1] 1180\n"
     ]
    }
   ],
   "source": [
    "%%R -o geneSets -i list_of_overlapping_paths\n",
    "\n",
    "# iterate over celltypes and obtain \n",
    "geneSets <- list()\n",
    "for(test_name in test_names){\n",
    "\n",
    "  geneSets[[test_name]] <- list()\n",
    "  for (cur_celltype in celltypes){\n",
    "    # get gene sets\n",
    "    geneSets[[test_name]][[cur_celltype]] <- getGmt(file.path('../results/gsva/', save_prefix, '/', cur_celltype, 'gs.gmt'))\n",
    "    geneSets[[test_name]][[cur_celltype]] <- geneIds(geneSets[[test_name]][[cur_celltype]])\n",
    "\n",
    "    # Check if test_name exists in list_of_overlapping_paths\n",
    "    if (test_name %in% names(list_of_overlapping_paths)) {\n",
    "      geneSets[[test_name]][[cur_celltype]] <-  geneSets[[test_name]][[cur_celltype]][names(geneSets[[test_name]][[cur_celltype]]) %in% list_of_overlapping_paths[[test_name]]]\n",
    "      } else {\n",
    "      geneSets[[test_name]][[cur_celltype]] <- character(0)  # Set an empty vector if no scores exist\n",
    "    }\n",
    "    geneSets[[test_name]][[cur_celltype]] <- unique(unlist(geneSets[[test_name]][[cur_celltype]], recursive = FALSE))\n",
    "\n",
    "    print(length(geneSets[[test_name]][[cur_celltype]]))\n",
    "\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f2d4d1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of geneSets object into Rpy2 object \n",
    "\n",
    "geneSets =  robjects.ListVector(\n",
    "                        {\n",
    "                            test_name: robjects.ListVector(\n",
    "                                    {cell_type: geneSets[test_name][cell_type]\n",
    "                                                 \n",
    "                                    for cell_type in celltypes\n",
    "                                    }\n",
    "                                )\n",
    "                            for test_name in test_names\n",
    "                        }\n",
    "\n",
    "                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "845fa1c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An object of class Seurat \n",
      "16585 features across 41432 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 0 variable features)\n",
      "[1] \"Creating hdWGNA Experiment for OLIGODENDROCYTE\"\n",
      "An object of class Seurat \n",
      "16585 features across 9606 samples within 1 assay \n",
      "Active assay: originalexp (16585 features, 2000 variable features)\n",
      " 3 dimensional reductions calculated: pca, harmony, umap\n",
      "[1] \"1180 WGCNA Genes\"\n"
     ]
    }
   ],
   "source": [
    "%%R -o seurat_obj -i geneSets\n",
    "\n",
    "seurat_obj <- as.Seurat(adata_annot, counts = \"X\", data = \"X\")\n",
    "\n",
    "print(seurat_obj)\n",
    "\n",
    "# Perform dimensionality reduction and plot\n",
    "\n",
    "seurat_obj <- FindVariableFeatures(seurat_obj)\n",
    "seurat_obj <- ScaleData(seurat_obj)\n",
    "seurat_obj <- RunPCA(seurat_obj)\n",
    "seurat_obj <- RunHarmony(seurat_obj, group.by.vars = subject_id)\n",
    "seurat_obj <- RunUMAP(seurat_obj, reduction='harmony', n.neighbors=15, dims=1:30, min.dist=0.1)\n",
    "\n",
    "p <- DimPlot(seurat_obj, group.by = \"cell_type\", label = TRUE) +\n",
    "     umap_theme() + ggtitle(\"leng\") + NoLegend()\n",
    "\n",
    "# Save plot to PDF\n",
    "pdf(file = paste0(\"../results/hdWGCNA/UMAP/\", save_prefix, \"_integrated_umap.pdf\"), width = 4, height = 4, useDingbats = FALSE)\n",
    "print(p)\n",
    "dev.off()\n",
    "\n",
    "# create a hdWGCNA experiment for each celltype\n",
    "seurat_dat <- seurat_obj\n",
    "seurat_obj <- list()\n",
    "\n",
    "\n",
    "for (cur_cell_type in celltypes){\n",
    "\n",
    "  seurat_obj[[cur_cell_type]] <- subset(seurat_dat, cell_type == cur_cell_type)\n",
    "\n",
    "  print(paste0('Creating hdWGNA Experiment for ', toupper(cur_cell_type)))\n",
    "\n",
    "  if (gene_selection == 'custom') {\n",
    "    seurat_obj[[cur_cell_type]] <- SetupForWGCNA(\n",
    "      seurat_obj[[cur_cell_type]],\n",
    "      gene_select = \"custom\",                                 # the gene selection approach\n",
    "      gene_list = geneSets[['ad_vs_no']][[cur_cell_type]],    # list of genes to be included\n",
    "      group.by = 'cell_type',                                 # grouping parameter\n",
    "      wgcna_name = toupper(substr(cur_cell_type, 1, 3))       # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  } else {\n",
    "    seurat_obj[[cur_cell_type]] <- SetupForWGCNA(\n",
    "      seurat_obj[[cur_cell_type]],\n",
    "      gene_select = \"fraction\",                               # the gene selection approach\n",
    "      fraction = gene_celltype_threshold,                     # fraction of cells for gene inclusion\n",
    "      group.by = 'cell_type',                                 # grouping parameter\n",
    "      wgcna_name = toupper(substr(cur_cell_type, 1, 3))       # the name of the hdWGCNA experiment\n",
    "    )\n",
    "  }\n",
    "\n",
    "  print(seurat_obj[[cur_cell_type]])\n",
    "  print(paste0(length(GetWGCNAGenes(seurat_obj[[cur_cell_type]])), \" WGCNA Genes\"))\n",
    "\n",
    "}\n",
    "\n",
    "rm(seurat_dat)\n",
    "rm(adata_annot)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9e89e217",
   "metadata": {},
   "source": [
    "Send list to R interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ff771eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "\n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ccd0e590",
   "metadata": {},
   "source": [
    "## **Construct metacells**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "07813598",
   "metadata": {},
   "source": [
    "After setting up our Seurat object, the first step in running the hdWGCNA pipeine in hdWGCNA is to construct metacells from the single-cell dataset. \n",
    "\n",
    "Briefly, `metacells are aggregates of small groups of similar cells originating from the same biological sample of origin.` The k-Nearest Neighbors (KNN) algorithm is used to identify groups of similar cells to aggregate, and then the average or summed expression of these cells is computed, thus yielding a metacell gene expression matrix. `The sparsity of the metacell expression matrix is considerably reduced` when compared to the original expression matrix, and therefore it is preferable to use. We were originally motivated to use metacells in place of the original single cells because correlation network approaches such as WGCNA are sensitive to data sparsity. Furthermore, single-cell epigenomic approaches, such as Cicero, employ a similar metacell aggregation approach prior to constructing co-accessibility networks.\n",
    "\n",
    "hdWGCNA includes a function MetacellsByGroups to construct metacell expression matrices given a single-cell dataset. This function constructs a new Seurat object for the metacell dataset which is stored internally in the hdWGCNA experiment. \n",
    "\n",
    "The `group.by` parameter determines which groups metacells will be constructed in. \n",
    "\n",
    "`We only want to construct metacells from cells that came from the same biological sample of origin, so it is critical to pass that information to hdWGCNA via the group.by parameter.` \n",
    "\n",
    "Additionally, we usually construct metacells for each cell type separately. `Thus, in this example, we are grouping by Sample and cell_type to achieve the desired result.`\n",
    "\n",
    "The number of cells to be aggregated k should be tuned based on the size of the input dataset, in general a lower number for k can be used for small datasets. `Generally use k values between 20 and 75 for 40,039 cells`, ranging from 890 to 8,188 in each biological sample, and here we used k=25. The amount of allowable overlap between metacells can be tuned using the max_shared argument.\n",
    "\n",
    "Note: The metacell aggregation approach does not yield good results for extremely underrepresented cell types. For example, the brain vascular cells (pericytes and endothelial cells), thus we have excluded them from this analysis. \n",
    "\n",
    "`MetacellsByGroups `has a parameter min_cells to exclude groups that are smaller than a specified number of cells. Errors are likely to arise if the selected value for min_cells is too low.\n",
    "\n",
    "`Here we construct metacells and normalize in each hdWGCNA experiment`\n",
    "\n",
    "The resulting expression matrix is obtained using the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e92860fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Constructing MetaCells in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# construct metacells in each group for each wgnca experiment\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "  print(paste0('Constructing MetaCells in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  seurat_obj[[cell_type]] <- MetacellsByGroups(\n",
    "      \n",
    "    seurat_obj = seurat_obj[[cell_type]],\n",
    "    group.by = c(\"cell_type\", subject_id), # specify the columns in seurat_obj@meta.data to group by\n",
    "    reduction = 'harmony',  # select the dimensionality reduction to perform KNN on\n",
    "    k = 25, # nearest-neighbors parameter\n",
    "    max_shared = 10, # maximum number of shared cells between two metacells\n",
    "    ident.group = 'cell_type', # set the Idents of the metacell seurat object\n",
    "    wgcna_name = toupper(substr(cell_type, 1, 3)),        # the name of the hdWGCNA experiment\n",
    "    )\n",
    "\n",
    "    # normalize metacell expression matrix:\n",
    "  seurat_obj[[cell_type]] <- NormalizeMetacells(seurat_obj[[cell_type]],  wgcna_name = toupper(substr(cell_type, 1, 3)))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "db92cb94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b70e5121",
   "metadata": {},
   "source": [
    "## **Process the Metacell Seurat Object**\n",
    "\n",
    "Since we store the Metacell expression information as its own Seurat object, we can run Seurat functions on the metacell data. We can get the metacell object from the hdWGCNA experiment using `GetMetacellObject.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "46d9dccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o metacell_obj\n",
    "\n",
    "# metacell_obj <- list()\n",
    "# for (cell_type in celltypes){\n",
    "#     print(paste0('Obtaining MetaCells in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "#     metacell_obj[[cell_type]] <- GetMetacellObject(seurat_obj[[cell_type]] , \n",
    "#                                         wgcna_name = toupper(substr(cell_type, 1, 3))       # the name of the hdWGCNA experiment\n",
    "#                                      )\n",
    "#     print(metacell_obj[[cell_type]])\n",
    "# }"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a6c0058a",
   "metadata": {},
   "source": [
    "Here we use wrapper functions to process the metacell object and visualize one of the aggregated expression profiles in two dimensions with UMAP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "db7b36f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Processing MetaCells in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "    print(paste0('Processing MetaCells in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "    seurat_obj[[cell_type]]  <- NormalizeMetacells(seurat_obj[[cell_type]])\n",
    "    seurat_obj[[cell_type]]  <- ScaleMetacells(seurat_obj[[cell_type]] , features=VariableFeatures(seurat_obj[[cell_type]]))\n",
    "    seurat_obj[[cell_type]]  <- RunPCAMetacells(seurat_obj[[cell_type]] , features=VariableFeatures(seurat_obj[[cell_type]]))\n",
    "    seurat_obj[[cell_type]]  <- RunHarmonyMetacells(seurat_obj[[cell_type]] , group.by.vars=subject_id)\n",
    "    seurat_obj[[cell_type]]  <- RunUMAPMetacells(seurat_obj[[cell_type]] , reduction='harmony', dims=1:15)\n",
    "\n",
    "\n",
    "    p1 <- DimPlotMetacells(seurat_obj[[cell_type]] , group.by='cell_type') + umap_theme() + ggtitle(\"Cell Type\")\n",
    "    p2 <- DimPlotMetacells(seurat_obj[[cell_type]] , group.by=subject_id) + umap_theme() + ggtitle(subject_id)\n",
    "\n",
    "    p1 | p2\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6085ce6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fac55be8",
   "metadata": {},
   "source": [
    "## **Co-expression network analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f03072ae",
   "metadata": {},
   "source": [
    "### **Set up the expression matrix**\n",
    "\n",
    "Here we specify the expression matrix that we will use for network analysis. We only include a single cell-type of interest per run/simulation, so we have to subset our expression data prior to constructing the network. hdWGCNA includes the `SetDatExpr` function to store the transposed expression matrix for a given group of cells that will be used for downstream network analysis.\n",
    "\n",
    "The metacell expression matrix is used by default `(use_metacells=TRUE)`, but hdWGCNA does allow for the single-cell expression matrix to be used if desired.. This function allows the user to specify which slot to take the expression matrix from, for example if the user wanted to apply SCTransform normalization instead of NormalizeData."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7830296b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Setting up expression matrix in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "source('../scripts/functions/hdWGCNA.R')\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "\n",
    "    print(paste0('Setting up expression matrix in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    seurat_obj[[cell_type]] <- SetDatExpr_Alt(\n",
    "      seurat_obj[[cell_type]],\n",
    "      group_name = cell_type, # the name of the group of interest in the group.by column\n",
    "      group.by = 'cell_type', # the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups\n",
    "      assay = 'originalexp', # using RNA assay\n",
    "      slot = 'data', # using normalized data\n",
    "      wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3c40b7ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a70cc5c4",
   "metadata": {},
   "source": [
    "### **Select soft-power threshold**\n",
    "\n",
    "Next we will select the “soft power threshold”. This is an extremely important step in the hdWGNCA pipleine (and for vanilla WGCNA). hdWGCNA constructs a gene-gene correlation adjacency matrix to infer co-expression relationships between genes. The correlations are raised to a power to reduce the amount of noise present in the correlation matrix, thereby retaining the strong connections and removing the weak connections. Therefore, it is critical to determine a proper value for the soft power threshold.\n",
    "\n",
    "We use the function `TestSoftPowers` to perform a parameter sweep for different soft power thresholds. This function helps us to guide our choice in a soft power threshold for constructing the co-expression network by inspecting the resulting network topology for different power values. The co-expression network should have a scale-free topology, therefore the TestSoftPowers function models how closely the co-expression network resembles a scale-free graph at different soft power thresholds. Furthermore, we use the function `PlotSoftPowers` to visualize the results of the parameter sweep.\n",
    "\n",
    "The following code performs the parameter sweep and outputs a summary figure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e63133c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estinmating Soft-Power Threshold in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "pickSoftThreshold: will use block size 1180.\n",
      " pickSoftThreshold: calculating connectivity for given powers...\n",
      "   ..working on genes 1 through 1180 of 1180\n",
      "   Power SFT.R.sq slope truncated.R.sq  mean.k. median.k.  max.k.\n",
      "1      1  0.21900 12.20          0.812 5.99e+02  6.01e+02 638.000\n",
      "2      2  0.00378  0.77          0.953 3.08e+02  3.08e+02 356.000\n",
      "3      3  0.08520 -2.37          0.937 1.60e+02  1.59e+02 203.000\n",
      "4      4  0.44200 -4.36          0.865 8.38e+01  8.23e+01 120.000\n",
      "5      5  0.74300 -4.79          0.887 4.44e+01  4.29e+01  73.400\n",
      "6      6  0.90600 -4.66          0.952 2.38e+01  2.25e+01  46.000\n",
      "7      7  0.94500 -4.20          0.962 1.29e+01  1.19e+01  29.500\n",
      "8      8  0.94700 -3.82          0.954 7.09e+00  6.35e+00  19.400\n",
      "9      9  0.96400 -3.37          0.969 3.95e+00  3.39e+00  13.000\n",
      "10    10  0.96700 -3.05          0.973 2.23e+00  1.83e+00   8.900\n",
      "11    12  0.91900 -2.70          0.921 7.48e-01  5.37e-01   4.390\n",
      "12    14  0.93700 -2.29          0.950 2.69e-01  1.62e-01   2.290\n",
      "13    16  0.96700 -1.97          0.976 1.05e-01  4.96e-02   1.250\n",
      "14    18  0.94800 -1.84          0.960 4.42e-02  1.56e-02   0.748\n",
      "15    20  0.92500 -1.82          0.924 2.01e-02  4.96e-03   0.540\n",
      "16    22  0.36300 -2.43          0.218 9.87e-03  1.59e-03   0.409\n",
      "17    24  0.32800 -3.09          0.140 5.17e-03  5.18e-04   0.321\n",
      "18    26  0.33800 -3.15          0.157 2.89e-03  1.70e-04   0.259\n",
      "19    28  0.33700 -3.10          0.231 1.71e-03  5.83e-05   0.213\n",
      "20    30  0.47500 -3.50          0.326 1.07e-03  2.00e-05   0.178\n",
      "  Power    SFT.R.sq     slope truncated.R.sq   mean.k. median.k.    max.k.\n",
      "1     1 0.219433223 12.238293      0.8118715 599.40594 600.79649 638.18183\n",
      "2     2 0.003782682  0.769770      0.9525082 307.94993 307.95797 355.62201\n",
      "3     3 0.085209714 -2.368681      0.9365322 159.84615 158.59490 203.07626\n",
      "4     4 0.442448736 -4.358650      0.8653303  83.83284  82.28715 120.34491\n",
      "5     5 0.743370588 -4.787600      0.8867003  44.43663  42.85912  73.39941\n",
      "6     6 0.905584226 -4.658810      0.9520990  23.81745  22.48148  45.96407\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/SoftPower/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in celltypes){\n",
    "  \n",
    "  print(paste0('Estinmating Soft-Power Threshold in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # Test different soft powers:\n",
    "  seurat_obj[[cell_type]] <- TestSoftPowers(\n",
    "    seurat_obj[[cell_type]],\n",
    "    networkType = 'signed', # you can also use \"unsigned\" or \"signed hybrid\"\n",
    "    setDatExpr = FALSE,\n",
    "    group.by = 'cell_type',\n",
    "    group_name = cell_type,\n",
    "  )\n",
    "\n",
    "  # plot the results:\n",
    "  plot_list <- PlotSoftPowers(seurat_obj[[cell_type]], \n",
    "                              # wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment\n",
    "                              )\n",
    "\n",
    "  # assemble with patchwork\n",
    "  pdf(paste0(fig_dir, cell_type, '_SoftPower.pdf'), width=12, height=8)\n",
    "  print(wrap_plots(plot_list, ncol=2))\n",
    "  dev.off()\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f5b7f474",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in celltypes\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2ad0b902",
   "metadata": {},
   "source": [
    "The general guidance for WGCNA and hdWGCNA is to `pick the lowest soft power threshold that has a Scale Free Topology Model Fit greater than or equal to 0.8`. Later on, the `ConstructNetwork `will automatically select the soft power threshold if we do not provide one.\n",
    "\n",
    "The output table from the parameter sweep is stored in the hdWGCNA experiment and can be accessed using the GetPowerTable function for further inspection:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "62c54e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# power_table <- GetPowerTable(seurat_obj)\n",
    "# head(power_table)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7c6251e4",
   "metadata": {},
   "source": [
    "### **Construct co-expression network**\n",
    "\n",
    "We now have everything that we need to construct our co-expression network. Here we use the hdWGCNA function ConstructNetwork, which calls the WGCNA function blockwiseConsensusModules under the hood. This function has quite a few parameters to play with if for advanced use cases, but we use default parameters that work well with many single-cell datasets. \n",
    "\n",
    "The parameters for blockwiseConsensusModules can be passed directly to ConstructNetwork with the same parameter names.\n",
    "\n",
    "The following code construtcts the co-expression network using the soft power threshold selected above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "14831950",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Constructing co-expression network in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Soft power not provided. Automatically using the lowest power that meets 0.8 scale-free topology fit. Using soft_power = 6\n",
      " Calculating consensus modules and module eigengenes block-wise from all genes\n",
      " Calculating topological overlaps block-wise from all genes\n",
      "   Flagging genes and samples with too many missing values...\n",
      "    ..step 1\n",
      "    TOM calculation: adjacency..\n",
      "    ..will not use multithreading.\n",
      "     Fraction of slow calculations: 0.000000\n",
      "    ..connectivity..\n",
      "    ..matrix multiplication (system BLAS)..\n",
      "    ..normalization..\n",
      "    ..done.\n",
      " ..Working on block 1 .\n",
      " ..Working on block 1 .\n",
      " ..merging consensus modules that are too close..\n",
      "[1] \"Oligodendrocyte\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "for (cell_type in celltypes) {\n",
    "  \n",
    "  print(paste0('Constructing co-expression network in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "  \n",
    "  tryCatch({\n",
    "    # construct co-expression network:\n",
    "    seurat_obj[[cell_type]] <- ConstructNetwork(\n",
    "      seurat_obj[[cell_type]], \n",
    "      soft_power = NULL, # Set to NULL so that the ConstructNetwork function obtains it automatically (Scale Free Topology Model Fit >= 0.8)\n",
    "      setDatExpr = FALSE,\n",
    "      overwrite_tom = TRUE,\n",
    "      tom_name = toupper(substr(cell_type, 1, 3)) # name of the topological overlap matrix written to disk set as the name of the hdWGCNA experiment \n",
    "    )\n",
    "  }, error = function(e) {\n",
    "    print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "    print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "    # Code to handle the error condition\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "  },  \n",
    "  message = function(m) {\n",
    "    print(paste0('Error encountered while processing ', toupper(cell_type)))\n",
    "    print(paste0(toupper(cell_type), ' dropped from experiment'))\n",
    "    # Code to handle the error condition\n",
    "    seurat_obj <- seurat_obj[names(seurat_obj) != cell_type]\n",
    "    assign(\"seurat_obj\", seurat_obj, envir = .GlobalEnv) # Update the modified seurat_obj in the global environment\n",
    "\n",
    "  }) \n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "print(names(seurat_obj))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b8e77379",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "49676183",
   "metadata": {},
   "source": [
    "hdWGCNA also includes a function PlotDendrogram to visualize the WGCNA dendrogram, a common visualization to show the different co-expression modules resulting from the network analysis. Each leaf on the dendrogram represents a single gene, and the color at the bottom indicates the co-expression module assignment.\n",
    "\n",
    "\n",
    "`Importantly, the “grey” module consists of genes that were not grouped into any co-expression module. The grey module should be ignored for all downstream analysis and interpretation.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8da7dcca",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/dendrogram/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "tryCatch({\n",
    "    for (cell_type in names(seurat_obj)){\n",
    "        # print(paste0('Constructing co-expression networkd in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "        pdf(paste0(fig_dir, cell_type, \"_dendro.pdf\"), height=2, width=4)\n",
    "        PlotDendrogram(seurat_obj[[cell_type]], main='hdWGCNA Dendrogram')\n",
    "        dev.off()\n",
    "    }\n",
    "}, error = function(e){\n",
    "    NULL\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ed9baba4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a5dfb822",
   "metadata": {},
   "source": [
    "### **Optional: inspect the topoligcal overlap matrix (TOM)**\n",
    "\n",
    "hdWGCNA represents the co-expression network as a `topoligcal overlap matrix (TOM)`. This is a square matrix of genes by genes, where each value is the topoligcal overlap between the genes. The TOM is written to the disk when running ConstructNetwork, and we can load it into R using the GetTOM function. \n",
    "\n",
    "We may wish to inspect the TOM for custom downstream analyses.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "fbea29f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj -o TOM\n",
    "\n",
    "# TOM <- GetTOM(seurat_obj)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5e1c23ce",
   "metadata": {},
   "source": [
    "### **Module Eigengenes and Connectivity**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "08779263",
   "metadata": {},
   "source": [
    "### Compute harmonized module eigengenes\n",
    "\n",
    "`Module Eigengenes (MEs)` are a commonly used metric to summarize the gene expression profile of an entire co-expression module. Briefly, module eigengenes are computed by performing principal component analysis (PCA) on the subset of the gene expression matrix comprising each module. `The first PC of each of these PCA matrices are the MEs.`\n",
    "\n",
    "Dimensionality reduction techniques are a very hot topic in single-cell genomics. It is well known that technical artifacts can muddy the analysis of single-cell datasets, and over the years there have been many methods that aim to reduce the effects of these artifacts. Therefore it stands to reason that `MEs would be subject to these technical artifacts as well`, and hdWGCNA seeks to alleviate these effects.\n",
    "\n",
    "`hdWGCNA includes a function ModuleEigengenes` to compute module eigengenes in single cells. Additionally, we can apply Harmony batch correction to the MEs, yielding harmonized module eigengenes (hMEs).\n",
    "\n",
    "The following code performs the module eigengene computation `harmonizing by the Sample of origin using the group.by.vars parameter.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "df854ccb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating module eigen-genes in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"turquoise\"\n",
      "[1] \"grey\"\n",
      "[1] \"blue\"\n",
      "[1] \"brown\"\n",
      "[1] \"----------------\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Estimating module eigen-genes in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "    # need to run ScaleData first or else harmony throws an error:\n",
    "    seurat_obj[[cell_type]] <- ScaleData(seurat_obj[[cell_type]], features=VariableFeatures(seurat_obj[[cell_type]]))\n",
    "\n",
    "    # compute all MEs in the full single-cell dataset\n",
    "    seurat_obj[[cell_type]] <- ModuleEigengenes(\n",
    "    seurat_obj[[cell_type]],\n",
    "    group.by.vars=subject_id\n",
    "    )\n",
    "    print('----------------')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "05f2fc35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "05cd78cf",
   "metadata": {},
   "source": [
    "The ME matrices are stored as a matrix where each row is a cell and each column is a module. This matrix can be extracted from the Seurat object using the GetMEs function, which retrieves the hMEs by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "93840f67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -o hMEs -o MEs\n",
    "\n",
    "hMEs <- list()\n",
    "MEs <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    print(paste0('Getting harmonized and non-harmonized MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "    # harmonized module eigengenes:\n",
    "    hMEs[[cell_type]] <- GetMEs(seurat_obj[[cell_type]])\n",
    "\n",
    "    # module eigengenes:\n",
    "    MEs[[cell_type]] <- GetMEs(seurat_obj[[cell_type]], harmonized=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "8075bc3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c011bbb2",
   "metadata": {},
   "source": [
    "### Compute module connectivity"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e9f2e4ad",
   "metadata": {},
   "source": [
    "In co-expression network analysis, we often want to focus on the `“hub genes”`,`those which are highly connected within each module.` Therefore we wish to determine the `eigengene-based connectivity, also known as kME`, of each gene. \n",
    "\n",
    "hdWGCNA includes the ModuleConnectivity to compute the kME values in the full single-cell dataset, rather than the metacell dataset. This function essentially computes pairwise correlations between genes and module eigengenes. kME can be computed for all cells in the dataset, but we recommend computing kME in the cell type or group that was previously used to run ConstructNetwork."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9ebaef5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating module connectivity in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Estimating module connectivity in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # compute eigengene-based connectivity (kME):\n",
    "  seurat_obj[[cell_type]] <- ModuleConnectivity(\n",
    "    seurat_obj[[cell_type]],\n",
    "    group.by = 'cell_type', \n",
    "    group_name = cell_type\n",
    "  )\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "72891253",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "95b46026",
   "metadata": {},
   "source": [
    "For convenience, we re-name the hdWGCNA modules to indicate that they are from the Excitatory neuron group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a52dff32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Renaming modules in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Renaming modules in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # rename the modules\n",
    "  seurat_obj[[cell_type]] <- ResetModuleNames(\n",
    "    seurat_obj[[cell_type]],\n",
    "    new_name = paste0(toupper(substr(cell_type, 1, 3)), \"-M\"),\n",
    "    wgcna_name = toupper(substr(cell_type, 1, 3)) # the name of the hdWGCNA experiment,\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "0f353a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "88ad5a62",
   "metadata": {},
   "source": [
    "We also reset module colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5e7726f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Resetting module colors in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj \n",
    "\n",
    "\n",
    "\n",
    "library(MetBrewer)\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Resetting module colors in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "\n",
    "  modules <- GetModules(seurat_obj[[cell_type]])\n",
    "  mods <- levels(modules$module)\n",
    "  mod_colors <- dplyr::select(modules, c(module, color)) %>%\n",
    "    distinct %>% arrange(module) %>% .$color\n",
    "  n_colors <- length(mod_colors) -1\n",
    "\n",
    "  new_colors <- paste0(met.brewer(\"Signac\", n=n_colors,))\n",
    "  new_colors <- sample(new_colors)\n",
    "  seurat_obj[[cell_type]] <- ResetModuleColors(seurat_obj[[cell_type]], new_colors)\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6b194589",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ae334b97",
   "metadata": {},
   "source": [
    "We can visualize the genes in each module ranked by kME using the PlotKMEs function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "03423498",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer dplyr::select over any other package.\n",
      "[1] \"Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "library(conflicted)\n",
    "conflicts_prefer(dplyr::select)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/ModuleConnectivity/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Visualize Module Connectivity (kMEs) in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # plot genes ranked by kME for each module\n",
    "\n",
    "    p <- PlotKMEs(seurat_obj[[cell_type]], ncol=5)\n",
    "\n",
    "    pdf(paste0(fig_dir, cell_type, '_KMEs.pdf'), width=12, height=8)\n",
    "    print(p)\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c6b1f8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b8d0658b",
   "metadata": {},
   "source": [
    "### Getting the module assignment table\n",
    "\n",
    "hdWGCNA allows for easy access of the module assignment table using the GetModules function. \n",
    "\n",
    "This table consists of three columns: \n",
    "\n",
    "- gene_name stores the gene’s symbol or ID, \n",
    "- module stores the gene’s module assignment, and \n",
    "- color stores a color mapping for each module, \n",
    "\n",
    "which is used in many downstream plotting steps. If ModuleConnectivity has been called on this hdWGCNA experiment, this table will have additional columns for the kME of each module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9121be04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Obtain Module Assignment Table in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "        gene_name module   color kME_OLI-M1  kME_grey kME_OLI-M2\n",
      "CPEB3       CPEB3 OLI-M1 #fe9b00  0.2819247 0.2914115  0.2233105\n",
      "RC3H1       RC3H1   grey    grey  0.1938126 0.2146999  0.1537275\n",
      "PAPSS1     PAPSS1   grey    grey  0.2007681 0.2236071  0.1768358\n",
      "TPST1       TPST1 OLI-M1 #fe9b00  0.2074987 0.2218625  0.1721418\n",
      "ATF6         ATF6   grey    grey  0.2238446 0.2429879  0.1517988\n",
      "HSP90B1   HSP90B1   grey    grey  0.2109174 0.2077080  0.1697903\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Obtain Module Assignment Table in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # get the module assignment table:\n",
    "    modules <- GetModules(seurat_obj[[cell_type]])\n",
    "\n",
    "    # show the first 6 columns:\n",
    "    print(head(modules[,1:6]))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "21bdbbf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "48d2e707",
   "metadata": {},
   "source": [
    "A table of the top N hub genes sorted by kME can be extracted using the GetHubGenes function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "3a0b8898",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Obtain Hub Genes in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "  gene_name module       kME\n",
      "1      RTN4 OLI-M1 0.5450107\n",
      "2     TCF12 OLI-M1 0.5516767\n",
      "3    SPOCK3 OLI-M1 0.5567026\n",
      "4  HSP90AA1 OLI-M1 0.5720840\n",
      "5     PDE4B OLI-M1 0.5923300\n",
      "6   PIP4K2A OLI-M1 0.5926874\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Obtain Hub Genes in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # get hub genes\n",
    "    hub_df <- GetHubGenes(seurat_obj[[cell_type]], n_hubs = 10)\n",
    "\n",
    "    print(head(hub_df))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8f478886",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c0739db5",
   "metadata": {},
   "source": [
    "This wraps up the critical analysis steps for hdWGCNA, let's save our output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ba8f4c61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Saving hdWGCNA object in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/SeuratObject/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "    print(paste0('Saving hdWGCNA object in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    saveRDS(seurat_obj[[cell_type]], file=paste0(dat_dir, cell_type, '_hdWGCNA_object.rds'))\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "05bbd46f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "27f2d80d",
   "metadata": {},
   "source": [
    "## **Compute hub gene signature scores**\n",
    "\n",
    "Gene scoring analysis is a popular method in single-cell transcriptomics for computing a score for the overall signature of a set of genes. Seurat implements their own gene scoring technique using the `AddModuleScore` function, but there are also alternative approaches such as UCell.\n",
    "\n",
    "hdWGCNA includes the function `ModuleExprScore` to compute gene scores for a give number of genes for each module, using either the Seurat or UCell algorithm. Gene scoring is an alternative way of summarizing the expression of a module from computing the module eigengene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e1f93a8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Compute hub gene scores in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n",
      "Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Compute hub gene scores in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    \n",
    "  # compute gene scoring for the top 25 hub genes by kME for each module\n",
    "  # with Seurat method\n",
    "  seurat_obj[[cell_type]] <- ModuleExprScore(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_genes = 25,\n",
    "    method='Seurat'\n",
    "  )\n",
    "\n",
    "  # compute gene scoring for the top 25 hub genes by kME for each module\n",
    "  # with UCell method\n",
    "  library(UCell)\n",
    "  seurat_obj[[cell_type]] <- ModuleExprScore(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_genes = 25,\n",
    "    method='UCell'\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "34de1c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "03dc3f32",
   "metadata": {},
   "source": [
    "## **Basic Visualization**\n",
    "\n",
    "Here we visualize some results using hdWGCNA, and we employ some of Seurat’s built-in plotting tools to visualize our hdWGCNA results.\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "368c8716",
   "metadata": {},
   "source": [
    "### **Module Feature Plots**\n",
    "\n",
    "FeaturePlot is a commonly used Seurat visualization to show a feature of interest directly on the dimensionality reduction. hdWGCNA includes the ModuleFeaturePlot function to consruct FeaturePlots for each co-expression module colored by each module’s uniquely assigned color."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c5865383",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module Feature Plots in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/FeaturePlots/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Module Feature Plots in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "     \n",
    "  # make a featureplot of hMEs for each module\n",
    "  plot_list <- ModuleFeaturePlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    features='hMEs', # plot the hMEs\n",
    "    order=TRUE, # order so the points with highest hMEs are on top\n",
    "    # raster=TRUE, \n",
    "    # raster_dpi=400, \n",
    "    # alpha=1,\n",
    "    # restrict_range=FALSE, \n",
    "    # raster_scale=0.25,\n",
    "  )\n",
    "\n",
    "  plot_list <- lapply(1:length(plot_list), function(x){\n",
    "    plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_featureplot_MEs.pdf\"), height=10, width=5)\n",
    "  print(wrap_plots(plot_list, ncol=3))\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "7d1a4b43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2e09999a",
   "metadata": {},
   "source": [
    "We can also plot the hub gene signature score using the same function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "07825465",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module Feature Plots for HUB GENES in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/FeaturePlots/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "    \n",
    "  print(paste0('Module Feature Plots for HUB GENES in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "     \n",
    "  # make a featureplot of hMEs for each module\n",
    "  plot_list <- ModuleFeaturePlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    features='scores',  # plot the hub gene scores\n",
    "    order='shuffle', # order so cells are shuffled\n",
    "    ucell = TRUE, # depending on Seurat vs UCell for gene scoring\n",
    "    # raster=TRUE, \n",
    "    # raster_dpi=400, \n",
    "    # alpha=1,\n",
    "    # restrict_range=FALSE, \n",
    "    # raster_scale=0.25,\n",
    "  )\n",
    "\n",
    "  plot_list <- lapply(1:length(plot_list), function(x){\n",
    "    plot_list[[x]] + NoLegend() + theme(plot.title=element_text(face='plain', vjust=0.25), plot.margin=margin(c(0,0,0,0)))\n",
    "  })\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_featureplot_hubs.pdf\"), height=10, width=5)\n",
    "  print(wrap_plots(plot_list, ncol=3))\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "dbbce473",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "219d5785",
   "metadata": {},
   "source": [
    "### **Module Correlations**\n",
    "\n",
    "hdWGCNA includes the ModuleCorrelogram function to visualize the correlation between each module based on their hMEs, MEs, or hub gene scores using the R package corrplot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9ac695b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module Correlation in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/ModuleCorrelations/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('Module Correlation in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "    # plot module correlagram\n",
    "    pdf(paste0(fig_dir, cell_type, \"_correlogram.pdf\"), height=10, width=5)\n",
    "    ModuleCorrelogram(seurat_obj[[cell_type]])\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "29692fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3cf396dd",
   "metadata": {},
   "source": [
    "### **Seurat plotting functions**\n",
    "\n",
    "\n",
    "The base Seurat plotting functions are also great for visualizing hdWGCNA outputs. Here we demonstrate plotting hMEs using DotPlot and VlnPlot. The key to using Seurat’s plotting functions to visualize the hdWGCNA data is to add it into the Seurat object’s @meta.data slot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4dd017bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -o mods\n",
    "\n",
    "mods <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    # get hMEs from seurat object\n",
    "    MEs <- GetMEs(seurat_obj[[cell_type]], harmonized=TRUE)\n",
    "    mods[[cell_type]] <- colnames(MEs); \n",
    "    mods[[cell_type]] <- mods[[cell_type]][mods[[cell_type]] != 'grey']\n",
    "    \n",
    "    \n",
    "    # add hMEs to Seurat meta-data:\n",
    "    seurat_obj[[cell_type]]@meta.data <- cbind(seurat_obj[[cell_type]]@meta.data, MEs)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "6c8c3009",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )\n",
    "\n",
    "\n",
    "# convert nested list of Seurat object into Rpy2 object \n",
    "mods =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: mods[cell_type]\n",
    "                     \n",
    "                            for cell_type in mods.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c1c37b40",
   "metadata": {},
   "source": [
    "Now we can easily use Seurat’s DotPlot function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4f45a3f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"DotPlot for average expression of MEs in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Scale for colour is already present.\n",
      "Adding another scale for colour, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -i mods -o seurat_obj\n",
    "\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/DotPlot/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('DotPlot for average expression of MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # plot with Seurat's DotPlot function\n",
    "  p <- DotPlot(seurat_obj[[cell_type]], features=mods[[cell_type]], group.by = 'cell_type')\n",
    "\n",
    "  # flip the x/y axes, rotate the axis labels, and change color scheme:\n",
    "  p <- p +\n",
    "    coord_flip() +\n",
    "    RotatedAxis() +\n",
    "    scale_color_gradient2(high='red', mid='grey95', low='blue')\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, \"_dotplot.pdf\"), height=10, width=5)\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "31592f99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b1b75ca0",
   "metadata": {},
   "source": [
    "Here is another example where we use Seurat’s VlnPlot function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "5c23ebda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "# for (cell_type in names(seurat_obj)){\n",
    "\n",
    "#   print(paste0('ViolinPlot for average expression of MEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "#   # Plot INH-M4 hME using Seurat VlnPlot function\n",
    "#   p <- VlnPlot(\n",
    "#     seurat_obj,\n",
    "#     features = as.vector(mods[[cell_type]]),\n",
    "#     group.by = 'cell_type',\n",
    "#     pt.size = 0 # don't show actual data points\n",
    "#   )\n",
    "\n",
    "#   # add box-and-whisker plots on top:\n",
    "#   p <- p + geom_boxplot(width=.25, fill='white')\n",
    "\n",
    "#   # change axis labels and remove legend:\n",
    "#   p <- p + xlab('') + ylab('hME') + NoLegend()\n",
    "\n",
    "#   # plot output\n",
    "#   pdf(paste0(\"../results/hdWGCNA/ViolinPlot/\", save_prefix, '_', cell_type, \"_violinplot.pdf\"), height=10, width=5)\n",
    "#   print(p)\n",
    "#   dev.off()\n",
    "\n",
    "# }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "8246a79f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # convert nested list of Seurat object into Rpy2 object \n",
    "# seurat_obj =  robjects.ListVector(\n",
    "#                         {\n",
    "#                             cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "#                             for cell_type in seurat_obj.keys()\n",
    "#                         }\n",
    "#                     )\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9b2ac7d7",
   "metadata": {},
   "source": [
    "# **Network Visualization**\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "20a45cef",
   "metadata": {},
   "source": [
    "Here, we visualize the co-expression networks made with hdWGCNA using the following functions:\n",
    "\n",
    "- `ModuleNetworkPlot`, visualizes a separate network plot for each module, showing the top N genes by kME.\n",
    "- `HubGeneNetworkPlot`, visualizes the network comprisng all modules with a given number of hub genes per module.\n",
    "- `ModuleUMAPPlot`, visualizes all of the genes in the co-expression simultaneously using the UMAP dimensionality reduction algorithm."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f720e63b",
   "metadata": {},
   "source": [
    "## **Individual module network plots**\n",
    "\n",
    "\n",
    "Here we demonstrate using the `ModuleNetworkPlot` function to visualize the network underlying the top 25 hub genes for each module. By default, this function creates a new folder called “ModuleNetworks”, and generates a .pdf figure for each module.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "8537d71c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Module network plots in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Writing output files to ../results/hdWGCNA/NetworkPlot/leng_etc/Oligodendrocyte_hubNetworks/Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0('../results/hdWGCNA/NetworkPlot/', save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Module network plots in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # individual module networks\n",
    "  ModuleNetworkPlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    mods = \"all\",\n",
    "    outdir = paste0(fig_dir, cell_type, '_hubNetworks/')\n",
    "  )\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "311f315f",
   "metadata": {},
   "source": [
    "In this network, each node represents a gene, and each edge represents the co-expression relationship between two genes in the network. Each of these module network plots are colored based on the color column in the hdWGCNA module assignment table GetModules(seurat_obj). The top 10 hub genes by kME are placed in the center of the plot, while the remaining 15 genes are placed in the outer circle.\n",
    "\n",
    "Optionally, certain visualization parameters can be changed in this plot:\n",
    "\n",
    "- `edge.alpha`: determines the opacity of the network edges\n",
    "- `vertex.size`: determines the size of the nodes\n",
    "- `vertex.label.cex`: determines the font size of the gene label\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "db743ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e00f5d04",
   "metadata": {},
   "source": [
    "## **Combined hub gene network plots**\n",
    "\n",
    "Here we will make a network plot combining all of the modules together using the `HubGeneNetworkPlot` function. \n",
    "\n",
    "This function takes the top n hub genes, and other randomly selected genes, and constructs a joint network using the force-directed graph drawing algorithm. For visual clarity, the number of edges in the network can be downsampled using the edge_prop parameter. In the following example, we visualize the top 3 hub genes and 6 other genes per module.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "eca4578b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Hubs Gene Network Plot in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n",
      "\n",
      "OLI-M1   grey OLI-M2 OLI-M3 \n",
      "     9      3      9      9 \n",
      "[1] \"#fe9b00\" \"grey90\"  \"#d8443c\" \"#f4c40f\"\n",
      "[1] \"here\"\n",
      "[1] 179   4\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "\n",
    "  fig_dir = paste0(\"../results/hdWGCNA/HubGeneNetworkPlot/\", save_prefix, '/')\n",
    "  if (!dir.exists(fig_dir)) {\n",
    "    dir.create(fig_dir, recursive=TRUE)\n",
    "  }\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Hubs Gene Network Plot in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # hubgene network\n",
    "\n",
    "  # Save plot to PDF\n",
    "  pdf(file = paste0(fig_dir,  cell_type, \"_allHubGeneNetwork.pdf\"), width = 4, height = 4, useDingbats = FALSE)\n",
    "  HubGeneNetworkPlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_hubs = 6,\n",
    "    n_other=3,\n",
    "    edge_prop = 0.75,\n",
    "    mods = 'all',\n",
    "    sample_edges = TRUE,\n",
    "    return_graph = FALSE,\n",
    "    edge.alpha = 0.25,\n",
    "    vertex.label.cex = 0.25,\n",
    "    hub.vertex.size = 6,\n",
    "    other.vertex.size = 2,\n",
    "    wgcna_name = NULL,\n",
    "\n",
    "  )\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d00bdc4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fc1be299",
   "metadata": {},
   "source": [
    "As in the previous network plot, each node represents a gene and each edge represents a co-expression relationship. In this network, `we color intramodular edges with the module’s color, and intermodular edges gray.` The opacity of edges in this network is scaled by the strength of the co-expression relationship."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "289931dc",
   "metadata": {},
   "source": [
    "## **Applying UMAP to co-expression networks**\n",
    "\n",
    "We have only visualized a subset of the co-expression network with an emphasis on the hub genes. Here, we use an alternative approach to visualize all genes in the co-expression network simultaneously. UMAP is a suitable method for visualizing high-dimensional data in two dimensions, and here we apply UMAP to embed the hdWGCNA network in a low-dimensional manifold.\n",
    "\n",
    "hdWGCNA includes the function `RunModuleUMAP to run the UMAP algorithm on the hdWGCNA topological overlap matrix (TOM)`. For the UMAP analysis, we subset the columns in the TOM to only contain the top n hub genes by kME for each module. \n",
    "\n",
    "Therefore, the organization of each gene in UMAP space is dependent on that gene’s connectivity with the network’s hub genes. This function leverages the UMAP implementation from the uwot R package, so additional UMAP parameters for the uwot::umap function such as min_dist or spread can be included in RunModuleUMAP.\n",
    "\n",
    "The following code demonstrates using the RunModuleUMAP function with 10 hub genes per module:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "179e3598",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Obtaining Module UMAP in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Obtaining Module UMAP in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  seurat_obj[[cell_type]] <- RunModuleUMAP(\n",
    "    seurat_obj[[cell_type]],\n",
    "    n_hubs = 5, # number of hub genes to include for the UMAP embedding\n",
    "    n_neighbors=15, # neighbors parameter for UMAP\n",
    "    min_dist=0.3, # min distance between points in UMAP space\n",
    "    spread = 2,\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "f507f011",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4a2e8943",
   "metadata": {},
   "source": [
    "Next we will make a simple visualization of the UMAP for Excitatory cells using ggplot2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "16d8a540",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Store Module UMAP in hdWGCNA Experiment for OLIGODENDROCYTE\"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAYAAAB91L6VAAAEDmlDQ1BrQ0dDb2xvclNwYWNlR2VuZXJpY1JHQgAAOI2NVV1oHFUUPpu5syskzoPUpqaSDv41lLRsUtGE2uj+ZbNt3CyTbLRBkMns3Z1pJjPj/KRpKT4UQRDBqOCT4P9bwSchaqvtiy2itFCiBIMo+ND6R6HSFwnruTOzu5O4a73L3PnmnO9+595z7t4LkLgsW5beJQIsGq4t5dPis8fmxMQ6dMF90A190C0rjpUqlSYBG+PCv9rt7yDG3tf2t/f/Z+uuUEcBiN2F2Kw4yiLiZQD+FcWyXYAEQfvICddi+AnEO2ycIOISw7UAVxieD/Cyz5mRMohfRSwoqoz+xNuIB+cj9loEB3Pw2448NaitKSLLRck2q5pOI9O9g/t/tkXda8Tbg0+PszB9FN8DuPaXKnKW4YcQn1Xk3HSIry5ps8UQ/2W5aQnxIwBdu7yFcgrxPsRjVXu8HOh0qao30cArp9SZZxDfg3h1wTzKxu5E/LUxX5wKdX5SnAzmDx4A4OIqLbB69yMesE1pKojLjVdoNsfyiPi45hZmAn3uLWdpOtfQOaVmikEs7ovj8hFWpz7EV6mel0L9Xy23FMYlPYZenAx0yDB1/PX6dledmQjikjkXCxqMJS9WtfFCyH9XtSekEF+2dH+P4tzITduTygGfv58a5VCTH5PtXD7EFZiNyUDBhHnsFTBgE0SQIA9pfFtgo6cKGuhooeilaKH41eDs38Ip+f4At1Rq/sjr6NEwQqb/I/DQqsLvaFUjvAx+eWirddAJZnAj1DFJL0mSg/gcIpPkMBkhoyCSJ8lTZIxk0TpKDjXHliJzZPO50dR5ASNSnzeLvIvod0HG/mdkmOC0z8VKnzcQ2M/Yz2vKldduXjp9bleLu0ZWn7vWc+l0JGcaai10yNrUnXLP/8Jf59ewX+c3Wgz+B34Df+vbVrc16zTMVgp9um9bxEfzPU5kPqUtVWxhs6OiWTVW+gIfywB9uXi7CGcGW/zk98k/kmvJ95IfJn/j3uQ+4c5zn3Kfcd+AyF3gLnJfcl9xH3OfR2rUee80a+6vo7EK5mmXUdyfQlrYLTwoZIU9wsPCZEtP6BWGhAlhL3p2N6sTjRdduwbHsG9kq32sgBepc+xurLPW4T9URpYGJ3ym4+8zA05u44QjST8ZIoVtu3qE7fWmdn5LPdqvgcZz8Ww8BWJ8X3w0PhQ/wnCDGd+LvlHs8dRy6bLLDuKMaZ20tZrqisPJ5ONiCq8yKhYM5cCgKOu66Lsc0aYOtZdo5QCwezI4wm9J/v0X23mlZXOfBjj8Jzv3WrY5D+CsA9D7aMs2gGfjve8ArD6mePZSeCfEYt8CONWDw8FXTxrPqx/r9Vt4biXeANh8vV7/+/16ffMD1N8AuKD/A/8leAvFY9bLAAAAOGVYSWZNTQAqAAAACAABh2kABAAAAAEAAAAaAAAAAAACoAIABAAAAAEAAAHgoAMABAAAAAEAAAHgAAAAAKWfY0oAAEAASURBVHgB7J0FfBzH9cd/u3uWZJmZObZjdsBBh9FhZmamNin806ZpG2ybNA0zM3McjpM47NiOmZlZsi243f2/N6c7i3Un3eno9z6xtbc7O/Adxe9m5oHli4BCAiRAAiRAAiTQqATsRm2NjZEACZAACZAACRgCVMD8RSABEiABEiCBJBCgAk4CdDZJAiRAAiRAAlTA/B0gARIgARIggSQQoAJOAnQ2SQIkQAIkQAJUwPwdIAESIAESIIEkEKACTgJ0NkkCJEACJEACVMD8HSABEiABEiCBJBCgAk4CdDZJAiRAAiRAAlTA/B0gARIgARIggSQQoAJOAnQ2SQIkQAIkQAJUwPwdIAESIAESIIEkEKACTgJ0NkkCJEACJEACVMD8HSABEiABEiCBJBCgAk4CdDZJAiRAAiRAAlTA/B0gARIgARIggSQQoAJOAnQ2SQIkQAIkQAJUwPwdIAESIAESIIEkEKACTgJ0NkkCJEACJEACVMD8HSABEiABEiCBJBCgAk4CdDZJAiRAAiRAAlTA/B0gARIgARIggSQQoAJOAnQ2SQIkQAIkQAJUwPwdIAESIAESIIEkEKACTgJ0NkkCJEACJEACVMCN8Duw+bffMO3kkzHtxBPhFRc3QotsggRIgARIINUJWL5IqncyXfu38sUXsfLxx1G6enVkCJ3OOw/dr7468pkXJEACJEAC2UmAK+AEzXvhhAlYcuedFZSvNrXuww/huy5WvfIKFt16K/zS0gT1gNWSAAmQAAmkMoFAKncunfsWaNsWdm4uvK1bKwzD3bwZ8//0J2z44gujiEuWLcN2991XoQw/kAAJkAAJZD4BroATNMd5vXtjwBNPoM2YMaaF/MGD0WL33THs3XdRvHQp/JISQFbCm376CbMuvhgzzjoLHlfDCZoNVksCJEACqUeAZ8CNPCcrnnkGS//zH9NqoH17+MEg3A0bgEAAnc85B92uvBLFy5djsWxP5/boge7XXw/Lshq5l2yOBEiABEgg0QS4Ak4Q4Y3jx2PKEUdAt5zDUrRwYUT5ilbFQDHQ6ijW0VZODizbRl6/fnALCzFFVs0bx43DqpdewuqXXw6/zp8kQAIkQAIZRIBnwAmazDmXXgpVsvNkBdv/gQdMK7qibTpoELZOn44eN9wA3abuIuWcli2R260bWu+3H9aNHQuIMtbtaf1TLGfEFBIgARIggcwjwC3oBM3p+k8/xfwbb0TbQw5B/sCB6HjaaVG1pOfA08VfuGjBAnN+3OeWW8zqOKqXWYgESIAESCBtCFABJ3CqJu69N9xNm8yKVs92O597bo2tqTu2bjdv/Oor5PbqZVbIui2tZ8Qrn3sOnc48E5bj1Pg+H5AACZAACaQXAZ4BJ3C+ArK1DI1zIlvJS//731pbWi7b1EvEOGuTnB2vee01LH/oIVN+0W23Yendd2PKUUfV+j4fkgAJkAAJpBcBKuAEzlc/8e+1xBdYxWnTptaWCidODLkmSSkNzqEGWCp6LqzSWSJoUUiABEiABDKHALegG3EuJx98MFqNHo1ef/1rlVaXP/II9E84MpbVtCl2/P77KuV4gwRIgARIIDMIcAXciPNYunIl1rzxRoUWS+TeiqeeQic5H2594IFo2r8/OojB1sivv65Qjh9IgARIgAQyiwBXwI04n+oTrD6/dpMmkVannXQSts6cif6y+m25666R+5UvNk+ejGbDh1e+zc8kQAIkQAJpSoAKOIkT50k4ysJJk7BZEjd0kXCUNcm8P/wB6z/6CJ0kUlb3a6+tqRjvkwAJkAAJpBEBbkEnabJ8z8OUI4/E7AsugC8BO7bOnh05/63cpc1TpogVl4NN335b+RE/kwAJkAAJpCkBroCTNHEalnLqMccAoojLy06yIq4sGpxj+YMPmlWyZliikAAJkAAJpD8BroCTNId5EmxDraFze/bEdmWhKpt07Fhtb/TMuNtVV5n0htUW4E0SIAESIIG0I8AVcNpNGTtMAiRAAiSQCQS4As6EWeQYSIAESIAE0o4AFXDaTRk7TAIkQAIkkAkEqIAzYRY5BhIgARIggbQjQAWcdlPGDpMACZAACWQCASrgTJhFjoEESIAESCDtCFABp8mUrf/iC0zcZx/8MmIEipctS5Nes5skQAIkQAI1EaACrolMit1f+Je/wN2wwfRqWZnfcIp1kd0hARIgARKIgQAVcAywklm0w8knhxI5SJrCtmPGJLMrbJsESIAESCAOBBiIIw4QG6uKjZKiMNCuHZoNHlxnkxpbOrd37wqZl+p8iQVIgARIgAQajQAVcKOhbryGVj7/PJbec49pcMTnn8Np3rzxGmdLJEACJEACURHgFnRUmNKr0Mqnn4ZfXGw6rekOKSRAAiRAAqlHgAo49eakwT3q97//mTpa7r47Wu25Z5X63IICLLj5Zkw9/ngULVpU5TlvkAAJkAAJJJ4At6ATzzjlWpi0//4Irl1r+tWkUycMfu01s02tK2ddMfe9807YOTkp1292iARIgAQyiQBXwBk0m77r4rfDDjO+wiXLl8P3fcy64AIs/Mc/KozSadEi8rlUFHHJqlVY+t//Qt2bNo4bh0W33BJ5zgsSIAESIIHEEKACTgzXpNSqVtKlK1cCgQBWPPkk1rz6Kgp++glrZIVb/iy49b77butfMIi1b78Ny3GgChzyp+DHH7c95xUJkAAJkEBCCAQSUisrTQqBVhIpq83BB5sVbdMBA7Duww+R0707Wu29N5pLBK2wNBs6FFZeHvyiInOr8OefMfCZZ7Dhm2+Q27Ur+vzzn+Gi/EkCJEACJJAgAlTACQKbjGoty0Kf227DurFjMf/PfwZkdWvl5lbxG26x666hM94mTUw3u117Lda8/jqKZs1CTseOKL9FnYxxsE0SIAESyAYCVMAZOMvFixcb5atDU3ekwokT0e7IIyMjDbRsieGffWa2mnPECKtp//4IrltnnreWVfSWmTPRtG9fWGUKOvIiL0iABEiABOJGgFbQcUOZOhUVLViAqUcfDTs/H77nocWoUdBt5jYSwrL3TTfV2tGNsg095/LLTdjLHeX8mEICJEACJJAYAlwBJ4ZrUmvNkxCUwyV7UuGECShZsQJL7rrLGFdtkKhYm489Fs2GD6+xf5t/+w1Qg6ySkhrL8AEJkAAJkEDDCVABN5xhytWgSle3odsceKBxSQp3ULMpaXzo2qTrpZci0KoVmu+wQ23F+IwESIAESKCBBLgF3UCAqfa6Kt5pkjlJnIChxlat9toLi/7+d3Q84wx0PO005HbrlmpdZn9IgARIICsJUAFn2LQvlu3mVRLRSiXQujVGfPVVho2QwyEBEiCBzCDAQByZMY+RUei2s4q6EjURlyIKCZAACZBAahLgCjg156VBvSpesgRbxae3pWw/23QlahBLvkwCJEACiSJABZwosqyXBEiABEiABGohwC3oWuDwEQmQAAmQAAkkigAVcKLIsl4SIAESIAESqIUAFXAtcPiIBEiABEiABBJFgAo4UWRZLwmQAAmQAAnUQoCRsGqBw0fVE/BKSzFXMiippXWT9u0xUPyOaW1dPSveJQESIIGaCHAFXBMZ3q+RwCLJF7xp/HiUrlxplPCa116rsSwfkAAJkAAJVE+ACrh6LrxbCwGnWTOT3EGL+JJz2KQ/rKU8H5EACZAACVQlQAVclUlG3tEsRxskQ1I8pNM555hqmkguYf3T9bLL4lEt6yABEiCBrCLAQBxZMN1bpk/HdEnEAMtCjxtuQMdTTql21IWTJyNHFKr+qUvcLVtQvHAhcnv2hFkR1/UCn5MACZAACVQgwBVwBRyZ+WGLGEvZOTkmQ9KGzz6rdpDLH3sMcy6/HL8dfDBUYau4BQUmneH6albOTn4+8gcNovKtliZvkgAJkEDdBKiA62aU9iXaH3002p9wAlrvuy8C7dpBrZgrixpSuZs2mdurXnnF/Fx4883m57xrrqlcPKrPuu29+vXXJTOiH1V5FiIBEiCBbCLALegsmu2JkpxBlWygQwcMe/992Lm5kdHr+fBcUbSBtm0x6MUXkdO5s1HU8//4R3S78krk9e4dKRvNReGvv2LWhRcCto1Wo0ejn6RJpJAACZAACWwjQAW8jUXGX/0yYkRkjC122QUDHn008lkvipcuNWkMAy1bVrhfnw9L778fK2RbG55nXt9p0qT6VMN3SIAESCBjCXALOmOnturAul59tbmpuYK7X399lQK53bqhOuXrixKdetxx5jx449dfV3mvuhudzjwTViAAR5R579tuq64I75EACZBAVhPgCjirpz+6wW8cNw7z/vAHeGL57LRqhZHymUICJEACJNAwAlwBN4xfVryt29V206awmjRBzz//OSvGzEGSAAmQQKIJcAWcaMKsnwRIgARIgASqIcAVcDVQeIsESIAESIAEEk2ACjjRhNOkft91seiOOzBxn30w8/zz4RUVpUnP2U0SIAESSE8CVMDpOW/16rVaM2+dPbvad5f8+99Y/cILcDdsQOHPP2Ph3/5WbbnyN9eKL7G6NpVIVqTGErewEMseeggFP/7YWE2yHRIgARJICAEq4IRgTc1KF8sKd5pExFLFWVk0QlZ52fjddyiYMAHTzzgDa999t/yjyPWCMoOste+8E7mX6IvJBx2E5Q8/jNmSAKJAvihQSIAESCBdCVABp+vM1aPf7SQkpfoAt5Zt5srS+dxzkdurF+A4sJs3h8Z6niX3tkg4yYXixxuOD13+vZHffouO4u/bWbasGyrrPvkktJpesaLWqtQS2wT3kMQSmyV5BIUESIAE0pUAFXC6zlw9+t1s8GCM/OYbOKJgK4slineorGS3u/deeLLNW7JsWaSIv3mzCagRuVF2ofX0+P3vYUm4yYaKbn+rVKfoy9fdoyyASLPhw9HprLPKP+I1CZAACaQVAbohpdV0Jb6zuqqcJykLS5YvjzSmK+IdZLWbSFEjsMKJE9Fip52qNKOpD7fOmIF8+QJh5+VFnhctWgRNGJEnK/eeN95Y4YuAnk23GTMGfW+/PVKeFyRAAiSQSgQCqdQZ9iX5BJpIEgZNoFBedEVcWfT8tcXOO1e+Xe/PugKvTvn6krlpyhFHQBW0t3UrhstWdUCicalMPfJI81OzLuUPG4YOxx5rPutfbeWdzuecE/nMCxIgARJINQIV/6VNtd6xP41OIKdjR/SWVaWeteb06AFLMib116QK5WT+n/6EWXLuq+5KiRZVrn4waKyzNaexfg6LJdG5VPziYpRIIony0ueWW9C0f//yt3hNAiRAAilFgCvglJqO1OhMi1GjMPzzz+Fu3IgmnTpBFV95KRTraIgRlLorJVJ0G3yBfBnQfljNmsGVs+jmI0eaJjdI/1AurzH9lhM5E6ybBEggEQS4Ak4E1QyoU7Mi5coKuLLy1aENfuMNtBJL6mFjxyZ0pNNOOgnFCxYYZd9CFK+2FzYg80T5WuW+GGydNSuufSn45RdojmTf9+NaLysjARIggTABKuAwiQz6Of2UU4xLT/HixQkZlSOr0e3uuQc5el6cQFG3Kd0KV0Xbco89oNvjYWm9//7I69sXgfbtkdu9O/rEMeXhymeewdxrroHZar/wQmgAEwoJkAAJxJsAt6DjTTQF6svt08e481QOrpECXYupC92vu86seJuKom0mK+CVzz9vfJhV4dqimAfJ5+IlSxBo29b4LcdUeS2FV2lEsE2bTAn9EqMuWdpmqohaha+T4CjtJEezcqCQAAmkJwG6IaXnvGVVr1XhTNx9dxMkRH2OB7/+unE9igXCxq+/Rss996zgqlTT+4tuvRVrpA01/lKL8B2+/x62GKOFRV2c8uRLzpC33grfqvJTw3NOOfxwDH7tNeT17l3lud6Yfuqp2DJtGnaaNCnyXM+5ta9af/7AgZH75S8myjjUKjyna1cMkeMACgmQQHoS4BZ0es5bVvW6aN48BFq3BkTp6Ha0fo5FVjz1FOZccQUW/OUv1b5WunYtgmUrXi3QQ/ygdfs7r18/DP3oowrKN1xBeX/k8L3yP9dJuE91odoi/ss1SW7PnhVcvsIuVwv//ndMl/PvDV9+idI1a7B5ypTQl4GyilRJq0tW0dy5NVXN+yRAAmlAgCvgNJikbO+iGkLN+93vsElWomoYNujFF6NayYa5bREDreknnogBTzxRxde4UJTbnEsvNVvOQ95+O7JanX7aadgydSpUSQ6tIRZ2uP6afur2eCxb1+piNeeqqxBct85U2Vys0VXJ6qpfz6G1H2qEpitwlS4SD3u1sAiuX2+u84RN28MOM8/4FwmQQOoToAJO/TliDxNAwCspMYpurqx2t5RtAbc59FD0lYQVKroNrKvmfvfdh9Z77ZWAHlSsUpNfrJAvCMa1S42+ZMs7RwzMwv7NViCAHuJ/3UGSaegXio3jxplV8fJHHwV0q1xFgpmo/3NbiQBGIQESSH0CNMJK/TliD+NMQFMazpQoWcGCgsjWriNuV10uuijSUitRuuXPZiMPEnCxVbbUdXs8uHp1qHYxrGorWZ90210NwsLJJ8IuYfkDBkD/6Mp3+YMPQgOS+LIlrVv0+UOGJKCHrJIESCARBKiAE0GVdaY0gQU33YStc+ZICC0furLscsklZmXZpEOHpPR7rWx9R5Sv9kDOjnXrecDjj2O9hN5UV6xWo0ej3VFHVehfoE2byJcEPcP25Gw4p0sX846ujPWsWA25+t55pxlnhZf5gQRIIOkEaISV9ClgBxqbQI5E91Llq6Lny2r8lCzlq30oH1BEP6t4El5T/a0HiCJVg69uV14ZelDD3xo4RZXvpp9+Msk0ts6ciRI5g9at6kVMSFEDNd4mgeQSoAJOLn+2ngQCnS+4wKwIVWGpj3GXiy9OQi+2NdlOEkeoS1FYNP52p7PPNh+nyqq3SFbrc8UILRpZdv/9oS3rssL65aJAlHJpmWFXNHWwDAmQQOMQoBFW43BmKylGQI2wSlasMKvGVAhmoX7Dyx9+2JxJtz7wQLTee29DTC2pZ4gyHvbee7DLkk/UhlJTSa6vFCJUM1x1FUtvDa/ZW7bfddudQgIkkHwCVMDJnwP2IMUJqJGUJoRoJikPU1156Rb0/D/8AUHxbVbR8+NmO+yAwh9/NFbS6qbU55//THHi7B4JZAcBfhXOjnnmKOtJYJ0E4lgsRkxhgy31FXby8+tZW+Jfaym+w70lkMeSf//bKNw2upqWP9NPPtl8eVCFTCEBEkgNAlwBp8Y81KsXmq5vhbipbJEADrnduoXy+HJ7sV4sa3ppomwF6+pXRc9me4q7UPsjj6ypeErd3zx9Omade67JXqUW0roF3UEVsaSSpJAACSSfAFfAyZ+DevWgVHxGZ5x1FkpXrTLvb544EZYEYtDVDyV+BPJ69cLmyZNNhbr9nArnxeVHp+5Heu6rX8K6X3892h9zTOTxDMmKpTJV7u0oW9BN+/ePPOMFCZBA8gnQCjr5c1CvHmh847DyNRWIO806Mb7RIP2U+BHQ6FMqOZINSVMitjnkkAqVr3z2WUw/4wx4RUUV7jfWh6kSM7pAomhpcJGl//sf9Lw6LKqQVfpIdK8t4pak2+ka95pCAiSQGgS4BZ0a8xBzL5aKu8mKRx6p8l7Xyy+vENGpSgHeiJmARpzSqFka11njMpeXcFzmbtdei84SXaux5RcxsDKRsqRhtZJuJ9vj7SVcZflMSoWyOp4trlfG51n8izWLk2ZbopAACSSXQMV/TZLbF7YeA4FWkpKuvFgSrEENbDrKaowSXwIacSpPkjJUVr7aSh8x0NI0h2G/3fi2XHdtA2UFrtKkY0dj7bz6lVdMJqXwm4WyfT5T3Jh0he6L8lVZ/+mn4cf8SQIkkEQCPANOIvz6Nj3///4P68QvtO9//4tN335rQhQ2Hz68vtXxvQYQaCtb0prTVzMq9bntNkzad19sL0qwWQ25fGNpSjMgbZFsTflDh1ar/LWu5vJspG5Bi6GYlp8ibkbNyrIl6fOZZ56pPyKi59jlg35EHvCCBEig0QlwC7rRkTeswU0//IDZZUkDdhCfz3CA/obVGv+3NfLSYlFIvf/xDxNKMf4tpEaNGz7/HHNl+1l3HzR5ghrHqWz//PNoJsqxITJLYlSrcV0TCZ055LXXTBux1rf4X//CqueeQ95225lUhpopqWOZcVasdbE8CZBAfAlQAceXZ8JrW/vOOyZzTodTT0XPP/4x4e3VtwFzNirnpTkShWnYhx/Wt5qEvbdVkh2oslQrZw1JWV9RK+TJ++9vzlfVR3iuxGwuKjOEar7jjmi1zz71OhvW89oJI0dGujVE5l37SiEBEsgcAlTAmTOXKTWSRbfeitUvv2z61OPGG9HxxBNTpn+rX389EvZRI0bpOWq8tvALJbfwTHEPE2fbSMKH+qQ1LFqwAFOPPtowU2vmTg082/fXzpQsS5thdd4xZeaBHSGBbCdABZzGvwG6+tKE7Wqd67RokZSRFP76q8mt2+vmmyv4oGpnNo0fj9lyNtpVEtt3ufDCpPSvcqPFwmvaccdVcBtqOmgQBr/0kilqjJXkLLUh0a4KZdtYYzur0ZbGnG53+OGVuxHV562zZ5tyDfXf9RaNg/eGfAEq2QRrzKNwhtFQL6oJYCESSDABGmElGHCiqtcoWDPPP9+4oOj10PffN4o4Ue3VVG/pmjXmkRoA6fa4SjhvrfrN1mf1ZypJ0F+ufGkJtG2LkmXLIi144kMbll933dVclu+3KtNp4tqj57zDhHNdSRGal9s6Dtdbn58NVbzhNv1fHzbKVz/7390GUAGH0fAnCSSVAN2Qkoq//o3Pueoqs/pV5auyWGP/NpLo+aS2P0nON9d98AF2lFXwxq++wgKJwrVQ/qx4+ulG6knszeT26AFbQkqWl/JBNNSlSLMHlZf5EoxDFbeGc1wj57zpJvZefwt1Oa8t7EMkXSGFBEggJQhQAafENMTeiSbt2lV4qXjRogqf6/tBzx7XSKCG2mTNq69ik7i+BDdsMLlmN379NTZ9/72cMZaa5PZr33yztteT8ky365fedx9sSaTQ/8EHjd9snuQCbn3QQRhSrr/9H3gAwyul8+spbl8q6g8cDvW4ReIs6590EKttfzhXr4Rz3s+we+2bDl1mH0kgKwhwCzpNp7mDuJKoElSxmzVD18sui8tIwoY/uZIgvsUuu1RbZ0CUv/qT+nK+6cqq0JH2u0s84sWy+tVnfcX1JdVkuvDS83K1UO53110Y/sknUXexab9+FbbSg+Jzq/WpDHjySbQQa+dUFyu3JaB/KCRAAilDgEZYKTMVsXdky4wZ0D8aVrB5ueALsde07Q31M14jVsJ9br+9xuAPWlqtnAskwH97sW7udPrp2ypI0av1n32Gedddh6ESwES3oeuSItlRWCwxlFVp5w8ebPyZNdmFird1K37dbTdzPUJW/4GWVGwGBv8iARKIiQAVcEy4WDidCOiW84pHH0UP2ULuKEkLohXdrp60116R4mp81VkM3rqKRTeFBEiABOJFgGfA8SLJelKGgK5efxPXH1W+KotvuSWmvm2VXQXdSg+LL2fb6lJFIQESIIF4EuAZcDxpsq6UIDBLMv+UiutQWDQsZCyixlZ6xl1e/GCw/EdekwAJkECDCXAF3GCErCBWAq6k9psh0aJWiAFTIqTpgAGRartdfXXMMZnV/7ajhPpUf18NpZkrmZC2u/feSJ28IAESIIF4EOAZcDwoso6YCEyWDEKlK1bAysnBgIcfhsZMjreoX7TmxNX8uHWJ+jVPO/54FEl86OFirNWkfXvzStH8+dAvC3liBa2W3hQSIAESiCcBroDjSZN1RUVAV5cq6sbUNIq0fRpla4mmXhQL7Wilx+9/X6PyXXDTTdBkEWE/Xj3zVeULCR25rpwPsFqXN5M0j1S+0VJnORIggVgIcAUcCy2WjRsBjSGtW7uVA4pU18CCv/4V6yQEpCriwZJrt3yIRk1+oIq8xahR1b1a4Z6mDdT0gWFR3+kuF19sPq568UVsmTkTvUQ5W5pIgUICJEACCSZABZxgwKy+4QRWi9JdVGbJrH63W+fMwaxzz0VzCRSyddYs+MXFJqJVH8k9XJuYFIlSQNMP6tZ018svr604n5EACZBAQglQAScULyuPFwHdftasT7ndumGCJEzwi4oqVK1KddhHH0XubZ4yxcRvbiEBMwp++gkt5R3NLrT4zjvR7557GpTtKNIIL0iABEigAQSogBsAL5tf1XCMsyQ4RS9ZmTaL4hw3nqx0q3i6BNbQWM4auxqyNa35d3eSNIBhCa92m++8Mwp//hkdzzwTei5MIQESIIFUIUAjrFSZiUbshxof6VmoWvnWV6bIFq6uKGfEEGGqtrY0reEccRmaKrl6pxx1FIrFSnrFU08ZK+Ty72389lujfHVlO+iFF8zZb7tjjsHwTz8tXyxiWa1GVCrRnBFXqIAfMpqA+/N9CN7TGf6WUDrNjB4sB5eyBLgCTtmpSUzHNJ3eUtmCNYkUxMWmv7gBtSyLaxxLi+HYyh1EAYezBcXyfvmymg5wooR+1D6FxZb4yp6EhMyRpBDDPvwwfBurX37ZxKFuIVvKAx55JHKfFyQQLQF/5US4T+8eKt5xOALnRG9dH20bLEcC0RCgAo6GUgaVmTh6dIVVZf6QIWYlmcwh6nmtWiiXrloV6YbdogWcvDz0vftuaM7j+ddfj6GiiDVLk6681TfXErchCgnESsAPFsF9eBCweSXsI56EPfjkWKtgeRKICwH+CxYXjOlTSTjIRLjHwXXrwpdJ+6nJDsKZhsKdsOWeGks1HzbMKF+9v/qll8xjdUOi8g2T4s9YCViBPDiXzoFzweSEKV8/WAz3uzvgfvEn+L7YKFBIoBoCVMDVQEn3Wxq3eLWkFFz6v//BK7etq+NqvtNOJsSiXlu5uWgpK+KGyhpJaD/5wAOxeerUelWlEavyhw41kbG0Ag0BqeEkm8nqXGXkN9+g/QknQMNKUkggHgQs24HVdrt4VFWhDn/tDLhf/QXuYyPgf3sL/F8egDfupgpl+IEEwgS4BR0mkUE/Z19yCQomTIAqYl0tDpJkBOHkAhrMYtn996Nk2TKjjNuL0VO0q8kS2SLWHMALJL1f/4ceQsvddzdKd8Zppxl62sYOYnFc30AWa9991wTD0D63O+KIKqviDJoiDiUDCfiFy+E+0LdsZBrMxQ9dt+yJwCUzYx6xv3ISvN+ekWpc2PvdIf8P58ZcB19IbQIVU76kdl/ZuygJFC1caIJTaPHg2rUolT+eBKtYfPvtKJo3zxhNtbryyihrCxUrnDwZ82+4Aep+pLL4X//CYFllN2nbFrac1aohlSr8LbIKbiar2fqIBseIJnZzfermOyRQHwLBp/cEVk6Ac+onsHrUvlvkL/9Zory0BEo2SVNlyrdVLzgnvhtz037BMrhvnAAULJGtKgfu8l/gnPGFfFnmP9kxw0zhF7gFncKTU9+utT7oIJhzVdlidsSYyc7Px1RRbpvEhccYNMkKVg2ZohFVrCuffdb4/Oq73pYt5rUSSfe3/pNPTFSpQa++inayktawjvVVvtH0hWVIoNEJFCwONZnbqs6mLbGoljOfULlAU1j7/BOBi2fIVnf/Ku96Ex6GN3ebdX/lAt48CSpTsDR0W1bAKBJbjfXzoFvclMwhwK9TmTOXkZH0uO465G+3nYmd3Paww8x2c45EkCpZWvY/tAStUAVaPqZy5OVKF1OPPRYlq1cDkpS+vHiFhdgqATFw8MHIk5jOvSWGMoUEMomAN198y7fI776I1XFYnUOzdLUrhl3er4/A6rwj7IHHmnd8sbaW/xlhNe9iPrvf/B3+D3fLtRhnHfca7D4HVa1bV9JN8uX/u82hZxvmwdWz5Nlvwdr+JDhHPV31Hd5JOwJcAafdlEXX4XYSzKK9BKiwJeWfKl9dCcNxzMvuhg3YOH68WGeWbZPVUOVSOSvW7evKyleLqwFXXq9eNbzJ2ySQ2gR8L4ig+AJ7M16vtqPenA/hfXxF6Fl+x2rLVHfTatkDzj7/2KZ8Swrh3t/bnA2bLWp9yVhFi/LV///cbb7v5pFsOXvLfob/wYUh5ZvbGugshpNnfWeUrykz4xX9UUV0TNqGv1VWy5S0IEAFnBbT1LBOelu3okgSGOQPGoRAhw4m7d4a2TbWJAdh0W1mDd+oFs1hWSGBLjTRgebtVctkFVW8Tps2aC5l28q2NoUE0pGA9+m1crYrRk7vnFGl+/662fDeOA7YuBDI74DAFfKzvlKs58FlIu5PKvbov8La+UpYhz4Ie7vDyx6KPt60CO4jQ+E9t5csjmXHSc5+oQr9xLdhdx4J+/AnTVlr3zsQfGJneLO3nS2rq5P73L5ybnwS3Hu7wV8v6TUpKU+AVtApP0UN7+DsSy/Fpu/kG7R8424zZgzWS85b9bvtLBmFwhmBwrGTtbWdJMWfyhJxYyqSs+KOZ5xhktTrOxrG0mneHC333DNq62lTGf8igRQi4BcXwH18JKw9/gRn5AUVeuZvXgX3UXGB09Vpk2YIXLWswvNYP/iFKwBRrlbXXWp91ZPwmMZlKSh2Fk3li7KTI1vgEirTthC4bn3k3eCdoS/D+tw5fyKs1n3gLfwK3lunAMUbTDlr+DlwRMFTUpsAz4BTe35i7p0qyLXvvIPuv/tdxPWor2QA+u3QQ9HtmmuMe4+6IKkLUTgXrjbSWs5yN3z8MdrJmW9Yul91Vfgy8jOvd+/INS9IIF0JWLktELhs2yrR91x4H1wAe7/bYTXrZBSbP+VZWDuVbUM3YKBW887igC9/6pKRsu08+Sn5ouyK1fXH8L78P/jTXxZXhuIKb1rbHQF/gZxP6xZ2i+7mmW59R8QRd6UWPSMfeZG6BLgCTt25iblnQTnbnbTPPoD447bZf3/0FVehyuKK8ZQmPqhOkRYtWmQMqiq/w88kkMkEvDnvw/v2VuNupOMM3LA1KcMNviDGWEu+gT3mEdjDzhT96sH/Tr4QDDwBVrsBFfrkzXobVq995Uhom3W2iXEtkbdUQds7XVZvf/wKDfFDQglwBZxQvI1cucRG1vNaTc0XDrxRvge+WDJP3Htv2Vpz0feuu9DmgAPKP6byrUCDH7KBgDv9NTF4ki1oV1aZcuZqHbvNLqLRx79p21mzWk67T+0G+/jXqihf7Zc94Ogq3bM6jUTglJpdm6q8wBtJJ0AFnPQpiF8HApJBSNPybfjsM2iKvspi3IlE+apy3jJjRhUFXLk8P5NAxhNY/lNI+epAxe3HlvPUZIlz8Uz4C7+E3Xs/BO9ub6ygvWdGw07SijxZHLKpXVpBZ9hsB1q1Qk3hJTWT0CBJ59floovQ7fLLM2zkHA4JxE7AHlVm5yCBM1BSkFTrYQ3hqspXxT7hrdBgOgyTNJ1lvsChO/X621/2I9R4K/jw9vV63/3+Xwg+vgNc479cryr4UjUEqICrgZLJt/K3376C8VUmj5VjI4G6CFgtxEf+GrE0Nv64FvwZr9b1SqM812QRsicOrJkK93VxiWqgeJMel+pkw1Ndq2IUb+qL8L/5ByBRuPzv74C34IsYa2DxmghwC7omMrxPAiSQFQSsnGZwrl4Jb/wtsPf+e0qM2S8Sd6Kc5mZVjiXjG9wn+xDJyuSWwhosrkoxiq/RuNQnGeKbLHWgSILzUOJCgFbQccHISkiABEggdgLulOcBCTNp7/l/YjtZcUNS8wn7i76Cc8STxjUq9torvuEv/loSSuxV8WYUn/xgEdwnxYdZtuitDkPgnPReFG+xSDQEqICjocQyJEACJBBnAq5s6/o/3iVRryRO9E6XwhEf5ESJ+9n1kpv4PlgjzodzyH2Jaob1xkig4leuGF9mcRIgARIggaoE/KXfw1/xa+SBL8E0gs/uBW/e2Mg9aLhIWV1qBiV/6gvb7ifiSmJSG0mhdIbe7PeMYZdJVpGIMadBnVwBp8EksYskQAL1J6Bbr+6LB8OSFaYz6ur6VxTlm5rgwfvwYuNGZB/1LOztT4A38w14b59uaggH+vCL1sN9eJDEe+4O5/g3YLVMbPQqDVdp95JAPSkg/tpZEgp0RKgnrXpL2sbpKdCrxu8CV8CNz5wtkgAJNCYBu4lpzWo/tFFa9SVKVTiNoDfxMdOmPfA42Ic+BOfS2ZE+WHltELh6BQLn/hw35euLkZSmQ6wu01lY+fpiCa2ZkxIpJjOTfPGpSUzGJnX90pSLTdvWVCzj73MFnPFTzAGSAAnURsBb8Bm8VyR842GPwR4aWqXWVr6uZ2Zl+5D427YbKJmM3oGV17quV+L2PPiKZCjTONE99602Kpa/bg7cx4aZ9sIr8bg1Xq6ioETxwqpJcuZ8npw531/uiey4L/wC3psnhyy8tzsS9qCTQmE18yX4SJYJV8BZNuEcLgmQQEUCVm5Lc8PqJkqjDtGzXPeTq+EXLK2xpFnZXrMSgTPHRZSv+8sDJhBGogNZhMegcaKrla2rq70d95uifFX8SU9Wqdqf8HBI+eqTOe/Ce+8cuPf1hLdhQZWymX6DK+BMn2GOjwRIIG4Egk/vLiu7KaJZgjElbYikEJSeJHLlGc1A/YIlQG4biRvfLJri9SpjchuPvQrOkeJCJVvt5UW/vLgPblf+Vui6WWc4F8+QeCGSzSlLhCvgLJloDpMEsomAZgvS7EDxFs00pMoXzbvGVLV9ysemvH3yBzG9l4jClqQwTKTy1T6rQVngxLeqKF/zTKKPmS8hu11fcXjFm+BNeqzivQz/xBVwhk8wh0cC2UZAt3n98f8UQ6gtJq9ufYJP1MZM4yqj44haV2p6rowVE2DtKnm5KwXYqK3uZD0zPsnjb4Vz2TxYzbskpBt+6Vb4vz0FiNWz3W+MaUNTLrrvXwBMe1E+S+jNQJ4ECxltzs4T0okUq5Qr4BSbEHaHBEiggQTWyBaxJrF3cuCvnlZjZeYf/3fPhr9pcY1lqntgdd3FKF9/8yq4X/xRLIrdKsXUqMsb91dROM9WeZZqN/w10+QLy62mW67029+yWuJPnwD3x3vi2lX3seHwJF+xGmC5U18ydeuXk8ART5gvSshpIfMmSnr+J3FtN5UrowJO5dlh30iABGImoHGP0W1XWX3+HvaO4o9bg/hijetPf0UU5U01lKj9tnt/L/g/31ttsgRrz78AcqZpDTuz9kpS4Wm7QbD6hlak9hFPiW/y9vDnvg//23/Cm/tR/HqoCS8077JXWrbiLVd19z1h9RQfZScX9uFVDbfKlcyoS25BZ9R0cjAkQAKxEPAWiaVyl1GwmohPaozijr1CrHwfh33UcxJs4/gY30694roj4L1xIswW+1bJECVRs+zjXofd9+C4dNZ971z4s98N1Xviu7C7jopLvelcCRVwOs8e+04CJEACcSLg/fhfeF/dKEZmsqXebjCsXa6BE+cVvEYlUwM2q02/OPU6vauhAk7v+WPvSYAEEkBAo0W5Lx8GbF5poldVdqVJQJNJr9IvLpDt5wGifAchcPrnpj8motX0V2EPOTXp/cvEDlABZ+KsckwkQAK1EvAmPQGr+x6w2knEqmokqH6qJtiGBUuiYzmHPVpNqcy/5Ur8al/iWKPbHrBHnAtryOli1S3WypS4EKARVlwwshISIIF0IKAWy+6Hl8Abe7kkA9ihxi5bIy8yBkHqFoNe+9VYLuMf5DQPDXHpeHgfXAj3iZ0zfsiNOUCugBuTNtsiARJIKgH3k2uN4ZSxxBWLW+eKhbByW1XbJ2+auMrkd4Dd+4Bqn2fLTXfay/DfP0/Ohj0z5GRH8sok7lwBZ9JsciwkQAK1E1ADo3AmIHWLKdlcY3l78ClZr3wVjjP4ZNhnfQu0Hwz7mJfhThZXpa/Ex7lMIdcIkA/qJMAVcJ2IWIAESCBTCGgaPBOHWOITW3vdJFa+Z2XK0BptHOG41uozbItyptSfABVw/dnxTRIggSwn4Eu4S++982ANPzsSXjHTkbhf/hn+gs/hnP0dDbIaONlUwA0EyNdJgAQyh4AvgTnQXiJDydlvNBJ8dDiwfo4JLuGc+2ONVtXR1MUy2UeAZ8DZN+ccMQlkJQFNRu8v/Q7+1rXVjt9fMx3uS4eY3LTVFqjmprX9cZJDQP4Z1fCKzbtVUyIzb/luqcTZlpjblAYRCDTobb5MAiRAAmlAwJ38tCQcuC0U5Uny4Trn/1p1tZrXOjSS3LKftYzLXzsTriRyUP9gv9PIUDjLXEkmkCXiPjJY/KSXwD7kfvEPPi9LRh3/YVIBx58payQBEkghAv76ufA/vUYy7RRFeqW+wIEzvox81gtNw1eXi42uoiGJ7N3HR5p33WdHI/C7jRXqyYoPxaExV5cJKivGH6dBUgHHCSSrIQESSFECW1YBTdubFVukh2INHav4y3+G+4L4BKv7UhuJlFW4THx0JFBHFopz1TL488bC3u7wLBx9/IbMM+D4sWRNJEACqUhAlaVk9qkgpTX7/1YoV+6DN+tt2cL2pa4cWOIjDA3QUbQO3pQXTCnNcRt8ajeoZXSmi6WZkqh8GzzNtIJuMEJWQAIkkOoEvFW/wXvxYDGUkhy9opBtObu1wme+UXZeDY/c14+R9/uLHhZFPPGR0Jt2EwR+vwkh/1iJHb3DhXAOuifKWlksmwlQAWfz7HPsJEACNRLQTEDufb1gdd0FzglvRsp5YtClsaRNaMZWPeGc+Y24LbWXCFFi6PX9nWLgNRGW0yRSnhckUBMBKuCayPA+CZBAVhNwf74P/ufXGwbljbPUjcm9tzvQrAucS2ZS2Wb1b0nDBl/pYKRhlfFtEiABEsgYAuJmY0TOfMuL1bRdndbS5cvzmgRqIkAjrJrI8D4JkEBWE3D2ux32mIdhXzQVwf91hadRsmoRf820Wp5m3iNvxhsIPjIEftH6zBtcI42ICriRQLMZEiCBxiWgbkPegi/gb1ld74ZtSdbgvXumWDuvhydRsmoSd9ITkit3JwSf3aumIml/31v4FYKvHg0/WAwNROK9czqwYR7cF6vn4hcXGMM0V2JlU6onQAVcPRfeJQESSGMC7gcXwX3rVEmUcI4JLelVCpvoiy+vKpFoxDn+DfEjbgf7lI9rLO7PFCMtywFE6WeqeC8fCiz4DO4bJ4o1eRcgkC9/8mD1GF3tkP2fQ5bg/rQXq33Om/IrI+b0Yk9PIQESIIHMIOBOfRH+R5eEAmaEh9RxBJxTP4aV29LksXUfHCCBNJbCOuwxOENlJddAURclb+xlsHf9vYS4HNjA2lLz9eAD4k8tzHDgfwGNLCZiHfsqnP5HVNthzRfsffY7WH0Phd23+lVytS9m0U2ugLNosjlUEsgKAou/qah8ddBFG4CNC0LD13CSqkhE/K//Zn5G85fZUn35cDnzlLoqibodaVzoTFW+Olzn0llwzv0JTsehkdHbbftHritfWJKkwjnw7grKV9d77qQn4X51o/kiVPmdbPtMK+hsm3GOlwQynUCrXqItckUJF28babEozaZtzWer3QBYhz8B/5f74Jz8wbYydVy5T44CNi2SLe0eEnijoI7SmfdYFSo6hJSv7iboF5JYv3B4qnh/ecAkxfDEz1oN3bJZuALO5tnn2EkgAwnYO10G5ImyDcdplmtrr5thtRDf3TJxhpyKwFnfypZ0q/CtOn/aspoTzQHroP/VWTbTC3iS4MJ743h4Eg9aRbeb/SXjxUBrW8KL8gy85b8Ygyx/5aTQFyNJ3+hPf7V8kay85hlwVk47B00CmU1Ao1h5k58CJGiG3XMfWN12i3nAfkkh3GdGwznpHVgte8b8fia/EAq7GRqhBikJviQGWqpcvRLZqp4bCfPpb14FV8/jxRodrihnNVTTnYhWveEc9yqsZp0yGVOdY+MWdJ2IWIAESCDdCGiyAGfkBQ3qtvvMnsC6WXAfGsjAG5VIWqOuhf/T3bDMroA8XPaTpHvcIqkaxchtxS+wNFGFiHu/HAeUE+uoZ+EMPLbcney+5BZ0ds8/R08CJFADAee4180T+4S3ayiRvbed/W41X0qcHWV1K2INE19pSUphkl302m8bmG67y8MyNWPJ80ruYNsKZucVt6Czc945ahIgASHgr54K70ex1B3zkOgJbgjG8kvhb1ljVrj2OT/B7jAY/oYFQMvuFThq0A5v/K3wF3wuSS1GwT7g3zDGXLE0lMFlqYAzeHI5NBIggeoJmLNJDSix/EdTwBp6hnEjqr4071Ym4H53RwUXrvLJKiqXLf/ZF+MtDdlp9z+y/O2sveZXvqydeg6cBLKXgPv8vhJGcX4EgDXwuMg1L6IgULgiUsg+67vIdV0X7qNlPsSH3A97xHl1Fc/45zwDzvgp5gBJgASqEGjVd9utnBah88ttd3hVBwHnIHHJ6hGKe211GlFH6XKPW/UxH6zy58TlHmfbJbegs23GOV4SIAF4KyfDe3pXObPsAavtQHE1epdU4kTAm/0evDdPhH3iO7D7HBSnWjOzGirgzJxXjooESKAOAmpEhM0rgfaDxTDIqqM0H0dLIPiUWD6vmghr5IVwDmbQktq4UQHXRofPSIAESIAEYiJgomJJuElLXJTqsiw3VtIfXQpLfIOz0TCLCjimXy0WJgESyGYCJrewhF3M9ghO8fodCD4j58grfzW+ws6Z42B1GhmvqtOiHhphpcU0sZMkQALJJqDJB9z7eorva2/4yyXyEyUmAu6X/wf38z9UeMfSNIW6/S+xofU8PtuEbkjZNuMcLwmQQP0IrJc0hmXib1oMq4tkR6JERcCb9Q78H+8SZevA67wD7MGnmPec0TfCa9MPVucdYTVtF1VdmVSIW9CZNJscCwmQQL0J+BsXwqQc7LwT7KOfg/esJHHY/ng4e98cqdNbNA4QRUwf1giSqC784k1wHxAXpCbN4VwyC1ZA0kVSQAXMXwISIAESEALBR4eLcp0NBJoC+R1M7l8DRnxWnZPep6U0f0viToBnwHFHygpJgATSkYB92COhbrfoBuuwx7YNYdFXkhVJFDOFBOJMgCvgOANldSRAAplBwP3pHvhf/FG2oU+Cc9TTmTEojiKlCHAFnFLTwc6QAAmkCgFn1NWhlHvVKF890wze3xfupCfgb10LTVDvTn4mVbqekv1wv7rRcPLXzkjJ/iWjU1TAyaDONkmABNKWgPvL/XDv6SRRtJbDH3s5vLFXmrH4H12ctmNqjI77P/wnxGnxN43RXFq0QQWcFtPETpIACaQCAX/dHPif/T7SFWu3G2Af+RTQ/yg451b0DfYWfoXgw4NkhbwuUj4dLvyCZQi+eDC8+Z/EtbvONWtC+YBHnB/XetO5Mp4Bp/Psse8kQAKNRsCb/a4kGTgJ0DR6KybAOf3LWt1pdFvaSOs+CFw0zVx6c96HJwEpTNSn3JaN1vdYGor0O5AP55zvJVlF/1heZ9kYCHAFHAMsFiUBEsheAlZXyZ7Ucx84+/wTgbO/q1X5GkrdR5sf1g6XRKB5b5wgFtUz4cvZccpKj70BJ0/8srYAzWSrPU7iFy4PnZX/9mycakz/argCTv855AhIgAQamYBfsATYuAhW9z1qbdlfOxNWu4GRMn7BUnjf/AP2oQ+mrF+x75bCFwtwa/sTYLXuHel7LBd+yWa4z+8HrJkG5/IFsPLbSxjKG+D/fC/QvCsCl82NpbqMLUsFnLFTy4GRAAkkikB4m9Y+6jnYEi2rMcVfMx1W+0GN2WTMbbmfXA1/ovhSa+IKCTvpHPEkfN8396zBJ8HKbRVznZn4AregM3FWOSYSIIHEEpBgHSpW110S206l2t1Pfwf3iR3hvntOpSfRfQze2y3kCiQr00SKvePlRvkipzns3a43TWnOZXuHC2tVvt7CLxF85Uj4waJEdi9l6uYKOGWmgh0hARIggdoJBP/VXBSbawoFbthae+FqnkZW7oc/DnvIadWUUL0ZhD/tZVjyXJVmfcUvWi/nyMWwmneOqgrd1ncfFIMvS9aFcg4dOOXDqN5L50JcAafz7LHvJEACWUXAOV9y5+Z3hH1WbL603sy3zMoXsh1s73dHjcpXYXpf/hneBxfA+/afDWJr5bWJWvmahnLbADktRAE7dZ6tN6hj8rIvW+PelOdhkms0tLIGvM8VcAPg8VUSIAESSAcCwbvaymo0tGKuaeXszfkA/pJvYe14Kbz3z4dz/OuwZAu5McV3S4BlPwJice4+vz+s7cbA2ePPMXXBGJHNeBXWoJNlMe1U+6775inwF3xmmNjHvgx7u8OrLZfom1wBJ5ow6ycBEiCBSgS87/8Nf8OCSncT99E+61tTuX3EU9U24s18E97bpxorZf/TaxE4dWyjK1/tmOXkwOoxGt5nvwNWToQvFuPezLer7XNNN91n94L34SUmtaRu2bvjb69S1F8zFSgtNNv5nhqMybZ7MoQKOBnU2SYJkEBaENAVoScrJV/OMmsTs+paNbm2IpFn3q+Pwhv3F7iPNJ4lsy1W07rytQefHOlH+Qt/5SQ9/JX9Z3FBmvNe+UdRX7ufXgf3dfFzjoe07GmUozhbyxeDU+BOeSH6WiVOt7wMrJ1u6vC/ubnKu/ZR5XyRC1fBfW4fBP/dEsFHhsAvDe0UVHkpATeogBMAlVWSAAmkPwH3w4vhvn06vHfOgntfD5N0oaZRuY8Ng/uUbJv+8kBNRSL3rR57hq677R65l+wLY6ksZ8to3g326V/Uqzv+hAfhz32/Xu+Wf8n76X+y8lWlKUq0bAvZH3dj+SK1XmtIUGv4ObDGPGzKWQOrfimwO40E2g8J1ePL6lcim+mXD2yYZ87Aa20gjg95BhxHmKyKBEggMwjoitB96RCgeGNoQGoYNPovcHb/Q7UDDP5bjIfEz9Xa6TI4+99ZbZlMv2lch8Ty2WrepUFD9Wa8Jl96zgzVsYMkuBDXJOe0T00wjwZVXOllf8N8uE/vBrTuJ9vdYtxWJs5pnyXcCCzcViB8wZ8kQAIkQAJlBDwxBmoiBkhhBayuP5WSKvhiLORNeU4Urwf78oXA9Fdh7ygKI0vFCkj4ygYqX0VnawQuXZ026wirabu40vQ9V7JXXQF/1ptwLpoK57IF8GWr2numbFeiZa9GU746MCrguE4vKyMBEsgIAp12MNGmQr6sW4BWvWCPujIUzWnm60Db7eG9dSpQuEw0hvwzKluX9kn1OzvNCF5lg9BsUSjZBKvzjlEPS1fOuttgdd4ZVq/9YHUanrBIXxoG1J/6vNludt8SP2gNSLJ1DTD8fGDVr3BOGRt1v+NRkFvQ8aDIOkiABDKOgPEVnfyMWMtuht3vUFht+sFkM9KECpVFtjEdCRxhtexR+UnWfPa3roV7b3czXvvo52EPPC6qsbtyxu7rlxo1ArObGOXoSPYoS7JIxVu8BV/Ae0Vcjpo0A7Y7DJglFtZusSxF8xC4TgKHNLJwBdzIwNkcCZBAehCwJCKTM+KcCp212g4Mfe6xlxjuyLmhurKobJgLSOCJhohftAHuCwcCW1abFaStfrgaFSpdZPPKUE9ViRpL5Og6bo/+K1zx24VuNxepBbOI+gMnQOzessI+8yv4GxearW7367/Bl4AczumfJ6C1uqvkCrhuRixBAiRAAlUIuBIxSkM2ok1fOGMeafCKzX3zZPiSc1iWgkBua9hHPCEr7zGmXV+Vm5xfWi26VulHKt3wl/8MX9yxbM2ZXIeYkJfi4mX3PcSU1GQN3re3yJePHZIWGKOOLsf9MRVw3JGyQhIgARKInYBZjX2nFtSigEXso1+Qbdxj4f7wH/iTn5LVtpxFtxsE58S3JMJTaPNSjYqw9DtAzlytJvnmvXT5K/jcfhL16ntYO18tluNVg2UkYhze0h/gPb+vuCANhi3nvZYY1lntBiSiqajqTKP9jajGw0IkQAIkkBIEfNmijiXWsL3bDdv6LVvPagHsLZTt0nFViF7GAABAAElEQVR/A9aLcZMafC35Bt6EhyPl3FeOgPvmSXDv6Vyrn3LkhRS6sFqGzovRcWjUvdIvHP7S76MuX7mg98GFoVuS0tF7fCTcZ/aA+8k1lYs12mcq4EZDzYZIgASyhYA343W4L48Rgx9RkONvi2rYuoK1T/7QlLWGnQ2r596AZAhS/+KIqMHQKolaFRbZ8oVmHQo0lTjOshJOE/HmfQxf/H2t414ziSG8724Xd6CZdfbefWa0xIjeD+7Xf6+zbHUFnBPeDN0WFyeI0Zga2Pmz34mq7erqa+g9KuCGEuT7JEACJFCJgPfrIyEfYg3t+Ou2FWulYlU+2r32MSEjnUMfMM8sCSGJvJbbysnZMLqM2vY5HE1LXH/8BFgNb2sovlfea0ebCn1x5fI+vgqexGt2ZUXqFy6vvaGNC0JuX6sn116uhqdqya4hOa09/gQ4uaFS2mZZfucaXkvYbSrghKFlxSRAAtlKwB5Vtq0p28hWFAZJNXFSf1r7kPslKIXk1O0wFNYu18KRpPYRWbjNetf/+MrI7VS/0PNtFfvMcRKJqm/IFUhv5LbSv2sU5zLxtx59E+xjxPitAWKPvAjWrtfBGnoGnItnJCXxhHafRlgNmES+SgIkQAI1EfDlnBFFYuTTvSzKUk0FG3A/qBGcNI6xSn4HBK5YFLqO09+eWGVrXmBHFJ7Vunecaq1ajTf3I1gdJPqVbqXLjoElKQgty6paMMPuUAFn2IRyOCRAAtlDwC9YCvfZvWFJ/lz7kHvjGrrRXztDtoV3CMGUFXjg8vkJBxvOW2wNPRPOYbKNX4e4k58GFo+DfeiDJpVhHcVT7jG3oFNuStghEiABEoiOgCVnl4HL5soK9YW4Kl/TuiOxnTUeNmQlqgZLjSDWQfeYVmxJfOGvn4fgnU3hfnR5tS27Yg3uay7f6a9IfOfqy1T7YgrdpAJOoclgV0iABEggVQjolrNGiLJ2vAROI6x+ddzOsDNDRlIS0tP9/PdyxxYf6CeqRyKRw4yFuBeUHMYNT4NYfSOJvcst6MTyZe0kQAIkQAL1IGCSNDx/AOy9xOiq78FVatBIWu6rYk0tMaSdo56F99Fl8Od9BOeKhbAaGBa0SmMJukEFnCCwrJYESIAESKBxCHgz34D37rkSrrME1sDj4Rz9XKRhTRKhPtJWn4NgBcpcjyJPk3sRimeW3D6wdRIgARIgARKoNwGr76HiwiT+0uIPbZcLa+mXboX7gLg56Xm2ZDzSHMBWjp5rp4ZQAafGPLAXJEACJJDVBDQZg/ukBBlZMzXkmys5mKMVjSIWuHJx1eIaNUxTD2q0MAmKgvWStarTiKrlknSHRlhJAs9mSYAESIAEyhHQKFeifNXq2hNfYG9uKCxnuRKxX3bdBZZGDhNrbmu7I2GlkPLVwVABxz6lfIMESIAESCDOBCwJpWkyFO36e/g/3g3v9ePgzfkg5lY0nKVfEsrTbPIpd9lJgpS0N37SMVeW4BdohJVgwKyeBEiABEggNgLBhyUGtqyInWvXxpRm0ZPkFJ4kwYAoYA0x6S/7Ed7754e2n3sfgMBJ78XWkQSX5go4wYBZPQmQAAmQQIwENPWiiK/ZnmIQ78s/G+VrtrF/vl+2nEeGlG8gH5Yo4FQTroBTbUbYHxIgARLIcgLe7PfgfX+HBAL5ApZdt62wv2423NeOhX3Av8zWtWY3cs75wUQH8zevBCTVoUnvmGJcqYBTbELYHRIgARIggRABX6JdeZ9eB3vMQ7XGetaQlWHRdIPpItyCTpeZYj9JgARIIMsIuM/tC3/ai3Cf36/WkVvbn2CeW/0Oq7Vcqj3kCjjVZoT9IQESIIEsJuAt/QHeC/vDuVR8dreshvvULnI9G1aL7rVS8Zf/BHTeOa3SGFIB1zqlfEgCJEACJNCYBMLbyRrdyjnhzcZsutHbogJudORskARIgARIoCYC6sfrvnkynFM/SbnYzTX1ub73qYDrS47vkQAJkAAJkEADCNAIqwHw+CoJkAAJpCMB9+3TELynE/zignTsfsb0mQo4Y6aSAyEBEiCB6Aj4M+VstXiTBLoQwyVK0ghwCzpp6NkwCZBAYxDQLDvej3cBqybDPvzxqAI7NEa/ktmGv3IS/Pkfw97t+mR2I+vbpgLO+l8BAiCBzCbgfv13Ce4vCth3YY24AM5Bd8dlwH7BEgny30kCRDSJS32sRKZoybfw18yAPVLiN2eBcAs6CyaZQySBrCZgErD7EhPYgz/jNdHDpQg+sxfcXx4wWLypL8LfvComRP7qKXAf7A/37rZyjrqp2nf9YBHcL/4I99tboatwSt0E3BcOhPfxFfAmPFR34QwoUXeQzQwYJIdAAiSQvQTsUVfDWzsd/pTnJDH7OriS5g4rfobftC1cJwf+J9fK0isI56plsPLaRAXKk3R5RrwgULgCyG1Z5T2N4oTVkt/WduDlt4Ozw8VVyvBGJQKSkhAb5sMacHSlB5n5kSvgzJxXjooESKCMgCUK0DnsUVi7XCeK1pPE7IdD4wUHTnxbPsvKVJSvESevCjNNbxd8ZCg02H95sYaHtkitPf4PVrsB5R9Frq28VqG6Zesby2LL6hOpJMsuAhdNM3NjNe+SFSPnGXBWTDMHSQIk4E16QrY3rxKl6MI5+3tJVTfCQPEWfG6UaHWhDsNRmRDIQ+C69TVC9BZ+JXloDwW67YaAZPBR8TWM4mMjzWrOPugenhXXSC97H3AFnL1zz5GTQFYRsLruYpQv7BygvSR8LxO79/41xxluPxiqfKF5ZWuTpeNDT5d+Hyll5XdA4KqlcA59IOnK1/3uDuiXCW/WW5H++ct/gTn/ll0BSnII8Aw4OdzZKgmQQCMR8OWc1n1mtChfH87vNtaa1s799RE5E74aGHo2Aoc9BOfcn+DL6tbquRd8cWNC+yHixuRU6bm9x59Cz3rsWeVZKtzw530U6sbW0CreWzRO8uYeG7r3w3/gnCM7AlHk3U2FsWRSH7gFnUmzybGQAAlUIeD9fC+8L/4sCjgofsBPwB5yapUy4RuRLWe54Zz3CyxdAYv4KyaIEg8p11jyzap1tdWso6kj2X95C7+E3Wtf043gE6OANVNCXWoqBmJnj4fVsmeyu5h17XMLOuumnAMmgewiYKn1ccehYnx1JPySwoj7UbUU+o4J3ZYt57DyNTeCZUne89pW+1p1N/0ta+De3wtBcVdKBQkrX9OXsPLVD236UfkmaYK4Ak4SeDZLAiTQuAS8Oe/De+NEWdo2gX3gf2GPODehHVB/Y/e+nrDEDcrZ448JbSvWyoMvyRcNPbeWLXXnzHHcfo4VYJzKUwHHCSSrIQESSG0Cvvjkuk/ubDrpnP+rWD5vn7QO+5sWw18rEZ/6HJS8PqiPshijWRY3QpM1CVTAySLPdkmABBqdgG4Lo3QzrFa9Et62Meia8CCsXvvDOfA/FdoLnzWXd4eqUCDLP/ilW+B9fj3sA/4jOYHFCj1DhV99MnRiOSwSIIGqBKz89glRvppEXv2Jw+IvGQ//U4mwJatcf8oz4v7zTviR+Wmf+A6s/kfJ2fTwyH1//VyJE1IWFCRyNzsv3Gf3hj/5abgPDcxoAFTAGT29HBwJkECiCWhOXfeBvvBePRru138zzWkcaOS2DjXtilItqhjEQ7eenWNflu1fy5Rxf3sW7hM7wf1fF2Mopje9yU/BleAh8RJvxuvGF1j9f1NdrB0vMT7b1g4X1burvufC/egyuD/eU+86Ev0iFXCiCbN+EiCBjCbgrxCF1qSZcXPyf7jLjNUE9xh4DCAuPtbAY2ENO6tWBv74WwC3WLSuC3/Zj6as99Gl8MdeDnUfiod4Y68I1SsJKVJdnJEXmJCUzp7/V++ualIHjf/tf30TvLllftD1ri0xL1IBJ4YrayUBEsgSAureYw0Xi+q2A+BcPD0yaueQ+xG4cgmcI56IrHQjDytd2Me+au5YPfeFKm8Va+Bxxkq5gvuQeVK/v5wrF8Pa9XrJ/LRSXKJF2We4WH0OkS80pfLFpgRW5x1ScrQ0wkrJaWGnSIAESCD+BIKaCWruh4D4M2uYzPqKL+ErvS//DI0AZuW2qm81CX/PXzNNjgJa1hxqNOE9qL0BroBr58OnJEACJJAxBJy9bjZjsY9/vUFj8jTH8U/3wL2nc4PqSfTLGkyluiQbiW432vq5Ao6WFMuRAAmQQJwI+AXLgJzmsnpsGacaG7catdh2H5XoYgfdK3mOL2jcxjOoNSrgDJpMDoUESCB+BHxJ3uD/9F9jQGWJMVVY/FIJS7luprgQjajzbDf8TvmfwRflbHLxOHPLkfy3aNVbAoRIbGbZynVO+4SBMcrDyvBrbkFn+ARzeCRAAvUjoBa0es7pPjosUoHJrPTYcLiiRN3n9xO/XTfyLJoLzUIUVr5a3tPsS7PflcQIEpVKQ0Mu/zmaauos4357i3E5cj/9XZ1lWSB5BKiAk8eeLZMACaQwAWOF3GEo7OPf2NbLZT8AxZuAEvmzVlbB6+dseyZXumr2pr4AjQNdnfhFG+R2yPdX8iKKEZMYMg04CtYe/wdrz7/C5Cyu7sUY7/migFX8CQ/U+qYnqRb9DQtqLcOHiSNABZw4tqyZBEggjQlYOc0QkHzAdrfdto2iw/CQ8tU74t6irkflxdfUh++fD/febuVvR65tjT+twTfUb7jrKDkDbmGeOaNvhLPnnyLlGnrhXDAJyO8A++zvaq3Ke/lQuI8MqrVMsh+qy1TwiZ3hLf462V2Je/tUwHFHygpJgAQylYBRmKOuMcOzjn6x6hlw676hZ/0OrYBAg2uY3MDtxFf4gt9g73cHnFM/qVAmnh+stv0RuGIRbEmrWJs4oqAT2Y/a2o72mb/gU7NF771/YbSvpE05GmGlzVSxoyRAAulCwIR7lMhWVvc94K+cBPfp0CrauXwBrGadGmUYxlf3nTOhCSick96THe8mjdJuIhrx5n0Mq9uuKe1zXJ9xcwVcH2p8hwRIgARqIOBvWQ332dFwXzgA3vxP4ZePAx1oWsNb8b/tSyYmY+C17Hs5C34w/g00Yo1234ON8jVfKn55QM7avUZsPXFNUQEnji1rJgESyEYCgfzIqC014pJQlZp/WF2OGtPv1xpyeui8Wc6CrZGZsX3riVW399nvKlimR2Cn4QW3oNNw0thlEiCB1CKgwf6tnnvDarJN+aZWDzOjN5rRyXvnDPHNPhfOmNotvNNhxFwBp8MssY8kQAIpS8CdLPl+Xz9WwjI2ztluyoJohI7Z2x8Pa8Cx8H97Eu5bpzZCi4ltggo4sXxZOwmQQKYTKC0MjdCTvL+UhBPwV/8G2E3gz/0g4W0lugFuQSeaMOsnARLIeALegi9gdZTYyHLeSkksAT9YBO/zG2Dv/XdYea1rbcxELntpDLDkGzjilpVq88MVcK3Tx4ckQAIkUDcBu/d+9frHXSNnuW+fDm96KB9w3S2xhBXIg3Pw/+pUvkrKn/gosCIU3tMbe2XKwaMCTrkpYYdIgASyhYD/413wZ74B792zEjpkX8JmBu9sKor+tYS2k2qVW9ufCMiKGU4u7H3+mWrdA7egU25K2CESIIFsIaCxod0H+sIS4yLnMFmtJUiCLx8GLPxCFFEeAr9bn6BWWG2sBKiAYyXG8iRAAiSQogTc354FVk2Gc8C/KvRQz0K9d2SV3edAOCPOq/CMH5JHgAo4eezZMgmQAAnElYBuM6s4F8+A1apXhbojz06VsI499qrwLB0/+FvXhZJetOiGwKUVs1Kly3gC6dJR9pMESIAEsomAr9mWNN1h24GwbCeqoTvnSNKHVZOqKF992T7mJZN/GN32iKquVCzkl26BP2+sySYV+YIR3JqKXY2qT1wBR4WJhUiABEig8QhorGNXUvBB4kpDkig4F07J+ihbxqXoYUmdqPmY9ctJjz0RkCQT5cVT5SwrY3tIegTpoBV0+dlL8WuvaAlK135hkn6neFfZPRIggQYQ8BdJ7tvCZaJM1gClW+Ev/LIBtWXGq74akZVuDuVjdsWyecFncL+6MTI4d9or8N46Bd5Hl8D96X+R+6l8kbEKWP3rghsnwC2cnpL8tW++G/3WiVe8CgXf7IItE8/C5l+OTskxsVMkQALxIWBJ3mB4pbL6zZEV3waTiq98zd6st+F+dCl0VZgtYuW1leQSFbfi7eHnwf3s98bFyp/+iqyMhYesjv0pYoyWBpKxCnjLxNNEWZ2Jwu/2RdHCB6udCuMEXzC12me13dTtoYLv9kHxgvtrK1bjsy2/XSpK9Dhs+ryn/D9WvUtA6fofsGnccGyd8SdTT+myF+R/SPnW5xbC3TChxrr5gARIIP0JWM27wLlkNuwD7w5tPzdtV2FQutLzJz8Fb9zNCD6zp/wZDT9YXKFMTR/c7/9lFJY7/vaaiqTkfavLTrD2kH8P89oAnXeCfe7PsNr0hf+L/jtsAWtlsdV+EKzeYul92meRMXiSjzn4n9YIPr6DLHpk6zqFJCPPgIPrx4uCO0kUlmxXiFhNeyPQdm/k9rwATothEfwlS5/D1qlXIq//Tcjtc03kfl0XxQvuRdGsv5pirQ6uXoHWVsfGT7uElKnTAvlDH0CTTkdUKb7xY/klUwm0RLMRz8h4jgl9lr+d1ruj+S7pHwc1MiBekAAJxETA/eHfkuP3YWDzKvm3RJSKxEa2drsBzuhtW7I1VRi2htbngRui34Wrqb5k3/cLlsL98BI4YmRm5TSLdMdfP88o6OBDspuwabEx3LKPeBJ2/yMjZZJ9kZkrYEuMu8uUrwL2ty5G6dJnZDW8dwXegbb7wM7vj0CXk7Hp6x2gSs/dPLdCmeo+5PS8GIGORyBv0N3VPa7zXk73c+VbQUD+a4ZAh4OrLW83Gyj3ZXqCm6BfFHK1rTxxK7By0ET6SyEBEsheAs6uvxfXm9mSGego+TdBtmUt+beieeeogNjHylatiFpFZ4JY6oZ00rsVlK/7wUVixLYTgi8eDGtMWYCTZh1h9ZOAJCkkGbkCVr4bP9YtG0+Wi62Q0/c6lMz5OwKdjkWz4VWjzZQseVa2ev8gxbfKSnlfNNv5zYRPkVswTZR/b/l/J7/atvzgZmydcwtKFz0Ip41sL5Wsku8Us0xZK6cjmo16D06z/tW+y5skQALZQUC3VL2PLgN67Qtn6BnZMegoRhn8V3NZebmyxJe40Wd+A6vDkCjeavwiGauAS9d9C1WsTQfchIJxgyNkq9sy9sWirmD8brLaLETzPb6FndspUj5VLjZPPAPBVe+HumPnofmun8l2+rZxpUo/2Q8SIAESSDYB3ZL2pzwnrkp7IXDKh8nuTo3tZ2QgDlcMq7b8LPv8VhOU5PdBTvfzRBk/JdkzulULwpL4qC33mljts1S52XTgrSgQBWzl9UBOt7OofFNlYtgPEkhhAv7mleLOtBxWp5HV9tIvETuZtdNgdRlV7fN0vemMeQhBT3dAJW+weMRYlhhppaBkpgIWFx9zfuqXiOJ9Ei33nYFA+wPgtB2dglMQXZfspj1R3eo9urdZigRIINMIaF5cf+qLsAadVOH803h3PL8fsOyHkMWwbFNbg06Ec+iDVRC4z8q/iWslbOXB98L/+V7Zrh0HK7dVlXLpdsP77RlgmniOSAQxv+soWMPPSckhZKQRltVUY6DK/r+Inb+dMazaMvF0FHzeC37pRnOff5EACZBAOhNwHxspiemvh/vo0ArD8Ge/KzlwfwndK1pvglf4cz+Ev2FBhXL6wT7gP0CHofA/vhJYNwvus3tVKZOKN9zxt4XSK86p3hvE6jcmdAYsu5smJWEqDkL5p2i/GtStgKx0nTbyi2Q3Rf6IJxFc96VsR4ccuL0SCe2WoVI07y5s+qIfSla+A3fL/AwdJYdFAiRgCJTIYkIjQ21eIQE5XGhwDk1QYG13ONBpB0BXsuKeJP/4SUhLiajVomsVcHbv/RE49yezitaH9u5/rlImJW80bW+65X1xA/ziqosqK7+DcbEKXLOywu5Aqo0lY42wKoMuXvwEbLEers7ntnLZdPxcuuZzbPn1ZPnWF4x0X7es/dINxo0pp/tZ4vbUMvKMFyRAAulNwN+4EO7bp0ODdXhfiuKUNIQoKYBzyUxYLXuYwbmTHpeEDvNh73qtxEOoGMwjvUcve5xvnQZ/1ptwThkLq2dFF9N0GVtGK2Bv1jvwl3wDtOwJe8dLo84oki6TV76fanhW+MMB4kol4etEAu3EnWqn17Hxk45m9W81aY2W+0wv/wqvSYAEUoyARtnTbD+2bqHGIMH/yv/nonwRaAr76Bfk/UNjeDv1i5osSLLCt2RclkbCElFWWC6r9667pv4AauhhRm5B61hdOdPwPrjQGBb4ErDbfWy4LA4llGOGitNiiPgGv4+c3leg5X5zjfI1Q7VzRCmLsUbxigwdOYdFAplDwH35cHhvnIDg8/vHNCj72JdNeWvAMRWUr/6bF3z1GGi0KJXgs/uE4iYv/c58Tpe/XMmC5H1yFdwHJPSkWnaLWCb4SJdQWM1XJSBJGkpGKmBftmL8Ga+HsmbopLgSI1UmzZuWPpFf9Ntd6SoxnIgh2Hqg1U7I2062ogLbrBib7fQ27OaDkD/qozT89WSXSSDLCGxQ2w05s41RQdq99jNnns4RTxhg7g93hRTTa8cB88fC/eDiEMgNc0M/xTipJvHlvDj47xYhRS1nyikhJgtSoS57xWp7ZqRLugWv4s//JHIvnS4yUgGbrZjKv2A6gVvFIjBNRK22t0w+FwXfSk7QGGTTp52x6RPJGlImgdY7ocUe49GkTfTbNJqlqWj2P2JS/uH2+JMESKAqAXeC+KU+uYus3lZVfVjujnPeL7AGnwrn8gXl7sZ+6ct2rZFFX4g+l3/mnZDHaeDKJXAumAyr8w41VmoCWJQZrfqLvqqxXGM+sPa7zTRnDTm94nlvm1A0QGvPvzZmd+LWVkb6AaO9RIjKk1Xg5uXlQFmwJYFzuohbOEO2jovFqnFhTF12JL615TSN6Z3KhQt/OBhe4TRZgX+AFnum11ZV5bHwMwk0JgF/8TfwVkyA1bqvBP0/wjRtbFE+/4P8/1wC99UjETjnhxq7pMkEnMMfrfF5tA+cMyRv+K8Pw/vuTrGAXgWr3baoeVbb2kPYWqOugrVqklhNd4c98Nhom2xwOU2egNa9Q1vLlWpzRl4I6J8ycSc+Cn/cTYC4Wdlnfy/jGwj3/fNN3mTnPMmSVHZOHC6fqj8z1gjLWzMd3tN7AM06mWgo9u5/gj30tFSdhyr9Un/lLb9diLwBN8ORLeTGlKK5/0Lx3FuR0+c6NO3/l8Zsmm2RQNoS8Ka9DE/9adUYSrMT7XItnL1vhic+uN47Z4ZchsRIKnBd423rqnuSv+hL2L3FQFPE37oW7r3dJUTjaAROTdy2ra/Zh3JbS1CPFlHNpzv5GfhjL5WEEl1D6RcDueY9b/F4qKuRc9qn4sWxbds8EutZtus1C5S/UiIZLpDxyNGd1fdQOCe8GVW7yS6UsQpYwfrF8j/CpkVA07bQ/JrJFs39W/D1SDTf+V04LYdHuqOxqEsWP4acHuc3ePUaqbSai+DGX8U1uoe4Y4V86KopwlskQAL1JBB8VP6fXj9729uBfEDy1TonvC0+um8Bi8fBPvg+WPnJ+//Pm/yUJG8QRSeSqFSE3oLP4b11qrHBcS6WKFutem1jUsNV8OHtJYNOaLfPOeMrsWzexZQMpU6UM3H1Vz7pvcjbevarwUVU4TqnfQJv/mfwv5NtavkMCasZOP3zSNlUvsjMM+Ay4vrtS7NgpILy1S6VrpBvZZJeUH2Sy0vhd6NRNOefct4rCSESJJpmcfMP+6Pgy/6ysx0KRqLpFwu+3TVBLbJaEsgyApLuroIEtwCrp8B9eQzsYWeH8tUmSPl6Cz4zRlPewi8rdCH8wZvxBoJ3t4P37S2wDn8Czvm/hh/F/acnXico2WTq9X55IKr6nZNFmYpYo66OKF/zeZfr5IcP56B79GNEnKOfh33oA6HQmaJw7d1vgDXiAlg7X4lwXZHCKXyRFgrYF9/Wojm3oWieJKFWK7g0ldwe5xlXoaaD/1NxBIHm5pubXySr9TIpXvykyU+8ZYpsacVF5JthmVhNWporDdMZaH9g+Lb5qdbXqpRLlkocVQoJkEDUBOzdrg+VVaOn8hIULwxJiJBI8V4pO28WZV+deO/qFrh8ISjeAEu3advJijNBooFBjEg0Klu24aMRq3WfkBX3frdXKO7se4u5b7XpV+G+frAHnwKrY2gnUV2SnAP/A2ffW2WrOrR9XeWFFLyRFlvQm385XsJJfm2s+fIG/BO5PS9IQZT175IvxlZqdZzb+2pJhdjBVKSrUyNijdh8j+/ikvvXxMGWrCC1RcQqWfoctk4NKX0mf6j/nPLN7CTgy4rXmzsW/vhbZLdrq0CQ7dMmYlh11TI5XmqSMCietvukrATPmwC7fVWbEXfcX+GLaxKcXDiXzRUjpdYJ6wsrjp5AeijgX09HcPUH8rscQJMuJyJ/aHTbGtFjSGxJP1gAt3A6Aq13iTRUuvojbJ31NwRaDEf+8Eci98MXRXNuR/G8O8wKtdmOr4ZvJ/ynroCL598trsRyjtJu74S3xwZIIBMJ+GoEqpbPbQfA3vPPEgZym2tgssbra3Yksc7WOMmU1CCQFgrYl3PTwu/F0bztvsjb/g5xa0sf7yld3W76or+4IGxBTs+L0XTgLShd/YnEbT4p8hugK023YJpssf8L+cMekfEl7ptypFFekAAJkAAJJJVApcOKpPalxsZ1y7TF6F+gZ6fppHx1QO6Gn+VvObf2XZQsDK3cw2ewliSHCG/zFs28EcGVbyEovrdeydoaWUT7oHT1WHOGrNG0KCRAAplJwC/eJKEmjxKPj5DRU2OM0v3+3wi+cqT8k1aasObMuO7rCfejyxPWRipUnBYK+P/bOw84K6rrj5878972pS0IUqSDgoogoigg2GPsvcQoGtFoTNTERBMTjdG/JtFoYou9YA/WKBYUgViQIlKkg4B0qdvLm7n/37mz+7a9t/W9fe3cz2ffm5l755bv7O6Ze+8p8QCqpX2wO47GzB1LuVY6ZY2YYqrxdTjcCN5246tdqmUdOoXS+91CxUt+Dk1ljmG8sqVNevfBmxUnld7VfPP+L+8rF8473Zw35UPDcUBg7/ymFJUyQkAIxICA8+QwuJr8mJxHB7RZ63oWfAOs+xj2xTOj1iY71eAQinrR06Th2CRZU1wKYNZ6Lll2M2xmR5jYto3BN/uW6x7B0m78+Ttm7bzs4S9S++O2kr/zsWGHwuWcvdijcSE4IawDO2eELcsZrA3OS9nhtML93c4wQt7XfoRXDzStrezBsDW+ssF6a2ayNnTRV8dhafy+mpflWAgIgTghYB3/AHqiSR3bdn+jNjx5sfMLq29tC4pIIrFPfNgb14EIo9qt8n9YJBuIk7ricg+4aOFlWIp9D79XAehddYDpzjtk5x4UFlnxoklwm/gOSyXKOuQF8nc5PmzZeM7g2L35eOnwdz0Vy+33h3TJVtX/4sVXUcWW18iCElfu6Oi8iXqa2NCaTutCNWfrVX3In3UQ6dKNwWX0quvyLQRSjUDgyUOIdq2AD+f1pOraA6caDBlvkwnE5QzYqO5XBpbX5GJPdAcmht/T3ul9MCOGYK6TtC7HzBH7ETCxcSEQEjVxzN72x6ylrKH/pPLNr1DRwolw7FIGm9wXyMlfXGtYbskGjNdPbsGiWtcjeZIz+jNoYR+P/Xfex66fMgZCi7vLSfUz5IoQSDUCEL4mFXtObqI9fGfO/cTuG+Mp6S3zEAYWIQM/vQVzoWq/A/HUx3jrS1zOgJ2iVVT4+SjsX3YnK6MHZY/6kDjKD7GgRapSXKqCydF7iuadAWFwAqXBTKl01R0woTmO0npcUFUk+M2zzNJ1DxmPVBmDYbQdhxrVZRufp9JlN5kVACv3QBMYgVcDco76CvbAg8xYOExh6co/Ujr8NVfZDgcHGYMD9rRViMhNOUfNRR/bbj8qBkOVJoVAPQIaYft4rzKay7KBfw/2XOvykmzlvmi03EnWG2AjFzTcSDovYYuN4w6zH+zhk8g+9t5G7pJsXzwisLMHUu74NWZ500KgeYWZbdaIV6l4/pkIOP+rel3m6D85h39orlc5sOBIPiqjK/nzJgTLs9DiZVNyijF7tOCScStlHxJfb5HcWY0Zv5cw+y9YgkPvbZKXgqsSvzhk7n931WnwW0Pb2i1YDF/TWBJrw1Q0/yzTWumKW6gt7ZbbcIjSlBAIS4DtfFUE90Sd6Teb4A3WCf8y//9Mw+zXntN2Xg1TRPtAAStOkrvkRXj72ur1BquRei3+HzdDAHPQCOeNc0j1O5HsEVfHyaii3424FMA8bCutExH/VCZ/3vh6M9+qvJrfyt+ZdAULMCxeY1ZGNQVw+Vas2naA+nwhsl2YCEHpKQ5Tep/roIT1CeRuGWUNew4CeSc8p+fixbLSO1YDfS766kQsV8+n9P5/oIz+v2mgZGSzcsfMgQet642pWGRrltqEQGoR0Jtmk573TzNo3fdYUoPOMMe8v6w3zyH3zXMRaah9XEX80Z//pfoh8apiRvX/7uqM8EfujN97mtVrPyC3+xHQO8kh1Sn5V9LiZgm6ZMUfYSf7EGUjUpCv05jwTypMTtnG5+At6yPydT2NSr/9Jfm7nUmZBz5a/faI+3hfomjuyRBQ3+AF0ud5mRr2bJgaE/OytwJgw/NOL2o3dkFiDkJ6LQSSgAD/v3Fn3U7WAZjZVfosbuqwnKmTSG+ZS7RzOdmXzyfFMc7jODnvQl9l6StE+yEeeffDyToS3r+a4ZNZF2yEKdVAou6jzeSeNn1JasyfyD7yljgedeu7FnMBXPrdP6HNiziacNXIyd/zcsqqG6ygcpyla/6KOLX3ULsJ6zCTbV95Fdu5u7+AYP2xEaqNuarkP4oKKDUpX3sI61MhoK1gPclw4JZtx97wbZ7TEhvh0CQJASHQJgQcdojx3TSyJy0lDi5ghCgLJSzJ2levINVuv2b1IxiKb9Dp5Dvj5Wbdm6iF9c4V5DwzEswCZgjxsscdLZ4xlT7lm1+lslV3BoWvyuhNmQf8LexYnb2eNq6u1JCuKmhBWcskBC6w21f7W67Kr/nNAjet508xQz49boUvB55g/9EtSVb6PnBniZm/CN+W4JN7hECLCLhrPoBjilnmXufN8706WFmKrTM4Ze3jfTfj075mLdGAH5N96vPNuCuxi6q8waQOvtwMwjoDLy9JnmI6Ay5b/xiVroCyQY2U1msShPBfa1xp2iHP/NzC5QkfQKBix8dU/DX2eJDqans3jUR8lCrGfnDFpuco44B/UHqvifHRKemFEIgSAV1RQg4HlYf3JouDwVcUwYHEcLOEzAEZVGZes1rW2xc1e9m6WQ1I4bggENMZcFqPCwGhdhcClUvRjdFxilbDreJZlD9zCFXs/gqmOPuQldXHaDnzPmj5ljcbqyI+8wNFDfareMkvjEvJ8q1vBcu5ZT8gmANrS8dPCuBFgpMjrizj56FIT6JGQPkzyfeL9YhdW0R62wJyp5xJzr964C26b7OFL/tadp493CxhR63DUnFcEKgt/dq4SxxkIa0/Ntnt7GDLunApAgetC56HOjDRkT4/jJxdn8IJ+RYqWXSZccBRvPAyY7rE95Su/EOoW+P+Gi+N54z+nHInYPkpRKrYDHV/sql8w2PB3IKZg6jwy7FUsQNv3nGScsctpIzBd1MmnIpIEgIpRWAblDw5AAv/lO1t1tADjx9Ixtcy3w1/y5KSm0DMBLCGLW7Jst9RRu+rEIQA5jJWpiGtK2By04hiFDvTUJl9qp8MXFCyqY6/65moJx3Xbfz2utX5OGLPUeVbXjf+k2tl4IRnj4G98aMxbOcOCWtylDt+Ffm7nwdt8eoZsJWNpS+ktrb9NY3W+GCHKIVzTyHWSFfYj0/vfbX5rlEEJmJ7zHNnu+GW7nPXrE+OhUCsCLAfdmf6bynwwniYNnpOgrgvFvwYqwMvJuvM/xDvaTYr7YHpJCcrjaxLvvCO5TNpCcRsD7gMM7jS5TcjGP0zUIg6g8rWPUzlW6dQxoDbEbTg6AaBsyZzyaIrqGInZnzQlmNvWTlHzcY9CsHk78M/9kLKgFBXCEDAyXhpmj3elGXhzr6Wsw76t8krQ4jAMmhi8wuBv9vZcAP5gLkuH80nUMg2yHthOmHe/n1wprIM/0c616po7/TeUFtnBTNNdpGfLLcdZYx6mqxe42qVkxMhEO8E3G+eInfa9egmgiGw56fj/tHqLrtw5+hOHmvqsU56lKyDL2t1nYlWgTFpWjfd0yZPq14dTbRxNKW/MRPA5u0R/6zt9iNbrI1cvhX7vBDG/n1+DK3fjLDjLZyH/ZhdM4L5KqMnIhS9ggAPQ2nvJ/vxFNjkqYxelH3om1gR7x8sKwdNJ1ABO+ziBZUaoLjNzjuGcg593cx6C74YjaAaU+Ey9HRsE3xPaRthmrHbIeV49dunPEfWkPOa3piUFAIxJuCueJPcty8yvbBOfpIszHojkZxvXyL6HpYQq94lKtlB9qWzSXUdFomq47aOqmAW1pmvwdEI/odgBU0Nm0g2PIElc4rZEjS7l/R1gL/nRpabG4KfBmcbafueHVb4ViCiUunqu+ub5CDObdUStZXVL9gECwb2lNXcxC4uC2cfCw+Xa5t7a8TK8woCK5+VbXgiYnU2tyI/fHFnj5pGGQfin1HWALj5fMFUUfDFkdir30qFn42Af+6fkCpVRvhaEL5wqGd+3Nnhzc+a2w8pLwTahACcR5CNLS8IC7LTItakPfQisjH7ZeHLSed/H7G647ainG6ma5p5+uG/AE48rLG3x213I9WxmM2AIzWAcPWUrr2fytb+HcvOZbw5ij8SKEQEoBCh8GBzBlHOETMh/BE9CXvDhVietrIGUcb+d5Gv/aHhqgx7vWTZb6j8+6dMfqxMh/JnDDY+pNlBSbswClxhBxChDI5aVTAD3myQco74NLgnXRWoIWvkO+TvNJbKVzxONPVXZFXUaBhOCnxwViBJCCQSAeezvxi3i/bIayPebV24hfT6GWQNvTDidcdjhUFvWKzDQ3g773IQZv9fxGXAnEjxi9kMOFIDCFeP50u5BNkuXlDxVkXwT8qfiJaUc/g0I3z53MrczwisnMM/aJHw5ToyBt5mhA3P/mKVcuF20tf1FAi+z8xLBfdDwwkA77lWbP+gTbqlS7cE2+G2q5KCD2uO2uTreJS55O9/KVl5I3jnzCuC6CnUoXolouo++RYC8U7AHvNHiobw5XGrnH1TRvia55wBW+k0TJZY+PLK6A+L8VNtXqlLd5O74Am4FMZkKklS0s6Ayze/RiVLfg4Jm26UsThMXmDHNCxZn5Mkj67+MFxoIRd82tfM+jOHv0rO7s+pfJ23h9LQzNyB7XUhlomzD3271Y5MHERicku3kr/L8cEOVkWo4tjC2SNeM9d1eRG571yM/WE4LOhxFFlHwXdsBJfxgo3LgRCIMAF3+RTSy6aQOu5+UkVb4HADHq+ikHTpHnIe7EEWPGFZ+58dhRbir0rtIJISnJC4035JavDZZB9+o+mk3vwVOS9MgGBWpIZeTPbJWEVLguRNC5NgIHWHkAZTHcwBYVO8Bq4nL4NZT/ukFr48/qI5P/KW3HHMLx+5Y78FAofSunuKIlwmVHLgQYyTroy3HKpMU6/ZuVg2wk/NxCZjumQdulLtZERBu9E+562axeRYCMQ9AY0ZmfvOJeinBSWpt01/1YmPkA2FoUgnvWEG/igR0GHuAwkvgN31M41NtDXotAYxKRuxhPc9lKyffh4spxGQwoGpl0k8+83oGMxL9IOkXYLmwA2la+6m8k0vUcGsoY0690j0BxnYOQNjrKEE5iLc4u5ZlDn4TgjEIQ0Oj5XZeIbs73xcg+VampnW42Jzq7/HRATO+BKzXuzFSxICiUggo1JJ0+J9SiQOvbfuE+84wp8WwhDal3xG9kXRqT/C3Q1bnTPnAWg2n0fuu5e20LuXCtatTnuJ7GP+GjxP9IOkXII20ZHmY8nGLQ0+H7vDkZQz6r3gebIdVOz4hIoXT4IP2l3e0Oxc2Fg/jqXgk0IOlZeKVVpX48IzZIEIX+R9m6I5J2F/eq1RFssdu9Dsv0e4GalOCESdgC7ajlCBCAyDfUq9+l2yjn8gqRWFGgLq7lpF7gfXEMH9pn3ZHFId6+tyBJ5BgBzez+WU2Zl81zVdq1u7DrlTf0bU+xiyBp5CKolmv4wjKWfAbukmXk/l8QUTB2vg6Eslq/+PqpZcg5kJfsDCt3zTi96YrSxofeMHyc6rntFyDOSSFX8AFgcuKz+G5vdxVDBzMEygsUzdBsktwjJS0SojfLm5wq+Og4L6D23QsjQhBCJLQGXD7/yAk8nqfxLZJz5khK+Gy8kAgjFwCEG9Y2lEGjR7wG+cC8dCsOSI0+Q+eTDRxs9M8AnnnZ+E7KU65EpPqQqrBwrxgpuT3E9+TZr33N/H5AJa4cmWklIAW5lwgO5nbbrqpEs3UMnSG6kcpknsBjGwxwttWF2i+Udu2TYIlZXNvzGCd/A4ir8+hwLb4JQEe6z+rqejdjxWdi7CLyJILtx0Fs6eAP/RT2CWfDUF4DAD0hp/FH5y8heaMtH+sLAPrOxMdC3DNKXLfzBbA9FuV+oXAm1BwP3qfqL8jaYph2eEEUjO40PMDNt5oAu5iDMclwmRnoKphsZy8BoO7OFXkn3hR2T9GB7vTptcM4uc+Y+Qs+i5WtdqnhjnJhwbOKc7Ud7+NbOS4jgpBbCvw0hK63MdbI7aw2d0b9j9Hug52HCLzUPjAA7ssrI1KbBnLgIgjMPPBCpZeXtrqmrVvU7BItxfuUcCJSq3ZD21G7+Sska8Dtvm3qZuxWY+RvApfPVEkIT/Ixv2uGwalNajYQWtVnWuxs0sfHPHzEfbd1J6/9+bnIwGYj/XuFUOhUDcE7BYCQurS+TPgQOJP0emv+zggxNM+tz/nIYZ9gHkfHGPdy1OPq1j/+H1pENfss6F564wSfWEpUP/HwXNP6uKaZ7hfnB11Wm9b7XvYYgwVUK+a9bg3uQTV0m5B1z1FN2S77F8swd7jX3hAvE0zPYWVGWZ7/S+v4YN7621rvGJhvOOkhW3koOl3Ywh95M/r/6ySf6sg41LRS6v0rtjf/n9mOxpGhMivAhgoBCymZQx6HZK3w/LNXUSBz4I7JyJGfIpdXLkVAgIgUgQ0OWFMD/YSaq99+Lbkjqd/91BGm4o7fOnEhVsIufxA2pUwy/a8KF++XxSnYfUuJ64h7oY21Cle0l1GpC4g2hFz5NOALPwLP0Otq/Y8/UhtJ+/0xiDJ1CwjIq+PJIUXE/q4vXmWlqvy+GLegT2TydDcN2JCfNwc71w3hnwHY19DRiEK38eZR/+EbZV+5m8qo/ib29AwPnnceqaS+2O2QB749yq7Db9Zk9TFZvh2xr99+9zcpu2LY0JgVQnwEKTdq2EvS6igFVpR7cQCu8hc7LOm0pWnwnBWtgrlvNIP1JDLiD7lGeC1xP5wP36UVL7HEw8O07VlFQCmBWM8mdgn6DC86HKDzV90F2U0cfbk+E4wvwDl0wU2PoGNOp6IXgAHHOYpaOOlHvEDDOLLfr6POO0w/xS2DlwHvEf3HJErd8R7ZRS0dfnYl8zGy4ssaRbKaA1/EyXLPstOXs8cxvOS+t2Vq17U/WENa952T7rkBflRSFVfwmSbNwu9i/dj7DdhW0gNeoGssfd3qoR6m0LyV35Ftljb2tVPYlwc9XLBi8xp2ryJdPAy9Zi5ltD+PLYytb8jdIQLYn3Q5Wvnfnh62k9L0F84NfMsi3Uok2IPM2+opEyBv6JCuE1yyTkueXbveManxx9Kb3vDUbxqUr4cnbRggvI2TkLR9gPQipZegP2XXsh8MRh5jwRPsq+f8aEFcw68JGIdZddUxYvvsrUx+EjJQmBZCCg9h1p9mjZHpi1olubOOqRneSRj6oYqWFXQEMrreo0Jb+tZBo125jWS85eKl3/cL3LfIHj/9rthsMetgulD7g16MHJzj2Q0gdCkQJuLPnNlrWH6yYnfxG0j8+m4kWXw7zp5WC2LmFNSE/4mouYcTsR0LgONtAGB6Urb8WSdvWYItFkALGbXbi8ZL/QdrthkahS6hACsSeQ2wOxgKFE1HMsues+hpmftyUV+47Ffw/YhCsSMZTjf6The5hUM2ArB8vPRiNY1xpxxdbXKQAf0Bz+sGZS0DLMOeydmpeCx+l9fmFmgay8lD28vjDy3CqyliLeYaq0FfnuunGJoRil6gSlDzYSowO2vy2YyeYDijKHPU9pdRSz2h/rmS9Fsnv+Lidi6fkFsjuMjmS1UpcQiBkBts91oJlMrHwFJUi9eTa5WEK2znw1KTV2YwY6iRtOqhlwGWZurCVYL2G/1jjnqJcR/gKrvGcfMplyRr7l2a/WKerryAHm/0uZQ/8FH9PnBXOzhlbNtiGcfR2w9N2H/IhZHE+p3CiPcY80lSy8JNg1dhEZ2DPHaIEHL0bwwM9bAWmdIlijVCUEYkfAXYoXc/xvMRYI3I1AifGQpdnvcZST+82T5C6ZHOVWpPpoE0iqGTC7VtTl2+ozY/tfOJ2IdGIhTPxTI9ntDqLccUuMIFNYwvbBv3K82a9l9IPtXfkuLK0/4i21o/86UEQFn7EWuA8CuITajYOrSn+l39sa45NDISAEKgkgkhcFqt3dmqsOvFYF8P8misn99iVyP74RLSgswKWRNeT8KLYmVUeTQFLNgNMH/KEOK/yCVqaShT+l0rWVRuNVF6P0zUpGHOCATYIU/kAaS6w5Xb5lCpVtnAyF7LbRCMzc/y4TgCGj7y9N9yq2/gdtFyNQAruHdKnihw8b67bkC4GUJmCxW8XMvNoMELOWnUdENbFXKI63jf8bqnNNO+GotiqVR4FAUpkhMR/2iVyy7CaysgdQep8bEZYPChIab6WclI8yD3oawvFU7zwOPjlIQeHs8eQWrUY/y9Eji3KP/hYvtp3btHfs0rJoHrhUBrDIGY0oLLlDI9YHFzPu4oWXYh8YsU2hiCVJCCQDAXf1e+S+9zNPEGd2IutHj5HVBkJR71xplr/ZjlZS4hJIOgFc81Fo+BDNn94rKFQ4z847iXIOra9UVfO+tjwO7J0PW+QLIHsrbZdVGrxvPUDpPS5sy26YtipgPlWx9U2jHe7P8xyYRKoTbI5VvvFZSu/3O8oYcHOkqpV6hEDMCXDQBOIfBGlQ/qyY90c6kDgEkmoJui52Bds8f7dza1129nyOSR6bCtVODmLpFi2cSGUbHq+dEeUz3ieGmnR1K1D+Yhvjtkq8/O3ke6HC/HlwkjGUXW9GVvjyWDL2/ztlHHAf/EDf1FZDk3aEQJsQUBzlp0OfNhG+zvTfehGXEBJRUuITSGoBzI/HD/Mjz57Xe1jsuco1trrVD4+1fws/OxQRhd6i0pW3U8X2qdWZUT5im2P/vnhJgMctld7D2CX7u57R4lbZDefejzoi1N/WJtVR8L/hWAIfh6X7F5pUvqWF+GUoHa4/2fRLkhAQAi0joOc9aG7U8JYlKfEJJJUWdKjHwVrJClGRdKU3Kw1PWXbuQbWKulj+ZWccHCKPTQo4RF9b+lTOHHQ7pXW/AEvlZWShb0pVK4/V6mhTThB8gpOC/XFTEvMJ/LAZfqSjrDjSlM5IGSEgBBokYF8BV5WLENZv+KQGy0lmYhBI6j3gqkfAArYUwejZTCm9z7WYEHetygp+l657mMpW3UE2fD5nj5gCARZ5s6VgY210wL6xi+acSD7Y32bAbaYOFMId5CTydRpH6b2hnCZJCAiBpCHgrp+BwAYIOFPp3tFd8wG5r59J6ph7yR55rRln4N9wwJP/PdkT55HqEjkly6SB2MYDSfoZMPNkjeKsgx5rEG0GBDP/xEtyy3dSCXwnc7jB3CO/wP5S+7Bd4whQxQt+AsuEHQiL+AFeHrCvjFS2+m548/oaP4jDCwFcOPsYcotXIyzhDKx4DyEf9nwlCQEhkNgE2P2lXvgMgkL8gqhdb/JdvdwbUAU8dHFKy/G++RPCF/8gTMjDWAlgd/tict+Ac6IO/ck+77/oTkqIoepnUOMo6feAa4w1oQ4LvxgNQfkJLKg2U/GyXzfY95Jvf4Wy8LVcsJTK1t4fLMvBIlRmH0rrNcnsC1NaHv74MLOHow0rq2+wXKQOeN+Z99BZ+1ySEBACbUPAeeVET/iyAudR1b4QrP3hfheRhuyDLw12xP7JTFKjbyFrROxWwNxnR3kvAlvmkl4SXd2T4MDj9EAEcJw+GF/eBPSMH48NIdywxiPHNYZUNW+2Nd1eKl82tRv7Nfm7enbPCsb7vs7HUvYRMxF2EeZZEU4FMw+gsvWPICKUeOaJMFqpLsEJuN9No8Brp1Lg6cNI71gW2dFsX4L6EA6RYwUfVO1aNlQjqvsossfcGiqrza6pgafh3xpW6SqKSXVl73upm1JiDzgRHy97pSqafw60og+hjEG3BZeVw43FLd1ssqwMeMkJkSp2zED0pjPxd5pmwiiGs8Xl5ayytfeZ2TfHQVa+3BC1hb6UP+sg87JghPzwl0IXkqtCIMUI6B1LyXn60OpRZ3Uh+1JsK+X2rL7WiiPtlJNe8BipEdfg/0T8Wxlo1yE95x+k9htHqvvhrRh54t+auovvcf7slJ2F/dymm0O5RSuNBnW4YVmZVX/sGl7CBoYrRqXwImZMkjTi935zKWWPfCNs2aoMp3C5WdLOHTOfKmDK5e8G0y9JQkAIGAJ66wIsZvHWD9xHmoQ9WHiyipQAZqUrNfK6yrob/tKwH+YZuNpvbMx81PNLgjpC/AHwkxIB3PDva0LkOoUrMFvG7Bap/Qm7Q/bZhmvOXARY4LJ+LEOHTWynC1MsfFBg16ywxaoyyrd/SCXfwIQKM+t241fWigxVVUa+hUAqE1Ad+8EhATxkle31MBRvI9Wxf4uR6OId5DzUywgxe9wdTa5HlyE2+ePwHQ2XvNRpIJn94ASYMTd5gAlY0ErAPqdkl93SLcRLvDUdbGiEQts7bR9iL16crMyGFausjJ4wwdrXKGRx4IdQKWPgbahnP7Kw9M0z2sZSYOtr+INmoV0O3a71jRWXfCGQcgRUj9GkjoNyZC70LnqOIeviGaTa9245hz1rzL161X8pcH8XCvyzG7Z+ChqtT3/3EcrgX355PhHXkb+h0XukQHQJiACOLt+I1V6++WXScKFZvPAyKt/+gamX/SsjfBFmoBeZmW8uFK5Klv6mQU9YTiErbGA1rGCx+a77wYpbuWMXUO4Rn2JZufF/EpkHPkJ2hyPhv/qf2K8+2ERzKpx3OuyNryEONCFJCAgB6BwNvZB8P19JvoumkdWjdfuevG9qX7uOaNdqKDLB1AhL23oZXoSReHk58LdMcuY+aM5rfiiO3sShEjMQk7vTYCyX9amZLccxICBKWDGA3pom2c0kL/e2P34bzH3KqXgJ7Jf7/454iZlTxc5PsY/7mwZnr27JBjPLrdkPNh2q2P4OPICdCkWO8E5IAnsXUNFXx+BFOh1LzqvqKWmx8HV2/Q/5GZQ55H54+BKN6Jqc5VgIRIqAM+MPZFxTwo7WvnoFKSh3BZ4bTbTtG9MEmyDVTRqaxyYf2tCpbH9bl0uszkUAx4p8C9t1ilbBsUgXmPN2aGENoW8rnHOSccGp4CWs3VjvDzhUyaIFF8F1pTcDz4LHMH9nCOMaqXz7+2ZGzvbGuUcvbfOwijW6IodCIOkJ6I1fwKFFP1I53cxYdXkhOZPHkf3jp0h1G07Ol3/FTHmlOU96GAk4QBHACfjQotHlfARl0CXrTNXhFLk4k91ZFnx+GAJInE/swzpU4pcE1PCEgAAAFudJREFU9r9tpe8TKluuCQEhEAECeu96cl6cQPbl80llYGUsROLlaE6hZsMhisulNiYge8BtDDxem8sdPZN8XX5EOUfNbbCLypdD7Y5eFlb4lq5/lAI7PhHh2yBFyUw1Aux+MfDYAeSuCG/W58y6jQL3dSDWcm5Kch7bn6hwCzlPDQ9b3L52PdmXzDL5umAjaXifkhQ/BGQGHD/PIil6YvaoMZLsUR+Sr8OopBiTDEIItJZA1UyULQZ8N0Fxqk7Sm2ab2ay5DCUr309m1ClR/9T4VIZbR+uEh0gNm9igXa8u2EQOXgBYYUuhvH3IFfUrlCttTkBmwG2OPLkbTO//ByxPnyfCN7kfs4yumQRU/5M994uZeaHv7HoIUc6+Js8a++fQZepctfZB6NJhPzN+oN2Zt9bJrX2qERmJKqMk6S/vqZ0pZzEjIDPgmKFPzoZ1yS5yXjmJ7Is+JpXeLjkHKaMSAi0gwL6gad3HZP98DbxghXYZ29xq9baF5Dx3BNlXLmnQuQdbOTgvHUdUtA1/m59ErP3m9lfK1yYgArg2DzlrJYHAg3B5WbLTeP7x3YBvSUJACJDevoicZ9n+VxH1Q4zuc95sERVnKex9t39D1hj4h/d5YUd5b9d58hAI4cXQhvZm0S2qPAI3uXAO4mKGbZ+LMIOZsDeW1CABEcAN4kn+THfLfHJfxfIY9obsa9ZCm7Jh8yZdAdtC/MFTxwHYc8I/kzrJ3foNuc+PJuunn5HV7dA6uXIqBFKXgDP9t6SXv072Fd9gdSi3ySDc9TPJ/eAqIgcuYvnlll3FwruWdei1cLyxlPTsez0HG9ldyQcHHXo3POMVbibVa0yT24hEQb1zBRTCsJTOqTPijUM7W1LDBEQAN8wn6XNZM5P2rjP7UxZiiVoNOEnnqCvOkwfDlV0RvOj0hi9ZuNSrDKbtvH426XVY2kLsUeuw641nHoXvqrf0pAcpAxQCUSJQrcAFlR1EKzOJ3b9ySD8O8MB+pUt3kz1xHpxxdDaesLiMdcGHZCHiUFslo+j1xIFoDi/meYPJd+mXbdV0wrYjSlgJ++gi03Fr/P95FfFyFhRFAk+P9FzZfXVfvQZM8GxEU6ESmEnABlFvgMcrJHfddBzPxBt6GWkskTmPDyH3i7vJefn4enXIBSEgBJpJAI42TFKV/64x06V2+3mzXhbAxT+YmS8LX07qwJ943z2PMt9t9aFye5D9s4VYHv+TCfTQVu0mcjsyA07kpxehvuudy4mgnek83Cf4hm2d8zZZ/U6o1QLbEDovQahC0HIyS9bYc9L535PzBGbGnDjPB+N/9jmLJA4ADAb5EAItJqALt5Iz9UpSQy4gq8+xZhma933dKWfg77YzPF6NwJ7r2y2uX26MHQERwLFjH3ctOx//mvTXj+KPuhP5roPRPsKXudNvIguRXBSHU0PSm74kd/VUsoZeRKozlq8rE/+T4OgsavAZpDeizFf3kn3q86Q69KkqIt9CQAhEkAAHXmBBrCCUVdXsuIH6nW8R0OW9y+E56+taf7sN3CJZUSYgAjjKgBOtev09lpW7jYTAzaTACxOINs9GCLWjEMXl40QbivRXCMQNAXfTHHJfPJqsU54ja8h5MemX89mdpL+4C0pgC0jlwYuWpJgTkD3gmD+C+OqA6jXWCF/TKxj6m9Q1vKu7+Oq99EYIxCcBd8qppmPuu5dGtYPOt68YHQ5+eXamTkJI0EqlLbRqj7nVbAmJ8I3qI2hW5TIDbhau1CvM0VZUzyNTb+AyYiEQQQJmHxcWBBacYLAHq2gl97uPyf0PhL3Rkk4ja8I9ZA2fFK3mpN5WEhAB3EqAcrsQEAJCIJ4IuCvfJvetC2CHhDjBvN/baWA8dU/6UoOACOAaMORQCAgBIZAMBIxlg5UG95SVJkzJMKgkHIPsASfhQ5UhCQEhEJ8EOIav/mEJnFl5pnyt7aW7ZDKsFfbWq4b3eVn4miXpJS/WynfeupCcD39R65qcxIaALzbNSqtCQAgIgdQi4Cx4jPTcf8FzjQNb3h/Ivmo5PFd1MRBYKKsu7EWq6cldPJnc93l/d1JYe3uzH8xVWja0r7EsjaRXvmW+6cSHvO86n4EFTxJNu85ctX+2SJaw6/CJ5KnMgCNJU+oSAkJACIQgoLcuIA1f0LRnLVH+eqKKYnKnXW9KOp/cRM4zh2FWem2IOxu4BHewJjVkUpTdzRRR/U4KVmTDxt/++argeb2DWX+qvKQQWOGv9bLlQuQIyB5w5FhKTUJACAiBkARMlCCYBVHZnur8PAQsuGI+BR6Aa8nyfHM9HjzHuWUF5E5GIAd4tLMvnh50wlPd8fpHGrN6F04+qH0fssf9uX4BuRKSgCxBh8QiF4WAEBACkSNgNJHZ33rV1i97rsK5xkxYwR+7nvMP0j++jwrnnE72qgWUlnEYWWf9p0keriLXS68mC5GaLPh0bk5yZ9xCegVCLCJCmrvvSLIGenbPzakjFcvKEnQqPnUZsxAQAm1KgJWirOMfJEprR9RpEKmBp5loRc6Tw0h/+jvSxRup+FsoR+2ZReWd8imwY4Yn0KLcS71rtRd8ZS761opkDbnQi8xEFqne8KAnqUkEZAm6SZikkBAQAkKg9QTYvzoFEFMbyld6zQfk/heesSoKyU1TVDIwjcivIMg0pW/Cz8XrG43P3doeOR/9kvQ3T6DdbKL0DmbWbXVrmec7zbGKLT9iHeMlQ1KTCMgMuEmYpJAQEAJCoPUEWDgphBP0gidoTxijWqtcU9pWuI1U6eTvcialX7i6nvB1v33JBENpfS+qa7BP+Bep0TfjJQAxvgs3kft8y73eKURUE+FbzbYpR7IH3BRKUkYICAEhEGECqvcxsAmCSRInOM3wq0GUMX5WtS92L8d8mvCD711hjiOtqMXLx848LEGzEJbUpgRkCbpNcUtjQkAICIFqAm4+wn7O/ruJOGb1PY4UQoGGS85nf8HS9T5kj7iqXhFdvIOcR/oZJSj7FxswE21fr0xDF9wN/zOhSK1j7yWV272hopIXQQIigCMIU6oSAkJACMSCgPvlPeR+cTeRU07Wma+JFnIsHkIL2pQl6BZAk1uEgBAQArEkwC4tnccQ0xd7yva1UNY6/CZSm+cS5R0gwjeWD6aZbYsSVjOBSXEhIASEQKwJ6OVTsNyM+RO0qnn52Hl2lKeg5VY0qWu6dDcFnjyE9I5lTSovhaJDQJago8NVahUCQkAIRJWA88U9pDoPIffdy4La1ARNZPuMl0n1Gttg24GH+xIVbTVlIq3U1WDDklmLgMyAa+GQEyEgBIRAYhCwj7yZrEGnEXWAMK1KTgXp8sa1ma3TXjB3qAn3VN0p3zEgIHvAMYAuTQoBISAEWksg8OrJWIZGlKMjfotZ8EQi2Bezgw/V74QGq9b5G8h9GRrXY/9M9mG/arCsZEaXgCxBR5ev1C4EhIAQiAqBwN8yTb28hMzhDAn7utT9CMhkf4PtBZ49gmi75+vZ3Ktd0stfJwUfzqrmbLrBWiQzEgREAEeCotQhBISAEGhjAnrbQmJtaLMM3Yy22WWk8+ggsn70OFkHnE3OlDNIb/zSRGSyL/2SVNdDmlGbFG0NAdkDbg09uVcICAEhECMCquuwZgtf7iq7jPTduJNU3kDTc45VbMIh2hmk101v8mi0GyBn9r3kvHEuvv9OWsO1pqRmEZA94GbhksJCQAgIgcQnEPgv9oyXvULU5UBihSz3Fewbdx9FauR1TR6c++Z5nsB2EGPxu4/I9WWQ3Yz7m9xQEhcUAZzED1eGJgSEgBAISWDV295l7B0rfya1xBRJ/7AUnrcqAxzDA5de/gaRCOCQuMNdlCXocGTkuhAQAkIgSQlY8J5FHbEEPerXRvmqRcPs0Kf2bRxmUVKzCIgSVrNwSWEhIASEQPwTMIpWD/YkdfRdZB9+Y1Q6HHSH2XEAGQcg57xVL4RiVBpOokplCTqJHqYMRQgIASHABFhD2nyvfJt07wnkLn6OTKQjK3L/8lX73mTfsIuoeDtRzr4wf0pDdEUsRS95kdQ+B7V8Zm16nhofMgNOjecsoxQCQiDFCOjda4ja9yHn3hwzcnXY9WRPQMSkRpKz+AWiddPIPvW5RkrWzw48PZJo7zrEFi4m6/z3yep9dP1CciVIQPaAgyjkQAgIASGQPARUx/6kLJvUiQ+bQVlj/tikwen3ryS97DXSu1Y3qXytQoWbIXw9V5i6StGrVgE5qUlAZsA1acixEBACQiDFCfDert78FZx0nNdsEoHJmPFumUOU25PsiXOwJ9yx2XWk0g0yA06lpy1jFQJCICUJsFKW+/mdpF2n0fHz3m5LhK+7eirRtq+9+rtgD1iEb6OsRQA3ikgKCAEhIAQSm4Az9SoI4LtIr3wzagNR+2H262P/1Irso++IWjvJVLEsQSfT05SxCAEhIATqEODwhO5X9xJBA9o68veklKpTQk5jRSByOumxGoG0KwSEgBAQAmEJOE8NJyr4nqgPQhCK8A3LKRYZsgQdC+rSphAQAkKgjQiogaeYllTf46PSonH68cE1pGF6JKl5BGQJunm8pLQQEAJCIGkJ6NI92MfNIIWfpiYTl5gdfEDz2XfVsqbeJuVAQGbA8msgBISAEBAChoDzr33J+UczTYf6YGaN0ISq81Ch2EwCsgfcTGBSXAgIASGQtAR6jYUSc/PmZb7z3iF3/QwyWtBJCyY6A5Ml6OhwlVqFgBAQAglFQOdvgA/pRWRV7hm3pvMaM2K96r+k+p+M5ez01lSV1Pc271UnqVHI4ISAEBACyU9Al+VT4JlR5Mz5pxms3rGMtHbJ+fdgct88l9z5nuvKUCT09kWhLte75rx0PLlTryTnsf3r5cmFagKyBF3NQo6EgBAQAklPwHnpOKIfFpOecTO5OV3JfXci0T7DiNLaEZXnI5LRwSEZuCux1PzW+UR5g8l3xTchywQv7lnr+YSu9AsdvC4HtQjIDLgWDjkRAkJACCQ2AXfRc6R3LjeD0LvXkg6U1hqQdezfvXNWmtpvvDlWB19G9q+2kA3BqngfOFQq3ORdrahdX6ii9qVfEsEdpXXhtFDZcq2SgOwBy6+CEBACQiBJCLgLnyZ32vXQSq4gGnkd0bcvE5XsIPua70jldGv1KPVW+HqG4JZ93VajNBXIDDgyHKUWISAEhEBMCeidK0nDD7MRvtyT+Y8Y4cuHLh8jGeWoLXPNcUs+VLcRInxbAi7MPSKAw4CRy0JACAiBRCGgt8wj56lhpD+8hqzz38dS8kKinkd5wRHsNLL6nWCG4kweR+Zn6qR6QzPKWch3l75aK8+Zcz8F7u9C7PFKUmQJiACOLE+pTQgIASHQ5gT0rlXBNq3e40nlDSL7/KlkHfN3srEPq3qN8fK3LTDfesnkYPmqA+eZwxDLdy6Usi4jvXuNVw5LznrG76FQVUjOa6dWFQ35rcv2kgvTI0lNJyACuOmspKQQEAJCIC4JWEMvJPuCD8m+ylO+Mp1EbF69YRZRp0HBPhulqPZ9yTr5KXK/+zh4nQ+scX/2ztv3IerQ1ztmRa2c7ubYOvIW71qYTwemTe6b55G7/PUwJeRyXQKihFWXiJwLASEgBBKcgC7YRM6jA7xR5O0PsyFv5ls1LOO/GSf2JZ+R2vfQqsukd60mat+blO0PXmvoQG+DORLMl1THfuRC2LsIymBPnEvKn0nO/EdJf3Ij2Vd+a/IbqidV82QGnKpPXsYtBIRA8hJg0yO26+VUaZLknXifavA5ZmZbU/hyjuo0oMnC15n+W3KmnEXOE0PJXfshWfuNI9+kJUb4cl0sfFEjubP+yKeSQhCQGXAIKHJJCAgBIZDoBNzvPyM970Gyjn8AJkj7Rnw4gQe6wHFHoVcvHHn4Lptdqw22QXY+xgz47ClwLy0+n2rBqTyRGXAoKnJNCAgBIZDgBCwoXtlnvtok4au1psArPyJemnY3/K9pI8+uFOrKRnlt7nGXTyET0hBnvCztO/ctEb4N0BQB3AAcyRICQkAIJCIB7ZQbcyJ3zQdN6r5e+SY0oOeYsu5bFzTpHvsieLnqMZrUyF+SffEMcmb9idz3fkbOQ72JNaIlNU5A1gUaZyQlhIAQEAIJQ4Bns87ksURYAianlPS4v5A9Ct6xGkiqJ8yUKkqwb5xLdPBEBGq4nzR8P6vcnmSd+kzIWazK7kq+i6dX11q0nT19eI5AEA1JUuMEZA+4cUZSQggIASGQMAR0wUZynodALdrm9TmEFnSowejyIihsLTWxffXnd0J4l8M2KY3UqBvIHnd7qFtqXeOISu6s28na/yxSXQ+plScnoQnIEnRoLnJVCAgBIZCYBLK6EmVBQUphgRMClLWg9Z51jY5FpWXDJAnOOBCe0AhfvsMtJ/01zIkCZY3erzFrtg69RoRvo6SqC4gArmYhR0JACAiBhCfANrzGvnf0TUaA8oD08teaPC41EB6v4L4ymBCisLF9YXfN++TOvJWcRzwHHtp1KPDmBeQWbA5WIwf1CYgArs9ErggBISAEEpoARyuyx/yJ7Ot3kDr8Jvz8psnjsQafSXTQZUQZed49vkyi3B4N3q96jYP3rH6kjr7LlHPegJ3xqrfJ/ffABu9L9UwRwKn+GyDjFwJCIGkJ8LKyffQdpFT1v3q9fTF8Nr8bdszGfeXKt4hKd6IMoivBxtc64cGw5U0GQh6q7qPgRasPBZ49HLGAK/eA8+DKUlJYAqIFHRaNZAgBISAEkouA3jKfnJeOxaA06cN+BeWqO+oN0N2NwA7FP1Reh31v0VYIcAjiGknnf2+iKtmXzyOVmUfOY/tDVkPIL30FpVDWD8F/CeyJEb5QUngC1a9F4ctIjhAQAkJACCQBAXf1e9gXhokQNJx1ZYzgusOyOIxhlecqXxap3hPqFiHn34OMYHYeG+LlQdMaPiyN4GXhbg2/CgpdI2vNvOtVIhdIzJDkl0AICAEhkCIEWCnKfRxCE8ESrLOmkBG2IcbO4Qjdb57EUvJQsoZeVE+Quhtnk/vSBLImziMLZdgESa+eStT/JFLQuGaf0pIaJyACuHFGUkIICAEhkPAENGa+ziP9PROjsj0kUYpi/0hlCTr2z0B6IASEgBCIPoFtCz1PVRC+vFysv4MrSUkxJSACOKb4pXEhIASEQBsRyBvsNcTuJiGA1YCT26hhaSYcAVmCDkdGrgsBISAEkowAB2nQiN2rOg9BtCIsR0uKKQERwDHFL40LASEgBIRAqhKQJehUffIybiEgBISAEIgpARHAMcUvjQsBISAEhECqEhABnKpPXsYtBISAEBACMSUgAjim+KVxISAEhIAQSFUCIoBT9cnLuIWAEBACQiCmBEQAxxS/NC4EhIAQEAKpSkAEcKo+eRm3EBACQkAIxJSACOCY4pfGhYAQEAJCIFUJiABO1Scv4xYCQkAICIGYEhABHFP80rgQEAJCQAikKgERwKn65GXcQkAICAEhEFMCIoBjil8aFwJCQAgIgVQlIAI4VZ+8jFsICAEhIARiSkAEcEzxS+NCQAgIASGQqgREAKfqk5dxCwEhIASEQEwJiACOKX5pXAgIASEgBFKVgAjgVH3yMm4hIASEgBCIKQERwDHFL40LASEgBIRAqhIQAZyqT17GLQSEgBAQAjElIAI4pvilcSEgBISAEEhVAiKAU/XJy7iFgBAQAkIgpgREAMcUvzQuBISAEBACqUpABHCqPnkZtxAQAkJACMSUgAjgmOKXxoWAEBACQiBVCYgATtUnL+MWAkJACAiBmBIQARxT/NK4EBACQkAIpCoBEcCp+uRl3EJACAgBIRBTAiKAY4pfGhcCQkAICIFUJfD/vHEcpChx7fAAAAAASUVORK5CYII=",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# get the hub gene UMAP table from the seurat object\n",
    "\n",
    "umap_df <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Store Module UMAP in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  umap_df[[cell_type]] <- GetModuleUMAP(seurat_obj[[cell_type]])\n",
    "\n",
    "}\n",
    "\n",
    "# plot sample for Excitatory cells with ggplot\n",
    "ggplot(umap_df[[celltypes[1]]], aes(x=UMAP1, y=UMAP2)) +\n",
    "  geom_point(\n",
    "   color=umap_df[[celltypes[1]]]$color, # color each point by WGCNA module\n",
    "   size=umap_df[[celltypes[1]]]$kME*2 # size of each point based on intramodular connectivity\n",
    "  ) +\n",
    "  umap_theme()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "f4962cdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c32ccef3",
   "metadata": {},
   "source": [
    "In this plot, each point represents a single gene. The size of each dot is scaled by the gene’s kME for it’s assigned module. ggplot2 is sufficient to visualize the genes in the module UMAP, but here we are not visualizing the underlying network. We can use the function ModuleUMAPPlot to plot the genes and their co-expression relationships."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "297033cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"UMAP Plot of Connectivities in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n",
      "[1] \"hub labels\"\n",
      "[1] \"SLC44A1\" \"PLP1\"    \"LINGO1\"  \"SYT1\"    \"DLG2\"    \"MAGI2\"  \n",
      "NULL\n",
      "[1] \"SLC44A1\" \"PLP1\"    \"LINGO1\"  \"SYT1\"    \"DLG2\"    \"MAGI2\"  \n",
      "[1] 10305     3\n",
      "[1] 343   4\n",
      "[1] \"making net\"\n",
      "# A tibble: 6 × 5\n",
      "# Groups:   color [3]\n",
      "  Var1    Var2     value color   color_alpha\n",
      "  <fct>   <fct>    <dbl> <chr>   <chr>      \n",
      "1 ABL1    DLGAP1  0      #d8443c #D8443C00  \n",
      "2 SLC44A1 SLC44A1 0      #fe9b00 #FE9B0000  \n",
      "3 DLG2    DLG2    0      #f4c40f #F4C40F00  \n",
      "4 XRCC5   SYT1    0.0137 #d8443c #D8443C03  \n",
      "5 AGPAT4  SLC26A3 0.0226 #d8443c #D8443C06  \n",
      "6 TMCC2   SYT1    0.0235 #d8443c #D8443C06  \n",
      "         gene_name module   color kME_OLI-M1  kME_grey kME_OLI-M2 kME_OLI-M3\n",
      "CPEB3        CPEB3 OLI-M1 #fe9b00  0.2819247 0.2914115 0.22331048  0.2671603\n",
      "TPST1        TPST1 OLI-M1 #fe9b00  0.2074987 0.2218625 0.17214183  0.1965290\n",
      "ALDOA        ALDOA OLI-M2 #d8443c  0.2330644 0.2261756 0.23697276  0.1540869\n",
      "ATP6V0A1  ATP6V0A1 OLI-M2 #d8443c  0.1992145 0.2246057 0.21959604  0.1803216\n",
      "SLC25A13  SLC25A13 OLI-M3 #f4c40f  0.2393341 0.2586805 0.17124180  0.2851446\n",
      "BPGM          BPGM OLI-M1 #fe9b00  0.1756432 0.1414640 0.09036278  0.1257324\n",
      "              UMAP1      UMAP2   hub        kME label fontcolor framecolor\n",
      "CPEB3      7.927951 -13.947054 other 0.26522245           black    #fe9b00\n",
      "TPST1      5.714437 -11.015360 other 0.12055038           black    #fe9b00\n",
      "ALDOA     -4.093190   9.852709 other 0.31897217           black    #d8443c\n",
      "ATP6V0A1  -5.187427   9.491945 other 0.27722245           black    #d8443c\n",
      "SLC25A13 -12.910704  -7.000625 other 0.24966817           black    #f4c40f\n",
      "BPGM       6.036702  -9.574216 other 0.05862842           black    #fe9b00\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/HubGeneUMAP/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('UMAP Plot of Connectivities in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # hubgene network\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_hubgene_umap_igraph.pdf'), width=10, height=10)\n",
    "\n",
    "  ModuleUMAPPlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    edge.alpha = 0.25,\n",
    "    sample_edges=TRUE,\n",
    "    keep_grey_edges=FALSE,\n",
    "    edge_prop=0.1,  # proportion of edges to sample (10% here) # taking the top 10% strongest edges in each module \n",
    "    #label_genes = label_genes,\n",
    "    label_hubs=2 # how many hub genes to plot per module?\n",
    "  )\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "182af47a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "65ac718d",
   "metadata": {},
   "source": [
    "Next, our hypothesis is that the hub genes are the differentially expressed genes that are in the overlapping pathways."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "563a2a2f",
   "metadata": {},
   "source": [
    "### **Visualizing Hub Genes: Overlapping Pathways and DEGs**\n",
    "\n",
    "The hub genes, which are also differentially expressed genes (DEGs), can be visualized in the pathways where they overlap. You can find the results of this analysis saved at `f'../results/{test_name}/{study}_{deg_method}_{cell_type}_DEGs_in_overlapping_pathways.csv.csv` from the previous analysis."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "786f4317",
   "metadata": {},
   "source": [
    "Save to pathway names to R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "bd43c3b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[conflicted] Will prefer base::intersect over any other package.\n",
      "[1] \"AD_VS_NO\"\n",
      "[1] \"UMAP Plot of Hub-DEG Connectivities in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Selecting by kME_OLI-M1\n",
      "Selecting by kME_OLI-M2\n",
      "Selecting by kME_OLI-M3\n",
      "[1] \"hub labels\"\n",
      "character(0)\n",
      "[1] \"LRP4\"     \"N4BP2L2\"  \"IL1RAPL1\" \"CHN1\"     \"IDS\"      \"SNAP25\"   \"GPM6A\"   \n",
      "[8] \"MAP2\"     \"SYT1\"    \n",
      "[1] \"LRP4\"     \"N4BP2L2\"  \"IL1RAPL1\" \"CHN1\"     \"IDS\"      \"SNAP25\"   \"GPM6A\"   \n",
      "[8] \"MAP2\"     \"SYT1\"    \n",
      "[1] 10305     3\n",
      "[1] 343   4\n",
      "[1] \"making net\"\n",
      "# A tibble: 6 × 5\n",
      "# Groups:   color [3]\n",
      "  Var1    Var2      value color   color_alpha\n",
      "  <fct>   <fct>     <dbl> <chr>   <chr>      \n",
      "1 NEGR1   NEGR1    0      #d8443c #D8443C00  \n",
      "2 SLC27A4 PIP4K2A  0      #fe9b00 #FE9B0000  \n",
      "3 IQGAP1  NKAIN2   0      #f4c40f #F4C40F00  \n",
      "4 ZBTB1   PIP4K2A  0.0109 #fe9b00 #FE9B0003  \n",
      "5 CNKSR3  NKAIN2   0.0203 #f4c40f #F4C40F05  \n",
      "6 HSPH1   IL1RAPL1 0.0317 #fe9b00 #FE9B0008  \n",
      "         gene_name module   color kME_OLI-M1  kME_grey kME_OLI-M2 kME_OLI-M3\n",
      "CPEB3        CPEB3 OLI-M1 #fe9b00  0.2819247 0.2914115 0.22331048  0.2671603\n",
      "TPST1        TPST1 OLI-M1 #fe9b00  0.2074987 0.2218625 0.17214183  0.1965290\n",
      "ALDOA        ALDOA OLI-M2 #d8443c  0.2330644 0.2261756 0.23697276  0.1540869\n",
      "ATP6V0A1  ATP6V0A1 OLI-M2 #d8443c  0.1992145 0.2246057 0.21959604  0.1803216\n",
      "SLC25A13  SLC25A13 OLI-M3 #f4c40f  0.2393341 0.2586805 0.17124180  0.2851446\n",
      "BPGM          BPGM OLI-M1 #fe9b00  0.1756432 0.1414640 0.09036278  0.1257324\n",
      "              UMAP1      UMAP2   hub        kME label fontcolor framecolor\n",
      "CPEB3      7.927951 -13.947054 other 0.26522245           black    #fe9b00\n",
      "TPST1      5.714437 -11.015360 other 0.12055038           black    #fe9b00\n",
      "ALDOA     -4.093190   9.852709 other 0.31897217           black    #d8443c\n",
      "ATP6V0A1  -5.187427   9.491945 other 0.27722245           black    #d8443c\n",
      "SLC25A13 -12.910704  -7.000625 other 0.24966817           black    #f4c40f\n",
      "BPGM       6.036702  -9.574216 other 0.05862842           black    #fe9b00\n"
     ]
    }
   ],
   "source": [
    "%%R -i test_names -i deg_method -i seurat_obj -o seurat_obj\n",
    "\n",
    "\n",
    "conflicts_prefer(base::intersect)\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/HubGeneUMAP/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "# label the DEGs in the overlapping pathways\n",
    "\n",
    "for (test in c('ad_vs_no')){\n",
    "\n",
    "  print(toupper(test))\n",
    "  for (cell_type in names(seurat_obj)){\n",
    "\n",
    "    print(paste0('UMAP Plot of Hub-DEG Connectivities in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "    # hubgene network\n",
    "\n",
    "    overlap_degs <- read.csv(paste0('../results/', test, '/', save_prefix, '_', deg_method, '_',\n",
    "                           cell_type, \"_DEGs_in_overlapping_pathways.csv\"), header=TRUE)\n",
    "\n",
    "    hub_genes <- GetHubGenes(seurat_obj[[cell_type]], 25)\n",
    "    label_genes <- intersect(hub_genes$gene_name, unique(overlap_degs$x))\n",
    "    \n",
    "    if (length(label_genes) == 0) {\n",
    "      print(paste0('No genes to label for ', toupper(cell_type)))\n",
    "      print('setting label_genes to NULL')\n",
    "      label_genes <- NULL\n",
    "    } else {\n",
    "      label_genes <- label_genes\n",
    "    }\n",
    "\n",
    "\n",
    "    pdf(paste0(fig_dir, test, \"_\", cell_type, '_overlap_hubDEGs_umap_igraph.pdf'), width=10, height=10)\n",
    "    ModuleUMAPPlot(\n",
    "      seurat_obj[[cell_type]],\n",
    "      edge.alpha=0.25,\n",
    "      sample_edges=TRUE,\n",
    "      keep_grey_edges=FALSE,\n",
    "      edge_prop=0.1,  # proportion of edges to sample (10% here) # taking the top 10% strongest edges in each module \n",
    "      label_genes = label_genes,\n",
    "      label_hubs=0 # how many hub genes to plot per module?\n",
    "    )\n",
    "    dev.off()\n",
    "  }\n",
    "\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "08af7c46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b534ca69",
   "metadata": {},
   "source": [
    "# **Enrichment analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4261d643",
   "metadata": {},
   "source": [
    "Next, we perform enrichment tests on the hdWGCNA modules. We leverage the R pacakge enrichR to perform enrichment tests on a wide range of curated gene lists. This analysis should point us towards biological processes that our hdWGCNA modules are involved in. Additionally, we perform a gene set overlap analysis to compare the genes in hdWGCNA modules with the marker genes identified using Seurat’s FindAllMarkers function.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a77a18cb",
   "metadata": {},
   "source": [
    "## **EnrichR**\n",
    "\n",
    "hdWGCNA includes the function RunEnrichr to compare the set of genes in each module with any of the gene lists hosted by Enrichr.\n",
    "\n",
    "The results of the enrichment tests are stored in the hdWGCNA experiment, so can be easily retrieved for downstream analysis or exporting to external applicaitons like Excel. In the following example, we perform the enrichment test with three Gene Ontology datbases:\n",
    "\n",
    "- `GO_Biological_Process_2021`\n",
    "- `GO_Cellular_Component_2021`\n",
    "- `GO_Molecular_Function_2021`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "f99c0a19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Estimating Module Enrichments in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "Selecting by kME_OLI-M1\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "  Querying GO_Cellular_Component_2021... Done.\n",
      "  Querying GO_Molecular_Function_2021... Done.\n",
      "Parsing results... Done.\n",
      "Selecting by kME_OLI-M2\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "  Querying GO_Cellular_Component_2021... Done.\n",
      "  Querying GO_Molecular_Function_2021... Done.\n",
      "Parsing results... Done.\n",
      "Selecting by kME_OLI-M3\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying GO_Biological_Process_2021... Done.\n",
      "  Querying GO_Cellular_Component_2021... Done.\n",
      "  Querying GO_Molecular_Function_2021... Done.\n",
      "Parsing results... Done.\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -o enrich_df\n",
    "\n",
    "# dbs <- c('GO_Biological_Process_2021')\n",
    "\n",
    "dbs <- c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021')\n",
    "\n",
    "enrich_df <- list()\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Estimating Module Enrichments in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # enrichr databases to test\n",
    "\n",
    "  # perform enrichment tests\n",
    "  seurat_obj[[cell_type]] <- RunEnrichr(\n",
    "    seurat_obj[[cell_type]],\n",
    "    dbs=dbs, # character vector of enrichr databases to test\n",
    "    max_genes = 100 # number of genes per module to test\n",
    "  )\n",
    "  # retrieve the output table\n",
    "  enrich_df[[cell_type]] <- GetEnrichrTable(seurat_obj[[cell_type]]) %>% subset(P.value < 0.05)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "22787bab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f2494516",
   "metadata": {},
   "source": [
    "## **Visualize enrichments**\n",
    "\n",
    "Now that we have done the enrichment tests, we can now visualize the results.\n",
    "\n",
    "### **EnrichrBarPlot**\n",
    "\n",
    "hdWGCNA includes the function EnrichrBarPlot to summarize the results of every Enrichr database and every module. This function outputs a .pdf figure for each module, containing a barplot showing the top N enriched terms. The following example will plot the top 10 terms in each module and will output the results to a folder called enrichr_plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e7d94ea7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Enrichr Plots for Modules in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"OLI-M1\"\n",
      "[1] \"OLI-M2\"\n",
      "[1] \"OLI-M3\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/Enrichr/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Enrichr Plots for Modules in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  # make GO term plots:\n",
    "  EnrichrBarPlot(\n",
    "    seurat_obj[[cell_type]],\n",
    "    outdir = paste0(fig_dir, cell_type, \"_enrichr_plots\"), # name of output directory\n",
    "    n_terms = 10, # number of enriched terms to show (sometimes more show if there are ties!!!)\n",
    "    plot_size = c(5,7), # width, height of the output .pdfs\n",
    "    logscale=TRUE # do you want to show the enrichment as a log scale?\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "d9adf73f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "678b520b",
   "metadata": {},
   "source": [
    "### **EnrichrDotPlot**\n",
    "\n",
    "hdWGCNA includes an additional visualization function for enrichment results, EnrichrDotPlot, which shows the top results for one Enrichr database in each module. In the following example, we plot the top term in the GO_Biological_Process_2021 database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "bcb62788",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%R -i seurat_obj -o seurat_obj\n",
    "\n",
    "# fig_dir = '../results/'\n",
    "# # enrichr dotplot\n",
    "# p <- EnrichrDotPlot(\n",
    "#   seurat_obj,\n",
    "#   mods='all',\n",
    "#   database = \"GO_Biological_Process_2021\",\n",
    "#   n_terms=3,\n",
    "#   break_ties=TRUE\n",
    "# )\n",
    "# pdf(paste0(fig_dir, 'EX_GO_dotplot.pdf'), width=10, height=10, useDingbats=FALSE)\n",
    "# p\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "fb1d1cbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # convert nested list of Seurat object into Rpy2 object \n",
    "# seurat_obj =  robjects.ListVector(\n",
    "#                         {\n",
    "#                             cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "#                             for cell_type in seurat_obj.keys()\n",
    "#                         }\n",
    "#                     )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d5bae359",
   "metadata": {},
   "source": [
    "# **Differential module eigengene (DME) analysis**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c5b93666",
   "metadata": {},
   "source": [
    "Here, we perform differential module eigengene (DME) analysis, revealing modules that are up- or down-regulated in given groups of cells."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b0ee2360",
   "metadata": {},
   "source": [
    "## **DME analysis comparing two groups**\n",
    "\n",
    "Here we perform DME testing between two different pathology groups (AD vs control). We use the hdWGCNA function FindDMEs, which is a special case of the Seurat function FindMarkers. We use the Mann-Whitney U test, also known as the Wilcoxon test, to compare two groups, but other tests can be used with the test.use parameter.\n",
    "\n",
    "FindDMEs requires a list of barcodes for group1 and for group2. We are only going to compare cells from the EX cluster since that is the group that we performed network analysis on.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "1e343be3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Differential Module Eigengene Analysis in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] 9606    3\n",
      "[1] \"OLI-M2\" \"OLI-M3\" \"OLI-M1\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -o DMEs\n",
    "\n",
    "group_col <- 'cell_type'\n",
    "groups <- celltypes\n",
    "DMEs <- list()\n",
    "\n",
    "\n",
    "dat_dir = paste0(\"../results/hdWGCNA/DMEs/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(dat_dir)) {\n",
    "  dir.create(dat_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Differential Module Eigengene Analysis in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "\n",
    "  # get cell barcodes for this pseudotime bin in control & AD\n",
    "  g1 <- seurat_obj[[cell_type]]@meta.data[seurat_obj[[cell_type]]$pathology.group %in% c('early', 'late'),] %>% rownames\n",
    "  g2 <- seurat_obj[[cell_type]]@meta.data[seurat_obj[[cell_type]]$pathology.group == 'no',] %>% rownames\n",
    "\n",
    "  DMEs[[cell_type]] <- FindDMEs(\n",
    "          seurat_obj[[cell_type]],\n",
    "          barcodes1 = g1,\n",
    "          barcodes2 = g2,\n",
    "          test.use='wilcox',\n",
    "          wgcna_name=toupper(substr(cell_type, 1, 3)), # the name of the hdWGCNA experiment,,\n",
    "          harmonized=TRUE\n",
    "      )\n",
    "  \n",
    "  DMEs[[cell_type]]$group <- cell_type\n",
    "\n",
    "  # fix infs:\n",
    "  DMEs[[cell_type]]$avg_log2FC <- ifelse(abs(DMEs[[cell_type]]$avg_log2FC) == Inf, 0, DMEs[[cell_type]]$avg_log2FC)\n",
    "\n",
    "  write.csv(DMEs[[cell_type]], row.names=FALSE, quote=FALSE, file=paste0(dat_dir, cell_type, '_AD_DMEs.csv'))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "4aababeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert nested list of Seurat object into Rpy2 object \n",
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )\n",
    "\n",
    "\n",
    "# convert nested list of Seurat object into Rpy2 object \n",
    "DMEs =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: DMEs[cell_type]\n",
    "                     \n",
    "                            for cell_type in DMEs.keys()\n",
    "                        }\n",
    "                    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4ed4d48c",
   "metadata": {},
   "source": [
    "We can now visualize the results using the hdWGNCA functions `PlotDMEsLollipop` or `PlotDMEsVolcano`. First we make a lollipop plot to visualize the DME results.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "2d11980c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Lollipop for DMEs in hdWGCNA Experiment for OLIGODENDROCYTE\"\n",
      "[1] \"Please be aware comparison group/groups are not provided, which may casue an ERROR. PlotDMEsLollipop function will automatically assume all values are within the same group.\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i seurat_obj -o seurat_obj -i DMEs\n",
    "\n",
    "\n",
    "fig_dir = paste0(\"../results/hdWGCNA/DMEs/\", save_prefix, '/')\n",
    "\n",
    "if (!dir.exists(fig_dir)) {\n",
    "  dir.create(fig_dir, recursive=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "for (cell_type in names(seurat_obj)){\n",
    "\n",
    "  print(paste0('Lollipop for DMEs in hdWGCNA Experiment for ', toupper(cell_type)))\n",
    "\n",
    "  pdf(paste0(fig_dir, cell_type, '_dmes_lollipop.pdf'), width=10, height=10)\n",
    "  p <- PlotDMEsLollipop(\n",
    "    seurat_obj[[cell_type]], \n",
    "    DMEs[[cell_type]], \n",
    "    wgcna_name = toupper(substr(cell_type, 1, 3)), # the name of the hdWGCNA experiment,, \n",
    "    pvalue = \"p_val_adj\"\n",
    "  )\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "8c6e9df6",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_obj =  robjects.ListVector(\n",
    "                        {\n",
    "                            cell_type: seurat_obj[cell_type]\n",
    "                     \n",
    "                            for cell_type in seurat_obj.keys()\n",
    "                        }\n",
    "                    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_seq_meta_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "2f414a000012e1980f5fd1119ed6b9bf8cc73b51e37b52eb52a16ff857effd1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
